// @spectrum-web-components/checkbox/src/CheckboxBase.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property, query, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import { Focusable } from '/web-component-designer-demo/node_modules/@spectrum-web-components/shared/src/focusable.js';
export class CheckboxBase extends Focusable {
    constructor() {
        super(...arguments);
        this.checked = false;
        this.readonly = false;
    }
    get focusElement() {
        return this.inputElement;
    }
    handleChange(event) {
        if (this.readonly) {
            this.inputElement.checked = this.checked;
            return;
        }
        this.checked = this.inputElement.checked;
        // Change events from the shadow DOM are not transmitted into
        // the parent light DOM
        const changeEvent = new CustomEvent('change', {
            detail: {
                sourceEvent: event,
            },
            bubbles: event.bubbles,
            cancelable: event.cancelable,
        });
        this.dispatchEvent(changeEvent);
    }
    render() {
        return html `
            <input
                id="input"
                aria-labelledby="label"
                type="checkbox"
                .checked=${this.checked}
                @change=${this.handleChange}
            />
        `;
    }
}
__decorate([
    property({ type: Boolean, reflect: true })
], CheckboxBase.prototype, "checked", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], CheckboxBase.prototype, "readonly", void 0);
__decorate([
    query('#input')
], CheckboxBase.prototype, "inputElement", void 0);
