// @node-projects/acad-ts/dist/IO/DWG/DwgStreamWriters/DwgFileHeaderWriterBase.js
import { CadUtils } from '../../../CadUtils.js';
import { DwgCheckSumCalculator } from '../DwgCheckSumCalculator.js';
export class DwgFileHeaderWriterBase {
    FileHeader;
    get bytesWritten() { return this._streamPosition; }
    _document;
    _encoding;
    _stream;
    _streamPosition = 0;
    _version;
    constructor(stream, encoding, model, fileHeader) {
        this._document = model;
        this._stream = stream;
        this._version = model.header.version;
        this._encoding = encoding;
        this.FileHeader = fileHeader;
    }
    applyMagicSequence(buffer, length) {
        for (let i = 0; i < length; i++) {
            buffer[i] ^= DwgCheckSumCalculator.MagicSequence[i];
        }
    }
    applyMask(buffer, offset, length) {
        const posValue = 0x4164536B ^ this._streamPosition;
        const maskBytes = new Uint8Array(4);
        const maskView = new DataView(maskBytes.buffer);
        maskView.setInt32(0, posValue, true);
        let diff = offset + length;
        while (offset < diff) {
            for (let i = 0; i < 4; i++) {
                buffer[offset + i] ^= maskBytes[i];
            }
            offset += 4;
        }
    }
    checkEmptyBytes(buffer, offset, spearBytes) {
        let result = true;
        const num = offset + spearBytes;
        for (let i = offset; i < num; i++) {
            if (buffer[i] !== 0) {
                result = false;
                break;
            }
        }
        return result;
    }
    getFileCodePage() {
        const codePage = CadUtils.getCodeIndex(CadUtils.getCodePage(this._document.header.codePage));
        if (codePage < 1) {
            return 30;
        }
        else {
            return codePage;
        }
    }
    writeMagicNumber() {
        const mod = this._streamPosition % 0x20;
        for (let i = 0; i < mod; i++) {
            this._stream[this._streamPosition++] = DwgCheckSumCalculator.MagicSequence[i];
        }
    }
    /** Write bytes to the output stream at current position */
    writeToStream(data, offset = 0, length) {
        const len = length ?? data.length;
        for (let i = 0; i < len; i++) {
            this._stream[this._streamPosition++] = data[offset + i];
        }
    }
    /** Write a 32-bit unsigned int to the output stream */
    writeUint32ToStream(value) {
        const view = new DataView(this._stream.buffer, this._stream.byteOffset + this._streamPosition, 4);
        view.setUint32(0, value >>> 0, true);
        this._streamPosition += 4;
    }
    /** Write a 32-bit signed int to the output stream */
    writeInt32ToStream(value) {
        const view = new DataView(this._stream.buffer, this._stream.byteOffset + this._streamPosition, 4);
        view.setInt32(0, value, true);
        this._streamPosition += 4;
    }
    /** Write a 16-bit unsigned int to the output stream */
    writeUint16ToStream(value) {
        const view = new DataView(this._stream.buffer, this._stream.byteOffset + this._streamPosition, 2);
        view.setUint16(0, value & 0xFFFF, true);
        this._streamPosition += 2;
    }
    /** Write a single byte to the output stream */
    writeByteToStream(value) {
        this._stream[this._streamPosition++] = value & 0xFF;
    }
}
