// @node-projects/acad-ts/dist/IO/DWG/DwgStreamWriters/DwgFileHeaderWriterAC15.js
import { DwgFileHeaderAC15 } from '../FileHeaders/DwgFileHeaderAC15.js';
import { DwgSectionLocatorRecord } from '../FileHeaders/DwgSectionLocatorRecord.js';
import { DwgSectionDefinition } from '../FileHeaders/DwgSectionDefinition.js';
import { CRC8StreamHandler } from '../CRC8StreamHandler.js';
import { DwgFileHeaderWriterBase } from './DwgFileHeaderWriterBase.js';
export class DwgFileHeaderWriterAC15 extends DwgFileHeaderWriterBase {
    get FileHeaderSize() { return 0x61; }
    get HandleSectionOffset() {
        return this.FileHeaderSize
            + (this._records.get(DwgSectionDefinition.AuxHeader).Stream?.length ?? 0)
            + (this._records.get(DwgSectionDefinition.Preview).Stream?.length ?? 0)
            + (this._records.get(DwgSectionDefinition.Header).Stream?.length ?? 0)
            + (this._records.get(DwgSectionDefinition.Classes).Stream?.length ?? 0);
    }
    static _nRecords = 6;
    _endSentinel = new Uint8Array([
        0x95, 0xA0, 0x4E, 0x28, 0x99, 0x82, 0x1A, 0xE5, 0x5E, 0x41, 0xE0, 0x5F, 0x9D, 0x3A, 0x4D, 0x00
    ]);
    _records = new Map();
    constructor(stream, encoding, document) {
        super(stream, encoding, document, new DwgFileHeaderAC15());
        this._records.set(DwgSectionDefinition.AuxHeader, new DwgSectionLocatorRecord(5));
        this._records.set(DwgSectionDefinition.Preview, new DwgSectionLocatorRecord(null));
        this._records.set(DwgSectionDefinition.Header, new DwgSectionLocatorRecord(0));
        this._records.set(DwgSectionDefinition.Classes, new DwgSectionLocatorRecord(1));
        this._records.set(DwgSectionDefinition.AcDbObjects, new DwgSectionLocatorRecord(null));
        this._records.set(DwgSectionDefinition.Handles, new DwgSectionLocatorRecord(2));
        this._records.set(DwgSectionDefinition.ObjFreeSpace, new DwgSectionLocatorRecord(3));
        this._records.set(DwgSectionDefinition.Template, new DwgSectionLocatorRecord(4));
    }
    addSection(name, stream, isCompressed, decompsize = 29696) {
        this._records.get(name).Stream = stream;
    }
    writeFile() {
        this.setSeekers();
        this.writeFileHeader();
        this.writeRecordStreams();
    }
    setSeekers() {
        let currOffset = this.FileHeaderSize;
        for (const [, record] of this._records) {
            record.Seeker = currOffset;
            currOffset += (record.Stream?.length ?? 0);
        }
    }
    writeFileHeader() {
        const ms = [];
        const msWriter = {
            pos: 0,
            writeByte(b) { ms.push(b & 0xFF); },
            writeBytes(arr, offset = 0, length) {
                const len = length ?? arr.length;
                for (let i = 0; i < len; i++)
                    ms.push(arr[offset + i]);
            },
            writeRawLong(value) {
                ms.push(value & 0xFF);
                ms.push((value >>> 8) & 0xFF);
                ms.push((value >>> 16) & 0xFF);
                ms.push((value >>> 24) & 0xFF);
            },
            writeRawShort(value) {
                ms.push(value & 0xFF);
                ms.push((value >>> 8) & 0xFF);
            },
        };
        //0x00	6	"ACXXXX" version string
        const versionBytes = new TextEncoder().encode(this._document.header.versionString);
        msWriter.writeBytes(versionBytes, 0, 6);
        //The next 7 starting at offset 0x06
        msWriter.writeBytes(new Uint8Array([0, 0, 0, 0, 0, 15, 1]));
        //At 0x0D is the preview seeker
        msWriter.writeRawLong(this._records.get(DwgSectionDefinition.Preview).Seeker);
        msWriter.writeByte(0x1B);
        msWriter.writeByte(0x19);
        //Code page at 0x13
        msWriter.writeRawShort(this.getFileCodePage());
        msWriter.writeRawLong(DwgFileHeaderWriterAC15._nRecords);
        this.writeRecord(msWriter, this._records.get(DwgSectionDefinition.Header));
        this.writeRecord(msWriter, this._records.get(DwgSectionDefinition.Classes));
        this.writeRecord(msWriter, this._records.get(DwgSectionDefinition.Handles));
        this.writeRecord(msWriter, this._records.get(DwgSectionDefinition.ObjFreeSpace));
        this.writeRecord(msWriter, this._records.get(DwgSectionDefinition.Template));
        this.writeRecord(msWriter, this._records.get(DwgSectionDefinition.AuxHeader));
        //Pad to align
        const bitPad = (8 - (ms.length * 8) % 8) % 8;
        if (bitPad > 0) {
            // writeSpearShift - pad to byte boundary (already at byte boundary in this array-based approach)
        }
        //CRC
        const msArr = new Uint8Array(ms);
        const crc = CRC8StreamHandler.GetCRCValue(0xC0C1, msArr, 0, msArr.length);
        msWriter.writeRawShort(crc);
        //End sentinel
        msWriter.writeBytes(this._endSentinel, 0, this._endSentinel.length);
        const finalArr = new Uint8Array(ms);
        this.writeToStream(finalArr);
    }
    writeRecord(writer, record) {
        writer.writeByte(record.Number);
        writer.writeRawLong(record.Seeker);
        writer.writeRawLong(record.Stream?.length ?? 0);
    }
    writeRecordStreams() {
        for (const [, item] of this._records) {
            if (item.Stream == null)
                continue;
            this.writeToStream(item.Stream);
        }
    }
}
