// @node-projects/acad-ts/dist/IO/DWG/DwgStreamWriters/DwgMergedStreamWriter.js
export class DwgMergedStreamWriter {
    get encoding() { return this.main.encoding; }
    main;
    textWriter;
    handleWriter;
    stream;
    savedPositionInBits = 0;
    positionInBits = 0;
    _savedPosition = false;
    constructor(stream, main, textwriter, handlewriter) {
        this.stream = stream.buffer;
        this.main = main;
        this.textWriter = textwriter;
        this.handleWriter = handlewriter;
    }
    handleReference(cadObject) {
        this.handleWriter.handleReference(cadObject);
    }
    handleReferenceTyped(type, cadObject) {
        this.handleWriter.handleReferenceTyped(type, cadObject);
    }
    handleReferenceHandle(handle) {
        this.handleWriter.handleReferenceHandle(handle);
    }
    handleReferenceTypedHandle(type, handle) {
        this.handleWriter.handleReferenceTypedHandle(type, handle);
    }
    resetStream() {
        this.main.resetStream();
        this.textWriter.resetStream();
        this.handleWriter.resetStream();
    }
    savePositonForSize() {
        this._savedPosition = true;
        this.positionInBits = this.main.positionInBits;
        //Save this position for the size in bits
        this.main.writeRawLong(0);
    }
    write2RawDouble(value) {
        this.main.write2RawDouble(value);
    }
    write2BitDouble(value) {
        this.main.write2BitDouble(value);
    }
    write3BitDouble(value) {
        this.main.write3BitDouble(value);
    }
    writeBit(value) {
        this.main.writeBit(value);
    }
    write2Bits(value) {
        this.main.write2Bits(value);
    }
    writeBitDouble(value) {
        this.main.writeBitDouble(value);
    }
    write2BitDoubleWithDefault(def, value) {
        this.main.write2BitDoubleWithDefault(def, value);
    }
    write3BitDoubleWithDefault(def, value) {
        this.main.write3BitDoubleWithDefault(def, value);
    }
    writeBitDoubleWithDefault(def, value) {
        this.main.writeBitDoubleWithDefault(def, value);
    }
    writeBitExtrusion(value) {
        this.main.writeBitExtrusion(value);
    }
    writeBitLong(value) {
        this.main.writeBitLong(value);
    }
    writeBitLongLong(value) {
        this.main.writeBitLongLong(value);
    }
    writeBitShort(value) {
        this.main.writeBitShort(value);
    }
    writeBitThickness(value) {
        this.main.writeBitThickness(value);
    }
    writeByte(value) {
        this.main.writeByte(value);
    }
    writeBytes(bytes) {
        this.main.writeBytes(bytes);
    }
    writeBytesOffset(bytes, offset, length) {
        this.main.writeBytesOffset(bytes, offset, length);
    }
    writeCmColor(value) {
        this.main.writeCmColor(value);
    }
    writeEnColor(color, transparency) {
        this.main.writeEnColor(color, transparency);
    }
    writeEnColorBook(color, transparency, isBookColor) {
        this.main.writeEnColorBook(color, transparency, isBookColor);
    }
    writeDateTime(value) {
        this.main.writeDateTime(value);
    }
    write8BitJulianDate(value) {
        this.main.write8BitJulianDate(value);
    }
    writeInt(value) {
        this.main.writeInt(value);
    }
    writeObjectType(value) {
        this.main.writeObjectType(value);
    }
    writeObjectTypeEnum(value) {
        this.main.writeObjectTypeEnum(value);
    }
    writeRawDouble(value) {
        this.main.writeRawDouble(value);
    }
    writeRawLong(value) {
        this.main.writeRawLong(value);
    }
    writeRawShort(value) {
        this.main.writeRawShort(value);
    }
    writeSpearShift() {
        const mainSizeBits = this.main.positionInBits;
        const textSizeBits = this.textWriter.positionInBits;
        this.main.writeSpearShift();
        if (this._savedPosition) {
            let mainTextTotalBits = mainSizeBits + textSizeBits + 1;
            if (textSizeBits > 0) {
                mainTextTotalBits += 16;
                if (textSizeBits >= 0x8000) {
                    mainTextTotalBits += 16;
                    if (textSizeBits >= 0x40000000) {
                        mainTextTotalBits += 16;
                    }
                }
            }
            this.main.setPositionInBits(this.positionInBits);
            //Write the total size in bits
            this.main.writeRawLong(mainTextTotalBits);
            this.main.writeShiftValue();
        }
        this.main.setPositionInBits(mainSizeBits);
        if (textSizeBits > 0) {
            this.textWriter.writeSpearShift();
            const textWrittenBytes = Math.ceil(this.textWriter.positionInBits / 8);
            const textBuffer = new Uint8Array(this.textWriter.stream).slice(0, textWrittenBytes);
            this.main.writeBytes(textBuffer);
            this.main.writeSpearShift();
            this.main.setPositionInBits(mainSizeBits + textSizeBits);
            this.main.setPositionByFlag(textSizeBits);
            this.main.writeBit(true);
        }
        else {
            this.main.writeBit(false);
        }
        this.handleWriter.writeSpearShift();
        this.savedPositionInBits = this.main.positionInBits;
        const handleWrittenBytes = Math.ceil(this.handleWriter.positionInBits / 8);
        const handleBuffer = new Uint8Array(this.handleWriter.stream).slice(0, handleWrittenBytes);
        this.main.writeBytes(handleBuffer);
        this.main.writeSpearShift();
    }
    writeTimeSpan(value) {
        this.main.writeTimeSpan(value);
    }
    writeVariableText(value) {
        this.textWriter.writeVariableText(value);
    }
    writeTextUnicode(value) {
        this.textWriter.writeTextUnicode(value);
    }
    setPositionInBits(posInBits) {
        this.main.setPositionInBits(posInBits);
    }
    setPositionByFlag(pos) {
        this.main.setPositionByFlag(pos);
    }
    writeShiftValue() {
        this.main.writeShiftValue();
    }
}
