// @node-projects/acad-ts/dist/ACadVersion.js
/**
 * Drawing format version.
 */
export var ACadVersion;
(function (ACadVersion) {
    /**
     * Version not identified.
     */
    ACadVersion[ACadVersion["Unknown"] = -1] = "Unknown";
    /**
     * Release 1.1
     */
    ACadVersion[ACadVersion["MC0_0"] = 0] = "MC0_0";
    /**
     * Release 1.2
     */
    ACadVersion[ACadVersion["AC1_2"] = 1] = "AC1_2";
    /**
     * Release 1.4
     */
    ACadVersion[ACadVersion["AC1_4"] = 2] = "AC1_4";
    /**
     * Release 2.0
     */
    ACadVersion[ACadVersion["AC1_50"] = 3] = "AC1_50";
    /**
     * Release 2.10
     */
    ACadVersion[ACadVersion["AC2_10"] = 4] = "AC2_10";
    /**
     * Release 2.5
     */
    ACadVersion[ACadVersion["AC1002"] = 5] = "AC1002";
    /**
     * Release 2.6
     */
    ACadVersion[ACadVersion["AC1003"] = 6] = "AC1003";
    /**
     * Release 9
     */
    ACadVersion[ACadVersion["AC1004"] = 7] = "AC1004";
    /**
     * Release 10
     */
    ACadVersion[ACadVersion["AC1006"] = 8] = "AC1006";
    /**
     * Release 11/12 (LT R1/R2)
     */
    ACadVersion[ACadVersion["AC1009"] = 9] = "AC1009";
    /**
     * Release 13 (LT95)
     */
    ACadVersion[ACadVersion["AC1012"] = 19] = "AC1012";
    /**
     * Release 14, 14.01 (LT97/LT98)
     */
    ACadVersion[ACadVersion["AC1014"] = 21] = "AC1014";
    /**
     * Release 2000/2000i/2002
     */
    ACadVersion[ACadVersion["AC1015"] = 23] = "AC1015";
    /**
     * Release 2004/2005/2006
     */
    ACadVersion[ACadVersion["AC1018"] = 25] = "AC1018";
    /**
     * Release 2007/2008/2009
     */
    ACadVersion[ACadVersion["AC1021"] = 27] = "AC1021";
    /**
     * Release 2010/2011/2012
     */
    ACadVersion[ACadVersion["AC1024"] = 29] = "AC1024";
    /**
     * Release 2013/2014/2015/2016/2017
     */
    ACadVersion[ACadVersion["AC1027"] = 31] = "AC1027";
    /**
     * Release 2018/2019/2020
     */
    ACadVersion[ACadVersion["AC1032"] = 33] = "AC1032";
})(ACadVersion || (ACadVersion = {}));
