// @node-projects/acad-ts/dist/CadDocument.js
import { CadObject } from './CadObject.js';
import { CadSummaryInfo } from './CadSummaryInfo.js';
import { CadHeader } from './Header/CadHeader.js';
import { CadDictionary } from './Objects/CadDictionary.js';
import { AppIdsTable } from './Tables/Collections/AppIdsTable.js';
import { BlockRecordsTable } from './Tables/Collections/BlockRecordsTable.js';
import { DimensionStylesTable } from './Tables/Collections/DimensionStylesTable.js';
import { LayersTable } from './Tables/Collections/LayersTable.js';
import { LineTypesTable } from './Tables/Collections/LineTypesTable.js';
import { TextStylesTable } from './Tables/Collections/TextStylesTable.js';
import { UCSTable } from './Tables/Collections/UCSTable.js';
import { ViewsTable } from './Tables/Collections/ViewsTable.js';
import { VPortsTable } from './Tables/Collections/VPortsTable.js';
// TODO: These imports reference types that need to be converted
// import { CadHeader } from './Header/CadHeader.js';
// import { CadDictionary } from './Objects/CadDictionary.js';
// import { BlockRecord } from './Tables/BlockRecord.js';
// ... and many more table/collection types
import { DxfClassCollection } from './Classes/DxfClassCollection.js';
export class CadDocument {
    appIds = null; // AppIdsTable
    blockRecords = null; // BlockRecordsTable
    classes = null; // DxfClassCollection
    colors = null; // ColorCollection
    dictionaryVariables = null; // DictionaryVariableCollection
    dimensionStyles = null; // DimensionStylesTable
    get entities() { return this.modelSpace?.entities; }
    groups = null; // GroupCollection
    get handle() { return 0; }
    header = null; // CadHeader
    imageDefinitions = null; // ImageDefinitionCollection
    layers = null; // LayersTable
    layouts = null; // LayoutCollection
    lineTypes = null; // LineTypesTable
    materials = null; // MaterialCollection
    mLeaderStyles = null; // MLeaderStyleCollection
    mLineStyles = null; // MLineStyleCollection
    get modelSpace() { return this.blockRecords?.tryGetValue("*Model_Space"); }
    get paperSpace() { return this.blockRecords?.tryGetValue("*Paper_Space"); }
    pdfDefinitions = null; // PdfDefinitionCollection
    get rootDictionary() { return this._rootDictionary; }
    set rootDictionary(value) {
        this._rootDictionary = value;
        this._rootDictionary.owner = this;
        this.registerCollection(this._rootDictionary);
    }
    scales = null; // ScaleCollection
    summaryInfo = null;
    tableStyles = null; // TableStyleCollection
    textStyles = null; // TextStylesTable
    uCSs = null; // UCSTable
    views = null; // ViewsTable
    vPorts = null; // VPortsTable
    /** @internal */
    vEntityControl = null;
    _cadObjects = new Map();
    _rootDictionary = null;
    constructor(version) {
        this._cadObjects.set(this.handle, this);
        this.createDefaults();
        if (version !== undefined && this.header != null) {
            this.header.version = version;
        }
    }
    createDefaults() {
        if (this.header == null) {
            this.header = new CadHeader(this);
        }
        this.summaryInfo = new CadSummaryInfo();
        if (this.classes == null) {
            this.classes = new DxfClassCollection();
        }
        DxfClassCollection.updateDxfClasses(this);
        if (this.rootDictionary == null) {
            this.rootDictionary = CadDictionary.createRoot();
        }
        if (this.appIds == null)
            this.registerCollection(new AppIdsTable());
        if (this.blockRecords == null)
            this.registerCollection(new BlockRecordsTable());
        if (this.dimensionStyles == null)
            this.registerCollection(new DimensionStylesTable());
        if (this.layers == null)
            this.registerCollection(new LayersTable());
        if (this.lineTypes == null)
            this.registerCollection(new LineTypesTable());
        if (this.textStyles == null)
            this.registerCollection(new TextStylesTable());
        if (this.uCSs == null)
            this.registerCollection(new UCSTable());
        if (this.views == null)
            this.registerCollection(new ViewsTable());
        if (this.vPorts == null)
            this.registerCollection(new VPortsTable());
        // Create default line types (ByLayer, ByBlock, Continuous)
        if (this.lineTypes && typeof this.lineTypes.createDefaultEntries === 'function') {
            this.lineTypes.createDefaultEntries();
        }
        // Create default layer ("0")
        if (this.layers && typeof this.layers.createDefaultEntries === 'function') {
            this.layers.createDefaultEntries();
        }
        // Create default block records (ModelSpace, PaperSpace)
        if (this.blockRecords && typeof this.blockRecords.createDefaultEntries === 'function') {
            this.blockRecords.createDefaultEntries();
        }
        // Create default text styles ("Standard")
        if (this.textStyles && typeof this.textStyles.createDefaultEntries === 'function') {
            this.textStyles.createDefaultEntries();
        }
        // Create default dimension style ("Standard")
        if (this.dimensionStyles && typeof this.dimensionStyles.createDefaultEntries === 'function') {
            this.dimensionStyles.createDefaultEntries();
        }
        // Create default app id ("ACAD")
        if (this.appIds && typeof this.appIds.createDefaultEntries === 'function') {
            this.appIds.createDefaultEntries();
        }
        // Create default viewport ("*Active")
        if (this.vPorts && typeof this.vPorts.createDefaultEntries === 'function') {
            this.vPorts.createDefaultEntries();
        }
    }
    getCadObject(handle) {
        const obj = this._cadObjects.get(handle);
        return obj ?? null;
    }
    tryGetCadObject(handle) {
        if (handle === this.handle)
            return { found: false, cadObject: null };
        const obj = this._cadObjects.get(handle);
        if (obj) {
            return { found: true, cadObject: obj };
        }
        return { found: false, cadObject: null };
    }
    restoreHandles() {
        const source = [...this._cadObjects.values()];
        this._cadObjects.clear();
        // TODO: this.header.handleSeed = 0;
        let nextHandle = 1;
        this._cadObjects.set(0, this);
        for (let i = 1; i < source.length; i++) {
            const item = source[i];
            item.handle = nextHandle;
            nextHandle++;
            this._cadObjects.set(item.handle, item);
        }
        // TODO: this.header.handleSeed = nextHandle;
    }
    toString() {
        return `${this.header?.toString() ?? 'CadDocument'}`;
    }
    updateCollections(createDictionaries, createDefaults) {
        // TODO: Implement when dictionary/collection types are converted
    }
    updateDxfClasses(reset) {
        if (reset && this.classes) {
            this.classes.clear();
        }
        DxfClassCollection.updateDxfClasses(this);
    }
    updateImageReactors() {
        // TODO: Implement when image types are converted
    }
    /** @internal */
    registerCollection(collection) {
        if (collection == null) {
            return;
        }
        if (collection instanceof CadObject) {
            collection.owner = this;
            this.addCadObject(collection);
        }
        if ('onAdd' in collection) {
            collection.onAdd = this.onAdd.bind(this);
        }
        if ('onRemove' in collection) {
            collection.onRemove = this.onRemove.bind(this);
        }
        if (collection instanceof AppIdsTable)
            this.appIds = collection;
        else if (collection instanceof BlockRecordsTable)
            this.blockRecords = collection;
        else if (collection instanceof DimensionStylesTable)
            this.dimensionStyles = collection;
        else if (collection instanceof LayersTable)
            this.layers = collection;
        else if (collection instanceof LineTypesTable)
            this.lineTypes = collection;
        else if (collection instanceof TextStylesTable)
            this.textStyles = collection;
        else if (collection instanceof UCSTable)
            this.uCSs = collection;
        else if (collection instanceof ViewsTable)
            this.views = collection;
        else if (collection instanceof VPortsTable)
            this.vPorts = collection;
    }
    /** @internal */
    unregisterCollection(collection) {
        // TODO: Implement when collection types are converted
    }
    addCadObject(cadObject) {
        if (cadObject.document != null) {
            throw new Error(`The item with handle ${cadObject.handle} is already assigned to a document`);
        }
        if (cadObject.handle === 0 || this._cadObjects.has(cadObject.handle)) {
            // TODO: const nextHandle = this.header.handleSeed;
            const nextHandle = this._cadObjects.size;
            cadObject.handle = nextHandle;
            // TODO: this.header.handleSeed = nextHandle + 1;
        }
        this._cadObjects.set(cadObject.handle, cadObject);
        cadObject.assignDocument(this);
    }
    onAdd(sender, e) {
        this.addCadObject(e.item);
        this.registerBlockMarkers(e.item);
        // Wire sub-collections (e.g., BlockRecord.entities) to this document
        if (e.item && 'entities' in e.item && e.item.entities) {
            const entities = e.item.entities;
            if ('onAdd' in entities) {
                entities.onAdd = this.onAdd.bind(this);
            }
            if ('onRemove' in entities) {
                entities.onRemove = this.onRemove.bind(this);
            }
        }
    }
    onRemove(sender, e) {
        this.unregisterBlockMarkers(e.item);
        this.removeCadObject(e.item);
    }
    registerBlockMarkers(item) {
        if (!item || !('blockEntity' in item) || !('blockEnd' in item)) {
            return;
        }
        for (const marker of [item.blockEntity, item.blockEnd]) {
            if (marker instanceof CadObject && marker.document == null) {
                this.addCadObject(marker);
            }
        }
    }
    unregisterBlockMarkers(item) {
        if (!item || !('blockEntity' in item) || !('blockEnd' in item)) {
            return;
        }
        for (const marker of [item.blockEntity, item.blockEnd]) {
            if (marker instanceof CadObject && marker.document === this) {
                this.removeCadObject(marker);
            }
        }
    }
    removeCadObject(cadObject) {
        const result = this.tryGetCadObject(cadObject.handle);
        if (!result.found || !this._cadObjects.delete(cadObject.handle)) {
            return;
        }
        cadObject.unassignDocument();
    }
    updateCollectionDict(dictName, createDictionary) {
        // TODO: Implement when CadDictionary is converted
        return { success: false, dictionary: null };
    }
}
