// @node-projects/acad-ts/dist/CadObject.js
import { ExtendedDataDictionary } from './XData/ExtendedDataDictionary.js';
export class CadObject {
    document = /* CadDocument */ null;
    extendedData;
    handle = 0;
    get hasDynamicSubclass() { return false; }
    get objectName() { return ""; }
    owner = null;
    get reactors() {
        return this._reactors;
    }
    get xDictionary() { return this._xdictionary; }
    set xDictionary(value) {
        if (value == null)
            return;
        this._xdictionary = value;
        this._xdictionary.owner = this;
        if (this.document != null) {
            this.document.registerCollection(this._xdictionary);
        }
    }
    _reactors = [];
    _xdictionary = null;
    constructor() {
        this.extendedData = new ExtendedDataDictionary(this);
    }
    addReactor(reactor) {
        this._reactors.push(reactor);
    }
    cleanReactors() {
        const reactors = [...this._reactors];
        for (const reactor of reactors) {
            if (reactor.document !== this.document) {
                const idx = this._reactors.indexOf(reactor);
                if (idx >= 0)
                    this._reactors.splice(idx, 1);
            }
        }
    }
    clone() {
        // TODO: MemberwiseClone equivalent - shallow copy
        const clone = Object.create(Object.getPrototypeOf(this));
        Object.assign(clone, this);
        clone.handle = 0;
        clone.document = null;
        clone.owner = null;
        clone._reactors = [];
        clone.extendedData = new ExtendedDataDictionary(clone);
        clone.xDictionary = this._xdictionary?.clone() ?? null;
        return clone;
    }
    createExtendedDictionary() {
        if (this._xdictionary == null) {
            const { CadDictionary: CadDictionaryImpl } = require('./Objects/CadDictionary');
            this.xDictionary = new CadDictionaryImpl();
        }
        return this._xdictionary;
    }
    removeReactor(reactor) {
        const idx = this._reactors.indexOf(reactor);
        if (idx >= 0) {
            this._reactors.splice(idx, 1);
            return true;
        }
        return false;
    }
    toString() {
        return `${this.objectName}:${this.handle}`;
    }
    /** @internal */
    assignDocument(doc) {
        this.document = doc;
        if (this.xDictionary != null) {
            doc.registerCollection(this.xDictionary);
        }
        if (this.extendedData.size > 0) {
            const entries = [...this.extendedData.entries()];
            this.extendedData.clear();
            for (const [key, value] of entries) {
                this.extendedData.add(key, value);
            }
        }
    }
    /** @internal */
    unassignDocument() {
        if (this.xDictionary != null) {
            this.document?.unregisterCollection(this.xDictionary);
        }
        this.handle = 0;
        this.document = null;
        if (this.extendedData.size > 0) {
            const entries = [...this.extendedData.entries()];
            this.extendedData.clear();
            for (const [key, value] of entries) {
                this.extendedData.add(key, value);
            }
        }
        this._reactors = [];
    }
    static updateCollection(entry, table) {
        if (table == null || entry == null)
            return entry;
        return table.tryAdd(entry);
    }
    static updateCollectionStatic(entry, table) {
        return CadObject.updateCollection(entry, table);
    }
    updateCollection(entry, table) {
        return CadObject.updateCollection(entry, table);
    }
}
