// @node-projects/acad-ts/dist/DxfClassMap.js
import { DxfMapBase } from './DxfMapBase.js';
export class DxfClassMap extends DxfMapBase {
    static _cache = new Map();
    constructor(arg) {
        super();
        if (arg instanceof DxfClassMap) {
            this.name = arg.name;
            for (const [k, v] of arg.dxfProperties) {
                this.dxfProperties.set(k, v);
            }
        }
        else if (typeof arg === 'string') {
            this.name = arg;
        }
    }
    // TODO: create requires C# reflection
    static createFromType(typeName, name) {
        if (DxfClassMap._cache.has(typeName)) {
            return new DxfClassMap(DxfClassMap._cache.get(typeName));
        }
        const classMap = new DxfClassMap();
        classMap.name = name ?? typeName;
        // TODO: Reflection-based property mapping
        DxfClassMap._cache.set(typeName, classMap);
        return new DxfClassMap(classMap);
    }
    static Create(type, name) {
        const typeName = typeof type === 'string' ? type : type.name;
        return DxfClassMap.createFromType(typeName, name);
    }
    clearCache() {
        DxfClassMap._cache.clear();
    }
    toString() {
        return `DxfClassMap:${this.name}`;
    }
}
