// @node-projects/acad-ts/dist/DxfCode.js
export var DxfCode;
(function (DxfCode) {
    DxfCode[DxfCode["Invalid"] = -9999] = "Invalid";
    DxfCode[DxfCode["XDictionary"] = -6] = "XDictionary";
    DxfCode[DxfCode["PReactors"] = -5] = "PReactors";
    DxfCode[DxfCode["Operator"] = -4] = "Operator";
    DxfCode[DxfCode["XDataStart"] = -3] = "XDataStart";
    DxfCode[DxfCode["HeaderId"] = -2] = "HeaderId";
    DxfCode[DxfCode["FirstEntityId"] = -2] = "FirstEntityId";
    DxfCode[DxfCode["End"] = -1] = "End";
    DxfCode[DxfCode["Start"] = 0] = "Start";
    DxfCode[DxfCode["Text"] = 1] = "Text";
    DxfCode[DxfCode["XRefPath"] = 1] = "XRefPath";
    DxfCode[DxfCode["ShapeName"] = 2] = "ShapeName";
    DxfCode[DxfCode["BlockName"] = 2] = "BlockName";
    DxfCode[DxfCode["AttributeTag"] = 2] = "AttributeTag";
    DxfCode[DxfCode["SymbolTableName"] = 2] = "SymbolTableName";
    DxfCode[DxfCode["MlineStyleName"] = 2] = "MlineStyleName";
    DxfCode[DxfCode["SymbolTableRecordName"] = 2] = "SymbolTableRecordName";
    DxfCode[DxfCode["AttributePrompt"] = 3] = "AttributePrompt";
    DxfCode[DxfCode["DimStyleName"] = 3] = "DimStyleName";
    DxfCode[DxfCode["LinetypeProse"] = 3] = "LinetypeProse";
    DxfCode[DxfCode["TextFontFile"] = 3] = "TextFontFile";
    DxfCode[DxfCode["Description"] = 3] = "Description";
    DxfCode[DxfCode["DimPostString"] = 3] = "DimPostString";
    DxfCode[DxfCode["TextBigFontFile"] = 4] = "TextBigFontFile";
    DxfCode[DxfCode["DimensionAlternativePrefixSuffix"] = 4] = "DimensionAlternativePrefixSuffix";
    DxfCode[DxfCode["CLShapeName"] = 4] = "CLShapeName";
    DxfCode[DxfCode["SymbolTableRecordComments"] = 4] = "SymbolTableRecordComments";
    DxfCode[DxfCode["Handle"] = 5] = "Handle";
    DxfCode[DxfCode["DimensionBlock"] = 5] = "DimensionBlock";
    DxfCode[DxfCode["DimBlk1"] = 6] = "DimBlk1";
    DxfCode[DxfCode["LinetypeName"] = 6] = "LinetypeName";
    DxfCode[DxfCode["DimBlk2"] = 7] = "DimBlk2";
    DxfCode[DxfCode["TextStyleName"] = 7] = "TextStyleName";
    DxfCode[DxfCode["LayerName"] = 8] = "LayerName";
    DxfCode[DxfCode["CLShapeText"] = 9] = "CLShapeText";
    DxfCode[DxfCode["XCoordinate"] = 10] = "XCoordinate";
    DxfCode[DxfCode["YCoordinate"] = 20] = "YCoordinate";
    DxfCode[DxfCode["ZCoordinate"] = 30] = "ZCoordinate";
    DxfCode[DxfCode["Elevation"] = 38] = "Elevation";
    DxfCode[DxfCode["Thickness"] = 39] = "Thickness";
    DxfCode[DxfCode["Real"] = 40] = "Real";
    DxfCode[DxfCode["ViewportHeight"] = 40] = "ViewportHeight";
    DxfCode[DxfCode["TxtSize"] = 40] = "TxtSize";
    DxfCode[DxfCode["TxtStyleXScale"] = 41] = "TxtStyleXScale";
    DxfCode[DxfCode["ViewWidth"] = 41] = "ViewWidth";
    DxfCode[DxfCode["ViewportAspect"] = 41] = "ViewportAspect";
    DxfCode[DxfCode["TxtStylePSize"] = 42] = "TxtStylePSize";
    DxfCode[DxfCode["ViewLensLength"] = 42] = "ViewLensLength";
    DxfCode[DxfCode["ViewFrontClip"] = 43] = "ViewFrontClip";
    DxfCode[DxfCode["ViewBackClip"] = 44] = "ViewBackClip";
    DxfCode[DxfCode["ShapeXOffset"] = 44] = "ShapeXOffset";
    DxfCode[DxfCode["ShapeYOffset"] = 45] = "ShapeYOffset";
    DxfCode[DxfCode["ViewHeight"] = 45] = "ViewHeight";
    DxfCode[DxfCode["ShapeScale"] = 46] = "ShapeScale";
    DxfCode[DxfCode["PixelScale"] = 47] = "PixelScale";
    DxfCode[DxfCode["LinetypeScale"] = 48] = "LinetypeScale";
    DxfCode[DxfCode["DashLength"] = 49] = "DashLength";
    DxfCode[DxfCode["MlineOffset"] = 49] = "MlineOffset";
    DxfCode[DxfCode["LinetypeElement"] = 49] = "LinetypeElement";
    DxfCode[DxfCode["Angle"] = 50] = "Angle";
    DxfCode[DxfCode["ViewportSnapAngle"] = 50] = "ViewportSnapAngle";
    DxfCode[DxfCode["ViewportTwist"] = 51] = "ViewportTwist";
    DxfCode[DxfCode["Visibility"] = 60] = "Visibility";
    DxfCode[DxfCode["LayerLinetype"] = 61] = "LayerLinetype";
    DxfCode[DxfCode["Color"] = 62] = "Color";
    DxfCode[DxfCode["HasSubentities"] = 66] = "HasSubentities";
    DxfCode[DxfCode["ViewportVisibility"] = 67] = "ViewportVisibility";
    DxfCode[DxfCode["ViewportActive"] = 68] = "ViewportActive";
    DxfCode[DxfCode["ViewportNumber"] = 69] = "ViewportNumber";
    DxfCode[DxfCode["Int16"] = 70] = "Int16";
    DxfCode[DxfCode["ViewMode"] = 71] = "ViewMode";
    DxfCode[DxfCode["CircleSides"] = 72] = "CircleSides";
    DxfCode[DxfCode["ViewportZoom"] = 73] = "ViewportZoom";
    DxfCode[DxfCode["ViewportIcon"] = 74] = "ViewportIcon";
    DxfCode[DxfCode["ViewportSnap"] = 75] = "ViewportSnap";
    DxfCode[DxfCode["ViewportGrid"] = 76] = "ViewportGrid";
    DxfCode[DxfCode["ViewportSnapStyle"] = 77] = "ViewportSnapStyle";
    DxfCode[DxfCode["ViewportSnapPair"] = 78] = "ViewportSnapPair";
    DxfCode[DxfCode["RegAppFlags"] = 71] = "RegAppFlags";
    DxfCode[DxfCode["TxtStyleFlags"] = 71] = "TxtStyleFlags";
    DxfCode[DxfCode["LinetypeAlign"] = 72] = "LinetypeAlign";
    DxfCode[DxfCode["LinetypePdc"] = 73] = "LinetypePdc";
    DxfCode[DxfCode["Int32"] = 90] = "Int32";
    DxfCode[DxfCode["Subclass"] = 100] = "Subclass";
    DxfCode[DxfCode["EmbeddedObjectStart"] = 101] = "EmbeddedObjectStart";
    DxfCode[DxfCode["ControlString"] = 102] = "ControlString";
    DxfCode[DxfCode["DimVarHandle"] = 105] = "DimVarHandle";
    DxfCode[DxfCode["UcsOrg"] = 110] = "UcsOrg";
    DxfCode[DxfCode["UcsOrientationX"] = 111] = "UcsOrientationX";
    DxfCode[DxfCode["UcsOrientationY"] = 112] = "UcsOrientationY";
    DxfCode[DxfCode["XReal"] = 140] = "XReal";
    DxfCode[DxfCode["ViewBrightness"] = 141] = "ViewBrightness";
    DxfCode[DxfCode["ViewContrast"] = 142] = "ViewContrast";
    DxfCode[DxfCode["Int64"] = 160] = "Int64";
    DxfCode[DxfCode["XInt16"] = 170] = "XInt16";
    DxfCode[DxfCode["NormalX"] = 210] = "NormalX";
    DxfCode[DxfCode["NormalY"] = 220] = "NormalY";
    DxfCode[DxfCode["NormalZ"] = 230] = "NormalZ";
    DxfCode[DxfCode["XXInt16"] = 270] = "XXInt16";
    DxfCode[DxfCode["Int8"] = 280] = "Int8";
    DxfCode[DxfCode["RenderMode"] = 281] = "RenderMode";
    DxfCode[DxfCode["Bool"] = 290] = "Bool";
    DxfCode[DxfCode["XTextString"] = 300] = "XTextString";
    DxfCode[DxfCode["BinaryChunk"] = 310] = "BinaryChunk";
    DxfCode[DxfCode["ArbitraryHandle"] = 320] = "ArbitraryHandle";
    DxfCode[DxfCode["SoftPointerId"] = 330] = "SoftPointerId";
    DxfCode[DxfCode["HardPointerId"] = 340] = "HardPointerId";
    DxfCode[DxfCode["SoftOwnershipId"] = 350] = "SoftOwnershipId";
    DxfCode[DxfCode["HardOwnershipId"] = 360] = "HardOwnershipId";
    DxfCode[DxfCode["LineWeight"] = 370] = "LineWeight";
    DxfCode[DxfCode["PlotStyleNameType"] = 380] = "PlotStyleNameType";
    DxfCode[DxfCode["PlotStyleNameId"] = 390] = "PlotStyleNameId";
    DxfCode[DxfCode["ExtendedInt16"] = 400] = "ExtendedInt16";
    DxfCode[DxfCode["LayoutName"] = 410] = "LayoutName";
    DxfCode[DxfCode["ColorRgb"] = 420] = "ColorRgb";
    DxfCode[DxfCode["ColorName"] = 430] = "ColorName";
    DxfCode[DxfCode["Alpha"] = 440] = "Alpha";
    DxfCode[DxfCode["GradientObjType"] = 450] = "GradientObjType";
    DxfCode[DxfCode["GradientPatType"] = 451] = "GradientPatType";
    DxfCode[DxfCode["GradientTintType"] = 452] = "GradientTintType";
    DxfCode[DxfCode["GradientColCount"] = 453] = "GradientColCount";
    DxfCode[DxfCode["GradientAngle"] = 460] = "GradientAngle";
    DxfCode[DxfCode["GradientShift"] = 461] = "GradientShift";
    DxfCode[DxfCode["GradientTintVal"] = 462] = "GradientTintVal";
    DxfCode[DxfCode["GradientColVal"] = 463] = "GradientColVal";
    DxfCode[DxfCode["GradientName"] = 470] = "GradientName";
    DxfCode[DxfCode["Comment"] = 999] = "Comment";
    DxfCode[DxfCode["ExtendedDataAsciiString"] = 1000] = "ExtendedDataAsciiString";
    DxfCode[DxfCode["ExtendedDataRegAppName"] = 1001] = "ExtendedDataRegAppName";
    DxfCode[DxfCode["ExtendedDataControlString"] = 1002] = "ExtendedDataControlString";
    DxfCode[DxfCode["ExtendedDataLayerName"] = 1003] = "ExtendedDataLayerName";
    DxfCode[DxfCode["ExtendedDataBinaryChunk"] = 1004] = "ExtendedDataBinaryChunk";
    DxfCode[DxfCode["ExtendedDataHandle"] = 1005] = "ExtendedDataHandle";
    DxfCode[DxfCode["ExtendedDataXCoordinate"] = 1010] = "ExtendedDataXCoordinate";
    DxfCode[DxfCode["ExtendedDataYCoordinate"] = 1020] = "ExtendedDataYCoordinate";
    DxfCode[DxfCode["ExtendedDataZCoordinate"] = 1030] = "ExtendedDataZCoordinate";
    DxfCode[DxfCode["ExtendedDataWorldXCoordinate"] = 1011] = "ExtendedDataWorldXCoordinate";
    DxfCode[DxfCode["ExtendedDataWorldYCoordinate"] = 1021] = "ExtendedDataWorldYCoordinate";
    DxfCode[DxfCode["ExtendedDataWorldZCoordinate"] = 1031] = "ExtendedDataWorldZCoordinate";
    DxfCode[DxfCode["ExtendedDataWorldXDisp"] = 1012] = "ExtendedDataWorldXDisp";
    DxfCode[DxfCode["ExtendedDataWorldYDisp"] = 1022] = "ExtendedDataWorldYDisp";
    DxfCode[DxfCode["ExtendedDataWorldZDisp"] = 1032] = "ExtendedDataWorldZDisp";
    DxfCode[DxfCode["ExtendedDataWorldXDir"] = 1013] = "ExtendedDataWorldXDir";
    DxfCode[DxfCode["ExtendedDataWorldYDir"] = 1023] = "ExtendedDataWorldYDir";
    DxfCode[DxfCode["ExtendedDataWorldZDir"] = 1033] = "ExtendedDataWorldZDir";
    DxfCode[DxfCode["ExtendedDataReal"] = 1040] = "ExtendedDataReal";
    DxfCode[DxfCode["ExtendedDataDist"] = 1041] = "ExtendedDataDist";
    DxfCode[DxfCode["ExtendedDataScale"] = 1042] = "ExtendedDataScale";
    DxfCode[DxfCode["ExtendedDataInteger16"] = 1070] = "ExtendedDataInteger16";
    DxfCode[DxfCode["ExtendedDataInteger32"] = 1071] = "ExtendedDataInteger32";
})(DxfCode || (DxfCode = {}));
