// @node-projects/acad-ts/dist/DxfPropertyBase.js
import { DxfCode } from './DxfCode.js';
import { GroupCodeValue } from './GroupCodeValue.js';
// TODO: DxfPropertyBase relies heavily on C# reflection (PropertyInfo, Attributes)
// This is a simplified version that stores the property mapping data
export class DxfPropertyBase {
    get assignedCode() {
        if (this._assignedCode !== null)
            return this._assignedCode;
        if (this.dxfCodes.length === 1)
            return this.dxfCodes[0];
        return DxfCode.Invalid;
    }
    dxfCodes = [];
    referenceType = 0; // DxfReferenceType
    storedValue = null;
    get groupCode() {
        const code = this.dxfCodes[0];
        return GroupCodeValue.transformValue(code);
    }
    _assignedCode = null;
    _propertyName = "";
    constructor(propertyName, dxfCodes) {
        this._propertyName = propertyName;
        this.dxfCodes = dxfCodes;
    }
    setValue(code, obj, value) {
        // TODO: Simplified - full implementation requires reflection
        if (obj && this._propertyName) {
            obj[this._propertyName] = value;
        }
    }
    getRawValue(obj) {
        // TODO: Simplified - full implementation requires reflection
        if (obj && this._propertyName) {
            return obj[this._propertyName];
        }
        return null;
    }
}
