// @node-projects/acad-ts/dist/GroupCodeValueType.js
export var GroupCodeValueType;
(function (GroupCodeValueType) {
    GroupCodeValueType[GroupCodeValueType["None"] = 0] = "None";
    GroupCodeValueType[GroupCodeValueType["String"] = 1] = "String";
    GroupCodeValueType[GroupCodeValueType["Point3D"] = 2] = "Point3D";
    GroupCodeValueType[GroupCodeValueType["Double"] = 3] = "Double";
    GroupCodeValueType[GroupCodeValueType["Byte"] = 4] = "Byte";
    GroupCodeValueType[GroupCodeValueType["Int16"] = 5] = "Int16";
    GroupCodeValueType[GroupCodeValueType["Int32"] = 6] = "Int32";
    GroupCodeValueType[GroupCodeValueType["Int64"] = 7] = "Int64";
    GroupCodeValueType[GroupCodeValueType["Handle"] = 8] = "Handle";
    GroupCodeValueType[GroupCodeValueType["ObjectId"] = 9] = "ObjectId";
    GroupCodeValueType[GroupCodeValueType["Bool"] = 10] = "Bool";
    GroupCodeValueType[GroupCodeValueType["Chunk"] = 11] = "Chunk";
    GroupCodeValueType[GroupCodeValueType["Comment"] = 12] = "Comment";
    GroupCodeValueType[GroupCodeValueType["ExtendedDataString"] = 13] = "ExtendedDataString";
    GroupCodeValueType[GroupCodeValueType["ExtendedDataChunk"] = 14] = "ExtendedDataChunk";
    GroupCodeValueType[GroupCodeValueType["ExtendedDataHandle"] = 15] = "ExtendedDataHandle";
    GroupCodeValueType[GroupCodeValueType["ExtendedDataDouble"] = 16] = "ExtendedDataDouble";
    GroupCodeValueType[GroupCodeValueType["ExtendedDataInt16"] = 17] = "ExtendedDataInt16";
    GroupCodeValueType[GroupCodeValueType["ExtendedDataInt32"] = 18] = "ExtendedDataInt32";
})(GroupCodeValueType || (GroupCodeValueType = {}));
