// @node-projects/acad-ts/dist/TextAttachmentType.js
export var TextAttachmentType;
(function (TextAttachmentType) {
    TextAttachmentType[TextAttachmentType["TopOfTopLine"] = 0] = "TopOfTopLine";
    TextAttachmentType[TextAttachmentType["MiddleOfTopLine"] = 1] = "MiddleOfTopLine";
    TextAttachmentType[TextAttachmentType["MiddleOfText"] = 2] = "MiddleOfText";
    TextAttachmentType[TextAttachmentType["MiddleOfBottomLine"] = 3] = "MiddleOfBottomLine";
    TextAttachmentType[TextAttachmentType["BottomOfBottomLine"] = 4] = "BottomOfBottomLine";
    TextAttachmentType[TextAttachmentType["BottomLine"] = 5] = "BottomLine";
    TextAttachmentType[TextAttachmentType["BottomOfTopLineUnderlineBottomLine"] = 6] = "BottomOfTopLineUnderlineBottomLine";
    TextAttachmentType[TextAttachmentType["BottomOfTopLineUnderlineTopLine"] = 7] = "BottomOfTopLineUnderlineTopLine";
    TextAttachmentType[TextAttachmentType["BottomofTopLineUnderlineAll"] = 8] = "BottomofTopLineUnderlineAll";
    TextAttachmentType[TextAttachmentType["CenterOfText"] = 9] = "CenterOfText";
    TextAttachmentType[TextAttachmentType["CenterOfTextOverline"] = 10] = "CenterOfTextOverline";
})(TextAttachmentType || (TextAttachmentType = {}));
