// @spectrum-web-components/swatch/stories/swatch.stories.js
/*
Copyright 2022 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { html } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { ifDefined } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/directives.js';
import '../sp-swatch.js';
export default {
    title: 'Swatch',
    component: 'sp-swatch',
    args: {
        color: '#ff0000',
    },
    argTypes: {
        color: { control: 'color' },
        border: {
            name: 'border',
            type: { name: 'string', required: false },
            description: 'The border to apply to the Swatch children.',
            table: {
                defaultValue: { summary: '' },
            },
            control: {
                type: 'inline-radio',
                options: ['normal', 'light', 'none'],
            },
        },
        rounding: {
            name: 'rounding',
            type: { name: 'string', required: false },
            description: 'The rounding to apply to the Swatch children.',
            table: {
                defaultValue: { summary: '' },
            },
            control: {
                type: 'inline-radio',
                options: ['normal', 'none', 'full'],
            },
        },
        shape: {
            name: 'shape',
            type: { name: 'string', required: false },
            description: 'The shape to apply to the Swatch children.',
            table: {
                defaultValue: { summary: '' },
            },
            control: {
                type: 'inline-radio',
                options: ['normal', 'rectangle'],
            },
        },
    },
};
const template = ({ color, border, rounding, shape, mixedValue, nothing, }) => {
    return html `
        <sp-swatch
            border=${ifDefined(border === 'normal' ? undefined : border)}
            rounding=${ifDefined(rounding === 'normal' ? undefined : rounding)}
            shape=${ifDefined(shape === 'normal' ? undefined : shape)}
            color=${color}
            ?mixed-value=${mixedValue}
            ?nothing=${nothing}
        ></sp-swatch>
    `;
};
export const Default = (args) => template(args);
export const mixedValue = (args) => template(args);
mixedValue.args = {
    mixedValue: true,
    color: '',
};
export const nothing = (args) => template(args);
nothing.args = {
    nothing: true,
    color: '',
};
export const borderLight = (args) => template(args);
borderLight.args = {
    border: 'light',
};
export const borderNone = (args) => template(args);
borderNone.args = {
    border: 'none',
};
export const roundingNone = (args) => template(args);
roundingNone.args = {
    rounding: 'none',
};
export const roundingFull = (args) => template(args);
roundingFull.args = {
    rounding: 'full',
};
export const shapeRectangle = (args) => template(args);
shapeRectangle.args = {
    shape: 'rectangle',
};
