// @spectrum-web-components/swatch/stories/swatch-group.stories.js
/*
Copyright 2022 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { html } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { ifDefined } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/directives.js';
import '../sp-swatch-group.js';
import '../sp-swatch.js';
export default {
    title: 'Swatch group',
    component: 'sp-swatch-group',
    args: {},
    argTypes: {
        border: {
            name: 'border',
            type: { name: 'string', required: false },
            description: 'The border to apply to the Swatch children.',
            table: {
                defaultValue: { summary: '' },
            },
            control: {
                type: 'inline-radio',
                options: ['normal', 'light', 'none'],
            },
        },
        density: {
            name: 'density',
            type: { name: 'string', required: false },
            description: 'The density at which to display the Swatch children.',
            table: {
                defaultValue: { summary: '' },
            },
            control: {
                type: 'inline-radio',
                options: ['normal', 'compact', 'spacious'],
            },
        },
        rounding: {
            name: 'rounding',
            type: { name: 'string', required: false },
            description: 'The rounding to apply to the Swatch children.',
            table: {
                defaultValue: { summary: '' },
            },
            control: {
                type: 'inline-radio',
                options: ['normal', 'none', 'full'],
            },
        },
        selects: {
            name: 'selects',
            type: { name: 'string', required: false },
            description: 'Whether the Swatch Group manages a selection, and whether it is a sinlge or multiple selection.',
            table: {
                defaultValue: { summary: '' },
            },
            control: {
                type: 'inline-radio',
                options: ['none', 'single', 'multiple'],
            },
        },
        shape: {
            name: 'shape',
            type: { name: 'string', required: false },
            description: 'The shape to apply to the Swatch children.',
            table: {
                defaultValue: { summary: '' },
            },
            control: {
                type: 'inline-radio',
                options: ['normal', 'rectangle'],
            },
        },
    },
    decorators: [
        (story) => html `
            <div
                @change=${async (event) => {
            await 0;
            if (event.defaultPrevented)
                return;
            const next = event.target
                .nextElementSibling;
            next.textContent = `Selected: ${JSON.stringify(event.target.selected)}`;
        }}
            >
                ${story()}
                <div>Selected: []</div>
            </div>
        `,
    ],
};
const colors = [
    '--spectrum-global-color-gray-500',
    '--spectrum-global-color-red-500',
    '--spectrum-global-color-orange-500',
    '--spectrum-global-color-yellow-500',
    '--spectrum-global-color-chartreuse-500',
    '--spectrum-global-color-celery-500',
    '--spectrum-global-color-green-500',
    '--spectrum-global-color-seafoam-500',
    '--spectrum-global-color-blue-500',
    '--spectrum-global-color-indigo-500',
    '--spectrum-global-color-purple-500',
    '--spectrum-global-color-fuchsia-500',
    '--spectrum-global-color-magenta-500',
];
const template = ({ border, density, rounding, selects, selected = [], shape, }) => {
    const groupLabel = !!selects
        ? selects === 'single'
            ? 'Select a color'
            : 'Selects color(s)'
        : undefined;
    return html `
        <sp-swatch-group
            border=${ifDefined(border === 'normal' ? undefined : border)}
            density=${ifDefined(density === 'normal' ? undefined : density)}
            rounding=${ifDefined(rounding === 'normal' ? undefined : rounding)}
            selects=${ifDefined(selects === 'none' ? undefined : selects)}
            .selected=${selected}
            shape=${ifDefined(shape === 'normal' ? undefined : shape)}
            aria-label=${ifDefined(groupLabel)}
        >
            ${colors.map((color) => html `
                    <sp-swatch
                        color="var(${color})"
                        label=${color}
                        value=${color}
                    ></sp-swatch>
                `)}
        </sp-swatch-group>
    `;
};
export const Default = (args) => template(args);
Default.args = {};
export const densityCompact = (args) => template(args);
densityCompact.args = {
    density: 'compact',
};
export const densitySpacious = (args) => template(args);
densitySpacious.args = {
    density: 'spacious',
};
export const selectsSingle = (args) => template(args);
selectsSingle.args = {
    selects: 'single',
    selected: ['--spectrum-global-color-yellow-500'],
};
export const selectsMultiple = (args) => template(args);
selectsMultiple.args = {
    selects: 'multiple',
    selected: [
        '--spectrum-global-color-celery-500',
        '--spectrum-global-color-red-500',
        '--spectrum-global-color-purple-500',
        '--spectrum-global-color-blue-500',
    ],
};
export const borderLight = (args) => template(args);
borderLight.args = {
    border: 'light',
};
export const borderNone = (args) => template(args);
borderNone.args = {
    border: 'none',
};
export const roundingNone = (args) => template(args);
roundingNone.args = {
    rounding: 'none',
};
export const roundingFull = (args) => template(args);
roundingFull.args = {
    rounding: 'full',
};
export const shapeRectangle = (args) => template(args);
shapeRectangle.args = {
    shape: 'rectangle',
};
