// @node-projects/acad-ts/dist/IO/DWG/FileHeaders/DwgSectionHash.js
export var DwgSectionHash;
(function (DwgSectionHash) {
    DwgSectionHash[DwgSectionHash["AcDb_Unknown"] = 0] = "AcDb_Unknown";
    DwgSectionHash[DwgSectionHash["AcDb_Security"] = 1241646314] = "AcDb_Security";
    DwgSectionHash[DwgSectionHash["AcDb_FileDepList"] = 1816266186] = "AcDb_FileDepList";
    DwgSectionHash[DwgSectionHash["AcDb_VBAProject"] = 1483605316] = "AcDb_VBAProject";
    DwgSectionHash[DwgSectionHash["AcDb_AppInfo"] = 1067451454] = "AcDb_AppInfo";
    DwgSectionHash[DwgSectionHash["AcDb_Preview"] = 1084884083] = "AcDb_Preview";
    DwgSectionHash[DwgSectionHash["AcDb_SummaryInfo"] = 1903822351] = "AcDb_SummaryInfo";
    DwgSectionHash[DwgSectionHash["AcDb_RevHistory"] = 1621231027] = "AcDb_RevHistory";
    DwgSectionHash[DwgSectionHash["AcDb_AcDbObjects"] = 1733035433] = "AcDb_AcDbObjects";
    DwgSectionHash[DwgSectionHash["AcDb_ObjFreeSpace"] = 2011301407] = "AcDb_ObjFreeSpace";
    DwgSectionHash[DwgSectionHash["AcDb_Template"] = 1242825934] = "AcDb_Template";
    DwgSectionHash[DwgSectionHash["AcDb_Handles"] = 1064174672] = "AcDb_Handles";
    DwgSectionHash[DwgSectionHash["AcDb_Classes"] = 1062470751] = "AcDb_Classes";
    DwgSectionHash[DwgSectionHash["AcDb_AuxHeader"] = 1425016074] = "AcDb_AuxHeader";
    DwgSectionHash[DwgSectionHash["AcDb_Header"] = 850920409] = "AcDb_Header";
    DwgSectionHash[DwgSectionHash["AcDb_Signature"] = -1] = "AcDb_Signature";
})(DwgSectionHash || (DwgSectionHash = {}));
