// @spectrum-web-components/accordion/src/Accordion.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, SpectrumElement, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property, queryAssignedNodes, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import { FocusGroupController } from '/web-component-designer-demo/node_modules/@spectrum-web-components/reactive-controllers/src/FocusGroup.js';
import styles from './accordion.css.js';
/**
 * @element sp-accordion
 * @slot - The sp-accordion-item children to display.
 */
export class Accordion extends SpectrumElement {
    constructor() {
        super(...arguments);
        /**
         * Allows multiple accordion items to be opened at the same time
         */
        this.allowMultiple = false;
        this.focusGroupController = new FocusGroupController(this, {
            direction: 'vertical',
            elements: () => this.items,
            isFocusableElement: (el) => !el.disabled,
        });
    }
    static get styles() {
        return [styles];
    }
    get items() {
        return [...(this.defaultNodes || [])].filter((node) => typeof node.tagName !== 'undefined');
    }
    focus() {
        this.focusGroupController.focus();
    }
    async onToggle(event) {
        const target = event.target;
        // Let the event pass through the DOM so that it can be
        // prevented from the outside if a user so desires.
        await 0;
        if (this.allowMultiple || event.defaultPrevented) {
            // No toggling when `allowMultiple` or the user prevents it.
            return;
        }
        const items = [...this.items];
        /* c8 ignore next 3 */
        if (items && !items.length) {
            // no toggling when there aren't items.
            return;
        }
        items.forEach((item) => {
            if (item !== target) {
                // Close all the items that didn't dispatch the event.
                item.open = false;
            }
        });
    }
    handleSlotchange() {
        this.focusGroupController.clearElementCache();
    }
    render() {
        return html `
            <slot
                @slotchange=${this.handleSlotchange}
                @sp-accordion-item-toggle=${this.onToggle}
            ></slot>
        `;
    }
}
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'allow-multiple' })
], Accordion.prototype, "allowMultiple", void 0);
__decorate([
    queryAssignedNodes()
], Accordion.prototype, "defaultNodes", void 0);
