// @spectrum-web-components/status-light/src/StatusLight.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, SizedMixin, SpectrumElement, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import statusLightStyles from './status-light.css.js';
/**
 * @element sp-status-light
 *
 * @slot - text label of the Status Light
 */
export class StatusLight extends SizedMixin(SpectrumElement) {
    constructor() {
        super(...arguments);
        /**
         * A status light in a disabled state shows that a status exists, but is not available in that circumstance. This can be used to maintain layout continuity and communicate that a status may become available later.
         */
        this.disabled = false;
        /**
         * The visual variant to apply to this status light.
         */
        this.variant = 'info';
    }
    static get styles() {
        return [statusLightStyles];
    }
    render() {
        return html `
            <slot></slot>
        `;
    }
    updated(changes) {
        super.updated(changes);
        if (changes.has('disabled')) {
            if (this.disabled) {
                this.setAttribute('aria-disabled', 'true');
            }
            else {
                this.removeAttribute('aria-disabled');
            }
        }
    }
}
__decorate([
    property({ type: Boolean, reflect: true })
], StatusLight.prototype, "disabled", void 0);
__decorate([
    property({ reflect: true })
], StatusLight.prototype, "variant", void 0);
