// @node-projects/web-component-designer/dist/elements/services/stylesheetService/CssTreeStylesheetService.js
import { TypedEvent } from '/web-component-designer-demo/node_modules/@node-projects/base-custom-webcomponent/./dist/index.js';
import { calculate as calculateSpecifity } from "./SpecificityCalculator.js";
export class CssTreeStylesheetService {
    _stylesheets = new Map();
    stylesheetChanged = new TypedEvent();
    constructor() { }
    setStylesheets(stylesheets) {
        if (stylesheets != null) {
            this._stylesheets = new Map();
            for (let stylesheet of stylesheets) {
                this._stylesheets.set(stylesheet.name, {
                    stylesheet: stylesheet,
                    ast: window.csstree.toPlainObject((window.csstree.parse(stylesheet.stylesheet, { positions: true, parseValue: false })))
                });
            }
        }
        else {
            this._stylesheets = null;
        }
    }
    getStylesheets() {
        let stylesheets = [];
        for (let item of this._stylesheets) {
            stylesheets.push(item[1].stylesheet);
        }
        ;
        return stylesheets;
    }
    /* Section covers the retrieval of rules and declarations */
    getAppliedRulesInternal(designItem) {
        let styles = [];
        for (let item of this._stylesheets) {
            if (!item[1].ast || !this.astHasChildren(item[1].ast))
                break;
            styles = styles.concat(Array.from(this.rulesFromAST(item[1].ast, item[1].stylesheet.stylesheet, item[0], designItem)));
        }
        ;
        return styles;
    }
    getAppliedRules(designItem) {
        let rules = this.getAppliedRulesInternal(designItem);
        if (!rules || rules.length == 0)
            return [];
        let retCollection = [];
        for (let rule of rules) {
            retCollection.push({
                ...rule,
                declarations: rule.ast.block.children.map((declaration) => {
                    return {
                        name: declaration.property,
                        value: declaration.value.value,
                        important: declaration.important == true,
                        specificity: rule.specificity,
                        parent: rule,
                        ast: declaration,
                    };
                })
            });
        }
        return retCollection;
    }
    getDeclarationInternal(designItem, prop) {
        let rules = this.getAppliedRulesInternal(designItem);
        if (!rules)
            return null;
        let declarations = [];
        for (let rule of rules) {
            let declaration = this.findDeclarationInRule(rule.ast, prop);
            if (!declaration)
                continue;
            declarations.push({
                ast: declaration,
                parent: rule,
                name: prop.name,
                value: declaration.value.value,
                important: declaration.important == true,
            });
        }
        ;
        return declarations;
    }
    getDeclarations(designItem, prop) {
        let declarations = this.getDeclarationInternal(designItem, prop);
        if (!declarations)
            return null;
        return this.sortDeclarations(declarations);
    }
    /* Section covers the update of rules and declarations */
    updateDeclarationWithProperty(designItem, property, value, important) {
        let sortedDecl = this.sortDeclarations(this.getDeclarationInternal(designItem, property));
        if (!sortedDecl) {
            // no declaration of property yet
            return false;
        }
        return false;
    }
    updateDeclarationWithDeclaration(declaration, value, important) {
        declaration.ast.value = window.csstree.toPlainObject(window.csstree.parse(declaration.name + ": " + value, { context: 'declaration', parseValue: false })).value;
        this._stylesheets.get(declaration.parent.stylesheetName).stylesheet.stylesheet = window.csstree.generate(window.csstree.fromPlainObject(this._stylesheets.get(declaration.parent.stylesheetName).ast));
        this.stylesheetChanged.emit({ stylesheet: this._stylesheets.get(declaration.parent.stylesheetName).stylesheet });
        return true;
    }
    /* Section covers the internal traversal creation of rules and declarations */
    *rulesFromAST(ast, stylesheet, source, designItem, previousCheck = '') {
        let parent = ast["children"] != null ? ast : ast.block;
        for (const child of parent.children) {
            if (child.type == "Atrule") {
                const currentCheck = this.buildAtRuleString(child, stylesheet);
                if (currentCheck.type === "@media" && !window.matchMedia(currentCheck.sel))
                    continue;
                if (currentCheck.type === "@supports" && !CSS.supports(currentCheck.sel))
                    continue;
                let ruleCollection = this.rulesFromAST(child, stylesheet, source, designItem, previousCheck + currentCheck.type + " " + currentCheck.sel + "\n");
                if (ruleCollection) {
                    for (const r of ruleCollection) {
                        if (!this.elementMatchesASelector(designItem, this.buildSelectorString(r.ast.prelude)))
                            continue;
                        yield r;
                    }
                }
            }
            if (child.type == "Rule") {
                let selectors = this.buildSelectorString(child.prelude);
                if (!this.elementMatchesASelector(designItem, selectors))
                    continue;
                yield ({
                    ast: child,
                    selector: previousCheck + this.buildSelectorString(child.prelude).join(", "),
                    specificity: this.getSpecificity(child.prelude),
                    stylesheetName: source,
                    declarations: null,
                });
            }
        }
        ;
    }
    /* Utility functions for building selectors, specificity and so on  */
    astHasChildren(ast) {
        return ast != null && ast["children"] != null && ast["children"].length > 0;
    }
    buildSelectorString(selectorsAST) {
        let selectors = [];
        for (let selector of selectorsAST.children) {
            let sel = "";
            for (let fraction of selector.children) {
                sel += fraction.name;
            }
            selectors.push(sel);
        }
        ;
        return selectors;
    }
    getSpecificity(selector) {
        const specificities = calculateSpecifity(selector);
        let sum = 0;
        specificities.forEach(specificity => sum += specificity.a * 10000 + specificity.b * 100 + specificity.c);
        return sum;
    }
    findDeclarationInRule(rule, property) {
        return rule.block.children.find(declaration => declaration.property == property.name);
    }
    elementMatchesASelector(designItem, selectors) {
        for (const selector of selectors)
            if (designItem.element.matches(selector))
                return true;
        return false;
    }
    buildAtRuleString(ast, stylesheet) {
        return {
            sel: stylesheet.slice(ast.prelude.loc.start.offset, ast.prelude.loc.end.offset),
            type: "@" + ast.name
        };
    }
    sortDeclarations(declarations) {
        if (declarations == null || declarations.length == 0)
            return null;
        return declarations.sort((dec1, dec2) => {
            if (dec1.parent.specificity > dec2.parent.specificity)
                return -1;
            return 1;
        });
    }
}
