// @node-projects/web-component-designer/dist/elements/services/stylesheetService/CssToolsStylesheetService.js
import { TypedEvent } from '/web-component-designer-demo/node_modules/@node-projects/base-custom-webcomponent/./dist/index.js';
export class CssToolsStylesheetService {
    _stylesheets = new Map();
    stylesheetChanged = new TypedEvent();
    stylesheetsChanged = new TypedEvent();
    async setStylesheets(stylesheets) {
        if (stylesheets != null) {
            let tools = await import('/web-component-designer-demo/node_modules/@adobe/css-tools/dist/cjs/cssTools.js');
            this._stylesheets = new Map();
            for (let stylesheet of stylesheets) {
                this._stylesheets.set(stylesheet.name, {
                    stylesheet: stylesheet,
                    ast: tools.parse(stylesheet.stylesheet)
                });
            }
            this.stylesheetsChanged.emit();
        }
        else {
            this._stylesheets = null;
        }
    }
    getStylesheets() {
        let stylesheets = [];
        for (let item of this._stylesheets) {
            stylesheets.push(item[1].stylesheet);
        }
        ;
        return stylesheets;
    }
    getAppliedRules(designItem) {
        return null;
    }
    getDeclarations(designItem, styleName) {
        return null;
    }
    updateDeclarationWithDeclaration(declaration, value, important) {
        return true;
    }
    insertDeclarationIntoRule(rule, declaration, important) {
        return true;
    }
    removeDeclarationFromRule(rule, declaration) {
        return true;
    }
}
