// @node-projects/acad-ts/dist/Header/CadHeader.js
import { ACadVersion } from '../ACadVersion.js';
import { CadUtils } from '../CadUtils.js';
import { Color } from '../Color.js';
import { LineWeightType } from '../Types/LineWeightType.js';
import { AngularDirection } from '../Types/Units/AngularDirection.js';
import { AngularUnitFormat } from '../Types/Units/AngularUnitFormat.js';
import { LinearUnitFormat } from '../Types/Units/LinearUnitFormat.js';
import { UnitsType } from '../Types/Units/UnitsType.js';
import { DimensionStyle } from '../Tables/DimensionStyle.js';
import { Layer } from '../Tables/Layer.js';
import { LineType } from '../Tables/LineType.js';
import { TextStyle } from '../Tables/TextStyle.js';
import { UCS } from '../Tables/UCS.js';
import { AttributeVisibilityMode } from './AttributeVisibilityMode.js';
import { DimensionAssociation } from './DimensionAssociation.js';
import { EntityPlotStyleType } from './EntityPlotStyleType.js';
import { IndexCreationFlags } from './IndexCreationFlags.js';
import { MeasurementUnits } from './MeasurementUnits.js';
import { ObjectSortingFlags } from './ObjectSortingFlags.js';
import { SpaceLineTypeScaling } from './SpaceLineTypeScaling.js';
import { SplineType } from './SplineType.js';
import { XClipFrameType } from './XClipFrameType.js';
import { XYZ } from '../Math/XYZ.js';
import { XY } from '../Math/XY.js';
// TODO: MLineStyle, VerticalAlignmentType not yet converted
export var VerticalAlignmentType;
(function (VerticalAlignmentType) {
    VerticalAlignmentType[VerticalAlignmentType["Top"] = 0] = "Top";
    VerticalAlignmentType[VerticalAlignmentType["Middle"] = 1] = "Middle";
    VerticalAlignmentType[VerticalAlignmentType["Bottom"] = 2] = "Bottom";
})(VerticalAlignmentType || (VerticalAlignmentType = {}));
export class CadHeader {
    // --- Simple properties ---
    angleBase = 0.0;
    angularDirection = AngularDirection.ClockWise;
    angularUnit = AngularUnitFormat.DecimalDegrees;
    get angularUnitPrecision() {
        return this._angularUnitPrecision;
    }
    set angularUnitPrecision(value) {
        if (value < 0 || value > 8) {
            throw new Error('AUPREC valid values are from 0 to 8');
        }
        this._angularUnitPrecision = value;
    }
    arrowBlockName = '';
    associatedDimensions = true;
    attributeVisibility = AttributeVisibilityMode.Normal;
    blipMode = false;
    cameraDisplayObjects = false;
    cameraHeight = 0;
    chamferAngle = 0.0;
    chamferDistance1 = 0.0;
    chamferDistance2 = 0.0;
    chamferLength = 0.0;
    codePage = 'ANSI_1252';
    createDateTime = new Date();
    createEllipseAsPolyline = false;
    currentEntityColor = Color.ByLayer;
    currentEntityLinetypeScale = 1.0;
    currentEntityLineWeight = LineWeightType.ByLayer;
    currentEntityPlotStyle = EntityPlotStyleType.ByLayer;
    get currentLayer() {
        if (this.document == null) {
            return this._currentLayer;
        }
        else {
            return this.document.layers.get(this.currentLayerName);
        }
    }
    get currentLayerName() {
        return this._currentLayer.name;
    }
    set currentLayerName(value) {
        if (this.document != null) {
            this._currentLayer = this.document.layers.get(value);
        }
        else {
            this._currentLayer = new Layer(value);
        }
    }
    get currentLineType() {
        if (this.document == null) {
            return this._currentLineType;
        }
        else {
            return this.document.lineTypes.get(this.currentLineTypeName);
        }
    }
    get currentLineTypeName() {
        return this._currentLineType.name;
    }
    set currentLineTypeName(value) {
        if (this.document != null) {
            this._currentLineType = this.document.lineTypes.get(value);
        }
        else {
            this._currentLineType = new LineType(value);
        }
    }
    currentMultiLineJustification = VerticalAlignmentType.Top;
    currentMultilineScale = 20.0;
    // TODO: MLineStyle not yet converted
    currentMLineStyleName = 'Standard';
    get currentTextStyle() {
        if (this.document == null) {
            return this._currentTextStyle;
        }
        else {
            return this.document.textStyles.get(this.currentTextStyleName);
        }
    }
    get currentTextStyleName() {
        return this._currentTextStyle.name;
    }
    set currentTextStyleName(value) {
        if (this.document != null) {
            this._currentTextStyle = this.document.textStyles.get(value);
        }
        else {
            this._currentTextStyle = new TextStyle(value);
        }
    }
    dgnUnderlayFramesVisibility = '';
    // --- Dimension style overrides (delegate to _dimensionStyleOverrides) ---
    get dimensionAlternateDimensioningSuffix() { return this._dimensionStyleOverrides.alternateDimensioningSuffix; }
    set dimensionAlternateDimensioningSuffix(v) { this._dimensionStyleOverrides.alternateDimensioningSuffix = v; }
    get dimensionAlternateUnitDecimalPlaces() { return this._dimensionStyleOverrides.alternateUnitDecimalPlaces; }
    set dimensionAlternateUnitDecimalPlaces(v) { this._dimensionStyleOverrides.alternateUnitDecimalPlaces = v; }
    get dimensionAlternateUnitDimensioning() { return this._dimensionStyleOverrides.alternateUnitDimensioning; }
    set dimensionAlternateUnitDimensioning(v) { this._dimensionStyleOverrides.alternateUnitDimensioning = v; }
    get dimensionAlternateUnitFormat() { return this._dimensionStyleOverrides.alternateUnitFormat; }
    set dimensionAlternateUnitFormat(v) { this._dimensionStyleOverrides.alternateUnitFormat = v; }
    get dimensionAlternateUnitRounding() { return this._dimensionStyleOverrides.alternateUnitRounding; }
    set dimensionAlternateUnitRounding(v) { this._dimensionStyleOverrides.alternateUnitRounding = v; }
    get dimensionAlternateUnitScaleFactor() { return this._dimensionStyleOverrides.alternateUnitScaleFactor; }
    set dimensionAlternateUnitScaleFactor(v) { this._dimensionStyleOverrides.alternateUnitScaleFactor = v; }
    get dimensionAlternateUnitToleranceDecimalPlaces() { return this._dimensionStyleOverrides.alternateUnitToleranceDecimalPlaces; }
    set dimensionAlternateUnitToleranceDecimalPlaces(v) { this._dimensionStyleOverrides.alternateUnitToleranceDecimalPlaces = v; }
    get dimensionAlternateUnitToleranceZeroHandling() { return this._dimensionStyleOverrides.alternateUnitToleranceZeroHandling; }
    set dimensionAlternateUnitToleranceZeroHandling(v) { this._dimensionStyleOverrides.alternateUnitToleranceZeroHandling = v; }
    get dimensionAlternateUnitZeroHandling() { return this._dimensionStyleOverrides.alternateUnitZeroHandling; }
    set dimensionAlternateUnitZeroHandling(v) { this._dimensionStyleOverrides.alternateUnitZeroHandling = v; }
    get dimensionAngularDimensionDecimalPlaces() { return this._dimensionStyleOverrides.angularDecimalPlaces; }
    set dimensionAngularDimensionDecimalPlaces(v) { this._dimensionStyleOverrides.angularDecimalPlaces = v; }
    get dimensionAngularUnit() { return this._dimensionStyleOverrides.angularUnit; }
    set dimensionAngularUnit(v) { this._dimensionStyleOverrides.angularUnit = v; }
    get dimensionAngularZeroHandling() { return this._dimensionStyleOverrides.angularZeroHandling; }
    set dimensionAngularZeroHandling(v) { this._dimensionStyleOverrides.angularZeroHandling = v; }
    get dimensionArcLengthSymbolPosition() { return this._dimensionStyleOverrides.arcLengthSymbolPosition; }
    set dimensionArcLengthSymbolPosition(v) { this._dimensionStyleOverrides.arcLengthSymbolPosition = v; }
    get dimensionArrowSize() { return this._dimensionStyleOverrides.arrowSize; }
    set dimensionArrowSize(v) { this._dimensionStyleOverrides.arrowSize = v; }
    dimensionAssociativity = DimensionAssociation.CreateAssociativeDimensions;
    dimensionBlockName = '';
    dimensionBlockNameFirst = null;
    dimensionBlockNameSecond = null;
    get dimensionCenterMarkSize() { return this._dimensionStyleOverrides.centerMarkSize; }
    set dimensionCenterMarkSize(v) { this._dimensionStyleOverrides.centerMarkSize = v; }
    get dimensionCursorUpdate() { return this._dimensionStyleOverrides.cursorUpdate; }
    set dimensionCursorUpdate(v) { this._dimensionStyleOverrides.cursorUpdate = v; }
    get dimensionDecimalPlaces() { return this._dimensionStyleOverrides.decimalPlaces; }
    set dimensionDecimalPlaces(v) { this._dimensionStyleOverrides.decimalPlaces = v; }
    get dimensionDecimalSeparator() { return this._dimensionStyleOverrides.decimalSeparator; }
    set dimensionDecimalSeparator(v) { this._dimensionStyleOverrides.decimalSeparator = v; }
    get dimensionDimensionTextArrowFit() { return this._dimensionStyleOverrides.dimensionTextArrowFit; }
    set dimensionDimensionTextArrowFit(v) { this._dimensionStyleOverrides.dimensionTextArrowFit = v; }
    get dimensionExtensionLineColor() { return this._dimensionStyleOverrides.extensionLineColor; }
    set dimensionExtensionLineColor(v) { this._dimensionStyleOverrides.extensionLineColor = v; }
    get dimensionExtensionLineExtension() { return this._dimensionStyleOverrides.extensionLineExtension; }
    set dimensionExtensionLineExtension(v) { this._dimensionStyleOverrides.extensionLineExtension = v; }
    get dimensionExtensionLineOffset() { return this._dimensionStyleOverrides.extensionLineOffset; }
    set dimensionExtensionLineOffset(v) { this._dimensionStyleOverrides.extensionLineOffset = v; }
    get dimensionFit() { return this._dimensionStyleOverrides.dimensionFit; }
    set dimensionFit(v) { this._dimensionStyleOverrides.dimensionFit = v; }
    get dimensionFixedExtensionLineLength() { return this._dimensionStyleOverrides.fixedExtensionLineLength; }
    set dimensionFixedExtensionLineLength(v) { this._dimensionStyleOverrides.fixedExtensionLineLength = v; }
    get dimensionFractionFormat() { return this._dimensionStyleOverrides.fractionFormat; }
    set dimensionFractionFormat(v) { this._dimensionStyleOverrides.fractionFormat = v; }
    get dimensionGenerateTolerances() { return this._dimensionStyleOverrides.generateTolerances; }
    set dimensionGenerateTolerances(v) { this._dimensionStyleOverrides.generateTolerances = v; }
    get dimensionIsExtensionLineLengthFixed() { return this._dimensionStyleOverrides.isExtensionLineLengthFixed; }
    set dimensionIsExtensionLineLengthFixed(v) { this._dimensionStyleOverrides.isExtensionLineLengthFixed = v; }
    get dimensionJoggedRadiusDimensionTransverseSegmentAngle() { return this._dimensionStyleOverrides.joggedRadiusDimensionTransverseSegmentAngle; }
    set dimensionJoggedRadiusDimensionTransverseSegmentAngle(v) { this._dimensionStyleOverrides.joggedRadiusDimensionTransverseSegmentAngle = v; }
    get dimensionLimitsGeneration() { return this._dimensionStyleOverrides.limitsGeneration; }
    set dimensionLimitsGeneration(v) { this._dimensionStyleOverrides.limitsGeneration = v; }
    get dimensionLinearScaleFactor() { return this._dimensionStyleOverrides.linearScaleFactor; }
    set dimensionLinearScaleFactor(v) { this._dimensionStyleOverrides.linearScaleFactor = v; }
    get dimensionLinearUnitFormat() { return this._dimensionStyleOverrides.linearUnitFormat; }
    set dimensionLinearUnitFormat(v) { this._dimensionStyleOverrides.linearUnitFormat = v; }
    get dimensionLineColor() { return this._dimensionStyleOverrides.dimensionLineColor; }
    set dimensionLineColor(v) { this._dimensionStyleOverrides.dimensionLineColor = v; }
    get dimensionLineExtension() { return this._dimensionStyleOverrides.dimensionLineExtension; }
    set dimensionLineExtension(v) { this._dimensionStyleOverrides.dimensionLineExtension = v; }
    get dimensionLineGap() { return this._dimensionStyleOverrides.dimensionLineGap; }
    set dimensionLineGap(v) { this._dimensionStyleOverrides.dimensionLineGap = v; }
    get dimensionLineIncrement() { return this._dimensionStyleOverrides.dimensionLineIncrement; }
    set dimensionLineIncrement(v) { this._dimensionStyleOverrides.dimensionLineIncrement = v; }
    dimensionLineType = 'ByBlock';
    get dimensionLineWeight() { return this._dimensionStyleOverrides.dimensionLineWeight; }
    set dimensionLineWeight(v) { this._dimensionStyleOverrides.dimensionLineWeight = v; }
    get dimensionMinusTolerance() { return this._dimensionStyleOverrides.minusTolerance; }
    set dimensionMinusTolerance(v) { this._dimensionStyleOverrides.minusTolerance = v; }
    get dimensionPlusTolerance() { return this._dimensionStyleOverrides.plusTolerance; }
    set dimensionPlusTolerance(v) { this._dimensionStyleOverrides.plusTolerance = v; }
    get dimensionPostFix() { return this._dimensionStyleOverrides.postFix; }
    set dimensionPostFix(v) { this._dimensionStyleOverrides.postFix = v; }
    get dimensionRounding() { return this._dimensionStyleOverrides.rounding; }
    set dimensionRounding(v) { this._dimensionStyleOverrides.rounding = v; }
    get dimensionScaleFactor() { return this._dimensionStyleOverrides.scaleFactor; }
    set dimensionScaleFactor(v) { this._dimensionStyleOverrides.scaleFactor = v; }
    get dimensionSeparateArrowBlocks() { return this._dimensionStyleOverrides.separateArrowBlocks; }
    set dimensionSeparateArrowBlocks(v) { this._dimensionStyleOverrides.separateArrowBlocks = v; }
    get currentDimensionStyle() {
        if (this.document == null) {
            return this._currentDimensionStyle;
        }
        else {
            return this.document.dimensionStyles.get(this.currentDimensionStyleName);
        }
    }
    get currentDimensionStyleName() {
        return this._currentDimensionStyle.name;
    }
    set currentDimensionStyleName(value) {
        if (this.document != null) {
            this._currentDimensionStyle = this.document.dimensionStyles.get(value);
        }
        else {
            this._currentDimensionStyle = new DimensionStyle(value);
        }
    }
    get dimensionSuppressFirstDimensionLine() { return this._dimensionStyleOverrides.suppressFirstDimensionLine; }
    set dimensionSuppressFirstDimensionLine(v) { this._dimensionStyleOverrides.suppressFirstDimensionLine = v; }
    get dimensionSuppressFirstExtensionLine() { return this._dimensionStyleOverrides.suppressFirstExtensionLine; }
    set dimensionSuppressFirstExtensionLine(v) { this._dimensionStyleOverrides.suppressFirstExtensionLine = v; }
    get dimensionSuppressOutsideExtensions() { return this._dimensionStyleOverrides.suppressOutsideExtensions; }
    set dimensionSuppressOutsideExtensions(v) { this._dimensionStyleOverrides.suppressOutsideExtensions = v; }
    get dimensionSuppressSecondDimensionLine() { return this._dimensionStyleOverrides.suppressSecondDimensionLine; }
    set dimensionSuppressSecondDimensionLine(v) { this._dimensionStyleOverrides.suppressSecondDimensionLine = v; }
    get dimensionSuppressSecondExtensionLine() { return this._dimensionStyleOverrides.suppressSecondExtensionLine; }
    set dimensionSuppressSecondExtensionLine(v) { this._dimensionStyleOverrides.suppressSecondExtensionLine = v; }
    dimensionTex1 = 'ByBlock';
    dimensionTex2 = 'ByBlock';
    get dimensionTextBackgroundColor() { return this._dimensionStyleOverrides.textBackgroundColor; }
    set dimensionTextBackgroundColor(v) { this._dimensionStyleOverrides.textBackgroundColor = v; }
    get dimensionTextBackgroundFillMode() { return this._dimensionStyleOverrides.textBackgroundFillMode; }
    set dimensionTextBackgroundFillMode(v) { this._dimensionStyleOverrides.textBackgroundFillMode = v; }
    get dimensionTextColor() { return this._dimensionStyleOverrides.textColor; }
    set dimensionTextColor(v) { this._dimensionStyleOverrides.textColor = v; }
    get dimensionTextDirection() { return this._dimensionStyleOverrides.textDirection; }
    set dimensionTextDirection(v) { this._dimensionStyleOverrides.textDirection = v; }
    get dimensionTextHeight() { return this._dimensionStyleOverrides.textHeight; }
    set dimensionTextHeight(v) { this._dimensionStyleOverrides.textHeight = v; }
    get dimensionTextHorizontalAlignment() { return this._dimensionStyleOverrides.textHorizontalAlignment; }
    set dimensionTextHorizontalAlignment(v) { this._dimensionStyleOverrides.textHorizontalAlignment = v; }
    get dimensionTextInsideExtensions() { return this._dimensionStyleOverrides.textInsideExtensions; }
    set dimensionTextInsideExtensions(v) { this._dimensionStyleOverrides.textInsideExtensions = v; }
    get dimensionTextInsideHorizontal() { return this._dimensionStyleOverrides.textInsideHorizontal; }
    set dimensionTextInsideHorizontal(v) { this._dimensionStyleOverrides.textInsideHorizontal = v; }
    get dimensionTextMovement() { return this._dimensionStyleOverrides.textMovement; }
    set dimensionTextMovement(v) { this._dimensionStyleOverrides.textMovement = v; }
    get dimensionTextOutsideExtensions() { return this._dimensionStyleOverrides.textOutsideExtensions; }
    set dimensionTextOutsideExtensions(v) { this._dimensionStyleOverrides.textOutsideExtensions = v; }
    get dimensionTextOutsideHorizontal() { return this._dimensionStyleOverrides.textOutsideHorizontal; }
    set dimensionTextOutsideHorizontal(v) { this._dimensionStyleOverrides.textOutsideHorizontal = v; }
    get dimensionTextStyle() {
        if (this.document == null) {
            return this._dimensionTextStyle;
        }
        else {
            return this.document.textStyles.get(this.dimensionTextStyleName);
        }
    }
    get dimensionTextStyleName() {
        return this._dimensionTextStyle.name;
    }
    set dimensionTextStyleName(value) {
        if (this.document != null) {
            this._dimensionTextStyle = this.document.textStyles.get(value);
        }
        else {
            this._dimensionTextStyle = new TextStyle(value);
        }
    }
    get dimensionTextVerticalAlignment() { return this._dimensionStyleOverrides.textVerticalAlignment; }
    set dimensionTextVerticalAlignment(v) { this._dimensionStyleOverrides.textVerticalAlignment = v; }
    get dimensionTextVerticalPosition() { return this._dimensionStyleOverrides.textVerticalPosition; }
    set dimensionTextVerticalPosition(v) { this._dimensionStyleOverrides.textVerticalPosition = v; }
    get dimensionTickSize() { return this._dimensionStyleOverrides.tickSize; }
    set dimensionTickSize(v) { this._dimensionStyleOverrides.tickSize = v; }
    get dimensionToleranceAlignment() { return this._dimensionStyleOverrides.toleranceAlignment; }
    set dimensionToleranceAlignment(v) { this._dimensionStyleOverrides.toleranceAlignment = v; }
    get dimensionToleranceDecimalPlaces() { return this._dimensionStyleOverrides.toleranceDecimalPlaces; }
    set dimensionToleranceDecimalPlaces(v) { this._dimensionStyleOverrides.toleranceDecimalPlaces = v; }
    get dimensionToleranceScaleFactor() { return this._dimensionStyleOverrides.toleranceScaleFactor; }
    set dimensionToleranceScaleFactor(v) { this._dimensionStyleOverrides.toleranceScaleFactor = v; }
    get dimensionToleranceZeroHandling() { return this._dimensionStyleOverrides.toleranceZeroHandling; }
    set dimensionToleranceZeroHandling(v) { this._dimensionStyleOverrides.toleranceZeroHandling = v; }
    get dimensionUnit() { return this._dimensionStyleOverrides.dimensionUnit; }
    set dimensionUnit(v) { this._dimensionStyleOverrides.dimensionUnit = v; }
    get dimensionZeroHandling() { return this._dimensionStyleOverrides.zeroHandling; }
    set dimensionZeroHandling(v) { this._dimensionStyleOverrides.zeroHandling = v; }
    get dimensionstyleOverrides() { return this._dimensionStyleOverrides; }
    displayLightGlyphs = '';
    displayLineWeight = false;
    displaySilhouetteCurves = false;
    document = null;
    draftAngleFirstCrossSection = 0;
    draftAngleSecondCrossSection = 0;
    draftMagnitudeFirstCrossSection = 0;
    draftMagnitudeSecondCrossSection = 0;
    dw3DPrecision = 0;
    dwgUnderlayFramesVisibility = '';
    get elevation() { return this.modelSpaceUcs.elevation; }
    set elevation(v) { this.modelSpaceUcs.elevation = v; }
    endCaps = 0;
    entitySortingFlags = ObjectSortingFlags.Disabled;
    extendedNames = true;
    get extensionLineWeight() { return this._dimensionStyleOverrides.extensionLineWeight; }
    set extensionLineWeight(v) { this._dimensionStyleOverrides.extensionLineWeight = v; }
    externalReferenceClippingBoundaryType = XClipFrameType.DisplayNotPlot;
    get facetResolution() { return this._facetResolution; }
    set facetResolution(value) {
        if (value < 0.01 || value > 10) {
            throw new Error('FACETRES valid values are from 0.01 to 10.0');
        }
        this._facetResolution = value;
    }
    filletRadius = 0.0;
    fillMode = true;
    fingerPrintGuid = crypto.randomUUID?.() ?? '';
    haloGapPercentage = 0;
    handleSeed = 0x01;
    hideText = 0;
    hyperLinkBase = null;
    indexCreationFlags = IndexCreationFlags.NoIndex;
    insUnits = UnitsType.Unitless;
    interfereColor = new Color(1);
    intersectionDisplay = 0;
    joinStyle = 0;
    lastSavedBy = 'ACadSharp';
    latitude = 37.7950;
    lensLength = 0;
    limitCheckingOn = false;
    linearUnitFormat = LinearUnitFormat.Decimal;
    get linearUnitPrecision() { return this._linearUnitPrecision; }
    set linearUnitPrecision(value) {
        if (value < 0 || value > 8) {
            throw new Error('LUPREC valid values are from 0 to 8');
        }
        this._linearUnitPrecision = value;
    }
    lineTypeScale = 1.0;
    loadOLEObject = false;
    loftedObjectNormals = '';
    longitude = -122.394;
    maintenanceVersion = 0;
    maxViewportCount = 64;
    measurementUnits = MeasurementUnits.Metric;
    menuFileName = '.';
    mirrorText = false;
    modelSpaceExtMax = new XYZ(0, 0, 0);
    modelSpaceExtMin = new XYZ(0, 0, 0);
    modelSpaceInsertionBase = new XYZ(0, 0, 0);
    modelSpaceLimitsMax = new XY(0, 0);
    modelSpaceLimitsMin = new XY(0, 0);
    get modelSpaceOrigin() { return this.modelSpaceUcs.origin; }
    set modelSpaceOrigin(v) { this.modelSpaceUcs.origin = v; }
    modelSpaceOrthographicBackDOrigin = new XYZ(0, 0, 0);
    modelSpaceOrthographicBottomDOrigin = new XYZ(0, 0, 0);
    modelSpaceOrthographicFrontDOrigin = new XYZ(0, 0, 0);
    modelSpaceOrthographicLeftDOrigin = new XYZ(0, 0, 0);
    modelSpaceOrthographicRightDOrigin = new XYZ(0, 0, 0);
    modelSpaceOrthographicTopDOrigin = new XYZ(0, 0, 0);
    modelSpaceUcs = new UCS();
    modelSpaceUcsBase = new UCS();
    get modelSpaceXAxis() { return this.modelSpaceUcs.xAxis; }
    set modelSpaceXAxis(v) { this.modelSpaceUcs.xAxis = v; }
    get modelSpaceYAxis() { return this.modelSpaceUcs.yAxis; }
    set modelSpaceYAxis(v) { this.modelSpaceUcs.yAxis = v; }
    northDirection = 0;
    numberOfSplineSegments = 8;
    objectSnapMode = 4133;
    obscuredColor = new Color(0);
    obscuredType = 0;
    orthoMode = false;
    get paperSpaceBaseName() { return this.paperSpaceUcsBase.name; }
    set paperSpaceBaseName(v) { this.paperSpaceUcsBase.name = v; }
    get paperSpaceElevation() { return this.paperSpaceUcs.elevation; }
    set paperSpaceElevation(v) { this.paperSpaceUcs.elevation = v; }
    paperSpaceExtMax = new XYZ(0, 0, 0);
    paperSpaceExtMin = new XYZ(0, 0, 0);
    paperSpaceInsertionBase = new XYZ(0, 0, 0);
    paperSpaceLimitsChecking = false;
    paperSpaceLimitsMax = new XY(0, 0);
    paperSpaceLimitsMin = new XY(0, 0);
    paperSpaceLineTypeScaling = SpaceLineTypeScaling.Normal;
    get paperSpaceName() { return this.paperSpaceUcs.name; }
    set paperSpaceName(v) { this.paperSpaceUcs.name = v; }
    paperSpaceOrthographicBackDOrigin = new XYZ(0, 0, 0);
    paperSpaceOrthographicBottomDOrigin = new XYZ(0, 0, 0);
    paperSpaceOrthographicFrontDOrigin = new XYZ(0, 0, 0);
    paperSpaceOrthographicLeftDOrigin = new XYZ(0, 0, 0);
    paperSpaceOrthographicRightDOrigin = new XYZ(0, 0, 0);
    paperSpaceOrthographicTopDOrigin = new XYZ(0, 0, 0);
    paperSpaceUcs = new UCS();
    paperSpaceUcsBase = new UCS();
    get paperSpaceXAxis() { return this.paperSpaceUcs.xAxis; }
    set paperSpaceXAxis(v) { this.paperSpaceUcs.xAxis = v; }
    get paperSpaceYAxis() { return this.paperSpaceUcs.yAxis; }
    set paperSpaceYAxis(v) { this.paperSpaceUcs.yAxis = v; }
    peditType = 0;
    pickStyle = 1;
    pointDisplayMode = 0;
    pointDisplaySize = 0.0;
    proxyGraphics = true;
    regenMode = true;
    get shadeEdge() { return this._shadeEdge; }
    set shadeEdge(v) { this._shadeEdge = v; }
    shadeDiffuse = 70;
    showModelSpaceInPaperSpace = false;
    sketchIncrement = 0.1;
    sketchPolylineType = SplineType.None;
    solidVisualStyleObjectType = 0;
    splineDegreeCurves = 3;
    stackedTextAlignment = 1;
    stackedTextSizePercentage = 70;
    get stepsPerSecond() { return this._stepsPerSecond; }
    set stepsPerSecond(v) { this._stepsPerSecond = v; }
    get surfaceIsolineCount() { return this._surfaceIsolineCount; }
    set surfaceIsolineCount(v) { this._surfaceIsolineCount = v; }
    surfaceDensityM = 6;
    surfaceDensityN = 6;
    surfaceType = 6;
    surfaceTabulation1 = 6;
    surfaceTabulation2 = 6;
    get textQuality() { return this._textQuality; }
    set textQuality(v) { this._textQuality = v; }
    textSize = 2.5;
    thickness = 0.0;
    tileModeEnabled = true;
    timeZone = 0;
    totalEditingTime = 0; // TimeSpan as milliseconds
    traceWidthDefault = 0;
    get ucsBaseName() { return this.modelSpaceUcsBase.name; }
    set ucsBaseName(v) { this.modelSpaceUcsBase.name = v; }
    get ucsName() { return this.modelSpaceUcs.name; }
    set ucsName(v) { this.modelSpaceUcs.name = v; }
    unitMode = 0;
    universalCreateDateTime = new Date();
    universalUpdateDateTime = new Date();
    updateDateTime = new Date();
    updateDimensionsWhileDragging = true;
    userDouble1 = 0;
    userDouble2 = 0;
    userDouble3 = 0;
    userDouble4 = 0;
    userDouble5 = 0;
    userElapsedTimeSpan = 0;
    userShort1 = 0;
    userShort2 = 0;
    userShort3 = 0;
    userShort4 = 0;
    userShort5 = 0;
    userTimer = false;
    get version() {
        return this._version;
    }
    set version(value) {
        this._version = value;
        switch (value) {
            case ACadVersion.AC1015:
                this.maintenanceVersion = 20;
                break;
            case ACadVersion.AC1018:
                this.maintenanceVersion = 104;
                break;
            case ACadVersion.AC1021:
                this.maintenanceVersion = 50;
                break;
            case ACadVersion.AC1024:
                this.maintenanceVersion = 226;
                break;
            case ACadVersion.AC1027:
                this.maintenanceVersion = 125;
                break;
            case ACadVersion.AC1032:
                this.maintenanceVersion = 228;
                break;
            default:
                this.maintenanceVersion = 0;
                break;
        }
    }
    versionGuid = crypto.randomUUID?.() ?? '';
    get versionString() {
        return CadUtils.getNameFromVersion(this.version);
    }
    set versionString(value) {
        this.version = CadUtils.getVersionFromName(value);
    }
    viewportDefaultViewScaleFactor = 0;
    worldView = true;
    xEdit = false;
    // --- Private fields ---
    _angularUnitPrecision = 0;
    _currentLayer = Layer.Default;
    _currentLineType = LineType.ByLayer;
    _currentTextStyle = TextStyle.Default;
    _dimensionStyleOverrides = new DimensionStyle('override');
    _currentDimensionStyle = DimensionStyle.Default;
    _dimensionTextStyle = TextStyle.Default;
    _facetResolution = 0.5;
    _linearUnitPrecision = 4;
    _stepsPerSecond = 2.0;
    _surfaceIsolineCount = 4;
    _textQuality = 50;
    _version = ACadVersion.AC1032;
    _shadeEdge = 0;
    constructor(versionOrDocument) {
        if (typeof versionOrDocument === 'number') {
            this.version = versionOrDocument;
        }
        else if (versionOrDocument && typeof versionOrDocument === 'object') {
            this.version = ACadVersion.AC1032;
            this.document = versionOrDocument;
        }
        else {
            this.version = ACadVersion.AC1032;
        }
    }
    // TODO: GetHeaderMap, GetValues, SetValue - require reflection which doesn't exist in TS
    requiredVersions = 0;
    DIMSAV = 0;
    polylineLineTypeGeneration = false;
    regenerationMode = 0;
    quickTextMode = false;
    showSplineControlPoints = false;
    showModelSpace = false;
    retainXRefDependentVisibilitySettings = false;
    spatialIndexMaxTreeDepth = 0;
    splineType = 0;
    shadeDiffuseToAmbientPercentage = 70;
    textHeightDefault = 0.2;
    thicknessDefault = 0;
    polylineWidthDefault = 0;
    dimensionAltMzf = 0;
    dimensionAltMzs = "";
    dimensionMzs = "";
    styleSheetName = "";
    plotStyleMode = 0;
    projectName = "";
    stepSize = 0;
    solidsRetainHistory = false;
    showSolidsHistory = false;
    sweptSolidWidth = 0;
    sweptSolidHeight = 0;
    solidLoftedShape = 0;
    shadowMode = 0;
    shadowPlaneLocation = 0;
    dimensionMzf = 0;
    static GetHeaderMap() {
        // TODO: Implement header map based on decorators/metadata
        return new Map();
    }
    SetValue(variable, parameters) {
        // TODO: Implement SetValue based on metadata
    }
    toString() {
        return `${this.version}`;
    }
}
export { SpaceLineTypeScaling } from './SpaceLineTypeScaling.js';
export { EntityPlotStyleType } from './EntityPlotStyleType.js';
