// @node-projects/acad-ts/dist/Header/ObjectSnapMode.js
export var ObjectSnapMode;
(function (ObjectSnapMode) {
    ObjectSnapMode[ObjectSnapMode["None"] = 0] = "None";
    ObjectSnapMode[ObjectSnapMode["EndPoint"] = 1] = "EndPoint";
    ObjectSnapMode[ObjectSnapMode["MidPoint"] = 2] = "MidPoint";
    ObjectSnapMode[ObjectSnapMode["Center"] = 4] = "Center";
    ObjectSnapMode[ObjectSnapMode["Node"] = 8] = "Node";
    ObjectSnapMode[ObjectSnapMode["Quadrant"] = 16] = "Quadrant";
    ObjectSnapMode[ObjectSnapMode["Intersection"] = 32] = "Intersection";
    ObjectSnapMode[ObjectSnapMode["Insertion"] = 64] = "Insertion";
    ObjectSnapMode[ObjectSnapMode["Perpendicular"] = 128] = "Perpendicular";
    ObjectSnapMode[ObjectSnapMode["Tangent"] = 256] = "Tangent";
    ObjectSnapMode[ObjectSnapMode["Nearest"] = 512] = "Nearest";
    ObjectSnapMode[ObjectSnapMode["ClearsAllObjectSnaps"] = 1024] = "ClearsAllObjectSnaps";
    ObjectSnapMode[ObjectSnapMode["ApparentIntersection"] = 2048] = "ApparentIntersection";
    ObjectSnapMode[ObjectSnapMode["Extension"] = 4096] = "Extension";
    ObjectSnapMode[ObjectSnapMode["Parallel"] = 8192] = "Parallel";
    ObjectSnapMode[ObjectSnapMode["AllModes"] = 16383] = "AllModes";
    ObjectSnapMode[ObjectSnapMode["SwitchedOff"] = 16384] = "SwitchedOff";
})(ObjectSnapMode || (ObjectSnapMode = {}));
