// @node-projects/acad-ts/dist/Types/Units/UnitStyleFormat.js
import { AngularZeroHandling, ZeroHandling } from '../../Tables/ZeroHandling.js';
import { FractionFormat } from '../../Tables/FractionFormat.js';
export class UnitStyleFormat {
    get angularDecimalPlaces() {
        return this._angularDecimalPlaces;
    }
    set angularDecimalPlaces(value) {
        if (value < 0) {
            throw new RangeError('The number of decimal places must be equals or greater than zero.');
        }
        this._angularDecimalPlaces = value;
    }
    angularZeroHandling = AngularZeroHandling.DisplayAll;
    decimalSeparator;
    degreesSymbol;
    feetInchesSeparator;
    feetSymbol;
    get fractionHeightScale() {
        return this._fractionHeightScale;
    }
    set fractionHeightScale(value) {
        if (value <= 0) {
            throw new RangeError('The fraction height scale must be greater than zero.');
        }
        this._fractionHeightScale = value;
    }
    fractionType;
    gradiansSymbol;
    inchesSymbol;
    get linearDecimalPlaces() {
        return this._linearDecimalPlaces;
    }
    set linearDecimalPlaces(value) {
        if (value < 0) {
            throw new RangeError('The number of decimal places must be equals or greater than zero.');
        }
        this._linearDecimalPlaces = value;
    }
    linearZeroHandling = ZeroHandling.SuppressDecimalTrailingZeroes;
    minutesSymbol;
    radiansSymbol;
    secondsSymbol;
    get suppressAngularLeadingZeros() {
        return this.linearZeroHandling === ZeroHandling.SuppressDecimalLeadingZeroes
            || this.linearZeroHandling === ZeroHandling.SuppressDecimalLeadingAndTrailingZeroes;
    }
    get suppressAngularTrailingZeros() {
        return this.linearZeroHandling === ZeroHandling.SuppressDecimalTrailingZeroes
            || this.linearZeroHandling === ZeroHandling.SuppressDecimalLeadingAndTrailingZeroes;
    }
    get suppressLinearLeadingZeros() {
        return this.linearZeroHandling === ZeroHandling.SuppressDecimalLeadingZeroes
            || this.linearZeroHandling === ZeroHandling.SuppressDecimalLeadingAndTrailingZeroes;
    }
    get suppressLinearTrailingZeros() {
        return this.linearZeroHandling === ZeroHandling.SuppressDecimalTrailingZeroes
            || this.linearZeroHandling === ZeroHandling.SuppressDecimalLeadingAndTrailingZeroes;
    }
    get suppressZeroFeet() {
        return this.linearZeroHandling === ZeroHandling.SuppressZeroFeetAndInches
            || this.linearZeroHandling === ZeroHandling.SuppressZeroFeetShowZeroInches;
    }
    get suppressZeroInches() {
        return this.linearZeroHandling === ZeroHandling.SuppressZeroFeetAndInches
            || this.linearZeroHandling === ZeroHandling.ShowZeroFeetSuppressZeroInches;
    }
    _angularDecimalPlaces;
    _fractionHeightScale;
    _linearDecimalPlaces;
    constructor() {
        this._linearDecimalPlaces = 2;
        this._angularDecimalPlaces = 0;
        this.decimalSeparator = '.';
        this.feetInchesSeparator = '-';
        this.degreesSymbol = '°';
        this.minutesSymbol = "'";
        this.secondsSymbol = '"';
        this.radiansSymbol = 'r';
        this.gradiansSymbol = 'g';
        this.feetSymbol = "'";
        this.inchesSymbol = '"';
        this._fractionHeightScale = 1.0;
        this.fractionType = FractionFormat.Horizontal;
    }
    getZeroHandlingFormat(isAngular = false) {
        let decimalPlaces;
        let leading;
        let trailing;
        if (isAngular) {
            decimalPlaces = this.angularDecimalPlaces;
            leading = this.angularZeroHandling === AngularZeroHandling.SuppressLeadingZeroes
                || this.angularZeroHandling === AngularZeroHandling.SupressAll ? '#' : '0';
            trailing = this.angularZeroHandling === AngularZeroHandling.SupressTrailingZeroes
                || this.angularZeroHandling === AngularZeroHandling.SupressAll ? '#' : '0';
        }
        else {
            decimalPlaces = this.linearDecimalPlaces;
            leading = this.linearZeroHandling === ZeroHandling.SuppressDecimalLeadingZeroes
                || this.linearZeroHandling === ZeroHandling.SuppressDecimalLeadingAndTrailingZeroes
                ? '#' : '0';
            trailing = this.linearZeroHandling === ZeroHandling.SuppressDecimalTrailingZeroes
                || this.linearZeroHandling === ZeroHandling.SuppressDecimalLeadingAndTrailingZeroes
                ? '#' : '0';
        }
        let zeroes = leading + '.';
        for (let i = 0; i < decimalPlaces; i++) {
            zeroes += trailing;
        }
        return zeroes;
    }
    toArchitectural(value) {
        const feet = Math.trunc(value / 12);
        const inchesDec = value - 12 * feet;
        const inches = Math.trunc(inchesDec);
        if (Math.abs(inchesDec) < 1e-12) {
            if (feet === 0) {
                if (this.suppressZeroFeet) {
                    return `0${this.inchesSymbol}`;
                }
                if (this.suppressZeroInches) {
                    return `0${this.feetSymbol}`;
                }
                return `0${this.feetSymbol}${this.feetInchesSeparator}0${this.inchesSymbol}`;
            }
            if (this.suppressZeroInches) {
                return `${feet}${this.feetSymbol}`;
            }
            return `${feet}${this.feetSymbol}${this.feetInchesSeparator}0${this.inchesSymbol}`;
        }
        const { numerator, denominator } = UnitStyleFormat.getFraction(inchesDec, Math.pow(2, this.linearDecimalPlaces));
        if (numerator === 0) {
            if (inches === 0) {
                if (feet === 0) {
                    if (this.suppressZeroFeet) {
                        return `0${this.inchesSymbol}`;
                    }
                    if (this.suppressZeroInches) {
                        return `0${this.feetSymbol}`;
                    }
                    return `0${this.feetSymbol}${this.feetInchesSeparator}0${this.inchesSymbol}`;
                }
                if (this.suppressZeroInches) {
                    return `${feet}${this.feetSymbol}`;
                }
                return `${feet}${this.feetSymbol}${this.feetInchesSeparator}0${this.inchesSymbol}`;
            }
            if (feet === 0) {
                if (this.suppressZeroFeet) {
                    return `${inches}${this.inchesSymbol}`;
                }
                return `0${this.feetSymbol}${this.feetInchesSeparator}${inches}${this.inchesSymbol}`;
            }
            return `${feet}${this.feetSymbol}${this.feetInchesSeparator}${inches}${this.inchesSymbol}`;
        }
        let feetStr;
        if (this.suppressZeroFeet && feet === 0) {
            feetStr = '';
        }
        else {
            feetStr = feet + this.feetSymbol + this.feetInchesSeparator;
        }
        let text = '';
        switch (this.fractionType) {
            case FractionFormat.Diagonal:
                text = `\\A1;${feetStr}${inches}{\\H${this.fractionHeightScale}x;\\S${numerator}#${denominator};}${this.inchesSymbol}`;
                break;
            case FractionFormat.Horizontal:
                text = `\\A1;${feetStr}${inches}{\\H${this.fractionHeightScale}x;\\S${numerator}/${denominator};}${this.inchesSymbol}`;
                break;
            case FractionFormat.None:
                text = `${feetStr}${inches} ${numerator}/${denominator}${this.inchesSymbol}`;
                break;
        }
        return text;
    }
    toDecimal(value, isAngular = false) {
        return UnitStyleFormat.formatNumber(value, this.getZeroHandlingFormat(isAngular), this.decimalSeparator);
    }
    toDegrees(angle) {
        const degrees = angle * (180.0 / Math.PI);
        return UnitStyleFormat.formatNumber(degrees, this.getZeroHandlingFormat(true), this.decimalSeparator) + this.degreesSymbol;
    }
    toDegreesMinutesSeconds(angle) {
        const degrees = angle * (180.0 / Math.PI);
        const minutes = (degrees - Math.trunc(degrees)) * 60;
        const seconds = (minutes - Math.trunc(minutes)) * 60;
        if (this.angularDecimalPlaces === 0) {
            return `${Math.round(degrees)}${this.degreesSymbol}`;
        }
        if (this.angularDecimalPlaces === 1 || this.angularDecimalPlaces === 2) {
            return `${Math.trunc(degrees)}${this.degreesSymbol}${Math.round(minutes)}${this.minutesSymbol}`;
        }
        if (this.angularDecimalPlaces === 3 || this.angularDecimalPlaces === 4) {
            return `${Math.trunc(degrees)}${this.degreesSymbol}${Math.trunc(minutes)}${this.minutesSymbol}${Math.round(seconds)}${this.secondsSymbol}`;
        }
        const f = '0.' + '0'.repeat(this.angularDecimalPlaces - 4);
        return `${Math.trunc(degrees)}${this.degreesSymbol}${Math.trunc(minutes)}${this.minutesSymbol}${UnitStyleFormat.formatNumber(seconds, f, this.decimalSeparator)}${this.secondsSymbol}`;
    }
    toEngineering(value) {
        const feet = Math.trunc(value / 12);
        const inches = value - 12 * feet;
        if (Math.abs(inches) < 1e-12) {
            if (feet === 0) {
                if (this.suppressZeroFeet) {
                    return `0${this.inchesSymbol}`;
                }
                if (this.suppressZeroInches) {
                    return `0${this.feetSymbol}`;
                }
                return `0${this.feetSymbol}${this.feetInchesSeparator}0${this.inchesSymbol}`;
            }
            if (this.suppressZeroInches) {
                return `${feet}${this.feetSymbol}`;
            }
            return `${feet}${this.feetSymbol}${this.feetInchesSeparator}0${this.inchesSymbol}`;
        }
        const inchesDec = UnitStyleFormat.formatNumber(inches, this.getZeroHandlingFormat(), this.decimalSeparator);
        if (feet === 0) {
            if (this.suppressZeroFeet) {
                return `${inches}${this.inchesSymbol}`;
            }
            return `0${this.feetSymbol}${this.feetInchesSeparator}${inchesDec}${this.inchesSymbol}`;
        }
        return `${feet}${this.feetSymbol}${this.feetInchesSeparator}${inchesDec}${this.inchesSymbol}`;
    }
    toFractional(value) {
        const num = Math.trunc(value);
        const { numerator, denominator } = UnitStyleFormat.getFraction(value, Math.pow(2, this.linearDecimalPlaces));
        if (numerator === 0) {
            return `${Math.trunc(value)}`;
        }
        let text = '';
        switch (this.fractionType) {
            case FractionFormat.Diagonal:
                text = `\\A1;${num}{\\H${this.fractionHeightScale}x;\\S${numerator}#${denominator};}`;
                break;
            case FractionFormat.Horizontal:
                text = `\\A1;${num}{\\H${this.fractionHeightScale}x;\\S${numerator}/${denominator};}`;
                break;
            case FractionFormat.None: {
                const prefix = num === 0 ? '' : `${num} `;
                text = `${prefix}${numerator}/${denominator}`;
                break;
            }
        }
        return text;
    }
    toGradians(angle) {
        const gradians = angle * (200.0 / Math.PI);
        return UnitStyleFormat.formatNumber(gradians, this.getZeroHandlingFormat(true), this.decimalSeparator) + this.gradiansSymbol;
    }
    toRadians(angle) {
        return UnitStyleFormat.formatNumber(angle, this.getZeroHandlingFormat(true), this.decimalSeparator) + this.radiansSymbol;
    }
    toScientific(value) {
        return value.toExponential(this.linearDecimalPlaces).toUpperCase();
    }
    static getFraction(number, precision) {
        let numerator = Math.round((number - Math.trunc(number)) * precision);
        let commonFactor = UnitStyleFormat.getGCD(numerator, precision);
        if (commonFactor <= 0) {
            commonFactor = 1;
        }
        numerator = Math.trunc(numerator / commonFactor);
        const denominator = Math.trunc(precision / commonFactor);
        return { numerator, denominator };
    }
    static getGCD(number1, number2) {
        let a = number1;
        let b = number2;
        while (b !== 0) {
            const count = a % b;
            a = b;
            b = count;
        }
        return a;
    }
    static formatNumber(value, format, decimalSeparator) {
        // Parse format string like "0.00", "#.##", "0.##" etc.
        const parts = format.split('.');
        const intPart = parts[0] || '0';
        const decPart = parts[1] || '';
        const decimalPlaces = decPart.length;
        const suppressLeading = intPart === '#';
        const suppressTrailing = decPart.length > 0 && decPart[0] === '#';
        let result = value.toFixed(decimalPlaces);
        if (decimalSeparator !== '.') {
            result = result.replace('.', decimalSeparator);
        }
        if (suppressTrailing && result.includes(decimalSeparator)) {
            while (result.endsWith('0')) {
                result = result.slice(0, -1);
            }
            if (result.endsWith(decimalSeparator)) {
                result = result.slice(0, -1);
            }
        }
        if (suppressLeading) {
            if (result.startsWith('0' + decimalSeparator)) {
                result = result.slice(1);
            }
            else if (result.startsWith('-0' + decimalSeparator)) {
                result = '-' + result.slice(2);
            }
        }
        return result;
    }
}
