// @node-projects/acad-ts/dist/Types/Units/UnitsType.js
export var UnitsType;
(function (UnitsType) {
    UnitsType[UnitsType["Unitless"] = 0] = "Unitless";
    UnitsType[UnitsType["Inches"] = 1] = "Inches";
    UnitsType[UnitsType["Feet"] = 2] = "Feet";
    UnitsType[UnitsType["Miles"] = 3] = "Miles";
    UnitsType[UnitsType["Millimeters"] = 4] = "Millimeters";
    UnitsType[UnitsType["Centimeters"] = 5] = "Centimeters";
    UnitsType[UnitsType["Meters"] = 6] = "Meters";
    UnitsType[UnitsType["Kilometers"] = 7] = "Kilometers";
    UnitsType[UnitsType["Microinches"] = 8] = "Microinches";
    UnitsType[UnitsType["Mils"] = 9] = "Mils";
    UnitsType[UnitsType["Yards"] = 10] = "Yards";
    UnitsType[UnitsType["Angstroms"] = 11] = "Angstroms";
    UnitsType[UnitsType["Nanometers"] = 12] = "Nanometers";
    UnitsType[UnitsType["Microns"] = 13] = "Microns";
    UnitsType[UnitsType["Decimeters"] = 14] = "Decimeters";
    UnitsType[UnitsType["Decameters"] = 15] = "Decameters";
    UnitsType[UnitsType["Hectometers"] = 16] = "Hectometers";
    UnitsType[UnitsType["Gigameters"] = 17] = "Gigameters";
    UnitsType[UnitsType["AstronomicalUnits"] = 18] = "AstronomicalUnits";
    UnitsType[UnitsType["LightYears"] = 19] = "LightYears";
    UnitsType[UnitsType["Parsecs"] = 20] = "Parsecs";
    UnitsType[UnitsType["USSurveyFeet"] = 21] = "USSurveyFeet";
    UnitsType[UnitsType["USSurveyInches"] = 22] = "USSurveyInches";
    UnitsType[UnitsType["USSurveyYards"] = 23] = "USSurveyYards";
})(UnitsType || (UnitsType = {}));
