// @spectrum-web-components/action-menu/stories/action-menu.stories.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { html } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/menu/sp-menu.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/menu/sp-menu-item.js';
import { ActionMenuMarkup } from './index.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/icons-workflow/icons/sp-icon-settings.js';
export default {
    component: 'sp-action-menu',
    title: 'Action menu',
    argTypes: {
        disabled: {
            name: 'disabled',
            type: { name: 'boolean', required: false },
            description: 'Disable this control. It will not receive focus or events.',
            table: {
                type: { summary: 'boolean' },
                defaultValue: { summary: false },
            },
            control: {
                type: 'boolean',
            },
        },
        open: {
            name: 'open',
            type: { name: 'boolean', required: false },
            description: 'Whether the menu is open or not.',
            table: {
                type: { summary: 'boolean' },
                defaultValue: { summary: false },
            },
            control: 'boolean',
        },
        visibleLabel: {
            name: 'Visible Label',
            description: 'The placeholder content for the picker.',
            type: { name: 'string', required: false },
            table: {
                type: { summary: 'string' },
                defaultValue: { summary: '' },
            },
            control: 'text',
        },
    },
    args: {
        visibleLabel: 'More Actions',
        disabled: false,
        open: false,
    },
};
const Template = (args = {}) => ActionMenuMarkup(args);
export const Default = (args = {}) => Template(args);
export const selects = (args = {}) => Template(Object.assign(Object.assign({}, args), { selects: 'single', selected: true }));
selects.args = {
    open: true,
};
export const iconOnly = (args = {}) => Template(args);
iconOnly.args = {
    visibleLabel: '',
};
export const customIcon = (args) => Template(args);
customIcon.args = {
    customIcon: html `
        <sp-icon-settings slot="icon"></sp-icon-settings>
    `,
    visibleLabel: '',
};
