// @node-projects/web-component-designer/dist/elements/services/undoService/transactionItems/CssStyleChangeAction.js
export class CssStyleChangeAction {
    constructor(designItem, name, newValue, oldValue) {
        this.title = "Change CSS Style";
        this.designItem = designItem;
        this.name = name;
        this.newValue = newValue;
        this.oldValue = oldValue;
    }
    get affectedItems() {
        return [this.designItem];
    }
    undo() {
        if (this.oldValue === '')
            this.designItem.styles.delete(this.name);
        else
            this.designItem.styles.set(this.name, this.oldValue);
        this.designItem.element.style[this.name] = this.oldValue;
    }
    do() {
        if (this.newValue === '')
            this.designItem.styles.delete(this.name);
        else
            this.designItem.styles.set(this.name, this.newValue);
        this.designItem.element.style[this.name] = this.newValue;
    }
    mergeWith(other) {
        if (other instanceof CssStyleChangeAction && this.designItem === other.designItem && this.name === other.name) {
            this.newValue = other.newValue;
            return true;
        }
        return false;
    }
}
