// @spectrum-web-components/top-nav/src/TopNavItem.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { ifDefined } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/directives.js';
import { property, query, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import { Focusable, LikeAnchor } from '/web-component-designer-demo/node_modules/@spectrum-web-components/shared/./src/index.js';
import itemStyles from '/web-component-designer-demo/node_modules/@spectrum-web-components/tabs/src/tab.css.js';
import topNavItemStyles from './top-nav-item.css.js';
/**
 * @element sp-top-nav-item
 *
 * @slot - text label of the Top Nav Item
 */
export class TopNavItem extends LikeAnchor(Focusable) {
    constructor() {
        super(...arguments);
        this.selected = false;
        this.value = '';
    }
    static get styles() {
        return [itemStyles, topNavItemStyles];
    }
    get focusElement() {
        return this.anchor;
    }
    click() {
        this.anchor.click();
    }
    render() {
        return html `
            <a
                id="item-label"
                href=${ifDefined(this.href)}
                download=${ifDefined(this.download)}
                target=${ifDefined(this.target)}
                aria-label=${ifDefined(this.label)}
                aria-current=${ifDefined(this.selected && this.href ? 'page' : undefined)}
                rel=${ifDefined(this.rel)}
            >
                <slot></slot>
            </a>
        `;
    }
    firstUpdated(changes) {
        super.firstUpdated(changes);
    }
    updated(changes) {
        super.updated(changes);
        this.value = this.anchor.href;
    }
}
__decorate([
    query('a')
], TopNavItem.prototype, "anchor", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], TopNavItem.prototype, "selected", void 0);
