// @spectrum-web-components/top-nav/src/TopNav.js
/* eslint-disable lit-a11y/click-events-have-key-events */
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, SizedMixin, SpectrumElement, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import { ifDefined } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/directives.js';
import tabStyles from '/web-component-designer-demo/node_modules/@spectrum-web-components/tabs/src/tabs.css.js';
const noSelectionStyle = 'transform: translateX(0px) scaleX(0) scaleY(0)';
/**
 * @element sp-top-nav
 *
 * @slot - Nav Items to display as a group
 * @attr {Boolean} quiet - The tabs border is a lot smaller
 * @attr {Boolean} compact - The collection of tabs take up less space
 */
export class TopNav extends SizedMixin(SpectrumElement) {
    constructor() {
        super(...arguments);
        this.selectionIndicatorStyle = noSelectionStyle;
        this.shouldAnimate = false;
        this.onClick = (event) => {
            const target = event.target;
            this.shouldAnimate = true;
            this.selectTarget(target);
        };
        this.items = [];
        this.updateSelectionIndicator = async () => {
            const selectedItem = this.items.find((item) => item.value === this.selected ||
                item.value === window.location.href);
            if (!selectedItem) {
                this.selectionIndicatorStyle = noSelectionStyle;
                return;
            }
            await Promise.all([
                selectedItem.updateComplete,
                document.fonts ? document.fonts.ready : Promise.resolve(),
            ]);
            const itemBoundingClientRect = selectedItem.getBoundingClientRect();
            const parentBoundingClientRect = this.getBoundingClientRect();
            const width = itemBoundingClientRect.width;
            const offset = this.dir === 'ltr'
                ? itemBoundingClientRect.left - parentBoundingClientRect.left
                : itemBoundingClientRect.right - parentBoundingClientRect.right;
            this.selectionIndicatorStyle = `transform: translateX(${offset}px) scaleX(${this.dir === 'ltr' ? width : -1 * width});`;
        };
    }
    static get styles() {
        return [tabStyles];
    }
    set selected(value) {
        const oldValue = this.selected;
        if (value === oldValue) {
            return;
        }
        this.updateCheckedState(value);
        this._selected = value;
        this.requestUpdate('selected', oldValue);
    }
    get selected() {
        return this._selected;
    }
    manageItems() {
        this.items = [
            ...this.querySelectorAll('sp-top-nav-item'),
        ];
        const selectedChild = this.items.find((item) => item.value === window.location.href);
        if (selectedChild) {
            this.selectTarget(selectedChild);
        }
    }
    render() {
        return html `
            <div @click=${this.onClick} id="list">
                <slot @slotchange=${this.onSlotChange}></slot>
                <div
                    id="selection-indicator"
                    class=${ifDefined(this.shouldAnimate ? undefined : 'first-position')}
                    style=${this.selectionIndicatorStyle}
                ></div>
            </div>
        `;
    }
    firstUpdated(changes) {
        super.firstUpdated(changes);
        this.setAttribute('direction', 'horizontal');
        this.manageItems();
    }
    updated(changes) {
        super.updated(changes);
        if (changes.has('dir')) {
            this.updateSelectionIndicator();
        }
        if (!this.shouldAnimate &&
            typeof changes.get('shouldAnimate') !== 'undefined') {
            this.shouldAnimate = true;
        }
    }
    selectTarget(target) {
        const { value } = target;
        if (value) {
            this.selected = value;
        }
    }
    onSlotChange() {
        this.manageItems();
    }
    updateCheckedState(value) {
        this.items.forEach((item) => {
            item.selected = false;
        });
        requestAnimationFrame(() => {
            if (value && value.length) {
                const currentItem = this.items.find((item) => item.value === value ||
                    item.value === window.location.href);
                if (currentItem) {
                    currentItem.selected = true;
                }
                else {
                    this.selected = '';
                }
            }
            this.updateSelectionIndicator();
        });
    }
    connectedCallback() {
        super.connectedCallback();
        window.addEventListener('resize', this.updateSelectionIndicator);
        if ('fonts' in document) {
            document.fonts.addEventListener('loadingdone', this.updateSelectionIndicator);
        }
    }
    disconnectedCallback() {
        window.removeEventListener('resize', this.updateSelectionIndicator);
        if ('fonts' in document) {
            document.fonts.removeEventListener('loadingdone', this.updateSelectionIndicator);
        }
        super.disconnectedCallback();
    }
}
__decorate([
    property()
], TopNav.prototype, "selectionIndicatorStyle", void 0);
__decorate([
    property({ attribute: false })
], TopNav.prototype, "shouldAnimate", void 0);
__decorate([
    property({ reflect: true })
], TopNav.prototype, "selected", null);
