// @spectrum-web-components/picker/stories/picker-sizes.stories.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { html } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import '../sp-picker.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/menu/sp-menu.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/menu/sp-menu-item.js';
export default {
    title: 'Picker/Sizes',
    component: 'sp-picker',
    argTypes: {
        onChange: { action: 'change' },
    },
};
const picker = ({ onChange, size, }) => {
    return html `
        <sp-field-label for="picker-${size}" size=${size}>
            Where do you live?
        </sp-field-label>
        <sp-picker
            id="picker-${size}"
            size=${size}
            @change="${(event) => {
        const picker = event.target;
        onChange(picker.value);
    }}"
            label="Select a Country with a very long label, too long, in fact"
        >
            <sp-menu-item>Deselect</sp-menu-item>
            <sp-menu-item>Select Inverse</sp-menu-item>
            <sp-menu-item>Feather...</sp-menu-item>
            <sp-menu-item>Select and Mask...</sp-menu-item>
            <sp-menu-divider></sp-menu-divider>
            <sp-menu-item>Save Selection</sp-menu-item>
            <sp-menu-item disabled>Make Work Path</sp-menu-item>
        </sp-picker>
    `;
};
export const s = (args) => picker(Object.assign(Object.assign({}, args), { size: 's' }));
export const m = (args) => picker(Object.assign(Object.assign({}, args), { size: 'm' }));
export const l = (args) => picker(Object.assign(Object.assign({}, args), { size: 'l' }));
export const XL = (args) => picker(Object.assign(Object.assign({}, args), { size: 'xl' }));
