// @spectrum-web-components/action-menu/src/ActionMenu.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { ifDefined } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/directives.js';
import { property } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/menu/sp-menu.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/popover/sp-popover.js';
import { PickerBase } from '/web-component-designer-demo/node_modules/@spectrum-web-components/picker/src/index.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/action-button/sp-action-button.js';
import { ObserveSlotText } from '/web-component-designer-demo/node_modules/@spectrum-web-components/shared/src/observe-slot-text.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/icons-workflow/icons/sp-icon-more.js';
import actionMenuStyles from './action-menu.css.js';
/**
 * @element sp-action-menu
 *
 * @slot - menu items to be listed in the Action Menu
 * @slot icon - The icon to use for Action Menu
 * @slot label - The label to use on for the Action Menu
 * @attr selects - By default `sp-action-menu` does not manage a selection. If
 *   you'd like for a selection to be held by the `sp-menu` that it presents in
 *   its overlay, use `selects="single" to activate this functionality.
 */
export class ActionMenu extends ObserveSlotText(PickerBase, 'label') {
    constructor() {
        super(...arguments);
        this.selects = undefined;
        this.listRole = 'menu';
        this.itemRole = 'menuitem';
    }
    static get styles() {
        return [actionMenuStyles];
    }
    get hasLabel() {
        return this.slotHasContent;
    }
    get buttonContent() {
        return [
            html `
                <slot name="icon" slot="icon" ?icon-only=${!this.hasLabel}>
                    <sp-icon-more class="icon"></sp-icon-more>
                </slot>
                <slot name="label" ?hidden=${!this.hasLabel}></slot>
            `,
        ];
    }
    render() {
        return html `
            <sp-action-button
                quiet
                ?selected=${this.open}
                aria-haspopup="true"
                aria-controls="popover"
                aria-expanded=${this.open ? 'true' : 'false'}
                aria-label=${ifDefined(this.label || undefined)}
                id="button"
                class="button"
                size=${this.size}
                @blur=${this.onButtonBlur}
                @click=${this.onButtonClick}
                @focus=${this.onButtonFocus}
                ?disabled=${this.disabled}
            >
                ${this.buttonContent}
            </sp-action-button>
        `;
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('invalid')) {
            this.invalid = false;
        }
        this.quiet = true;
    }
}
__decorate([
    property({ type: String })
], ActionMenu.prototype, "selects", void 0);
