// @spectrum-web-components/field-group/src/FieldGroup.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, SpectrumElement, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import { ManageHelpText } from '/web-component-designer-demo/node_modules/@spectrum-web-components/help-text/src/manage-help-text.js';
import styles from './field-group.css.js';
/**
 * @element sp-field-group
 * @slot - the form controls that make up the group
 * @slot help-text - default or non-negative help text to associate to your form element
 * @slot negative-help-text - negative help text to associate to your form element when `invalid`
 */
export class FieldGroup extends ManageHelpText(SpectrumElement, {
    mode: 'external',
}) {
    constructor() {
        super(...arguments);
        this.horizontal = false;
        this.invalid = false;
        this.label = '';
        this.vertical = false;
    }
    static get styles() {
        return [styles];
    }
    handleSlotchange() {
        // Surface this functionality for easy composition in extensions.
        return;
    }
    render() {
        return html `
            <div class="group" role="presentation">
                <slot @slotchange=${this.handleSlotchange}></slot>
            </div>
            ${this.renderHelpText(this.invalid)}
        `;
    }
    updated(changes) {
        super.updated(changes);
        if (changes.has('label')) {
            if (this.label) {
                this.setAttribute('aria-label', this.label);
            }
            else {
                this.removeAttribute('aria-label');
            }
        }
    }
}
__decorate([
    property({ type: Boolean, reflect: true })
], FieldGroup.prototype, "horizontal", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], FieldGroup.prototype, "invalid", void 0);
__decorate([
    property()
], FieldGroup.prototype, "label", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], FieldGroup.prototype, "vertical", void 0);
