// @spectrum-web-components/checkbox/stories/checkbox-sizes.stories.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { html } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import '../sp-checkbox.js';
export default {
    component: 'sp-checkbox',
    title: 'Checkbox/Sizes',
    argTypes: {
        onClick: { action: 'click' },
        onChange: { action: 'change' },
    },
};
const checkbox = ({ size, checked, indeterminate, onClick, onChange, }) => {
    return html `
        <sp-checkbox
            size=${size}
            ?checked=${checked}
            ?indeterminate=${indeterminate}
            @click="${onClick}"
            @change="${onChange}"
        >
            Checkbox
        </sp-checkbox>
    `;
};
export const s = (args) => checkbox(Object.assign(Object.assign({}, args), { size: 's' }));
export const sChecked = (args) => checkbox(Object.assign(Object.assign({}, args), { size: 's', checked: true }));
export const sIndeterminate = (args) => checkbox(Object.assign(Object.assign({}, args), { size: 's', indeterminate: true }));
export const m = (args) => checkbox(Object.assign(Object.assign({}, args), { size: 'm' }));
export const mChecked = (args) => checkbox(Object.assign(Object.assign({}, args), { size: 'm', checked: true }));
export const mIndeterminate = (args) => checkbox(Object.assign(Object.assign({}, args), { size: 'm', indeterminate: true }));
export const l = (args) => checkbox(Object.assign(Object.assign({}, args), { size: 'l' }));
export const lChecked = (args) => checkbox(Object.assign(Object.assign({}, args), { size: 'l', checked: true }));
export const lIndeterminate = (args) => checkbox(Object.assign(Object.assign({}, args), { size: 'l', indeterminate: true }));
export const XL = (args) => checkbox(Object.assign(Object.assign({}, args), { size: 'xl' }));
export const XLChecked = (args) => checkbox(Object.assign(Object.assign({}, args), { size: 'xl', checked: true }));
export const XLIndeterminate = (args) => checkbox(Object.assign(Object.assign({}, args), { size: 'xl', indeterminate: true }));
