// @lit-labs/observers/development/intersection_controller.js
/**
 * IntersectionController is a ReactiveController that integrates an
 * IntersectionObserver with a ReactiveControllerHost's reactive update
 * lifecycle. This is typically a ReactiveElement or LitElement.
 * IntersectionObservers can be used to detect when a target element
 * "intersects" is visible inside of) another element or the viewport by
 * default, where intersect means "visible inside of."
 *
 * The controller can specify a `target` element to observe and the
 * configuration options to pass to the IntersectionObserver. The `observe`
 * method can be called to observe additional elements.
 *
 * When a change is detected, the controller's given `callback` function is
 * used to process the result into a value which is stored on the controller.
 * The controller's `value` is usable during the host's update cycle.
 */
export class IntersectionController {
    constructor(host, { target, config, callback, skipInitial }) {
        this._skipInitial = false;
        /**
         * Flag used to help manage calling the `callback` when observe is called
         * and `skipInitial` is set to true. Note that unlike the other observers
         * IntersectionObserver *does* report its initial state (e.g. whether or not
         * there is an intersection). This flag is used to avoid handling this
         * state if `skipInitial` is true.
         */
        this._unobservedUpdate = false;
        /**
         * Function that returns a value processed from the observer's changes.
         * The result is stored in the `value` property.
         */
        this.callback = () => true;
        (this._host = host).addController(this);
        // Target defaults to `host` unless explicitly `null`.
        this._target =
            target === null ? target : target !== null && target !== void 0 ? target : this._host;
        this._skipInitial = skipInitial !== null && skipInitial !== void 0 ? skipInitial : this._skipInitial;
        this.callback = callback !== null && callback !== void 0 ? callback : this.callback;
        // Check browser support.
        if (!window.IntersectionObserver) {
            console.warn(`IntersectionController error: browser does not support IntersectionObserver.`);
            return;
        }
        this._observer = new IntersectionObserver((entries) => {
            const unobservedUpdate = this._unobservedUpdate;
            this._unobservedUpdate = false;
            if (this._skipInitial && unobservedUpdate) {
                return;
            }
            this.handleChanges(entries);
            this._host.requestUpdate();
        }, config);
    }
    /**
     * Process the observer's changes with the controller's `callback`
     * function to produce a result stored in the `value` property.
     */
    handleChanges(entries) {
        this.value = this.callback(entries, this._observer);
    }
    hostConnected() {
        if (this._target) {
            this.observe(this._target);
        }
    }
    hostDisconnected() {
        this.disconnect();
    }
    async hostUpdated() {
        // Eagerly deliver any changes that happened during update.
        const pendingRecords = this._observer.takeRecords();
        if (pendingRecords.length) {
            this.handleChanges(pendingRecords);
        }
    }
    /**
     * Observe the target element. The controller's `target` is automatically
     * observed when the host connects.
     * @param target Element to observe
     */
    observe(target) {
        // Note, this will always trigger the callback since the initial
        // intersection state is reported.
        this._observer.observe(target);
        this._unobservedUpdate = true;
    }
    /**
     * Disconnects the observer. This is done automatically when the host
     * disconnects.
     */
    disconnect() {
        this._observer.disconnect();
    }
}
