// @spectrum-web-components/sidenav/src/SidenavItem.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import { ifDefined } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/directives.js';
import { LikeAnchor } from '/web-component-designer-demo/node_modules/@spectrum-web-components/shared/src/like-anchor.js';
import { Focusable } from '/web-component-designer-demo/node_modules/@spectrum-web-components/shared/src/focusable.js';
import sidenavItemStyles from './sidenav-item.css.js';
/**
 * @element sp-sidenav-item
 *
 * @slot - the Sidenav Items to display as children of this item
 */
export class SideNavItem extends LikeAnchor(Focusable) {
    constructor() {
        super(...arguments);
        this.value = undefined;
        this.selected = false;
        this.expanded = false;
    }
    static get styles() {
        return [sidenavItemStyles];
    }
    get parentSideNav() {
        if (!this._parentSidenav) {
            this._parentSidenav = this.closest('sp-sidenav');
        }
        return this._parentSidenav;
    }
    get hasChildren() {
        return !!this.querySelector('sp-sidenav-item');
    }
    get depth() {
        let depth = 0;
        let element = this.parentElement;
        while (element instanceof SideNavItem) {
            depth++;
            element = element.parentElement;
        }
        return depth;
    }
    handleSideNavSelect(event) {
        this.selected = event.target === this;
    }
    handleClick(event) {
        if (!this.href && event) {
            event.preventDefault();
        }
        if (!this.disabled) {
            if (this.hasChildren) {
                this.expanded = !this.expanded;
            }
            else if (this.value) {
                this.announceSelected(this.value);
            }
        }
    }
    announceSelected(value) {
        const selectDetail = {
            value,
        };
        const selectionEvent = new CustomEvent('sidenav-select', {
            bubbles: true,
            composed: true,
            detail: selectDetail,
        });
        this.dispatchEvent(selectionEvent);
    }
    click() {
        this.handleClick();
    }
    get focusElement() {
        return this.shadowRoot.querySelector('#item-link');
    }
    update(changes) {
        if (!this.hasAttribute('slot')) {
            this.slot = 'descendant';
        }
        super.update(changes);
    }
    render() {
        return html `
            <a
                href=${this.href || '#'}
                target=${ifDefined(this.target)}
                download=${ifDefined(this.download)}
                rel=${ifDefined(this.rel)}
                data-level="${this.depth}"
                @click="${this.handleClick}"
                id="item-link"
                aria-current=${ifDefined(this.selected && this.href ? 'page' : undefined)}
            >
                <slot name="icon"></slot>
                ${this.label}
                <slot></slot>
            </a>
            ${this.expanded
            ? html `
                      <slot name="descendant"></slot>
                  `
            : html ``}
        `;
    }
    updated(changes) {
        if (this.hasChildren && this.expanded && !this.selected) {
            this.focusElement.tabIndex = -1;
        }
        super.updated(changes);
    }
    connectedCallback() {
        super.connectedCallback();
        this.startTrackingSelection();
    }
    disconnectedCallback() {
        this.stopTrackingSelection();
        super.disconnectedCallback();
    }
    async startTrackingSelection() {
        const parentSideNav = this.parentSideNav;
        if (parentSideNav) {
            await parentSideNav.updateComplete;
            parentSideNav.startTrackingSelectionForItem(this);
            this.selected =
                this.value != null && this.value === parentSideNav.value;
        }
    }
    stopTrackingSelection() {
        const parentSideNav = this.parentSideNav;
        if (parentSideNav) {
            parentSideNav.stopTrackingSelectionForItem(this);
        }
        this._parentSidenav = undefined;
    }
}
__decorate([
    property()
], SideNavItem.prototype, "value", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], SideNavItem.prototype, "selected", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], SideNavItem.prototype, "expanded", void 0);
