// @node-projects/web-component-designer/dist/elements/widgets/treeView/treeViewExtended.js
import { BaseCustomWebComponent, css, html } from "../../controls/BaseCustomWebComponent.js";

let TreeViewExtended =
/** @class */
(() => {
  class TreeViewExtended extends BaseCustomWebComponent {
    constructor() {
      super();
      this._treeDiv = document.createElement('div');

      this._treeDiv.setAttribute('id', 'tree');

      this.shadowRoot.appendChild(this._treeDiv);
    }

    async ready() {
      this._treeDiv.classList.add('fancytree-connectors');

      $(this._treeDiv).fancytree({
        icon: false,
        extensions: ['dnd5'],
        source: []
      }); //@ts-ignore

      this._tree = $.ui.fancytree.getTree(this._treeDiv);
    }

    createTree(rootItem) {
      this._recomputeTree(rootItem.element, null);
    }

    _recomputeTree(rootElement, activeElement) {
      this._tree.getRootNode().removeChildren();

      this._index = 0;
      this.items = this._getChildren(rootElement, null);

      this._tree.expandAll();

      this._highlight(activeElement);
    }

    _getChildren(element, currentNode) {
      if (currentNode == null) {
        currentNode = this._tree.getRootNode();
      }

      const data = {
        tag: element.tagName.toLowerCase(),
        id: element.id ? '#' + element.id : '',
        index: this._index,
        ref: element
      };
      this._index++;
      let nodes = [data];
      const newNode = currentNode.addChildren({
        title: data.tag + " " + data.id,
        folder: element.children.length > 0 ? true : false
      });

      for (let i = 0; i < element.children.length; i++) {
        let childElement = element.children[i];
        nodes = nodes.concat(this._getChildren(childElement, newNode));
      }

      return nodes;
    }

    _highlight(activeElement) {
      if (activeElement != null) {
        this._tree.getRootNode().getChildren().forEach(function (node) {
          //@ts-ignore
          if (node.ref === activeElement) {
            node.setSelected(true);
          }
        });
      }
    }

  }

  TreeViewExtended.style = css`
      span.drag-source {
        border: 1px solid grey;
        border-radius: 3px;
        padding: 2px;
        background-color: silver;
      }

      span.fancytree-node.fancytree-drag-source {
        outline: 1px dotted grey;
      }
      span.fancytree-node.fancytree-drop-accept {
        outline: 1px dotted green;
      }
      span.fancytree-node.fancytree-drop-reject {
        outline: 1px dotted red;
      }
    `;
  TreeViewExtended.template = html`
      <link rel="stylesheet" href="/node_modules/jquery.fancytree/dist/skin-xp/ui.fancytree.css">
    `;
  return TreeViewExtended;
})();

export { TreeViewExtended };
customElements.define('node-projects-tree-view-extended', TreeViewExtended);
