// @spectrum-web-components/search/src/Search.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property, query, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import { ifDefined } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/directives.js';
import { Textfield } from '/web-component-designer-demo/node_modules/@spectrum-web-components/textfield/./src/index.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/button/sp-clear-button.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/icons-workflow/icons/sp-icon-magnify.js';
import searchStyles from './search.css.js';
const stopPropagation = (event) => event.stopPropagation();
/**
 * @element sp-search
 * @slot help-text - default or non-negative help text to associate to your form element
 * @slot negative-help-text - negative help text to associate to your form element when `invalid`
 *
 * @fires submit - The search form has been submitted.
 */
export class Search extends Textfield {
    constructor() {
        super(...arguments);
        this.action = '';
        this.label = 'Search';
        this.placeholder = 'Search';
    }
    static get styles() {
        return [...super.styles, searchStyles];
    }
    handleSubmit(event) {
        const applyDefault = this.dispatchEvent(new Event('submit', {
            cancelable: true,
            bubbles: true,
        }));
        if (!applyDefault) {
            event.preventDefault();
        }
    }
    handleKeydown(event) {
        const { code } = event;
        if (!this.value || code !== 'Escape') {
            return;
        }
        this.reset();
    }
    async reset() {
        this.value = '';
        await this.updateComplete;
        this.focusElement.dispatchEvent(new InputEvent('input', {
            bubbles: true,
            composed: true,
        }));
        // The native `change` event on an `input` element is not composed,
        // so this synthetic replication of a `change` event must not be
        // either as the `Textfield` baseclass should only need to handle
        // the native variant of this interaction.
        this.focusElement.dispatchEvent(new InputEvent('change', {
            bubbles: true,
        }));
    }
    renderField() {
        return html `
            <form
                action=${this.action}
                id="form"
                method=${ifDefined(this.method)}
                @submit=${this.handleSubmit}
                @reset=${this.reset}
                @keydown=${this.handleKeydown}
            >
                <sp-icon-magnify
                    class="icon magnifier icon-workflow"
                ></sp-icon-magnify>
                ${super.renderField()}
                ${this.value
            ? html `
                          <sp-clear-button
                              id="button"
                              label="Reset"
                              tabindex="-1"
                              type="reset"
                              @keydown=${stopPropagation}
                          ></sp-clear-button>
                      `
            : html ``}
            </form>
        `;
    }
    firstUpdated(changedProperties) {
        super.firstUpdated(changedProperties);
        this.inputElement.setAttribute('type', 'search');
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        this.multiline = false;
    }
}
__decorate([
    property()
], Search.prototype, "action", void 0);
__decorate([
    property()
], Search.prototype, "label", void 0);
__decorate([
    property()
], Search.prototype, "method", void 0);
__decorate([
    property()
], Search.prototype, "placeholder", void 0);
__decorate([
    query('#form')
], Search.prototype, "form", void 0);
