// @node-projects/web-component-designer/dist/elements/services/propertiesService/services/CssCurrentPropertiesService.js
import { PropertyType } from '../PropertyType.js';
import { RefreshMode } from '../IPropertiesService.js';
import { CommonPropertiesService } from './CommonPropertiesService.js';
import cssProperties from './CssProperties.json' assert { type: 'json' };
import { ValueType } from '../ValueType.js';
import { NodeType } from '../../../item/NodeType.js';
const localName = '&lt;local&gt;';
export class CssCurrentPropertiesService extends CommonPropertiesService {
    getRefreshMode(designItem) {
        return RefreshMode.fullOnValueChange;
    }
    constructor() {
        super();
        this.name = 'styles';
    }
    isHandledElement(designItem) {
        return true;
    }
    getProperty(designItem, name) {
        return { name: name, type: 'string', service: this, propertyType: PropertyType.cssValue };
    }
    getProperties(designItem) {
        if (!designItem || designItem.nodeType != NodeType.Element)
            return [];
        let styles = designItem.getAllStyles();
        let arr = styles.map(x => ({
            name: x.selector ?? localName, description: x.stylesheetName ?? '', properties: [
                ...x.declarations.map(y => ({
                    name: y.name,
                    renamable: true,
                    type: cssProperties[y.name]?.type ?? 'string',
                    values: cssProperties[y.name]?.values ? [...cssProperties[y.name]?.values, 'initial', 'inherit', 'unset'] : ['initial', 'inherit', 'unset'],
                    service: this,
                    propertyType: PropertyType.cssValue,
                    styleRule: x,
                    styleDeclaration: y
                })),
                { name: '', type: 'addNew', service: this, propertyType: PropertyType.complex, styleRule: x }
            ]
        }));
        return arr;
    }
    setValue(designItems, property, value) {
        // No selector means local style, styleDeclaration is null means new property
        if (property.styleRule?.selector !== null && property.styleDeclaration) {
            designItems[0].instanceServiceContainer.stylesheetService.updateDeclarationValue(property.styleDeclaration, value, false);
            this._notifyChangedProperty(designItems[0], property, value);
            return;
        }
        if (property.styleRule?.selector !== null && !property.styleDeclaration) {
            designItems[0].instanceServiceContainer.stylesheetService.insertDeclarationIntoRule(property.styleRule, property.name, value, false);
            this._notifyChangedProperty(designItems[0], property, value);
            return;
        }
        for (const d of designItems) {
            // Local style
            d.setStyle(property.name, value);
            //unkown css property names do not trigger the mutation observer of property grid, 
            //fixed by assinging stle again to the attribute
            d.element.setAttribute('style', d.element.getAttribute('style'));
        }
    }
    clearValue(designItems, property) {
        if (property.styleRule?.selector !== null && property.styleDeclaration) {
            designItems[0].instanceServiceContainer.stylesheetService.removeDeclarationFromRule(property.styleRule, property.styleDeclaration.name);
            return;
        }
        super.clearValue(designItems, property);
    }
    getValue(designItems, property) {
        if (property.styleRule?.selector && property.styleDeclaration)
            return property.styleDeclaration.value;
        return super.getValue(designItems, property);
    }
    getUnsetValue(designItems, property) {
        if (property.styleRule?.selector && property.styleDeclaration)
            return property.styleDeclaration.value;
        return super.getUnsetValue(designItems, property);
    }
    isSet(designItems, property) {
        if (property.styleRule?.selector && property.styleDeclaration) {
            if (designItems[0].hasStyle(property.name))
                return ValueType.none;
            //TODO: we need to check if this is the dec. with the highest specifity
            return ValueType.all;
        }
        return super.isSet(designItems, property);
    }
}
