// @node-projects/web-component-designer/dist/elements/services/propertiesService/services/NativeElementsPropertiesService.js
import { ValueType } from "../ValueType.js";
export class NativeElementsPropertiesService {
  constructor() {
    //@ts-ignore
    this.inputProperties = [{
      name: "type",
      type: "list",
      values: ["text", "number", "button", "checkbox", "color", "date", "datetime-local", "email", "file", "hidden", "image", "month", "password", "radio", "range", "reset", "search", "submit", "tel", "time", "url", "week"],
      service: this,
      defaultValue: "text"
    }, {
      name: "value",
      type: "string",
      service: this
    }, {
      name: "checked",
      type: "boolean",
      service: this
    }, {
      name: "min",
      type: "number",
      service: this
    }, {
      name: "max",
      type: "number",
      service: this
    }];
    this.buttonProperties = [{
      name: "type",
      type: "list",
      values: ["button", "submit", "reset"],
      service: this,
      defaultValue: "button"
    }, {
      name: "value",
      type: "string",
      service: this
    }, {
      name: "disabled",
      type: "boolean",
      service: this
    }];
    this.name = "native";
  }

  isHandledElement(designItem) {
    switch (designItem.element.localName) {
      case 'input':
      case 'button':
      case 'a':
      case 'div':
        return true;
    }

    return false;
  }

  getProperties(designItem) {
    if (!this.isHandledElement(designItem)) return null;

    switch (designItem.element.localName) {
      case 'input':
        return this.inputProperties;

      case 'button':
        return this.buttonProperties;

      case 'a':
      case 'div':
    }

    return null;
  }

  setValue(designItems, property, value) {
    for (let d of designItems) {
      if (property.type == 'boolean' && !value) {
        d.removeAttribute(property.name);
      } else d.setAttribute(property.name, value);
    }
  }

  clearValue(designItems, property) {
    for (let d of designItems) {
      d.removeAttribute(property.name);
    }
  }

  isSet(designItems, property) {
    let all = true;
    let some = false;

    if (designItems != null && designItems.length !== 0) {
      let attributeName = property.name;
      designItems.forEach(x => {
        let has = x.attributes.has(attributeName);
        all = all && has;
        some = some || has;
      });
    } else return ValueType.none;

    return all ? ValueType.all : some ? ValueType.some : ValueType.none;
  }

  getValue(designItems, property) {
    if (designItems != null && designItems.length !== 0) {
      let attributeName = property.name;
      let lastValue = designItems[0].attributes.get(attributeName);

      for (const x of designItems) {
        let value = x.attributes.get(attributeName);

        if (value != lastValue) {
          lastValue = null;
          break;
        }
      }

      return lastValue;
    }

    return null;
  }

  getUnsetValue(designItems, property) {
    return property.defaultValue;
  }

}
