// @spectrum-web-components/action-bar/src/ActionBar.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, SpectrumElement, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import actionBarStyles from './action-bar.css.js';
export const actionBarVariants = ['sticky', 'fixed'];
/**
 * @element sp-action-bar
 * @slot - Content to display with the Action Bar
 */
export class ActionBar extends SpectrumElement {
    constructor() {
        super(...arguments);
        /**
         * When `flexible` the action bar sizes itself to its content
         * rather than a specific width.
         *
         * @param {Boolean} flexible
         */
        this.flexible = false;
        this.open = false;
        this._variant = '';
    }
    static get styles() {
        return [actionBarStyles];
    }
    /**
     * The variant applies specific styling when set to `sticky` or `fixed`.
     * `variant` attribute is removed when not matching one of the above.
     *
     * @param {String} variant
     */
    set variant(variant) {
        if (variant === this.variant) {
            return;
        }
        if (actionBarVariants.includes(variant)) {
            this.setAttribute('variant', variant);
            this._variant = variant;
            return;
        }
        this.removeAttribute('variant');
        this._variant = '';
    }
    get variant() {
        return this._variant;
    }
    render() {
        return html `
            <sp-popover ?open=${this.open} id="popover">
                <slot></slot>
            </sp-popover>
        `;
    }
}
__decorate([
    property({ type: Boolean, reflect: true })
], ActionBar.prototype, "flexible", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], ActionBar.prototype, "open", void 0);
__decorate([
    property({ type: String, reflect: true })
], ActionBar.prototype, "variant", null);
