// @node-projects/web-component-designer/dist/elements/services/selectionService/SelectionService.js
import { TypedEvent } from '../../../basic/TypedEvent';
export class SelectionService {
    constructor() {
        this.selectedElements = [];
        this.onSelectionChanged = new TypedEvent();
    }
    setSelectedElements(designItems) {
        let oldSelectedElements = this.selectedElements;
        //f (oldSelectedElements && oldSelectedElements.length==2 && designItems!=null && designItems.length<2)
        //debugger;
        if (!designItems) {
            this.selectedElements = [];
            this.primarySelection = null;
        }
        else {
            this.selectedElements = designItems;
            if (designItems && designItems.length > 0)
                this.primarySelection = designItems[0];
            else
                this.primarySelection = null;
        }
        this.onSelectionChanged.emit({ selectedElements: this.selectedElements, oldSelectedElements: oldSelectedElements });
    }
    isSelected(designItem) {
        return this.selectedElements.indexOf(designItem) >= 0;
    }
}
