// @node-projects/acad-ts/dist/IO/DWG/CRC8StreamHandler.js
import { CRC } from './CRC.js';
export class CRC8StreamHandler {
    _data;
    _position = 0;
    Seed;
    get Length() { return this._data.length; }
    get Position() { return this._position; }
    set Position(value) { this._position = value; }
    constructor(data, seed) {
        this._data = data;
        this.Seed = seed;
    }
    Read(buffer, offset, count) {
        const nbytes = Math.min(count, this._data.length - this._position);
        const length = offset + nbytes;
        for (let index = offset; index < length; ++index) {
            buffer[index] = this._data[this._position + (index - offset)];
            this.Seed = CRC8StreamHandler.decode(this.Seed, buffer[index]);
        }
        this._position += nbytes;
        return nbytes;
    }
    Write(buffer, offset, count) {
        const length = offset + count;
        for (let index = offset; index < length; ++index) {
            this.Seed = CRC8StreamHandler.decode(this.Seed, buffer[index]);
        }
        for (let i = 0; i < count; i++) {
            this._data[this._position + i] = buffer[offset + i];
        }
        this._position += count;
    }
    static GetCRCValue(seed, buffer, startPos, endPos) {
        let currValue = seed;
        let index = startPos;
        let remaining = endPos;
        while (remaining-- > 0) {
            currValue = CRC8StreamHandler.decode(currValue, buffer[index]);
            index++;
        }
        return currValue;
    }
    static decode(key, value) {
        const index = value ^ (key & 0xFF);
        key = ((key >>> 8) ^ CRC.CrcTable[index]) & 0xFFFF;
        return key;
    }
}
