// @node-projects/acad-ts/dist/IO/DWG/DwgSectionIO.js
import { ACadVersion } from '../../ACadVersion.js';
import { NotificationEventArgs, NotificationType } from '../NotificationEventHandler.js';
export class DwgSectionIO {
    OnNotification = null;
    R13_14Only;
    R13_15Only;
    R2000Plus;
    R2004Pre;
    R2007Pre;
    R2004Plus;
    R2007Plus;
    R2010Plus;
    R2013Plus;
    R2018Plus;
    _version;
    constructor(version) {
        this._version = version;
        this.R13_14Only = version === ACadVersion.AC1014 || version === ACadVersion.AC1012;
        this.R13_15Only = version >= ACadVersion.AC1012 && version <= ACadVersion.AC1015;
        this.R2000Plus = version >= ACadVersion.AC1015;
        this.R2004Pre = version < ACadVersion.AC1018;
        this.R2007Pre = version <= ACadVersion.AC1021;
        this.R2004Plus = version >= ACadVersion.AC1018;
        this.R2007Plus = version >= ACadVersion.AC1021;
        this.R2010Plus = version >= ACadVersion.AC1024;
        this.R2013Plus = version >= ACadVersion.AC1027;
        this.R2018Plus = version >= ACadVersion.AC1032;
    }
    static CheckSentinel(actual, expected) {
        if (expected.length !== actual.length) {
            return false;
        }
        for (let i = 0; i < expected.length; i++) {
            if (actual[i] !== expected[i]) {
                return false;
            }
        }
        return true;
    }
    checkSentinel(sreader, expected) {
        const sn = sreader.readSentinel();
        if (!DwgSectionIO.CheckSentinel(sn, expected)) {
            this.notify(`Invalid section sentinel found in ${this.SectionName}`, NotificationType.Warning);
        }
    }
    notify(message, type, ex = null) {
        this.OnNotification?.(this, new NotificationEventArgs(message, type, ex));
    }
}
