// @node-projects/acad-ts/dist/IO/DWG/DwgWriter.js
import { ACadVersion } from '../../ACadVersion.js';
import { CadUtils } from '../../CadUtils.js';
import { CadNotSupportedException } from '../../Exceptions/CadNotSupportedException.js';
import { CadWriterBase } from '../CadWriterBase.js';
import { DwgFileHeader } from './FileHeaders/DwgFileHeader.js';
import './FileHeaders/DwgFileHeaderFactory.js';
import { DwgSectionDefinition } from './FileHeaders/DwgSectionDefinition.js';
import { DwgWriterConfiguration } from './DwgWriterConfiguration.js';
import { DwgFileHeaderWriterAC15 } from './DwgStreamWriters/DwgFileHeaderWriterAC15.js';
import { DwgFileHeaderWriterAC18 } from './DwgStreamWriters/DwgFileHeaderWriterAC18.js';
import { DwgHeaderWriter } from './DwgStreamWriters/DwgHeaderWriter.js';
import { DwgClassesWriter } from './DwgStreamWriters/DwgClassesWriter.js';
import { DwgObjectWriter } from './DwgStreamWriters/DwgObjectWriter.js';
import { DwgHandleWriter } from './DwgStreamWriters/DwgHandleWriter.js';
import { DwgAuxHeaderWriter } from './DwgStreamWriters/DwgAuxHeaderWriter.js';
import { DwgPreviewWriter } from './DwgStreamWriters/DwgPreviewWriter.js';
import { DwgAppInfoWriter } from './DwgStreamWriters/DwgAppInfodWriter.js';
import { DwgSummaryInfoWriter } from './DwgStreamWriters/DwgSummaryInfoWriter.js';
import './DwgStreamWriters/DwgStreamWriterFactory.js';
export class DwgWriter extends CadWriterBase {
    Preview = null;
    get _version() { return this._document.header.version; }
    _fileHeader;
    _fileHeaderWriter;
    _handlesMap = new Map();
    /** The actual number of bytes written to the output stream after Write() completes. */
    get bytesWritten() { return this._fileHeaderWriter?.bytesWritten ?? 0; }
    constructor(stream, document) {
        super(stream, document);
        this._fileHeader = DwgFileHeader.CreateFileHeader(this._version);
    }
    createDefaultConfiguration() {
        return new DwgWriterConfiguration();
    }
    Write() {
        super.Write();
        this.getFileHeaderWriter();
        this.writeHeader();
        this.writeClasses();
        this.writeSummaryInfo();
        this.writePreview();
        this.writeAppInfo();
        this.writeFileDepList();
        this.writeRevHistory();
        this.writeAuxHeader();
        this.writeObjects();
        this.writeObjFreeSpace();
        this.writeTemplate();
        this.writeHandles();
        this._fileHeaderWriter.writeFile();
    }
    Dispose() {
        // No-op in TS
    }
    static WriteToStream(stream, document, configuration = null, notification = null) {
        const writer = new DwgWriter(stream, document);
        if (configuration) {
            writer.Configuration = configuration;
        }
        writer.OnNotification = notification;
        writer.Write();
        writer.Dispose();
    }
    getFileHeaderWriter() {
        switch (this._document.header.version) {
            case ACadVersion.MC0_0:
            case ACadVersion.AC1_2:
            case ACadVersion.AC1_4:
            case ACadVersion.AC1_50:
            case ACadVersion.AC2_10:
            case ACadVersion.AC1002:
            case ACadVersion.AC1003:
            case ACadVersion.AC1004:
            case ACadVersion.AC1006:
            case ACadVersion.AC1009:
            case ACadVersion.AC1012:
                throw new CadNotSupportedException(this._document.header.version);
            case ACadVersion.AC1014:
            case ACadVersion.AC1015:
                this._fileHeaderWriter = new DwgFileHeaderWriterAC15(new Uint8Array(this._stream), this._encoding, this._document);
                break;
            case ACadVersion.AC1018:
                this._fileHeaderWriter = new DwgFileHeaderWriterAC18(new Uint8Array(this._stream), this._encoding, this._document);
                break;
            case ACadVersion.AC1021:
                // AC1021 writer currently uses AC18-compatible section/page layout.
                this._fileHeaderWriter = new DwgFileHeaderWriterAC18(new Uint8Array(this._stream), this._encoding, this._document);
                break;
            case ACadVersion.AC1024:
            case ACadVersion.AC1027:
            case ACadVersion.AC1032:
                this._fileHeaderWriter = new DwgFileHeaderWriterAC18(new Uint8Array(this._stream), this._encoding, this._document);
                break;
            default:
                throw new CadNotSupportedException();
        }
    }
    writeHeader() {
        const stream = new Uint8Array(0);
        const writer = new DwgHeaderWriter(stream, this._document, this._encoding);
        writer.OnNotification = (sender, e) => this.triggerNotification(sender, e);
        writer.write();
        const data = new Uint8Array(writer.startWriterStream).slice(0, writer.bytesWritten);
        this._fileHeaderWriter.addSection(DwgSectionDefinition.Header, data, true);
    }
    writeClasses() {
        const stream = new Uint8Array(0);
        const writer = new DwgClassesWriter(stream, this._document, this._encoding);
        writer.write();
        const data = new Uint8Array(writer.startWriterStream).slice(0, writer.bytesWritten);
        this._fileHeaderWriter.addSection(DwgSectionDefinition.Classes, data, true);
    }
    writeSummaryInfo() {
        if (this._fileHeader.AcadVersion < ACadVersion.AC1018)
            return;
        if (this._document.summaryInfo) {
            const stream = new Uint8Array(8192);
            const writer = new DwgSummaryInfoWriter(this._version, stream);
            writer.write(this._document.summaryInfo);
            const data = new Uint8Array(writer.writerStream).slice(0, writer.bytesWritten);
            this._fileHeaderWriter.addSection(DwgSectionDefinition.SummaryInfo, data, false, 0x100);
        }
        else {
            this._fileHeaderWriter.addSection(DwgSectionDefinition.SummaryInfo, new Uint8Array(0), false, 0x100);
        }
    }
    writePreview() {
        const stream = new Uint8Array(4096);
        const writer = new DwgPreviewWriter(this._version, stream);
        if (this.Preview && this.Preview.code !== 0) {
            writer.writePreview(this.Preview, 0);
        }
        else {
            writer.write();
        }
        const data = new Uint8Array(writer.writerStream).slice(0, writer.bytesWritten);
        this._fileHeaderWriter.addSection(DwgSectionDefinition.Preview, data, false, 0x400);
    }
    writeAppInfo() {
        if (this._fileHeader.AcadVersion < ACadVersion.AC1018)
            return;
        const stream = new Uint8Array(4096);
        const writer = new DwgAppInfoWriter(this._version, stream);
        writer.write();
        const data = new Uint8Array(writer.writerStream).slice(0, writer.bytesWritten);
        this._fileHeaderWriter.addSection(DwgSectionDefinition.AppInfo, data, false, 0x80);
    }
    writeFileDepList() {
        if (this._fileHeader.AcadVersion < ACadVersion.AC1018)
            return;
        const data = new Uint8Array(8);
        const view = new DataView(data.buffer);
        // UInt32: Feature count
        view.setUint32(0, 0, true);
        // UInt32: File count
        view.setUint32(4, 0, true);
        this._fileHeaderWriter.addSection(DwgSectionDefinition.FileDepList, data, false, 0x80);
    }
    writeRevHistory() {
        if (this._fileHeader.AcadVersion < ACadVersion.AC1018)
            return;
        const data = new Uint8Array(12);
        const view = new DataView(data.buffer);
        view.setUint32(0, 0, true);
        view.setUint32(4, 0, true);
        view.setUint32(8, 0, true);
        this._fileHeaderWriter.addSection(DwgSectionDefinition.RevHistory, data, true);
    }
    writeObjects() {
        const stream = new Uint8Array(0);
        const writer = new DwgObjectWriter(stream, this._document, this._encoding, this.Configuration.WriteXRecords, this.Configuration.WriteXData, this.Configuration.WriteShapes);
        writer.OnNotification = (sender, e) => this.triggerNotification(sender, e);
        writer.write();
        this._handlesMap = writer.Map;
        this._fileHeaderWriter.addSection(DwgSectionDefinition.AcDbObjects, writer.getWrittenData(), true);
    }
    writeObjFreeSpace() {
        const data = new Uint8Array(37);
        const view = new DataView(data.buffer);
        let offset = 0;
        // Int32: 0
        view.setInt32(offset, 0, true);
        offset += 4;
        // UInt32: Approximate number of objects (handle count)
        view.setUint32(offset, this._handlesMap.size, true);
        offset += 4;
        // Julian datetime (8 bytes)
        const dateToUse = this._version >= ACadVersion.AC1015
            ? this._document.header.universalUpdateDateTime
            : this._document.header.updateDateTime;
        const { jdate, milliseconds } = CadUtils.dateToJulian(dateToUse);
        view.setInt32(offset, jdate, true);
        offset += 4;
        view.setInt32(offset, milliseconds, true);
        offset += 4;
        // UInt32: Offset of objects section in the stream
        view.setUint32(offset, 0, true);
        offset += 4;
        // UInt8: Number of 64-bit values that follow (ODA writes 4)
        data[offset] = 4;
        offset += 1;
        // 4 x UInt32 values (ODA standard)
        view.setUint32(offset, 0x00000032, true);
        offset += 4;
        view.setUint32(offset, 0x00000000, true);
        offset += 4;
        view.setUint32(offset, 0x00000064, true);
        offset += 4;
        view.setUint32(offset, 0x00000000, true);
        offset += 4;
        this._fileHeaderWriter.addSection(DwgSectionDefinition.ObjFreeSpace, data, true);
    }
    writeTemplate() {
        const data = new Uint8Array(4);
        const view = new DataView(data.buffer);
        // Int16: Template description string length (always 0)
        view.setInt16(0, 0, true);
        // UInt16: MEASUREMENT system variable (0 = English, 1 = Metric)
        view.setUint16(2, this._document.header.measurementUnits, true);
        this._fileHeaderWriter.addSection(DwgSectionDefinition.Template, data, true);
    }
    writeHandles() {
        const writer = new DwgHandleWriter(this._version, this._handlesMap);
        const data = writer.write(this._fileHeaderWriter.HandleSectionOffset);
        this._fileHeaderWriter.addSection(DwgSectionDefinition.Handles, data, true);
    }
    writeAuxHeader() {
        const stream = new Uint8Array(4096);
        const writer = new DwgAuxHeaderWriter(stream, this._encoding, this._document.header);
        writer.write();
        const data = new Uint8Array(writer.writerStream).slice(0, writer.bytesWritten);
        this._fileHeaderWriter.addSection(DwgSectionDefinition.AuxHeader, data, true);
    }
}
