// @papyrs/stylo/dist/components/config.store.js
import { c as createStore } from './index2.js';
import { h as h1, a as h2, b as h3, u as ul, i as img, c as code, d as hr } from './ul.plugin.js';
import { D as DEFAULT_PALETTE } from './palette.js';

const DEFAULT_PLUGINS = [h1, h2, h3, ul, img, code, hr];
const DEFAULT_TOOLBAR = {
  palette: DEFAULT_PALETTE,
  command: 'native',
  style: {
    list: false,
    align: true,
    fontSize: true,
    backgroundColor: true
  }
};
const DEFAULT_PLACEHOLDERS = ['div', 'p', 'span'];
const DEFAULT_TEXT_PARAGRAPHS = ['h1', 'h2', 'h3', 'div', 'p'];
const DEFAULT_EXCLUDE_ATTRIBUTES = [
  'placeholder',
  'data-gramm',
  'class',
  'spellcheck',
  'contenteditable'
];
const { state, onChange } = createStore({
  plugins: DEFAULT_PLUGINS,
  toolbar: DEFAULT_TOOLBAR,
  placeholders: DEFAULT_PLACEHOLDERS,
  textParagraphs: DEFAULT_TEXT_PARAGRAPHS,
  menus: undefined,
  excludeAttributes: DEFAULT_EXCLUDE_ATTRIBUTES
});
const configStore = { state, onChange };

export { DEFAULT_PLUGINS as D, DEFAULT_TOOLBAR as a, DEFAULT_PLACEHOLDERS as b, configStore as c, DEFAULT_TEXT_PARAGRAPHS as d, DEFAULT_EXCLUDE_ATTRIBUTES as e };
