// @papyrs/stylo/dist/components/list.js
import { proxyCustomElement, HTMLElement, createEvent, h, Fragment } from '@stencil/core/internal/client';
import { e as configStore } from './config.store.js';
import { i as i18n } from './i18n.store.js';
import { r as renderIcon } from './icon.utils.js';
import { d as toHTMLElement } from './paragraph.utils.js';

const listCss = ":host{display:flex;flex-direction:column;--light:#f4f5f8;--light-contrast:#000000;--medium:#6e6d6f;--medium-contrast:#000000;--black:#000000;--black-contrast:#ffffff;--white:#ffffff;--white-contrast:#000000;--highlight:#3880ff;--highlight-rgb:56, 128, 255;--highlight-contrast:#ffffff;--background-active:var(--stylo-list-background-active, var(--light));--border-active:var(--stylo-list-border-active, 1px solid var(--background-active));--color-active:var(--stylo-list-color-active, var(--light-contrast))}button{display:flex;position:relative;isolation:isolate;overflow:hidden;background:transparent;border:none;outline:none;cursor:pointer;transition:color 0.25s ease-out, background 0.25s ease-out, transform 0.15s ease-out;justify-content:flex-start;align-items:center;font-size:1rem;color:var(--stylo-toolbar-button-color, black)}button:active{box-shadow:none;transform:translateX(1px) translateY(1px)}button:hover,button:focus{background:var(--background-active);color:var(--color-active)}svg,span.placeholder{width:1.4rem;height:1.4rem;border:var(--border-active);border-radius:var(--stylo-border, 4px);margin:4px 12px 4px 8px;background:var(--white)}svg{padding:8px}span.placeholder{padding:10px 8px 6px;font-size:0.8rem}span.placeholder::first-letter{font-size:125%}";

const List = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.applyPlugin = createEvent(this, "applyPlugin", 7);
    this.cancelPlugins = createEvent(this, "cancelPlugins", 7);
  }
  emitPlugin($event, plugin) {
    $event.stopPropagation();
    this.applyPlugin.emit(plugin);
  }
  onKeyDown($event) {
    const { code } = $event;
    if (['Escape'].includes(code)) {
      this.cancelPlugins.emit();
      return;
    }
    if (['ArrowDown'].includes(code)) {
      this.focusNext();
      return;
    }
    if (['ArrowUp'].includes(code)) {
      this.focusPrevious();
    }
    if (['Enter'].includes(code)) {
      $event.stopPropagation();
    }
  }
  async focusFirstButton() {
    var _a;
    this.focusButton = this.el.shadowRoot.querySelector('button');
    (_a = this.focusButton) === null || _a === void 0 ? void 0 : _a.focus();
  }
  focusNext() {
    var _a, _b;
    this.focusButton = toHTMLElement((_a = (this.focusButton || this.el.shadowRoot.firstElementChild)) === null || _a === void 0 ? void 0 : _a.nextElementSibling);
    (_b = this.focusButton) === null || _b === void 0 ? void 0 : _b.focus();
  }
  focusPrevious() {
    var _a, _b;
    this.focusButton = toHTMLElement((_a = (this.focusButton || this.el.shadowRoot.lastElementChild)) === null || _a === void 0 ? void 0 : _a.previousElementSibling);
    (_b = this.focusButton) === null || _b === void 0 ? void 0 : _b.focus();
  }
  render() {
    return (h(Fragment, null, configStore.state.plugins.map((plugin) => this.renderPlugin(plugin))));
  }
  renderPlugin(plugin) {
    var _a, _b;
    const { text, icon: iconSrc } = plugin;
    const icon = renderIcon(iconSrc);
    return (h("button", Object.assign({ onClick: ($event) => this.emitPlugin($event, plugin) }, (icon === undefined && { innerHTML: iconSrc })), icon, (_b = (_a = i18n.state.plugins[text]) !== null && _a !== void 0 ? _a : i18n.state.custom[text]) !== null && _b !== void 0 ? _b : text));
  }
  get el() { return this; }
  static get style() { return listCss; }
}, [1, "stylo-list", {
    "focusFirstButton": [64]
  }, [[1, "keydown", "onKeyDown"]]]);
function defineCustomElement() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["stylo-list"];
  components.forEach(tagName => { switch (tagName) {
    case "stylo-list":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, List);
      }
      break;
  } });
}
defineCustomElement();

export { List as L, defineCustomElement as d };
