// @papyrs/stylo/dist/components/toolbar.js
var ToolbarActions;
(function (ToolbarActions) {
  ToolbarActions[ToolbarActions["STYLE"] = 0] = "STYLE";
  ToolbarActions[ToolbarActions["LINK"] = 1] = "LINK";
  ToolbarActions[ToolbarActions["COLOR"] = 2] = "COLOR";
  ToolbarActions[ToolbarActions["ALIGNMENT"] = 3] = "ALIGNMENT";
  ToolbarActions[ToolbarActions["LIST"] = 4] = "LIST";
  ToolbarActions[ToolbarActions["FONT_SIZE"] = 5] = "FONT_SIZE";
  ToolbarActions[ToolbarActions["BACKGROUND_COLOR"] = 6] = "BACKGROUND_COLOR";
})(ToolbarActions || (ToolbarActions = {}));
var ToolbarList;
(function (ToolbarList) {
  ToolbarList["ORDERED"] = "insertOrderedList";
  ToolbarList["UNORDERED"] = "insertUnorderedList";
})(ToolbarList || (ToolbarList = {}));
var ToolbarAlign;
(function (ToolbarAlign) {
  ToolbarAlign["LEFT"] = "left";
  ToolbarAlign["CENTER"] = "center";
  ToolbarAlign["RIGHT"] = "right";
})(ToolbarAlign || (ToolbarAlign = {}));
var ToolbarFontSize;
(function (ToolbarFontSize) {
  ToolbarFontSize["X_SMALL"] = "1";
  ToolbarFontSize["SMALL"] = "2";
  ToolbarFontSize["MEDIUM"] = "3";
  ToolbarFontSize["LARGE"] = "4";
  ToolbarFontSize["X_LARGE"] = "5";
  ToolbarFontSize["XX_LARGE"] = "6";
  ToolbarFontSize["XXX_LARGE"] = "7";
})(ToolbarFontSize || (ToolbarFontSize = {}));

export { ToolbarActions as T, ToolbarList as a, ToolbarAlign as b, ToolbarFontSize as c };
