// @spectrum-web-components/split-button/stories/split-button-accent-more.stories.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { ElementSizes } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { args, argTypes, renderSplitButtonSet, splitbutton } from './index.js';
import '../sp-split-button.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/menu/sp-menu-item.js';
const variant = 'accent';
const type = 'more';
export default {
    title: 'Split Button/Accent/More',
    component: 'sp-split-button',
    args: Object.assign(Object.assign({}, args), { variant,
        type }),
    argTypes,
};
export const s = (args) => renderSplitButtonSet(args);
s.args = { size: ElementSizes.s };
export const m = (args) => renderSplitButtonSet(args);
m.args = { size: ElementSizes.m };
export const l = (args) => renderSplitButtonSet(args);
l.args = { size: ElementSizes.l };
export const XL = (args) => renderSplitButtonSet(args);
XL.args = { size: ElementSizes.xl };
export const open = (args) => splitbutton(args);
open.args = { open: true };
