// @node-projects/acad-ts/dist/Entities/AecObjects/Wall.js
import { Entity } from '../Entity.js';
import { ObjectType } from '../../Types/ObjectType.js';
import { DxfFileToken } from '../../DxfFileToken.js';
import { DxfSubclassMarker } from '../../DxfSubclassMarker.js';
import { XYZ } from '../../Math/XYZ.js';
export var WallJustification;
(function (WallJustification) {
    WallJustification[WallJustification["Left"] = 0] = "Left";
    WallJustification[WallJustification["Center"] = 1] = "Center";
    WallJustification[WallJustification["Right"] = 2] = "Right";
    WallJustification[WallJustification["Baseline"] = 3] = "Baseline";
})(WallJustification || (WallJustification = {}));
export class Wall extends Entity {
    baseHeight = 0;
    binRecord = null;
    binRecordHandle = 0;
    cleanupGroup = null;
    cleanupGroupHandle = 0;
    endPoint = new XYZ();
    height = 0;
    justification = WallJustification.Baseline;
    length = 0;
    normal = new XYZ(0, 0, 1);
    rawData = null;
    startPoint = new XYZ();
    style = null;
    version = 0;
    width = 0;
    get objectName() { return DxfFileToken.EntityAecWall; }
    get objectType() { return ObjectType.UNLISTED; }
    get subclassMarker() { return DxfSubclassMarker.AecWall; }
    getBoundingBox() {
        throw new Error('Not implemented');
    }
    applyTransform(transform) { }
}
