// @node-projects/acad-ts/dist/Tables/Collections/BlockRecordsTable.js
import { ObjectType } from '../../Types/ObjectType.js';
import { DxfFileToken } from '../../DxfFileToken.js';
import { BlockRecord } from '../BlockRecord.js';
import { Table } from './Table.js';
export class BlockRecordsTable extends Table {
    get objectType() {
        return ObjectType.BLOCK_CONTROL_OBJ;
    }
    get objectName() {
        return DxfFileToken.TableBlockRecord;
    }
    get defaultEntries() {
        return [BlockRecord.ModelSpaceName, BlockRecord.PaperSpaceName];
    }
    constructor() {
        super();
    }
    createDefaultEntries() {
        if (!this.contains(BlockRecord.ModelSpaceName)) {
            this.add(BlockRecord.ModelSpace);
        }
        if (!this.contains(BlockRecord.PaperSpaceName)) {
            this.add(BlockRecord.PaperSpace);
        }
    }
    add(item) {
        if (item.isAnonymous && this.contains(item.name)) {
            const existing = this.tryGetValue(item.name);
            if (existing && existing === item) {
                throw new Error(`The BlockRecord with name ${item.name} has already been added.`);
            }
            item.name = this.createName(BlockRecord.AnonymousPrefix);
        }
        super.add(item);
    }
}
