// @node-projects/acad-ts/dist/Tables/Collections/LineTypesTable.js
import { ObjectType } from '../../Types/ObjectType.js';
import { DxfFileToken } from '../../DxfFileToken.js';
import { LineType } from '../LineType.js';
import { Table } from './Table.js';
export class LineTypesTable extends Table {
    get objectType() {
        return ObjectType.LTYPE_CONTROL_OBJ;
    }
    get objectName() {
        return DxfFileToken.TableLinetype;
    }
    get byLayer() {
        return this.get(LineType.ByLayerName);
    }
    get byBlock() {
        return this.get(LineType.ByBlockName);
    }
    get continuous() {
        return this.get(LineType.ContinuousName);
    }
    get defaultEntries() {
        return [LineType.ByLayerName, LineType.ByBlockName, LineType.ContinuousName];
    }
    constructor() {
        super();
    }
    createDefaultEntries() {
        if (!this.contains(LineType.ByLayerName)) {
            this.add(LineType.ByLayer);
        }
        if (!this.contains(LineType.ByBlockName)) {
            this.add(LineType.ByBlock);
        }
        if (!this.contains(LineType.ContinuousName)) {
            this.add(LineType.Continuous);
        }
    }
}
