// @node-projects/acad-ts/dist/Tables/Collections/Table.js
import { CadObject } from '../../CadObject.js';
import { CollectionChangedEventArgs } from '../../CollectionChangedEventArgs.js';
import { DxfFileToken } from '../../DxfFileToken.js';
import { DxfSubclassMarker } from '../../DxfSubclassMarker.js';
export class Table extends CadObject {
    onAdd = null;
    onRemove = null;
    get count() {
        return this.entries.size;
    }
    get objectName() {
        return DxfFileToken.TableEntry;
    }
    get subclassMarker() {
        return DxfSubclassMarker.Table;
    }
    entries = new Map();
    constructor() {
        super();
    }
    add(item) {
        if (!item.name || !item.name.trim()) {
            item.name = this.createName('unnamed');
        }
        this.addEntry(item.name, item);
    }
    addRange(items) {
        for (const item of items) {
            this.add(item);
        }
    }
    tryAdd(item) {
        const existing = this.tryGetValue(item.name);
        if (existing !== undefined) {
            return existing;
        }
        else {
            this.add(item);
            return item;
        }
    }
    contains(key) {
        return this.entries.has(key.toUpperCase());
    }
    createDefaultEntries() {
        for (const entry of this.defaultEntries) {
            if (this.contains(entry))
                continue;
            // TODO: Activator.CreateInstance equivalent - subclasses should override
        }
    }
    [Symbol.iterator]() {
        return this.entries.values()[Symbol.iterator]();
    }
    remove(key) {
        if (this.defaultEntries.some(d => d.toUpperCase() === key.toUpperCase())) {
            return null;
        }
        const upperKey = key.toUpperCase();
        const item = this.entries.get(upperKey);
        if (item) {
            this.entries.delete(upperKey);
            item.owner = null;
            this.onRemove?.(this, new CollectionChangedEventArgs(item));
            item.onNameChanged = null;
            return item;
        }
        return null;
    }
    tryGetValue(key) {
        return this.entries.get(key.toUpperCase());
    }
    addEntry(key, item) {
        this.entries.set(key.toUpperCase(), item);
        item.owner = this;
        item.onNameChanged = (sender, e) => {
            this.onEntryNameChanged(sender, e);
        };
        this.onAdd?.(this, new CollectionChangedEventArgs(item));
    }
    addHandlePrefix(item) {
        item.owner = this;
        item.onNameChanged = (sender, e) => {
            this.onEntryNameChanged(sender, e);
        };
        this.onAdd?.(this, new CollectionChangedEventArgs(item));
        const key = `${item.handle}:${item.name}`;
        this.entries.set(key.toUpperCase(), item);
    }
    createName(prefix) {
        let i = 0;
        while (this.entries.has(`${prefix}${i}`.toUpperCase())) {
            i++;
        }
        return `${prefix}${i}`;
    }
    onEntryNameChanged(sender, e) {
        if (this.defaultEntries.some(d => d.toUpperCase() === e.oldName.toUpperCase())) {
            throw new Error(`The name ${e.oldName} belongs to a default entry.`);
        }
        const entry = this.entries.get(e.oldName.toUpperCase());
        if (entry) {
            this.entries.set(e.newName.toUpperCase(), entry);
            this.entries.delete(e.oldName.toUpperCase());
        }
    }
    get(name) {
        const entry = this.entries.get(name.toUpperCase());
        if (entry === undefined) {
            throw new Error(`Entry '${name}' not found in table.`);
        }
        return entry;
    }
}
