// @spectrum-web-components/avatar/src/Avatar.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, SpectrumElement, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import { ifDefined } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/directives.js';
import avatarStyles from './avatar.css.js';
const validSizes = [50, 75, 100, 200, 300, 400, 500, 600, 700];
/**
 * @element sp-avatar
 */
export class Avatar extends SpectrumElement {
    constructor() {
        super(...arguments);
        this.label = '';
        this.src = '';
        this._size = 100;
    }
    static get styles() {
        return [avatarStyles];
    }
    get size() {
        return this._size || 100;
    }
    set size(value) {
        const defaultSize = 100;
        const size = value;
        const validSize = (validSizes.includes(size) ? size : defaultSize);
        if (validSize) {
            this.setAttribute('size', `${validSize}`);
        }
        if (this._size === validSize) {
            return;
        }
        const oldSize = this._size;
        this._size = validSize;
        this.requestUpdate('size', oldSize);
    }
    render() {
        return html `
            <img
                class="image"
                alt=${ifDefined(this.label || undefined)}
                src=${this.src}
            />
        `;
    }
    firstUpdated(changes) {
        super.firstUpdated(changes);
        if (!this.hasAttribute('size')) {
            this.setAttribute('size', `${this.size}`);
        }
    }
}
__decorate([
    property()
], Avatar.prototype, "label", void 0);
__decorate([
    property()
], Avatar.prototype, "src", void 0);
__decorate([
    property({ type: Number, reflect: true })
], Avatar.prototype, "size", null);
