// @spectrum-web-components/switch/src/Switch.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import { CheckboxBase } from '/web-component-designer-demo/node_modules/@spectrum-web-components/checkbox/src/CheckboxBase.js';
import switchStyles from './switch.css.js';
import legacyStyles from './switch-legacy.css.js';
/**
 * @element sp-switch
 *
 * @slot - text label of the Switch
 */
export class Switch extends CheckboxBase {
    constructor() {
        super(...arguments);
        this.emphasized = false;
    }
    static get styles() {
        /* c8 ignore next 4 */
        if (window.hasOwnProperty('ShadyDOM')) {
            // Override some styles if we are using the web component polyfill
            return [switchStyles, legacyStyles];
        }
        return [switchStyles];
    }
    render() {
        return html `
            ${super.render()}
            <span id="switch"></span>
            <label id="label" for="input"><slot></slot></label>
        `;
    }
    firstUpdated(changes) {
        super.firstUpdated(changes);
        this.inputElement.setAttribute('role', 'switch');
    }
    updated(changes) {
        if (changes.has('checked')) {
            this.inputElement.setAttribute('aria-checked', this.checked ? 'true' : 'false');
        }
    }
}
__decorate([
    property({ type: Boolean, reflect: true })
], Switch.prototype, "emphasized", void 0);
