// @node-projects/acad-ts/dist/IO/DXF/DxfStreamReader/DxfStreamReaderBase.js
import { DxfCode } from '../../../DxfCode.js';
import { DxfFileToken } from '../../../DxfFileToken.js';
import { GroupCodeValue, GroupCodeValueType } from '../../../GroupCodeValue.js';
import { DxfException } from '../../../Exceptions/DxfException.js';
import { MathHelper } from '../../../Math/MathHelper.js';
export class DxfStreamReaderBase {
    DxfCode = DxfCode.Invalid;
    GroupCodeValue = GroupCodeValueType.None;
    get Code() {
        return this.DxfCode;
    }
    Value = '';
    Position = 0;
    ValueRaw = '';
    get ValueAsString() {
        return String(this.Value)
            .replace(/\^J/g, '\n')
            .replace(/\^M/g, '\r')
            .replace(/\^I/g, '\t')
            .replace(/\^ /g, '^');
    }
    get ValueAsBool() {
        return Boolean(this.Value);
    }
    get ValueAsShort() {
        return Number(this.Value) & 0xFFFF;
    }
    get ValueAsUShort() {
        return (Number(this.Value) & 0xFFFF) >>> 0;
    }
    get ValueAsInt() {
        return Number(this.Value) | 0;
    }
    get ValueAsLong() {
        return Number(this.Value);
    }
    get ValueAsDouble() {
        return Number(this.Value);
    }
    get ValueAsAngle() {
        return MathHelper.DegToRad(Number(this.Value));
    }
    get ValueAsHandle() {
        return Number(this.Value);
    }
    get ValueAsBinaryChunk() {
        return this.Value;
    }
    ReadNext() {
        this.DxfCode = this.readCode();
        this.GroupCodeValue = GroupCodeValue.transformValue(this.Code);
        this.Value = this.transformValue(this.GroupCodeValue);
        if (this.DxfCode === DxfCode.Comment) {
            this.ReadNext();
        }
    }
    Find(dxfEntry) {
        this.Start();
        do {
            this.ReadNext();
        } while (this.ValueAsString !== dxfEntry && this.ValueAsString !== DxfFileToken.EndOfFile);
        return this.ValueAsString === dxfEntry;
    }
    ExpectedCode(code) {
        this.ReadNext();
        if (this.Code !== code) {
            throw new DxfException(code, this.Position);
        }
    }
    toString() {
        return `${this.Code} | ${this.Value}`;
    }
    Start() {
        this.DxfCode = DxfCode.Invalid;
        this.Value = '';
        this._streamPosition = 0;
        this.Position = 0;
    }
    _streamPosition = 0;
    transformValue(code) {
        switch (code) {
            case GroupCodeValueType.String:
            case GroupCodeValueType.Comment:
            case GroupCodeValueType.ExtendedDataString:
                return this.readStringLine();
            case GroupCodeValueType.Point3D:
            case GroupCodeValueType.Double:
            case GroupCodeValueType.ExtendedDataDouble:
                return this.lineAsDouble();
            case GroupCodeValueType.Byte:
            case GroupCodeValueType.Int16:
            case GroupCodeValueType.ExtendedDataInt16:
                return this.lineAsShort();
            case GroupCodeValueType.Int32:
            case GroupCodeValueType.ExtendedDataInt32:
                return this.lineAsInt();
            case GroupCodeValueType.Int64:
                return this.lineAsLong();
            case GroupCodeValueType.Handle:
            case GroupCodeValueType.ObjectId:
            case GroupCodeValueType.ExtendedDataHandle:
                return this.lineAsHandle();
            case GroupCodeValueType.Bool:
                return this.lineAsBool();
            case GroupCodeValueType.Chunk:
            case GroupCodeValueType.ExtendedDataChunk:
                return this.lineAsBinaryChunk();
            case GroupCodeValueType.None:
            default:
                throw new DxfException(code, this.Position);
        }
    }
}
