// @node-projects/acad-ts/dist/IO/DXF/DxfStreamReader/DxfTextReader.js
import { DxfStreamReaderBase } from './DxfStreamReaderBase.js';
import { DxfCode } from '../../../DxfCode.js';
export class DxfTextReader extends DxfStreamReaderBase {
    get baseStream() {
        return this._data;
    }
    _data;
    _text;
    _textPos = 0;
    constructor(stream) {
        super();
        this._data = stream;
        this._text = new TextDecoder('utf-8').decode(stream);
        this.Start();
    }
    Start() {
        super.Start();
        this._textPos = 0;
    }
    ReadNext() {
        super.ReadNext();
        this.Position += 2;
    }
    readStringLine() {
        let end = this._text.indexOf('\n', this._textPos);
        if (end === -1) {
            end = this._text.length;
        }
        let line = this._text.substring(this._textPos, end);
        this._textPos = end + 1;
        // Remove trailing \r
        if (line.endsWith('\r')) {
            line = line.substring(0, line.length - 1);
        }
        // Trim whitespace like C# StreamReader.ReadLine()
        line = line.trim();
        this.ValueRaw = line;
        return this.ValueRaw;
    }
    readCode() {
        const line = this.readStringLine();
        const value = parseInt(line, 10);
        if (!isNaN(value)) {
            return value;
        }
        this.Position++;
        return DxfCode.Invalid;
    }
    lineAsBool() {
        const str = this.readStringLine();
        const result = parseInt(str, 10);
        if (!isNaN(result)) {
            return result > 0;
        }
        return false;
    }
    lineAsDouble() {
        const str = this.readStringLine();
        const result = parseFloat(str);
        if (!isNaN(result)) {
            return result;
        }
        return 0.0;
    }
    lineAsShort() {
        const str = this.readStringLine();
        const result = parseInt(str, 10);
        if (!isNaN(result)) {
            return result;
        }
        return 0;
    }
    lineAsInt() {
        const str = this.readStringLine();
        const result = parseInt(str, 10);
        if (!isNaN(result)) {
            return result;
        }
        return 0;
    }
    lineAsLong() {
        const str = this.readStringLine();
        const result = parseInt(str, 10);
        if (!isNaN(result)) {
            return result;
        }
        return 0;
    }
    lineAsHandle() {
        const str = this.readStringLine();
        const result = parseInt(str, 16);
        if (!isNaN(result)) {
            return result;
        }
        return 0;
    }
    lineAsBinaryChunk() {
        const str = this.readStringLine();
        const bytes = [];
        for (let i = 0; i < str.length; i += 2) {
            const hex = str.substring(i, i + 2);
            const result = parseInt(hex, 16);
            if (!isNaN(result)) {
                bytes.push(result);
            }
            else {
                return new Uint8Array(0);
            }
        }
        return new Uint8Array(bytes);
    }
}
