// @node-projects/acad-ts/dist/IO/DXF/DxfStreamReader/DxfTablesSectionReader.js
import { DxfSectionReaderBase } from './DxfSectionReaderBase.js';
import { CadTableTemplate } from '../../Templates/CadTableTemplate.js';
import { CadTableEntryTemplate } from '../../Templates/CadTableEntryTemplate.js';
import { CadBlockRecordTemplate } from '../../Templates/CadBlockRecordTemplate.js';
import { CadBlockCtrlObjectTemplate } from '../../Templates/CadBlockCtrlObjectTemplate.js';
import { CadDimensionStyleTemplate } from '../../Templates/CadDimensionStyleTemplate.js';
import { CadLayerTemplate } from '../../Templates/CadLayerTemplate.js';
import { CadLineTypeTemplate } from '../../Templates/CadLineTypeTemplate.js';
import { CadUcsTemplate } from '../../Templates/CadUcsTemplate.js';
import { CadViewTemplate } from '../../Templates/CadViewTemplate.js';
import { CadVPortTemplate } from '../../Templates/CadVPortTemplate.js';
import { AppId } from '../../../Tables/AppId.js';
import { BlockRecord } from '../../../Tables/BlockRecord.js';
import { TextStyle } from '../../../Tables/TextStyle.js';
import { AppIdsTable } from '../../../Tables/Collections/AppIdsTable.js';
import { BlockRecordsTable } from '../../../Tables/Collections/BlockRecordsTable.js';
import { VPortsTable } from '../../../Tables/Collections/VPortsTable.js';
import { LineTypesTable } from '../../../Tables/Collections/LineTypesTable.js';
import { LayersTable } from '../../../Tables/Collections/LayersTable.js';
import { TextStylesTable } from '../../../Tables/Collections/TextStylesTable.js';
import { ViewsTable } from '../../../Tables/Collections/ViewsTable.js';
import { UCSTable } from '../../../Tables/Collections/UCSTable.js';
import { DimensionStylesTable } from '../../../Tables/Collections/DimensionStylesTable.js';
import { Color } from '../../../Color.js';
import { DxfCode } from '../../../DxfCode.js';
import { DxfFileToken } from '../../../DxfFileToken.js';
import { DxfSubclassMarker } from '../../../DxfSubclassMarker.js';
import { DxfMap } from '../../../DxfMap.js';
import { NotificationType } from '../../NotificationEventHandler.js';
import { DxfException } from '../../../Exceptions/DxfException.js';
import { XY } from '../../../Math/XY.js';
import { MathHelper } from '../../../Math/MathHelper.js';
export class DxfTablesSectionReader extends DxfSectionReaderBase {
    constructor(reader, builder) {
        super(reader, builder);
    }
    Read() {
        this._reader.ReadNext();
        while (this._reader.ValueAsString !== DxfFileToken.EndSection) {
            if (this._reader.ValueAsString === DxfFileToken.TableEntry) {
                this.readTable();
            }
            else {
                throw new DxfException(`Unexpected token at the beginning of a table: ${this._reader.ValueAsString}`, this._reader.Position);
            }
            if (this._reader.ValueAsString === DxfFileToken.EndTable) {
                this._reader.ReadNext();
            }
            else {
                throw new DxfException(`Unexpected token at the end of a table: ${this._reader.ValueAsString}`, this._reader.Position);
            }
        }
    }
    readTable() {
        // Debug.Assert(this._reader.ValueAsString == DxfFileToken.TableEntry);
        this._reader.ReadNext();
        let nentries = 0;
        let template = null;
        const edata = new Map();
        const commonData = this.readCommonObjectData();
        const { name, handle, ownerHandle, xdictHandle, reactors } = commonData;
        if (this._reader.DxfCode === DxfCode.Subclass) {
            while (this._reader.Code !== DxfCode.Start) {
                switch (this._reader.Code) {
                    case 70:
                        nentries = this._reader.ValueAsInt;
                        break;
                    case 100:
                        if (this._reader.ValueAsString === DxfSubclassMarker.DimensionStyleTable) {
                            while (this._reader.Code !== DxfCode.Start) {
                                this._reader.ReadNext();
                            }
                            break;
                        }
                        // Debug.Assert(this._reader.ValueAsString == DxfSubclassMarker.Table);
                        break;
                    case 1001:
                        this.readExtendedData(edata);
                        break;
                    default:
                        this._builder.Notify(`[AcDbSymbolTable] Unhandled dxf code ${this._reader.Code} at line ${this._reader.Position}.`);
                        break;
                }
                if (this._reader.Code === DxfCode.Start) {
                    break;
                }
                this._reader.ReadNext();
            }
        }
        else if (this._reader.ValueAsString === DxfFileToken.EndTable) {
            return;
        }
        else {
            this._reader.ReadNext();
        }
        switch (name) {
            case DxfFileToken.TableAppId:
                template = new CadTableTemplate(new AppIdsTable());
                this.readEntries(template);
                template.CadObject.handle = handle;
                this._builder.AppIds = template.CadObject;
                break;
            case DxfFileToken.TableBlockRecord:
                template = new CadBlockCtrlObjectTemplate(new BlockRecordsTable());
                this.readEntries(template);
                template.CadObject.handle = handle;
                this._builder.BlockRecords = template.CadObject;
                break;
            case DxfFileToken.TableVport:
                template = new CadTableTemplate(new VPortsTable());
                this.readEntries(template);
                template.CadObject.handle = handle;
                this._builder.VPorts = template.CadObject;
                break;
            case DxfFileToken.TableLinetype:
                template = new CadTableTemplate(new LineTypesTable());
                this.readEntries(template);
                template.CadObject.handle = handle;
                this._builder.LineTypesTable = template.CadObject;
                break;
            case DxfFileToken.TableLayer:
                template = new CadTableTemplate(new LayersTable());
                this.readEntries(template);
                template.CadObject.handle = handle;
                this._builder.Layers = template.CadObject;
                break;
            case DxfFileToken.TableStyle:
                template = new CadTableTemplate(new TextStylesTable());
                this.readEntries(template);
                template.CadObject.handle = handle;
                this._builder.TextStyles = template.CadObject;
                break;
            case DxfFileToken.TableView:
                template = new CadTableTemplate(new ViewsTable());
                this.readEntries(template);
                template.CadObject.handle = handle;
                this._builder.Views = template.CadObject;
                break;
            case DxfFileToken.TableUcs:
                template = new CadTableTemplate(new UCSTable());
                this.readEntries(template);
                template.CadObject.handle = handle;
                this._builder.UCSs = template.CadObject;
                break;
            case DxfFileToken.TableDimstyle:
                template = new CadTableTemplate(new DimensionStylesTable());
                this.readEntries(template);
                template.CadObject.handle = handle;
                this._builder.DimensionStyles = template.CadObject;
                break;
            default:
                throw new DxfException(`Unknown table name ${name}`);
        }
        // Debug.Assert(ownerHandle == null || ownerHandle === 0);
        template.OwnerHandle = ownerHandle ?? null;
        template.XDictHandle = xdictHandle;
        template.ReactorsHandles = reactors;
        template.EDataTemplateByAppName = edata;
        this._builder.AddTemplate(template);
    }
    readEntries(tableTemplate) {
        while (this._reader.ValueAsString !== DxfFileToken.EndTable) {
            this._reader.ReadNext();
            let template = null;
            switch (tableTemplate.CadObject.objectName) {
                case DxfFileToken.TableAppId:
                    template = this.readTableEntry(new CadTableEntryTemplate(new AppId()), this.readAppId.bind(this));
                    break;
                case DxfFileToken.TableBlockRecord: {
                    const block = new CadBlockRecordTemplate();
                    template = this.readTableEntry(block, this.readBlockRecord.bind(this));
                    if (block.CadObject.name.toUpperCase() === BlockRecord.ModelSpaceName.toUpperCase()) {
                        this._builder.ModelSpaceTemplate = block;
                    }
                    break;
                }
                case DxfFileToken.TableDimstyle:
                    template = this.readTableEntry(new CadDimensionStyleTemplate(), this.readDimensionStyle.bind(this));
                    break;
                case DxfFileToken.TableLayer:
                    template = this.readTableEntry(new CadLayerTemplate(), this.readLayer.bind(this));
                    break;
                case DxfFileToken.TableLinetype:
                    template = this.readTableEntry(new CadLineTypeTemplate(), this.readLineType.bind(this));
                    break;
                case DxfFileToken.TableStyle:
                    template = this.readTableEntry(new CadTableEntryTemplate(new TextStyle()), this.readTextStyle.bind(this));
                    break;
                case DxfFileToken.TableUcs:
                    template = this.readTableEntry(new CadUcsTemplate(), this.readUcs.bind(this));
                    break;
                case DxfFileToken.TableView:
                    template = this.readTableEntry(new CadViewTemplate(), this.readView.bind(this));
                    break;
                case DxfFileToken.TableVport:
                    template = this.readTableEntry(new CadVPortTemplate(), this.readVPort.bind(this));
                    break;
                default:
                    // Debug.Fail
                    break;
            }
            if (template !== null) {
                if (tableTemplate.CadObject.contains(template.Name) && this._builder.Configuration.Failsafe) {
                    this._builder.Notify(`Duplicated entry with name ${template.Name} found in ${template.CadObject.objectName}`, NotificationType.Warning);
                    tableTemplate.CadObject.remove(template.Name);
                    tableTemplate.CadObject.add(template.CadObject);
                }
                else {
                    tableTemplate.CadObject.add(template.CadObject);
                }
                this._builder.AddTemplate(template);
            }
        }
    }
    readTableEntry(template, readEntry) {
        const map = DxfMap.create(template.CadObject.constructor);
        while (this._reader.DxfCode !== DxfCode.Start) {
            if (!readEntry(template, map.subClasses.get(template.CadObject.SubclassMarker))) {
                const isExtendedData = { value: false };
                this.readCommonTableEntryCodes(template, isExtendedData, map);
                if (isExtendedData.value) {
                    continue;
                }
            }
            if (this._reader.Code !== DxfCode.Start) {
                this._reader.ReadNext();
            }
        }
        return template;
    }
    readCommonTableEntryCodes(template, isExtendedData, map) {
        isExtendedData.value = false;
        switch (this._reader.Code) {
            case 2:
                template.CadObject.Name = this._reader.ValueAsString;
                break;
            case 70:
                template.CadObject.Flags = this._reader.ValueAsUShort;
                break;
            case 100:
                // Debug.Assert
                break;
            default:
                this.readCommonCodes(template, isExtendedData, map);
                break;
        }
    }
    readAppId(template, map) {
        switch (this._reader.Code) {
            default:
                return this.tryAssignCurrentValue(template.CadObject, map);
        }
    }
    readBlockRecord(template, map) {
        const tmp = template;
        switch (this._reader.Code) {
            case 340:
                tmp.LayoutHandle = this._reader.ValueAsHandle;
                return true;
            default:
                return this.tryAssignCurrentValue(template.CadObject, map);
        }
    }
    readDimensionStyle(template, map) {
        const tmp = template;
        switch (this._reader.Code) {
            case 3:
                template.CadObject.PostFix = this._reader.ValueAsString;
                return true;
            case 4:
                template.CadObject.AlternateDimensioningSuffix = this._reader.ValueAsString;
                return true;
            case 5:
                tmp.DIMBL_Name = this._reader.ValueAsString;
                return true;
            case 6:
                tmp.DIMBLK1_Name = this._reader.ValueAsString;
                return true;
            case 7:
                tmp.DIMBLK2_Name = this._reader.ValueAsString;
                return true;
            case 40:
                try {
                    template.CadObject.ScaleFactor = this._reader.ValueAsDouble;
                }
                catch (ex) {
                    template.CadObject.ScaleFactor = MathHelper.Epsilon;
                    this._builder.Notify(`[${template.CadObject.SubclassMarker}] Assignation error for ScaleFactor.`, NotificationType.Warning, ex);
                }
                return true;
            case 41:
                try {
                    template.CadObject.ArrowSize = this._reader.ValueAsDouble;
                }
                catch (ex) {
                    this._builder.Notify(`[${template.CadObject.SubclassMarker}] Assignation error for ArrowSize.`, NotificationType.Warning, ex);
                }
                return true;
            case 42:
                template.CadObject.ExtensionLineOffset = this._reader.ValueAsDouble;
                return true;
            case 43:
                template.CadObject.DimensionLineIncrement = this._reader.ValueAsDouble;
                return true;
            case 44:
                template.CadObject.ExtensionLineExtension = this._reader.ValueAsDouble;
                return true;
            case 45:
                template.CadObject.Rounding = this._reader.ValueAsDouble;
                return true;
            case 46:
                template.CadObject.DimensionLineExtension = this._reader.ValueAsDouble;
                return true;
            case 47:
                template.CadObject.PlusTolerance = this._reader.ValueAsDouble;
                return true;
            case 48:
                template.CadObject.MinusTolerance = this._reader.ValueAsDouble;
                return true;
            case 49:
                template.CadObject.FixedExtensionLineLength = this._reader.ValueAsDouble;
                return true;
            case 50:
                try {
                    template.CadObject.JoggedRadiusDimensionTransverseSegmentAngle = this._reader.ValueAsAngle;
                }
                catch (ex) {
                    this._builder.Notify(`[${template.CadObject.SubclassMarker}] Assignation error for JoggedRadiusDimensionTransverseSegmentAngle.`, NotificationType.Warning, ex);
                }
                return true;
            case 69:
                template.CadObject.TextBackgroundFillMode = this._reader.ValueAsShort;
                return true;
            case 70:
                if (!tmp.DxfFlagsAssigned) {
                    tmp.DxfFlagsAssigned = true;
                    return true;
                }
                else if (this._reader.ValueAsShort >= 0) {
                    template.CadObject.TextBackgroundColor = new Color(this._reader.ValueAsShort);
                }
                return true;
            case 71:
                template.CadObject.GenerateTolerances = this._reader.ValueAsBool;
                return true;
            case 72:
                template.CadObject.LimitsGeneration = this._reader.ValueAsBool;
                return true;
            case 73:
                template.CadObject.TextInsideHorizontal = this._reader.ValueAsBool;
                return true;
            case 74:
                template.CadObject.TextOutsideHorizontal = this._reader.ValueAsBool;
                return true;
            case 75:
                template.CadObject.SuppressFirstExtensionLine = this._reader.ValueAsBool;
                return true;
            case 76:
                template.CadObject.SuppressSecondExtensionLine = this._reader.ValueAsBool;
                return true;
            case 77:
                template.CadObject.TextVerticalAlignment = this._reader.ValueAsShort;
                return true;
            case 78:
                template.CadObject.ZeroHandling = this._reader.ValueAsShort;
                return true;
            case 79:
                template.CadObject.AngularZeroHandling = this._reader.ValueAsShort;
                return true;
            case 90:
                template.CadObject.ArcLengthSymbolPosition = this._reader.ValueAsShort;
                return true;
            case 105:
                template.CadObject.Handle = this._reader.ValueAsHandle;
                return true;
            case 140:
                try {
                    template.CadObject.TextHeight = this._reader.ValueAsDouble;
                }
                catch (ex) {
                    this._builder.Notify(`[${template.CadObject.SubclassMarker}] Assignation error for TextHeight.`, NotificationType.Warning, ex);
                }
                return true;
            case 141:
                template.CadObject.CenterMarkSize = this._reader.ValueAsDouble;
                return true;
            case 142:
                template.CadObject.TickSize = this._reader.ValueAsDouble;
                return true;
            case 143:
                template.CadObject.AlternateUnitScaleFactor = this._reader.ValueAsDouble;
                return true;
            case 144:
                template.CadObject.LinearScaleFactor = this._reader.ValueAsDouble;
                return true;
            case 145:
                template.CadObject.TextVerticalPosition = this._reader.ValueAsDouble;
                return true;
            case 146:
                template.CadObject.ToleranceScaleFactor = this._reader.ValueAsDouble;
                return true;
            case 147:
                template.CadObject.DimensionLineGap = this._reader.ValueAsDouble;
                return true;
            case 148:
                template.CadObject.AlternateUnitRounding = this._reader.ValueAsDouble;
                return true;
            case 170:
                template.CadObject.AlternateUnitDimensioning = this._reader.ValueAsBool;
                return true;
            case 171:
                template.CadObject.AlternateUnitDecimalPlaces = this._reader.ValueAsShort;
                return true;
            case 172:
                template.CadObject.TextOutsideExtensions = this._reader.ValueAsBool;
                return true;
            case 173:
                template.CadObject.SeparateArrowBlocks = this._reader.ValueAsBool;
                return true;
            case 174:
                template.CadObject.TextInsideExtensions = this._reader.ValueAsBool;
                return true;
            case 175:
                template.CadObject.SuppressOutsideExtensions = this._reader.ValueAsBool;
                return true;
            case 176:
                template.CadObject.DimensionLineColor = new Color(this._reader.ValueAsShort);
                return true;
            case 177:
                template.CadObject.ExtensionLineColor = new Color(this._reader.ValueAsShort);
                return true;
            case 178:
                template.CadObject.TextColor = new Color(this._reader.ValueAsShort);
                return true;
            case 179:
                template.CadObject.AngularDecimalPlaces = this._reader.ValueAsShort;
                return true;
            case 270:
                template.CadObject.LinearUnitFormat = this._reader.ValueAsShort;
                return true;
            case 271:
                template.CadObject.DecimalPlaces = this._reader.ValueAsShort;
                return true;
            case 272:
                template.CadObject.ToleranceDecimalPlaces = this._reader.ValueAsShort;
                return true;
            case 273:
                template.CadObject.AlternateUnitFormat = this._reader.ValueAsShort;
                return true;
            case 274:
                template.CadObject.AlternateUnitToleranceDecimalPlaces = this._reader.ValueAsShort;
                return true;
            case 275:
                template.CadObject.AngularUnit = this._reader.ValueAsShort;
                return true;
            case 276:
                template.CadObject.FractionFormat = this._reader.ValueAsShort;
                return true;
            case 277:
                template.CadObject.LinearUnitFormat = this._reader.ValueAsShort;
                return true;
            case 278:
                template.CadObject.DecimalSeparator = String.fromCharCode(this._reader.ValueAsShort);
                return true;
            case 279:
                template.CadObject.TextMovement = this._reader.ValueAsShort;
                return true;
            case 280:
                template.CadObject.TextHorizontalAlignment = this._reader.ValueAsShort;
                return true;
            case 281:
                template.CadObject.SuppressFirstDimensionLine = this._reader.ValueAsBool;
                return true;
            case 282:
                template.CadObject.SuppressSecondDimensionLine = this._reader.ValueAsBool;
                return true;
            case 283:
                template.CadObject.ToleranceAlignment = this._reader.ValueAsShort;
                return true;
            case 284:
                template.CadObject.ToleranceZeroHandling = this._reader.ValueAsShort;
                return true;
            case 285:
                template.CadObject.AlternateUnitZeroHandling = this._reader.ValueAsShort;
                return true;
            case 286:
                template.CadObject.AlternateUnitToleranceZeroHandling = this._reader.ValueAsShort;
                return true;
            case 287:
                template.CadObject.DimensionFit = this._reader.ValueAsShort;
                return true;
            case 288:
                template.CadObject.CursorUpdate = this._reader.ValueAsBool;
                return true;
            case 289:
                template.CadObject.DimensionTextArrowFit = this._reader.ValueAsShort;
                return true;
            case 290:
                template.CadObject.IsExtensionLineLengthFixed = this._reader.ValueAsBool;
                return true;
            case 340:
                tmp.TextStyleHandle = this._reader.ValueAsHandle;
                return true;
            case 341:
                tmp.DIMLDRBLK = this._reader.ValueAsHandle;
                return true;
            case 342:
                tmp.DIMBLK = this._reader.ValueAsHandle;
                return true;
            case 343:
                tmp.DIMBLK1 = this._reader.ValueAsHandle;
                return true;
            case 344:
                tmp.DIMBLK2 = this._reader.ValueAsHandle;
                return true;
            case 345:
                tmp.Dimltype = this._reader.ValueAsHandle;
                return true;
            case 346:
                tmp.Dimltex1 = this._reader.ValueAsHandle;
                return true;
            case 347:
                tmp.Dimltex2 = this._reader.ValueAsHandle;
                return true;
            case 371:
                template.CadObject.DimensionLineWeight = this._reader.ValueAsShort;
                return true;
            case 372:
                template.CadObject.ExtensionLineWeight = this._reader.ValueAsShort;
                return true;
            default:
                return false;
        }
    }
    readLayer(template, map) {
        const tmp = template;
        switch (this._reader.Code) {
            case 6:
                tmp.LineTypeName = this._reader.ValueAsString;
                return true;
            case 62: {
                let index = this._reader.ValueAsShort;
                if (index < 0) {
                    template.CadObject.IsOn = false;
                    index = Math.abs(index);
                }
                const color = new Color(index);
                if (color.isByBlock || color.isByLayer) {
                    this._builder.Notify(`Wrong index ${index} for layer ${template.CadObject.Name}`, NotificationType.Warning);
                }
                else {
                    template.CadObject.Color = new Color(index);
                }
                return true;
            }
            case 347:
                tmp.MaterialHandle = this._reader.ValueAsHandle;
                return true;
            case 348:
                return true;
            case 390:
                template.CadObject.PlotStyleName = this._reader.ValueAsHandle;
                return true;
            case 430:
                tmp.TrueColorName = this._reader.ValueAsString;
                return true;
            default:
                return this.tryAssignCurrentValue(template.CadObject, map);
        }
    }
    readLineType(template, map) {
        const tmp = template;
        switch (this._reader.Code) {
            case 40:
                tmp.TotalLen = this._reader.ValueAsDouble;
                return true;
            case 49:
                do {
                    tmp.SegmentTemplates.push(this.readLineTypeSegment());
                } while (this._reader.Code === 49);
                return true;
            case 73:
                return true;
            default:
                return this.tryAssignCurrentValue(template.CadObject, map);
        }
    }
    readLineTypeSegment() {
        const template = new CadLineTypeTemplate.SegmentTemplate();
        template.Segment.length = this._reader.ValueAsDouble;
        this._reader.ReadNext();
        while (this._reader.Code !== 49 && this._reader.Code !== 0) {
            switch (this._reader.Code) {
                case 9:
                    template.Segment.text = this._reader.ValueAsString;
                    break;
                case 44:
                    template.Segment.offset = new XY(this._reader.ValueAsDouble, template.Segment.offset.y);
                    break;
                case 45:
                    template.Segment.offset = new XY(template.Segment.offset.x, this._reader.ValueAsDouble);
                    break;
                case 46:
                    template.Segment.scale = this._reader.ValueAsDouble;
                    break;
                case 50:
                    template.Segment.rotation = this._reader.ValueAsAngle;
                    break;
                case 74:
                    template.Segment.shapeFlags = this._reader.ValueAsUShort;
                    break;
                case 75:
                    template.Segment.shapeNumber = this._reader.ValueAsInt;
                    break;
                case 340:
                    break;
                default:
                    this._builder.Notify(`[LineTypeSegment] Unhandled dxf code ${this._reader.Code} with value ${this._reader.ValueAsString}, position ${this._reader.Position}`, NotificationType.None);
                    break;
            }
            this._reader.ReadNext();
        }
        return template;
    }
    readTextStyle(template, map) {
        switch (this._reader.Code) {
            case 2:
                if (this._reader.ValueAsString) {
                    template.CadObject.Name = this._reader.ValueAsString;
                }
                return true;
            default:
                return this.tryAssignCurrentValue(template.CadObject, map);
        }
    }
    readUcs(template, map) {
        switch (this._reader.Code) {
            default:
                return this.tryAssignCurrentValue(template.CadObject, map);
        }
    }
    readView(template, map) {
        const tmp = template;
        switch (this._reader.Code) {
            case 348:
                tmp.VisualStyleHandle = this._reader.ValueAsHandle;
                return true;
            default:
                return this.tryAssignCurrentValue(template.CadObject, map);
        }
    }
    readVPort(template, map) {
        const tmp = template;
        switch (this._reader.Code) {
            case 65:
            case 73:
                return true;
            case 348:
                tmp.StyleHandle = this._reader.ValueAsHandle;
                return true;
            default:
                return this.tryAssignCurrentValue(template.CadObject, map);
        }
    }
}
