// @node-projects/web-component-designer/dist/elements/widgets/paletteView/paletteView.js
import '../../controls/DesignerTabControl.js';
import { PaletteElements } from "./paletteElements.js";
import { DesignerTabControl } from "../../controls/DesignerTabControl.js";
import { BaseCustomWebComponentLazyAppend, css } from "../../../../../base-custom-webcomponent/dist/index.js";
export class PaletteView extends BaseCustomWebComponentLazyAppend {
  constructor() {
    super();
    this.selected = 'native';
    this._designerTabControl = new DesignerTabControl();
    this._designerTabControl.selectedIndex = 0;
    this.shadowRoot.appendChild(this._designerTabControl);
  }

  static get style() {
    return css`
    :host {
      display: flex;
      flex-direction: column;
      flex: 1;
      height: 100%;
    }`;
  }

  async loadControls(serviceContainer, elementsServices) {
    for (const s of elementsServices) {
      let elements = await s.getElements();
      let paletteElement = new PaletteElements();
      paletteElement.title = s.name;

      this._designerTabControl.appendChild(paletteElement);

      paletteElement.loadElements(serviceContainer, elements);
    }
  }

}
customElements.define('node-projects-palette-view', PaletteView);
