// @node-projects/web-component-designer/dist/elements/widgets/paletteView/paletteView.js
import '../../controls/DesignerTabControl.js';
import { PaletteElements } from './paletteElements';
import { DesignerTabControl } from '../../controls/DesignerTabControl';
import { BaseCustomWebComponentLazyAppend, css } from '@node-projects/base-custom-webcomponent';
export class PaletteView extends BaseCustomWebComponentLazyAppend {
    constructor() {
        super();
        this.selected = 'native';
        this._designerTabControl = new DesignerTabControl();
        this._designerTabControl.selectedIndex = 0;
        this.shadowRoot.appendChild(this._designerTabControl);
    }
    static get style() {
        return css `
    :host {
      display: flex;
      flex-direction: column;
      flex: 1;
      height: 100%;
    }`;
    }
    async loadControls(elementsServices) {
        for (const s of elementsServices) {
            let elements = await s.getElements();
            let paletteElement = new PaletteElements();
            paletteElement.title = s.name;
            this._designerTabControl.appendChild(paletteElement);
            paletteElement.loadElements(elements);
        }
    }
}
customElements.define('node-projects-palette-view', PaletteView);
