// @spectrum-web-components/button/stories/button-primary-outline.stories.js
import { renderButtonSet, renderIconSizeOverridden, renderLink, renderLinkWithTarget, renderMinWidthButton, renderWithIcon, } from './index.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/icons-workflow/icons/sp-icon-help.js';
import { args, argTypes } from './index.js';
const variant = 'primary';
const treatment = 'outline';
export default {
    component: 'sp-button',
    title: 'Button/Primary/Outline',
    args: Object.assign(Object.assign({}, args), { variant,
        treatment }),
    argTypes,
};
export const Default = (props) => renderButtonSet(props);
export const withIcon = (props) => renderWithIcon(props);
export const iconSizeOverridden = () => renderIconSizeOverridden(variant, treatment);
export const minWidthButton = (props) => renderMinWidthButton(props);
minWidthButton.story = {
    name: 'min-width',
};
export const link = (props) => renderLink(props);
link.story = {
    name: 'href',
};
export const linkWithTarget = (props) => renderLinkWithTarget(props);
linkWithTarget.story = {
    name: 'href with target="_blank"',
};
