// @spectrum-web-components/color-area/stories/color-area.stories.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { html } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/color-slider/sp-color-slider.js';
import '../sp-color-area.js';
export default {
    title: 'Color/Area',
    component: 'sp-color-area',
    argTypes: {
        onInput: { action: 'input' },
        onChange: { action: 'change' },
        color: {
            name: 'color',
            type: { name: 'ColorValue', required: 'true' },
            description: 'The color displayed by the ColorArea.',
            table: {
                type: { summary: 'ColorValue' },
                defaultValue: { summary: '' },
            },
            control: 'text',
        },
    },
};
export const Default = ({ onChange, onInput }) => {
    return html `
        <sp-color-area
            @input=${({ target }) => {
        const next = target.nextElementSibling;
        next.textContent = target.color;
        next.style.color = target.color;
        onInput(target.value);
    }}
            @change=${({ target }) => {
        onChange(target.value);
    }}
        ></sp-color-area>
        <div style="color: #ff0000" aria-live="off">#ff0000</div>
    `;
};
export const joint = () => {
    return html `
        <div>
            <sp-color-area
                color="hsv (120 0% 100%)"
                @input=${({ target }) => {
        const next = target.nextElementSibling;
        const display = next.nextElementSibling;
        display.textContent = target.color;
        display.style.color = target.color;
        next.color = target.color;
    }}
            ></sp-color-area>
            <sp-color-slider
                color="hsv(120 0% 1)"
                @input=${({ target: { color, previousElementSibling, nextElementSibling, }, }) => {
        previousElementSibling.color = color;
        nextElementSibling.textContent = color;
        nextElementSibling.style.color = color;
    }}
            ></sp-color-slider>
            <div style="color: hsv(120, 0, 1)">hsv(120, 0, 1)</div>
        </div>
    `;
};
export const disabled = () => {
    return html `
        <sp-color-area disabled></sp-color-area>
    `;
};
export const sized = () => {
    return html `
        <sp-color-area style="width: 72px; height: 72px"></sp-color-area>
    `;
};
export const canvas = () => {
    return html `
        <sp-color-area>
            <canvas slot="gradient"></canvas>
        </sp-color-area>
    `;
};
class CanvasWriter extends HTMLElement {
    constructor() {
        super();
        this.writeStatePromise = Promise.resolve(false);
        this.writeStatePromise = new Promise((res) => {
            requestAnimationFrame(() => {
                this.writeToCanvas();
                res(true);
            });
        });
    }
    writeToCanvas() {
        const { previousElementSibling } = this;
        if (previousElementSibling) {
            const canvas = previousElementSibling.querySelector('canvas[slot="gradient"]');
            if (canvas) {
                canvas.width = canvas.offsetWidth;
                canvas.height = canvas.offsetHeight;
                const context = canvas.getContext('2d');
                if (context) {
                    context.rect(0, 0, canvas.width, canvas.height);
                    const gradB = context.createLinearGradient(0, 0, 0, canvas.height);
                    gradB.addColorStop(0, 'white');
                    gradB.addColorStop(1, 'black');
                    const gradC = context.createLinearGradient(0, 0, canvas.width, 0);
                    gradC.addColorStop(0, 'hsla(0,100%,50%,0)');
                    gradC.addColorStop(1, 'hsla(0,100%,50%,1)');
                    context.fillStyle = gradB;
                    context.fillRect(0, 0, canvas.width, canvas.height);
                    context.fillStyle = gradC;
                    context.globalCompositeOperation = 'multiply';
                    context.fillRect(0, 0, canvas.width, canvas.height);
                    context.globalCompositeOperation = 'source-over';
                }
            }
        }
    }
    get updateComplete() {
        return this.writeStatePromise;
    }
}
customElements.define('area-canvas-writer', CanvasWriter);
canvas.decorators = [
    (story) => {
        return html `
            ${story()}
            <area-canvas-writer></area-canvas-writer>
        `;
    },
];
