// @node-projects/acad-ts/dist/IO/DWG/DwgStreamReaders/DwgLZ77AC18Decompressor.js
export class DwgLZ77AC18Decompressor {
    static decompress(compressed, compressedOffset, decompressedSize) {
        const result = new Uint8Array(decompressedSize);
        DwgLZ77AC18Decompressor.decompressToDest(compressed, compressedOffset, result);
        return result;
    }
    static decompressToDest(src, srcOffset, dst) {
        let srcPos = srcOffset;
        let dstPos = 0;
        let tempBuf = new Uint8Array(128);
        if (srcPos >= src.length)
            return;
        let opcode1 = src[srcPos++];
        if ((opcode1 & 0xF0) === 0) {
            const litCount = DwgLZ77AC18Decompressor.literalCount(opcode1, src, { pos: srcPos });
            srcPos = DwgLZ77AC18Decompressor._lastSrcPos;
            const result = DwgLZ77AC18Decompressor.copy(litCount + 3, src, srcPos, dst, dstPos, tempBuf);
            srcPos = result.srcPos;
            dstPos = result.dstPos;
            tempBuf = result.tempBuf;
            opcode1 = result.nextByte;
        }
        while (opcode1 !== 0x11) {
            if (srcPos >= src.length) {
                throw new Error(`LZ77AC18: read past end of source at srcPos=${srcPos}, src.length=${src.length}`);
            }
            let compOffset = 0;
            let compressedBytes = 0;
            if (opcode1 < 0x10 || opcode1 >= 0x40) {
                compressedBytes = (opcode1 >> 4) - 1;
                const opcode2 = src[srcPos++];
                compOffset = ((opcode1 >> 2 & 3) | (opcode2 << 2)) + 1;
            }
            else if (opcode1 < 0x20) {
                const rbResult = DwgLZ77AC18Decompressor.readCompressedBytes(opcode1, 0b0111, src, { pos: srcPos });
                compressedBytes = rbResult.value;
                srcPos = rbResult.srcPos;
                compOffset = (opcode1 & 8) << 11;
                const tboResult = DwgLZ77AC18Decompressor.twoByteOffset(compOffset, 0x4000, src, srcPos);
                compOffset = tboResult.offset;
                opcode1 = tboResult.firstByte;
                srcPos = tboResult.srcPos;
            }
            else if (opcode1 >= 0x20) {
                const rbResult = DwgLZ77AC18Decompressor.readCompressedBytes(opcode1, 0b00011111, src, { pos: srcPos });
                compressedBytes = rbResult.value;
                srcPos = rbResult.srcPos;
                const tboResult = DwgLZ77AC18Decompressor.twoByteOffset(compOffset, 1, src, srcPos);
                compOffset = tboResult.offset;
                opcode1 = tboResult.firstByte;
                srcPos = tboResult.srcPos;
            }
            // Copy compressed bytes from earlier in dst
            const position = dstPos;
            if (tempBuf.length < compressedBytes) {
                tempBuf = new Uint8Array(compressedBytes);
            }
            // Read from dst at (position - compOffset)
            const readStart = position - compOffset;
            const readLen = Math.min(compressedBytes, compOffset);
            for (let i = 0; i < readLen; i++) {
                tempBuf[i] = dst[readStart + i];
            }
            let remaining = compressedBytes;
            let writePos = position;
            while (remaining > 0) {
                const writeLen = Math.min(remaining, compOffset);
                for (let i = 0; i < writeLen; i++) {
                    dst[writePos + i] = tempBuf[i];
                }
                writePos += writeLen;
                remaining -= compOffset;
            }
            dstPos = position + compressedBytes;
            let litCount = opcode1 & 3;
            if (litCount === 0) {
                opcode1 = src[srcPos++];
                if ((opcode1 & 0b11110000) === 0) {
                    const lcResult = DwgLZ77AC18Decompressor.literalCount(opcode1, src, { pos: srcPos });
                    srcPos = DwgLZ77AC18Decompressor._lastSrcPos;
                    litCount = lcResult + 3;
                }
            }
            if (litCount > 0) {
                const result = DwgLZ77AC18Decompressor.copy(litCount, src, srcPos, dst, dstPos, tempBuf);
                srcPos = result.srcPos;
                dstPos = result.dstPos;
                tempBuf = result.tempBuf;
                opcode1 = result.nextByte;
            }
        }
    }
    static _lastSrcPos = 0;
    static copy(count, src, srcPos, dst, dstPos, tempBuf) {
        if (tempBuf.length < count) {
            tempBuf = new Uint8Array(count);
        }
        for (let i = 0; i < count; i++) {
            tempBuf[i] = src[srcPos + i];
            dst[dstPos + i] = src[srcPos + i];
        }
        srcPos += count;
        dstPos += count;
        const nextByte = src[srcPos++];
        return { srcPos, dstPos, tempBuf, nextByte };
    }
    static literalCount(code, src, ref) {
        let lowbits = code & 0b1111;
        if (lowbits === 0) {
            let lastByte;
            lastByte = src[ref.pos++];
            while (lastByte === 0) {
                lowbits += 0xFF;
                lastByte = src[ref.pos++];
            }
            lowbits += 0xF + lastByte;
        }
        DwgLZ77AC18Decompressor._lastSrcPos = ref.pos;
        return lowbits;
    }
    static readCompressedBytes(opcode1, validBits, compressed, ref) {
        let compressedBytes = opcode1 & validBits;
        if (compressedBytes === 0) {
            let lastByte;
            lastByte = compressed[ref.pos++];
            while (lastByte === 0) {
                compressedBytes += 0xFF;
                lastByte = compressed[ref.pos++];
            }
            compressedBytes += lastByte + validBits;
        }
        return { value: compressedBytes + 2, srcPos: ref.pos };
    }
    static twoByteOffset(offset, addedValue, stream, srcPos) {
        const firstByte = stream[srcPos++];
        offset |= firstByte >> 2;
        offset |= stream[srcPos++] << 6;
        offset += addedValue;
        return { offset, firstByte, srcPos };
    }
}
