// @node-projects/acad-ts/dist/IO/DWG/DwgStreamReaders/DwgMergedReader.js
import { Color } from '../../../Color.js';
import { DwgStreamReaderAC18 } from './DwgStreamReaderAC18.js';
export class DwgMergedReader {
    get encoding() {
        return this._mainReader.encoding;
    }
    set encoding(value) {
        this._mainReader.encoding = value;
        this._textReader.encoding = value;
        this._handleReader.encoding = value;
    }
    get stream() {
        throw new Error('InvalidOperation');
    }
    get bitShift() {
        throw new Error('InvalidOperation');
    }
    set bitShift(_value) {
        throw new Error('InvalidOperation');
    }
    get position() {
        return this._mainReader.position;
    }
    set position(_value) {
        throw new Error('InvalidOperation');
    }
    isEmpty = false;
    _mainReader;
    _textReader;
    _handleReader;
    constructor(manReader, textReader, handleReader) {
        this._mainReader = manReader;
        this._textReader = textReader;
        this._handleReader = handleReader;
    }
    advance(offset) {
        this._mainReader.advance(offset);
    }
    advanceByte() {
        throw new Error('InvalidOperation');
    }
    handleReference() {
        return this._handleReader.handleReference();
    }
    handleReferenceWithRef(referenceHandle) {
        return this._handleReader.handleReferenceWithRef(referenceHandle);
    }
    handleReferenceWithRefAndType(referenceHandle) {
        return this._handleReader.handleReferenceWithRefAndType(referenceHandle);
    }
    positionInBits() {
        return this._mainReader.positionInBits();
    }
    read2Bits() {
        return this._mainReader.read2Bits();
    }
    read2RawDouble() {
        return this._mainReader.read2RawDouble();
    }
    read3RawDouble() {
        return this._mainReader.read3RawDouble();
    }
    read3BitDouble() {
        return this._mainReader.read3BitDouble();
    }
    readBit() {
        return this._mainReader.readBit();
    }
    readBitAsShort() {
        return this._mainReader.readBitAsShort();
    }
    readBitDouble() {
        return this._mainReader.readBitDouble();
    }
    read2BitDouble() {
        return this._mainReader.read2BitDouble();
    }
    readBitLong() {
        return this._mainReader.readBitLong();
    }
    readBitLongLong() {
        return this._mainReader.readBitLongLong();
    }
    readBitShort() {
        return this._mainReader.readBitShort();
    }
    readBitShortAsBool() {
        return this._mainReader.readBitShortAsBool();
    }
    readByte() {
        return this._mainReader.readByte();
    }
    readBytes(length) {
        return this._mainReader.readBytes(length);
    }
    read2BitDoubleWithDefault(defValues) {
        return this._mainReader.read2BitDoubleWithDefault(defValues);
    }
    read3BitDoubleWithDefault(defValues) {
        return this._mainReader.read3BitDoubleWithDefault(defValues);
    }
    readCmColor(useTextStream = false) {
        if (!(this._mainReader instanceof DwgStreamReaderAC18) && !useTextStream) {
            return this._mainReader.readCmColor();
        }
        // CMC:
        // BS: color index (always 0)
        const colorIndex = this.readBitShort();
        // BL: RGB value - always negative
        const rgb = this.readBitLong() >>> 0;
        const arr = new Uint8Array(4);
        arr[0] = rgb & 0xFF;
        arr[1] = (rgb >>> 8) & 0xFF;
        arr[2] = (rgb >>> 16) & 0xFF;
        arr[3] = (rgb >>> 24) & 0xFF;
        let color;
        if (rgb === 0xC0000000) {
            color = Color.ByLayer;
        }
        else if ((rgb & 0b0000_0001_0000_0000_0000_0000_0000_0000) !== 0) {
            // Indexed color
            color = new Color(arr[0]);
        }
        else {
            // True color
            color = Color.fromTrueColor((arr[2] << 16) | (arr[1] << 8) | arr[0]);
        }
        // RC: Color Byte
        const id = this.readByte();
        let colorName = '';
        // RC: Color Byte(&1 => color name follows(TV))
        if ((id & 1) === 1) {
            colorName = this.readVariableText();
        }
        let bookName = '';
        // &2 => book name follows(TV))
        if ((id & 2) === 2) {
            bookName = this.readVariableText();
        }
        return color;
    }
    readEnColor() {
        return this._mainReader.readEnColor();
    }
    read8BitJulianDate() {
        return this._mainReader.read8BitJulianDate();
    }
    readDateTime() {
        return this._mainReader.readDateTime();
    }
    readDouble() {
        return this._mainReader.readDouble();
    }
    readInt() {
        return this._mainReader.readInt();
    }
    readModularChar() {
        return this._mainReader.readModularChar();
    }
    readSignedModularChar() {
        return this._mainReader.readSignedModularChar();
    }
    readModularShort() {
        return this._mainReader.readModularShort();
    }
    readColorByIndex() {
        return new Color(this.readBitShort());
    }
    readObjectType() {
        return this._mainReader.readObjectType();
    }
    readBitExtrusion() {
        return this._mainReader.readBitExtrusion();
    }
    readBitDoubleWithDefault(def) {
        return this._mainReader.readBitDoubleWithDefault(def);
    }
    readBitThickness() {
        return this._mainReader.readBitThickness();
    }
    readRawChar() {
        return this._mainReader.readRawChar();
    }
    readRawLong() {
        return this._mainReader.readRawLong();
    }
    readRawULong() {
        return this._mainReader.readRawULong();
    }
    readSentinel() {
        return this._mainReader.readSentinel();
    }
    readShort() {
        return this._mainReader.readShort();
    }
    readShortBigEndian() {
        return this._mainReader.readShortBigEndian();
    }
    readTextUnicode() {
        // Handle the text section if is empty
        if (this._textReader.isEmpty) {
            return '';
        }
        return this._textReader.readTextUnicode();
    }
    readTimeSpan() {
        return this._mainReader.readTimeSpan();
    }
    readUInt() {
        return this._mainReader.readUInt();
    }
    readVariableText() {
        // Handle the text section if is empty
        if (this._textReader.isEmpty) {
            return '';
        }
        return this._textReader.readVariableText();
    }
    resetShift() {
        return this._mainReader.resetShift();
    }
    setPositionInBits(position) {
        this._mainReader.setPositionInBits(position);
    }
    setPositionByFlag(position) {
        throw new Error('InvalidOperation');
    }
}
