// @node-projects/acad-ts/dist/IO/CadDocumentBuilder.js
import { UnknownEntity } from '../Entities/UnknownEntity.js';
import { CadDictionary } from '../Objects/CadDictionary.js';
import { UnknownNonGraphicalObject } from '../Objects/UnknownNonGraphicalObject.js';
import { AppIdsTable } from '../Tables/Collections/AppIdsTable.js';
import { BlockRecordsTable } from '../Tables/Collections/BlockRecordsTable.js';
import { DimensionStylesTable } from '../Tables/Collections/DimensionStylesTable.js';
import { LayersTable } from '../Tables/Collections/LayersTable.js';
import { LineTypesTable } from '../Tables/Collections/LineTypesTable.js';
import { TextStylesTable } from '../Tables/Collections/TextStylesTable.js';
import { UCSTable } from '../Tables/Collections/UCSTable.js';
import { ViewsTable } from '../Tables/Collections/ViewsTable.js';
import { VPortsTable } from '../Tables/Collections/VPortsTable.js';
import { NotificationEventArgs, NotificationType } from './NotificationEventHandler.js';
export class CadDocumentBuilder {
    OnNotification = null;
    AppIds = new AppIdsTable();
    BlockRecords = new BlockRecordsTable();
    DimensionStyles = new DimensionStylesTable();
    DocumentToBuild;
    InitialHandSeed = 0;
    Layers = new LayersTable();
    LineTypesTable = new LineTypesTable();
    TextStyles = new TextStylesTable();
    UCSs = new UCSTable();
    Version;
    Views = new ViewsTable();
    VPorts = new VPortsTable();
    cadObjects = new Map();
    cadObjectsTemplates = new Map();
    dictionaryTemplates = new Map();
    tableEntryTemplates = new Map();
    tableTemplates = new Map();
    templatesMap = new Map();
    unassignedObjects = [];
    constructor(version, document) {
        this.Version = version;
        this.DocumentToBuild = document;
    }
    AddTemplate(template) {
        if (!this.addToMap(template)) {
            return;
        }
        if (this.isDictionaryTemplate(template)) {
            this.dictionaryTemplates.set(template.CadObject.handle, template);
        }
        else if (this.isTableTemplate(template)) {
            this.tableTemplates.set(template.CadObject.handle, template);
        }
        else if (this.isTableEntryTemplate(template)) {
            this.tableEntryTemplates.set(template.CadObject.handle, template);
        }
        else {
            this.cadObjectsTemplates.set(template.CadObject.handle, template);
        }
    }
    BuildDocument() {
        for (const template of this.tableEntryTemplates.values()) {
            template.Build(this);
        }
        for (const template of this.cadObjectsTemplates.values()) {
            template.Build(this);
        }
    }
    BuildTable(table) {
        const template = this.tableTemplates.get(table.handle);
        if (template) {
            template.Build(this);
        }
        else {
            this.Notify(`Table ${table.objectName} not found in the document`, NotificationType.Warning);
        }
    }
    BuildTables() {
        this.BuildTable(this.AppIds);
        this.BuildTable(this.TextStyles);
        this.BuildTable(this.LineTypesTable);
        this.BuildTable(this.Layers);
        this.BuildTable(this.UCSs);
        this.BuildTable(this.Views);
        this.BuildTable(this.BlockRecords);
        this.BuildTable(this.DimensionStyles);
        this.BuildTable(this.VPorts);
    }
    GetObjectTemplate(handle) {
        const template = this.templatesMap.get(handle);
        if (template) {
            return template;
        }
        return null;
    }
    Notify(message, notificationType = NotificationType.None, exception = null) {
        this.OnNotification?.(this, new NotificationEventArgs(message, notificationType, exception));
    }
    RegisterTables() {
        this.DocumentToBuild.registerCollection(this.AppIds);
        this.DocumentToBuild.registerCollection(this.TextStyles);
        this.DocumentToBuild.registerCollection(this.LineTypesTable);
        this.DocumentToBuild.registerCollection(this.Layers);
        this.DocumentToBuild.registerCollection(this.UCSs);
        this.DocumentToBuild.registerCollection(this.Views);
        this.DocumentToBuild.registerCollection(this.BlockRecords);
        this.DocumentToBuild.registerCollection(this.DimensionStyles);
        this.DocumentToBuild.registerCollection(this.VPorts);
    }
    TryGetCadObject(handle) {
        if (handle == null || handle === 0) {
            return null;
        }
        const obj = this.cadObjects.get(handle);
        if (obj) {
            if (obj instanceof UnknownEntity && !this.KeepUnknownEntities) {
                return null;
            }
            if (obj instanceof UnknownNonGraphicalObject && !this.KeepUnknownNonGraphicalObjects) {
                return null;
            }
            return obj;
        }
        return null;
    }
    TryGetObjectTemplate(handle) {
        if (handle == null || handle === 0) {
            return null;
        }
        const template = this.templatesMap.get(handle);
        if (template) {
            return template;
        }
        return null;
    }
    TryGetTableEntry(name) {
        if (!name || name.length === 0) {
            return null;
        }
        let table = null;
        // Determine the table based on T's type at runtime
        // We'll try each table in order
        const tables = [
            this.AppIds,
            this.Layers,
            this.LineTypesTable,
            this.UCSs,
            this.Views,
            this.DimensionStyles,
            this.TextStyles,
            this.VPorts,
            this.BlockRecords,
        ];
        for (const t of tables) {
            const entry = t.tryGetValue(name);
            if (entry) {
                return entry;
            }
        }
        return null;
    }
    buildDictionaries() {
        for (const dictionaryTemplate of this.dictionaryTemplates.values()) {
            dictionaryTemplate.Build(this);
        }
        if (!this.DocumentToBuild.rootDictionary) {
            const roots = [];
            for (const t of this.dictionaryTemplates.values()) {
                if (t.CadObject instanceof CadDictionary && t.CadObject.owner == null) {
                    roots.push(t.CadObject);
                }
            }
            if (roots.length !== 1) {
                this.Notify(`The root dictionary could not be found.`, NotificationType.Warning);
            }
            else {
                this.DocumentToBuild.rootDictionary = roots[0];
            }
        }
        this.DocumentToBuild.updateCollections(true, false);
    }
    createMissingHandles() {
        for (const template of this.unassignedObjects) {
            template.CadObject.handle = this.InitialHandSeed + 1;
            this.AddTemplate(template);
        }
        this.unassignedObjects = [];
    }
    registerTable(table, tableConstructor) {
        if (!table) {
            this.DocumentToBuild.registerCollection(new tableConstructor());
        }
        else {
            this.DocumentToBuild.registerCollection(table);
        }
    }
    addToMap(template) {
        if (template.CadObject.handle === 0) {
            this.pushUnassigned(template);
            return false;
        }
        if (this.templatesMap.has(template.CadObject.handle)) {
            this.Notify(`Repeated handle found ${template.CadObject.handle}.`, NotificationType.Warning);
            template.CadObject.handle = 0;
            this.pushUnassigned(template);
            return false;
        }
        if (template.CadObject.handle > this.InitialHandSeed) {
            this.InitialHandSeed = template.CadObject.handle;
        }
        this.templatesMap.set(template.CadObject.handle, template);
        this.cadObjects.set(template.CadObject.handle, template.CadObject);
        return true;
    }
    pushUnassigned(template) {
        if (!Array.isArray(this.unassignedObjects)) {
            this.unassignedObjects = [];
        }
        try {
            this.unassignedObjects.push(template);
        }
        catch (e) {
            if (e instanceof RangeError) {
                this.Notify('Resetting unassigned object queue due to invalid array length.', NotificationType.Warning, e);
                this.unassignedObjects = [template];
                return;
            }
            throw e;
        }
    }
    isDictionaryTemplate(template) {
        return 'CadObject' in template && template.CadObject instanceof CadDictionary;
    }
    isTableTemplate(template) {
        return 'EntryHandles' in template;
    }
    isTableEntryTemplate(template) {
        return 'Type' in template && 'Name' in template && !('EntryHandles' in template);
    }
}
