// @node-projects/acad-ts/dist/IO/CadReaderBase.js
import { CadDocument } from '../CadDocument.js';
import { NotificationEventArgs } from './NotificationEventHandler.js';
export class CadReaderBase {
    OnNotification = null;
    Configuration;
    _document = new CadDocument();
    _encoding = 'utf-8';
    _fileStream;
    constructor(stream, notification = null) {
        if (notification) {
            this.OnNotification = notification;
        }
        this._fileStream = stream;
        this.Configuration = this.createDefaultConfiguration();
    }
    Dispose() {
        // No-op in TS; ArrayBuffer doesn't need disposal
    }
    getListedEncoding(code) {
        // In TypeScript/browser, use TextDecoder encoding labels
        // Simplified: return a label or default
        return 'utf-8';
    }
    triggerNotification(message, notificationType, ex = null) {
        this.onNotificationEvent(null, new NotificationEventArgs(message, notificationType, ex));
    }
    onNotificationEvent(sender, e) {
        this.OnNotification?.(this, e);
    }
}
