// @node-projects/acad-ts/dist/IO/CadWriterBase.js
import { CadUtils } from '../CadUtils.js';
import { Dimension } from '../Entities/Dimension.js';
import { BlockRecord } from '../Tables/BlockRecord.js';
import { NotificationEventArgs } from './NotificationEventHandler.js';
export class CadWriterBase {
    OnNotification = null;
    Configuration;
    _document;
    _encoding = 'utf-8';
    _stream;
    constructor(stream, document) {
        this._stream = stream;
        this._document = document;
        this.Configuration = this.createDefaultConfiguration();
    }
    Write() {
        this._document.updateImageReactors();
        this._document.updateDxfClasses(this.Configuration.ResetDxfClasses);
        if (this.Configuration.UpdateDimensionsInModel) {
            this.updateDimensions(this._document.modelSpace);
        }
        if (this.Configuration.UpdateDimensionsInBlocks) {
            for (const item of this._document.blockRecords) {
                if (item.Name.toLowerCase() === BlockRecord.ModelSpaceName.toLowerCase()) {
                    continue;
                }
                this.updateDimensions(item);
            }
        }
        this._encoding = this.getListedEncoding(this._document.header.CodePage);
    }
    getListedEncoding(codePage) {
        const code = CadUtils.getCodePage(codePage);
        // Simplified encoding lookup for TypeScript
        return 'utf-8';
    }
    triggerNotification(messageOrSender, notificationTypeOrArgs, ex) {
        if (typeof messageOrSender === 'string') {
            this.OnNotification?.(this, new NotificationEventArgs(messageOrSender, notificationTypeOrArgs, ex ?? null));
        }
        else {
            this.OnNotification?.(messageOrSender, notificationTypeOrArgs);
        }
    }
    updateDimensions(record) {
        for (const item of record.entities) {
            if (item instanceof Dimension) {
                item.updateBlock();
            }
        }
    }
}
