// @node-projects/acad-ts/dist/IO/HugeMemoryStream.js
export class HugeMemoryStream {
    static _maxChunkSize = 1048576 * 1024;
    static _maxChunkShift = 30;
    static _maxChunkMask = (1048576 * 1024) - 1;
    _currentChunk;
    _currentInChunk = 0;
    _length;
    _position = 0;
    _chunks;
    static Create(length) {
        if (length <= 0x7FFFFFFF) {
            return new Uint8Array(length);
        }
        else {
            return new HugeMemoryStream(length);
        }
    }
    constructor(arg) {
        if (typeof arg === 'number') {
            const length = arg;
            this._length = length;
            this._position = 0;
            this._chunks = [];
            let lengthLeft = length;
            while (lengthLeft > 0) {
                const chunkSize = Math.min(lengthLeft, HugeMemoryStream._maxChunkSize);
                this._chunks.push(new Uint8Array(chunkSize));
                lengthLeft -= chunkSize;
            }
            this.Position = 0;
        }
        else {
            this._chunks = arg;
            this._length = arg.reduce((sum, x) => sum + x.length, 0);
            this.Position = 0;
        }
    }
    get CanRead() { return true; }
    get CanSeek() { return true; }
    get CanWrite() { return true; }
    get Length() { return this._length; }
    get Position() { return this._position; }
    set Position(value) {
        this._position = value;
        this._currentChunk = this._chunks[value >> HugeMemoryStream._maxChunkShift];
        this._currentInChunk = value & HugeMemoryStream._maxChunkMask;
    }
    Read(buffer, offset, count) {
        if (count === 1) {
            buffer[offset] = this._currentChunk[this._currentInChunk];
            this.Position++;
            return 1;
        }
        if (this._currentInChunk + count > HugeMemoryStream._maxChunkSize) {
            const toRead = HugeMemoryStream._maxChunkSize - this._currentInChunk;
            buffer.set(this._currentChunk.subarray(this._currentInChunk, this._currentInChunk + toRead), offset);
            this.Position += toRead;
            return toRead + this.Read(buffer, offset + toRead, count - toRead);
        }
        else {
            buffer.set(this._currentChunk.subarray(this._currentInChunk, this._currentInChunk + count), offset);
            this.Position += count;
            return count;
        }
    }
    ReadByte() {
        const value = this._currentChunk[this._currentInChunk];
        this.Position++;
        return value;
    }
    Write(buffer, offset, count) {
        if (count === 1) {
            this._currentChunk[this._currentInChunk] = buffer[offset];
            this.Position++;
            return;
        }
        if (this._currentInChunk + count > HugeMemoryStream._maxChunkSize) {
            const toWrite = HugeMemoryStream._maxChunkSize - this._currentInChunk;
            this._currentChunk.set(buffer.subarray(offset, offset + toWrite), this._currentInChunk);
            this.Position += toWrite;
            this.Write(buffer, offset + toWrite, count - toWrite);
        }
        else {
            this._currentChunk.set(buffer.subarray(offset, offset + count), this._currentInChunk);
            this.Position += count;
        }
    }
    WriteByte(value) {
        this._currentChunk[this._currentInChunk] = value;
        this.Position++;
    }
    Clone() {
        return new HugeMemoryStream(this._chunks);
    }
    static CloneStream(stream) {
        if (stream instanceof HugeMemoryStream) {
            return stream.Clone();
        }
        else if (stream instanceof Uint8Array) {
            return new Uint8Array(stream);
        }
        else {
            throw new Error('The provided stream type is not supported for cloning.');
        }
    }
}
