// @spectrum-web-components/dialog/src/Dialog.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, SpectrumElement, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property, query, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import { ifDefined } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/directives.js';
import { conditionAttributeWithId } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/condition-attribute-with-id.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/divider/sp-divider.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/button/sp-close-button.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/button-group/sp-button-group.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/icons-workflow/icons/sp-icon-alert.js';
import { FocusVisiblePolyfillMixin, ObserveSlotPresence, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/shared/./src/index.js';
import styles from './dialog.css.js';
function gatherAppliedIdsFromSlottedChildren(slot, idBase) {
    const assignedElements = slot.assignedElements();
    const ids = [];
    assignedElements.forEach((el, i) => {
        if (el.id) {
            ids.push(el.id);
        }
        else {
            const id = idBase + `-${i}`;
            el.id = id;
            ids.push(id);
        }
    });
    return ids;
}
/**
 * @element sp-dialog
 *
 * @slot hero - Accepts a hero image to display at the top of the dialog
 * @slot heading - Acts as the heading of the dialog. This should be an actual heading tag `<h1-6 />`
 * @slot - Content not addressed to a specific slot will be interpreted as the main content of the dialog
 * @slot footer - Content addressed to the `footer` will be placed below the main content and to the side of any `[slot='button']` content
 * @slot button - Button elements addressed to this slot may be placed below the content when not delivered in a fullscreen mode
 * @fires close - Announces that the dialog has been closed.
 */
export class Dialog extends FocusVisiblePolyfillMixin(ObserveSlotPresence(SpectrumElement, [
    '[slot="hero"]',
    '[slot="footer"]',
    '[slot="button"]',
])) {
    constructor() {
        super(...arguments);
        this.error = false;
        this.dismissable = false;
        this.noDivider = false;
        this.shouldManageTabOrderForScrolling = () => {
            const { offsetHeight, scrollHeight } = this.contentElement;
            if (offsetHeight < scrollHeight) {
                this.contentElement.tabIndex = 0;
            }
            else {
                this.contentElement.removeAttribute('tabindex');
            }
        };
        this.labelledbyId = `sp-dialog-label-${Dialog.instanceCount++}`;
        this.describedbyId = `sp-dialog-description-${Dialog.instanceCount++}`;
    }
    static get styles() {
        return [styles];
    }
    get hasFooter() {
        return this.getSlotContentPresence('[slot="footer"]');
    }
    get hasButtons() {
        return this.getSlotContentPresence('[slot="button"]');
    }
    get hasHero() {
        return this.getSlotContentPresence('[slot="hero"]');
    }
    close() {
        this.dispatchEvent(new Event('close', {
            bubbles: true,
            composed: true,
            cancelable: true,
        }));
    }
    render() {
        return html `
            <div class="grid">
                <slot name="hero"></slot>
                <slot
                    name="heading"
                    class=${ifDefined(this.hasHero ? this.hasHero : undefined)}
                    @slotchange=${this.onHeadingSlotchange}
                ></slot>
                ${this.error
            ? html `
                          <sp-icon-alert class="type-icon"></sp-icon-alert>
                      `
            : html ``}
                ${this.noDivider
            ? html ``
            : html `
                          <sp-divider size="m" class="divider"></sp-divider>
                      `}
                <div class="content">
                    <slot @slotchange=${this.onContentSlotChange}></slot>
                </div>
                ${this.hasFooter
            ? html `
                          <div class="footer">
                              <slot name="footer"></slot>
                          </div>
                      `
            : html ``}
                ${this.hasButtons
            ? html `
                          <sp-button-group
                              class="button-group ${this.hasFooter
                ? ''
                : 'button-group--noFooter'}"
                          >
                              <slot name="button"></slot>
                          </sp-button-group>
                      `
            : html ``}
                ${this.dismissable
            ? html `
                          <sp-close-button
                              class="close-button"
                              label="Close"
                              quiet
                              size="m"
                              @click=${this.close}
                          ></sp-close-button>
                      `
            : html ``}
            </div>
        `;
    }
    shouldUpdate(changes) {
        if (changes.has('mode') && !!this.mode) {
            this.dismissable = false;
        }
        if (changes.has('dismissable') && this.dismissable) {
            this.dismissable = !this.mode;
        }
        return super.shouldUpdate(changes);
    }
    firstUpdated(changes) {
        super.firstUpdated(changes);
        this.setAttribute('role', 'dialog');
    }
    onHeadingSlotchange({ target, }) {
        if (this.conditionLabelledby) {
            this.conditionLabelledby();
            delete this.conditionLabelledby;
        }
        const ids = gatherAppliedIdsFromSlottedChildren(target, this.labelledbyId);
        if (ids.length) {
            this.conditionLabelledby = conditionAttributeWithId(this, 'aria-labelledby', ids);
        }
    }
    onContentSlotChange({ target, }) {
        if (this.conditionDescribedby) {
            this.conditionDescribedby();
            delete this.conditionDescribedby;
        }
        const ids = gatherAppliedIdsFromSlottedChildren(target, this.describedbyId);
        if (ids.length && ids.length < 4) {
            this.conditionDescribedby = conditionAttributeWithId(this, 'aria-describedby', ids);
        }
        else if (!ids.length) {
            const idProvided = !!this.id;
            if (!idProvided)
                this.id = this.describedbyId;
            const conditionDescribedby = conditionAttributeWithId(this, 'aria-describedby', this.id);
            this.conditionDescribedby = () => {
                conditionDescribedby();
                if (!idProvided) {
                    this.removeAttribute('id');
                }
            };
        }
    }
    connectedCallback() {
        super.connectedCallback();
        this.tabIndex = 0;
        window.addEventListener('resize', this.shouldManageTabOrderForScrolling);
    }
    disconnectedCallback() {
        window.removeEventListener('resize', this.shouldManageTabOrderForScrolling);
        super.disconnectedCallback();
    }
}
Dialog.instanceCount = 0;
__decorate([
    query('.close-button')
], Dialog.prototype, "closeButton", void 0);
__decorate([
    query('.content')
], Dialog.prototype, "contentElement", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], Dialog.prototype, "error", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], Dialog.prototype, "dismissable", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'no-divider' })
], Dialog.prototype, "noDivider", void 0);
__decorate([
    property({ type: String, reflect: true })
], Dialog.prototype, "mode", void 0);
__decorate([
    property({ type: String, reflect: true })
], Dialog.prototype, "size", void 0);
