// @node-projects/acad-ts/dist/Entities/CadWipeoutBase.js
import { Entity } from './Entity.js';
import { CadObject } from '../CadObject.js';
import { ClipMode } from './ClipMode.js';
import { ClipType } from './ClipType.js';
import { ImageDisplayFlags } from './ImageDisplayFlags.js';
import { XYZ } from '../Math/XYZ.js';
import { XY } from '../Math/XY.js';
export class CadWipeoutBase extends Entity {
    get brightness() {
        return this._brightness;
    }
    set brightness(value) {
        if (value < 0 || value > 100) {
            throw new Error(`Invalid Brightness value: ${value}, must be in range 0-100`);
        }
        this._brightness = value;
    }
    classVersion = 0;
    clipBoundaryVertices = [];
    clipMode = ClipMode.Outside;
    clippingState = false;
    clipType = ClipType.Rectangular;
    get contrast() {
        return this._contrast;
    }
    set contrast(value) {
        if (value < 0 || value > 100) {
            throw new Error(`Invalid Contrast value: ${value}, must be in range 0-100`);
        }
        this._contrast = value;
    }
    get definition() {
        return this._definition;
    }
    set definition(value /* ImageDefinition */) {
        if (this.document != null) {
            this._definition = CadObject.updateCollection(value, this.document.imageDefinitions);
        }
        else {
            this._definition = value;
        }
    }
    get fade() {
        return this._fade;
    }
    set fade(value) {
        if (value < 0 || value > 100) {
            throw new Error(`Invalid Fade value: ${value}, must be in range 0-100`);
        }
        this._fade = value;
    }
    get flags() {
        return this._flags;
    }
    set flags(value) {
        this._flags = value;
    }
    insertPoint = new XYZ(0, 0, 0);
    get showImage() {
        return (this._flags & ImageDisplayFlags.ShowImage) !== 0;
    }
    set showImage(value) {
        if (value) {
            this._flags = this._flags | ImageDisplayFlags.ShowImage;
        }
        else {
            this._flags = this._flags & ~ImageDisplayFlags.ShowImage;
        }
    }
    size = new XY(0, 0);
    uVector = new XYZ(1, 0, 0);
    vVector = new XYZ(0, 1, 0);
    /** @internal */
    definitionReactor = null;
    _brightness = 50;
    _contrast = 50;
    _definition = null;
    _fade = 0;
    _flags = ImageDisplayFlags.None;
    applyTransform(transform) {
        // TODO: transform operations not available
    }
    clone() {
        const clone = super.clone();
        clone._definition = this._definition?.clone?.() ?? null;
        return clone;
    }
    getBoundingBox() {
        if (this.clipBoundaryVertices.length === 0) {
            return null;
        }
        let minX = Infinity, minY = Infinity;
        let maxX = -Infinity, maxY = -Infinity;
        for (const v of this.clipBoundaryVertices) {
            if (v.x < minX)
                minX = v.x;
            if (v.y < minY)
                minY = v.y;
            if (v.x > maxX)
                maxX = v.x;
            if (v.y > maxY)
                maxY = v.y;
        }
        const min = new XYZ(minX + this.insertPoint.x, minY + this.insertPoint.y, this.insertPoint.z);
        const max = new XYZ(maxX + this.insertPoint.x, maxY + this.insertPoint.y, this.insertPoint.z);
        return { min, max };
    }
    /** @internal */
    assignDocument(doc) {
        super.assignDocument(doc);
        this._definition = CadObject.updateCollection(this._definition, doc.imageDefinitions);
    }
    /** @internal */
    unassignDocument() {
        super.unassignDocument();
        this._definition = this._definition?.clone?.() ?? null;
    }
    imageDefinitionsOnRemove(sender, e) {
        if (e.item === this._definition) {
            this._definition = null;
        }
    }
}
export { ImageDisplayFlags } from './ImageDisplayFlags.js';
export { ClipMode } from './ClipMode.js';
export { ClipType } from './ClipType.js';
