// @node-projects/acad-ts/dist/Entities/Circle.js
import { Entity } from './Entity.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { XYZ } from '../Math/XYZ.js';
export class Circle extends Entity {
    center = new XYZ(0, 0, 0);
    normal = new XYZ(0, 0, 1);
    get objectName() {
        return DxfFileToken.EntityCircle;
    }
    get objectType() {
        return ObjectType.CIRCLE;
    }
    get radius() {
        return this._radius;
    }
    set radius(value) {
        if (value <= 0) {
            throw new Error('The radius must be greater than 0.');
        }
        this._radius = value;
    }
    get radiusRatio() {
        return 1;
    }
    get subclassMarker() {
        return DxfSubclassMarker.Circle;
    }
    thickness = 0.0;
    _radius = 1.0;
    constructor() {
        super();
    }
    applyTransform(transform) {
        // TODO: Transform operations not available
    }
    getBoundingBox() {
        const min = new XYZ(this.center.x - this._radius, this.center.y - this._radius, this.center.z);
        const max = new XYZ(this.center.x + this._radius, this.center.y + this._radius, this.center.z);
        return { min, max };
    }
    polarCoordinateRelativeToCenter(angle) {
        // TODO: CurveExtensions.PolarCoordinate not available
        return new XYZ(0, 0, 0);
    }
    polygonalVertexes(precision) {
        if (precision < 2) {
            throw new Error('The arc precision must be equal or greater than two.');
        }
        // TODO: CurveExtensions.PolygonalVertexes not available
        return [];
    }
}
