// @node-projects/acad-ts/dist/Entities/DimensionLinear.js
import { DimensionAligned } from './DimensionAligned.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { DimensionType } from './DimensionType.js';
import { XYZ } from '../Math/XYZ.js';
export class DimensionLinear extends DimensionAligned {
    get measurement() {
        const angle = new XYZ(Math.cos(this.rotation), Math.sin(this.rotation), 0);
        const diff = new XYZ(this.secondPoint.x - this.firstPoint.x, this.secondPoint.y - this.firstPoint.y, this.secondPoint.z - this.firstPoint.z);
        const len = Math.sqrt(diff.x * diff.x + diff.y * diff.y + diff.z * diff.z);
        if (len === 0)
            return 0;
        const norm = new XYZ(diff.x / len, diff.y / len, diff.z / len);
        const dot = Math.abs(angle.x * norm.x + angle.y * norm.y + angle.z * norm.z);
        return super.measurement * dot;
    }
    get objectName() {
        return DxfFileToken.EntityDimension;
    }
    get objectType() {
        return ObjectType.DIMENSION_LINEAR;
    }
    get offset() {
        return super.offset;
    }
    set offset(value) {
        // TODO: Transform.CreateRotation not available
    }
    rotation = 0;
    get subclassMarker() {
        return DxfSubclassMarker.LinearDimension;
    }
    constructor() {
        super();
        this._flags = DimensionType.Linear | DimensionType.BlockReference;
    }
    updateBlock() {
        this.createBlock();
        // TODO: Complex block generation with rotation
    }
}
