// @node-projects/acad-ts/dist/Entities/Dimension.js
import { Entity } from './Entity.js';
import { CadObject } from '../CadObject.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { DimensionStyle } from '../Tables/DimensionStyle.js';
import { BlockRecord } from '../Tables/BlockRecord.js';
import { DimensionType } from './DimensionType.js';
import { AttachmentPointType } from './AttachmentPointType.js';
import { LineSpacingStyleType } from './LineSpacingStyleType.js';
import { XYZ } from '../Math/XYZ.js';
export class Dimension extends Entity {
    attachmentPoint = AttachmentPointType.TopLeft;
    get block() {
        return this._block;
    }
    set block(value) {
        if (this.document != null) {
            this._block = CadObject.updateCollection(value, this.document.blockRecords);
        }
        else {
            this._block = value;
        }
    }
    definitionPoint = new XYZ(0, 0, 0);
    get flags() {
        return this._flags;
    }
    /** @internal */
    set flags(value) {
        this._flags = value;
    }
    flipArrow1 = false;
    flipArrow2 = false;
    get hasStyleOverride() {
        // TODO: ExtendedData access not available
        return false;
    }
    horizontalDirection = 0;
    insertionPoint = new XYZ(0, 0, 0);
    get isAngular() {
        return (this._flags & DimensionType.Angular3Point) !== 0 || (this._flags & DimensionType.Angular) !== 0;
    }
    get isTextUserDefinedLocation() {
        return (this._flags & DimensionType.TextUserDefinedLocation) !== 0;
    }
    set isTextUserDefinedLocation(value) {
        if (value) {
            this._flags = this._flags | DimensionType.TextUserDefinedLocation;
        }
        else {
            this._flags = this._flags & ~DimensionType.TextUserDefinedLocation;
        }
    }
    lineSpacingFactor = 0;
    lineSpacingStyle = LineSpacingStyleType.AtLeast;
    normal = new XYZ(0, 0, 1);
    get style() {
        return this._style;
    }
    set style(value) {
        if (value == null) {
            throw new Error('value cannot be null');
        }
        if (this.document != null) {
            this._style = CadObject.updateCollection(value, this.document.dimensionStyles);
        }
        else {
            this._style = value;
        }
    }
    get subclassMarker() {
        return DxfSubclassMarker.Dimension;
    }
    text = '';
    textMiddlePoint = new XYZ(0, 0, 0);
    textRotation = 0;
    version = 0;
    _block = null;
    _flags;
    _style = DimensionStyle.Default;
    constructor(type) {
        super();
        this._flags = type;
        this._flags |= DimensionType.BlockReference;
    }
    applyTransform(transform) {
        // TODO: Complex transform with world matrix
    }
    clone() {
        const clone = super.clone();
        clone._style = this._style.clone();
        clone._block = this._block?.clone() ?? null;
        return clone;
    }
    getActiveDimensionStyle() {
        if (!this.hasStyleOverride) {
            return this.style;
        }
        // TODO: Style override map not available
        return this.style;
    }
    getMeasurementText(style) {
        // TODO: Complex measurement text formatting not available
        return this.measurement.toString();
    }
    getStyleOverrideMap() {
        // TODO: DxfClassMap not available
        return null;
    }
    setDimensionOverride(styleOverride) {
        // TODO: DxfClassMap not available
    }
    setStyleOverrideMap(map) {
        // TODO: ExtendedData not available
    }
    updateBlock() {
        this.createBlock();
    }
    /** @internal */
    assignDocument(doc) {
        super.assignDocument(doc);
        this._style = CadObject.updateCollection(this._style, doc.dimensionStyles);
        this._block = CadObject.updateCollection(this._block, doc.blockRecords);
        if (this._block != null) {
            this._block.name = this.generateBlockName();
        }
        this._block = CadObject.updateCollection(this._block, doc.blockRecords);
    }
    /** @internal */
    unassignDocument() {
        super.unassignDocument();
        this._style = this._style?.clone();
        this._block = this._block?.clone() ?? null;
    }
    createBlock() {
        if (this._block == null) {
            this._block = new BlockRecord(this.generateBlockName());
            this._block.isAnonymous = true;
        }
        if (this.document != null) {
            this._block = CadObject.updateCollection(this._block, this.document.blockRecords);
        }
        this._block.entities?.clear?.();
    }
    createDefinitionPoint(location) {
        // TODO: Point with Defpoints layer
        return null;
    }
    createTextEntity(insertPoint, text) {
        // TODO: MText creation
        return null;
    }
    _tableOnRemove(sender, e) {
        super._tableOnRemove(sender, e);
        if (e.item === this._style) {
            this._style = this.document.dimensionStyles.get(DimensionStyle.DefaultName);
        }
        if (e.item === this._block) {
            this._block = null;
        }
    }
    generateBlockName() {
        return `*D${this.handle}`;
    }
}
export { DimensionType } from './DimensionType.js';
