// @node-projects/acad-ts/dist/Entities/DimensionAligned.js
import { Dimension } from './Dimension.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { DimensionType } from './DimensionType.js';
import { XYZ } from '../Math/XYZ.js';
function distanceFrom(a, b) {
    const dx = a.x - b.x;
    const dy = a.y - b.y;
    const dz = a.z - b.z;
    return Math.sqrt(dx * dx + dy * dy + dz * dz);
}
export class DimensionAligned extends Dimension {
    extLineRotation = 0;
    firstPoint = new XYZ(0, 0, 0);
    get measurement() {
        return distanceFrom(this.firstPoint, this.secondPoint);
    }
    get objectName() {
        return DxfFileToken.EntityDimension;
    }
    get objectType() {
        return ObjectType.DIMENSION_ALIGNED;
    }
    get offset() {
        return distanceFrom(this.secondPoint, this.definitionPoint);
    }
    set offset(value) {
        // TODO: XYZ.Cross not available for perpendicular calculation
    }
    secondPoint = new XYZ(0, 0, 0);
    get subclassMarker() {
        return DxfSubclassMarker.AlignedDimension;
    }
    constructor(firstPoint, secondPoint) {
        super(DimensionType.Aligned);
        if (firstPoint) {
            this.firstPoint = firstPoint;
        }
        if (secondPoint) {
            this.secondPoint = secondPoint;
        }
    }
    /** @internal */
    static createWithType(type) {
        const d = new DimensionAligned();
        d._flags = type;
        d.definitionPoint = d.secondPoint;
        return d;
    }
    applyTransform(transform) {
        super.applyTransform(transform);
        // TODO: Transform FirstPoint and SecondPoint
    }
    getBoundingBox() {
        return {
            min: {
                x: Math.min(this.firstPoint.x, this.secondPoint.x),
                y: Math.min(this.firstPoint.y, this.secondPoint.y),
                z: Math.min(this.firstPoint.z, this.secondPoint.z),
            },
            max: {
                x: Math.max(this.firstPoint.x, this.secondPoint.x),
                y: Math.max(this.firstPoint.y, this.secondPoint.y),
                z: Math.max(this.firstPoint.z, this.secondPoint.z),
            },
        };
    }
    updateBlock() {
        super.updateBlock();
        // TODO: Complex block generation
    }
}
