// @node-projects/acad-ts/dist/Entities/DimensionAngular2Line.js
import { Dimension } from './Dimension.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { DimensionType } from './DimensionType.js';
import { XYZ } from '../Math/XYZ.js';
export class DimensionAngular2Line extends Dimension {
    angleVertex = new XYZ(0, 0, 0);
    get center() {
        // TODO: Line3D intersection not available
        return new XYZ(0, 0, 0);
    }
    dimensionArc = new XYZ(0, 0, 0);
    firstPoint = new XYZ(0, 0, 0);
    get measurement() {
        // TODO: AngleBetweenVectors not available
        return 0;
    }
    get objectName() {
        return DxfFileToken.EntityDimension;
    }
    get objectType() {
        return ObjectType.DIMENSION_ANG_2_Ln;
    }
    get offset() {
        const dx = this.secondPoint.x - this.definitionPoint.x;
        const dy = this.secondPoint.y - this.definitionPoint.y;
        const dz = this.secondPoint.z - this.definitionPoint.z;
        return Math.sqrt(dx * dx + dy * dy + dz * dz);
    }
    set offset(value) {
        // TODO: Cross product not available
    }
    secondPoint = new XYZ(0, 0, 0);
    get subclassMarker() {
        return DxfSubclassMarker.Angular2LineDimension;
    }
    constructor() {
        super(DimensionType.Angular);
    }
    applyTransform(transform) {
        super.applyTransform(transform);
        // TODO: Transform points
    }
    getBoundingBox() {
        return {
            min: {
                x: Math.min(this.firstPoint.x, this.secondPoint.x),
                y: Math.min(this.firstPoint.y, this.secondPoint.y),
                z: Math.min(this.firstPoint.z, this.secondPoint.z),
            },
            max: {
                x: Math.max(this.firstPoint.x, this.secondPoint.x),
                y: Math.max(this.firstPoint.y, this.secondPoint.y),
                z: Math.max(this.firstPoint.z, this.secondPoint.z),
            },
        };
    }
    updateBlock() {
        // Not yet implemented in C# source
        return;
    }
}
