// @node-projects/acad-ts/dist/Entities/DimensionAngular3Pt.js
import { Dimension } from './Dimension.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { DimensionType } from './DimensionType.js';
import { XYZ } from '../Math/XYZ.js';
export class DimensionAngular3Pt extends Dimension {
    angleVertex = new XYZ(0, 0, 0);
    firstPoint = new XYZ(0, 0, 0);
    get measurement() {
        // TODO: AngleBetweenVectors / IsParallel not available
        return 0;
    }
    get objectName() {
        return DxfFileToken.EntityDimension;
    }
    get objectType() {
        return ObjectType.DIMENSION_ANG_3_Pt;
    }
    secondPoint = new XYZ(0, 0, 0);
    get subclassMarker() {
        return DxfSubclassMarker.Angular3PointDimension;
    }
    constructor() {
        super(DimensionType.Angular3Point);
    }
    applyTransform(transform) {
        super.applyTransform(transform);
        // TODO: Transform points
    }
    getBoundingBox() {
        return {
            min: {
                x: Math.min(this.firstPoint.x, this.secondPoint.x),
                y: Math.min(this.firstPoint.y, this.secondPoint.y),
                z: Math.min(this.firstPoint.z, this.secondPoint.z),
            },
            max: {
                x: Math.max(this.firstPoint.x, this.secondPoint.x),
                y: Math.max(this.firstPoint.y, this.secondPoint.y),
                z: Math.max(this.firstPoint.z, this.secondPoint.z),
            },
        };
    }
    updateBlock() {
        super.updateBlock();
    }
}
