// @node-projects/acad-ts/dist/Entities/DimensionRadius.js
import { Dimension } from './Dimension.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { DimensionType } from './DimensionType.js';
import { XYZ } from '../Math/XYZ.js';
function distanceFrom(a, b) {
    const dx = a.x - b.x;
    const dy = a.y - b.y;
    const dz = a.z - b.z;
    return Math.sqrt(dx * dx + dy * dy + dz * dz);
}
export class DimensionRadius extends Dimension {
    angleVertex = new XYZ(0, 0, 0);
    leaderLength = 0;
    get measurement() {
        return distanceFrom(this.definitionPoint, this.angleVertex);
    }
    get objectName() {
        return DxfFileToken.EntityDimension;
    }
    get objectType() {
        return ObjectType.DIMENSION_RADIUS;
    }
    get subclassMarker() {
        return DxfSubclassMarker.RadialDimension;
    }
    constructor() {
        super(DimensionType.Radius);
    }
    applyTransform(transform) {
        super.applyTransform(transform);
        // TODO: Transform AngleVertex
    }
    getBoundingBox() {
        return {
            min: new XYZ(this.insertionPoint.x - this.angleVertex.x, this.insertionPoint.y - this.angleVertex.y, this.insertionPoint.z - this.angleVertex.z),
            max: new XYZ(this.insertionPoint.x + this.angleVertex.x, this.insertionPoint.y + this.angleVertex.y, this.insertionPoint.z + this.angleVertex.z),
        };
    }
    updateBlock() {
        super.updateBlock();
        // TODO: Complex block generation via angularBlock
    }
}
