// @node-projects/acad-ts/dist/Entities/DimensionOrdinate.js
import { Dimension } from './Dimension.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { DimensionType } from './DimensionType.js';
import { XYZ } from '../Math/XYZ.js';
import { XY } from '../Math/XY.js';
export class DimensionOrdinate extends Dimension {
    featureLocation = new XYZ(0, 0, 0);
    get isOrdinateTypeX() {
        return (this._flags & DimensionType.OrdinateTypeX) !== 0;
    }
    set isOrdinateTypeX(value) {
        if (value) {
            this._flags = this._flags | DimensionType.OrdinateTypeX;
        }
        else {
            this._flags = this._flags & ~DimensionType.OrdinateTypeX;
        }
    }
    leaderEndpoint = new XYZ(0, 0, 0);
    get measurement() {
        const dirBase = this.isOrdinateTypeX ? new XY(0, 1) : new XY(1, 0);
        const sin = Math.sin(this.horizontalDirection);
        const cos = Math.cos(this.horizontalDirection);
        const dir = new XY(dirBase.x * cos - dirBase.y * sin, dirBase.x * sin + dirBase.y * cos);
        const diff = new XY(this.definitionPoint.x - this.featureLocation.x, this.definitionPoint.y - this.featureLocation.y);
        const t = dir.x * diff.x + dir.y * diff.y;
        const pr = new XY(this.featureLocation.x + t * dir.x, this.featureLocation.y + t * dir.y);
        const v = new XY(this.definitionPoint.x - pr.x, this.definitionPoint.y - pr.y);
        const distSqrt = v.x * v.x + v.y * v.y;
        return Math.sqrt(distSqrt);
    }
    get objectName() {
        return DxfFileToken.EntityDimension;
    }
    get objectType() {
        return ObjectType.DIMENSION_ORDINATE;
    }
    get subclassMarker() {
        return DxfSubclassMarker.OrdinateDimension;
    }
    constructor() {
        super(DimensionType.Ordinate);
    }
    applyTransform(transform) {
        super.applyTransform(transform);
        // TODO: Transform FeatureLocation and LeaderEndpoint
    }
    getBoundingBox() {
        return {
            min: {
                x: Math.min(this.featureLocation.x, this.leaderEndpoint.x),
                y: Math.min(this.featureLocation.y, this.leaderEndpoint.y),
                z: Math.min(this.featureLocation.z, this.leaderEndpoint.z),
            },
            max: {
                x: Math.max(this.featureLocation.x, this.leaderEndpoint.x),
                y: Math.max(this.featureLocation.y, this.leaderEndpoint.y),
                z: Math.max(this.featureLocation.z, this.leaderEndpoint.z),
            },
        };
    }
    updateBlock() {
        super.updateBlock();
        // TODO: Complex block generation
    }
}
