// @node-projects/acad-ts/dist/Entities/Ellipse.js
import { Entity } from './Entity.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { XYZ } from '../Math/XYZ.js';
export class Ellipse extends Entity {
    center = new XYZ(0, 0, 0);
    endParameter = Math.PI * 2;
    get isFullEllipse() {
        return this.startParameter === 0 && this.endParameter === Math.PI * 2;
    }
    get majorAxis() {
        return 2 * Math.sqrt(this.majorAxisEndPoint.x ** 2 +
            this.majorAxisEndPoint.y ** 2 +
            this.majorAxisEndPoint.z ** 2);
    }
    majorAxisEndPoint = new XYZ(1, 0, 0);
    get minorAxis() {
        return this.majorAxis * this.radiusRatio;
    }
    get minorAxisEndpoint() {
        const len = Math.sqrt(this.majorAxisEndPoint.x ** 2 +
            this.majorAxisEndPoint.y ** 2 +
            this.majorAxisEndPoint.z ** 2);
        if (len === 0)
            return new XYZ(0, 0, 0);
        const nx = this.majorAxisEndPoint.x / len;
        const ny = this.majorAxisEndPoint.y / len;
        const nz = this.majorAxisEndPoint.z / len;
        // Cross product of Normal and normalized MajorAxisEndPoint
        const dx = this.normal.y * nz - this.normal.z * ny;
        const dy = this.normal.z * nx - this.normal.x * nz;
        const dz = this.normal.x * ny - this.normal.y * nx;
        const dLen = Math.sqrt(dx * dx + dy * dy + dz * dz);
        if (dLen === 0)
            return new XYZ(0, 0, 0);
        const scale = this.radiusRatio * len;
        return new XYZ((dx / dLen) * scale, (dy / dLen) * scale, (dz / dLen) * scale);
    }
    normal = new XYZ(0, 0, 1);
    get objectName() {
        return DxfFileToken.EntityEllipse;
    }
    get objectType() {
        return ObjectType.ELLIPSE;
    }
    get radiusRatio() {
        return this._radiusRatio;
    }
    set radiusRatio(value) {
        if (value <= 0 || value > 1) {
            throw new Error('Radius ratio must be a value between 0 (not included) and 1.');
        }
        this._radiusRatio = value;
    }
    get rotation() {
        return Math.atan2(this.majorAxisEndPoint.y, this.majorAxisEndPoint.x);
    }
    startParameter = 0.0;
    get subclassMarker() {
        return DxfSubclassMarker.Ellipse;
    }
    thickness = 0.0;
    _radiusRatio = 1.0;
    applyTransform(transform) {
        // TODO: Transform operations not available
    }
    getBoundingBox() {
        // TODO: PolygonalVertexes and BoundingBox.FromPoints not available
        return null;
    }
    getEndVertices() {
        const start = this.polarCoordinateRelativeToCenter(this.startParameter);
        const end = this.polarCoordinateRelativeToCenter(this.endParameter);
        return { start, end };
    }
    polarCoordinateRelativeToCenter(angle) {
        // TODO: CurveExtensions.PolarCoordinate not available
        return new XYZ(0, 0, 0);
    }
    polygonalVertexes(precision) {
        // TODO: CurveExtensions.PolygonalVertexes not available
        return [];
    }
}
