// @node-projects/acad-ts/dist/Entities/Entity.js
import { CadObject } from '../CadObject.js';
import { Color } from '../Color.js';
import { Layer } from '../Tables/Layer.js';
import { LineType } from '../Tables/LineType.js';
import { LineWeightType } from '../Types/LineWeightType.js';
import { Transparency } from '../Transparency.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { XY } from '../Math/XY.js';
function isBlockRecordOwner(value) {
    return value != null && typeof value === 'object' && 'blockEntity' in value;
}
export class Entity extends CadObject {
    bookColor = null;
    color = Color.ByLayer;
    isInvisible = false;
    get layer() {
        return this._layer;
    }
    set layer(value) {
        if (value == null) {
            throw new Error('value cannot be null');
        }
        this._layer = CadObject.updateCollection(value, this.document?.layers ?? null);
    }
    get lineType() {
        return this._lineType;
    }
    set lineType(value) {
        if (value == null) {
            throw new Error('value cannot be null');
        }
        this._lineType = CadObject.updateCollection(value, this.document?.lineTypes ?? null);
    }
    lineTypeScale = 1.0;
    lineWeight = LineWeightType.ByLayer;
    material = null;
    get subclassMarker() {
        return DxfSubclassMarker.Entity;
    }
    transparency = Transparency.ByLayer;
    _bookColor = null;
    _layer = Layer.Default;
    _lineType = LineType.ByLayer;
    constructor() {
        super();
    }
    applyRotation(axis, rotation) {
        // TODO: Transform.CreateRotation not available
        // const transform = Transform.CreateRotation(axis, rotation);
        // this.applyTransform(transform);
    }
    applyScaling(scale, origin) {
        // TODO: Transform.CreateScaling not available
    }
    applyTranslation(translation) {
        // TODO: Transform.CreateTranslation not available
    }
    clone() {
        const clone = super.clone();
        clone._layer = this.layer.clone();
        clone._lineType = this.lineType.clone();
        clone.material = this.material?.clone?.() ?? null;
        return clone;
    }
    getActiveColor() {
        let color;
        if (this.color.isByLayer) {
            color = this.layer.color;
        }
        else if (this.color.isByBlock && isBlockRecordOwner(this.owner)) {
            color = this.owner.blockEntity.color;
        }
        else {
            color = this.color;
        }
        return color;
    }
    getActiveLineType() {
        if (this.lineType.name.toUpperCase() === LineType.ByLayerName.toUpperCase()) {
            return this.layer.lineType;
        }
        else if (this.lineType.name.toUpperCase() === LineType.ByBlockName.toUpperCase() &&
            isBlockRecordOwner(this.owner)) {
            return this.owner.blockEntity.lineType;
        }
        return this.lineType;
    }
    getActiveLineWeightType() {
        switch (this.lineWeight) {
            case LineWeightType.ByLayer:
                return this.layer.lineWeight;
            case LineWeightType.ByBlock:
                if (isBlockRecordOwner(this.owner)) {
                    return this.owner.blockEntity.getActiveLineWeightType();
                }
                else {
                    return this.lineWeight;
                }
            default:
                return this.lineWeight;
        }
    }
    matchProperties(entity) {
        if (entity == null) {
            throw new Error('entity cannot be null');
        }
        if (entity.handle === 0 || this.document !== entity.document) {
            this.layer = entity.layer.clone();
            this.lineType = entity.lineType.clone();
            this.material = entity.material?.clone?.() ?? null;
        }
        else {
            this.layer = entity.layer;
            this.lineType = entity.lineType;
            this.material = entity.material;
        }
        this.color = entity.color;
        this.lineWeight = entity.lineWeight;
        this.lineTypeScale = entity.lineTypeScale;
        this.isInvisible = entity.isInvisible;
        this.transparency = entity.transparency;
    }
    /** @internal */
    assignDocument(doc) {
        super.assignDocument(doc);
        this._layer = CadObject.updateCollection(this.layer, doc.layers);
        this._lineType = CadObject.updateCollection(this.lineType, doc.lineTypes);
        doc.layers.onRemove = this._tableOnRemove.bind(this);
        doc.lineTypes.onRemove = this._tableOnRemove.bind(this);
    }
    /** @internal */
    unassignDocument() {
        // TODO: Remove event listeners properly
        super.unassignDocument();
        this._layer = this.layer.clone();
        this._lineType = this.lineType.clone();
    }
    applyRotationToPoints(points, rotation) {
        if (points == null) {
            throw new Error('points cannot be null');
        }
        if (Math.abs(rotation) < 1e-12) {
            return [...points];
        }
        const sin = Math.sin(rotation);
        const cos = Math.cos(rotation);
        const transPoints = [];
        for (const p of points) {
            transPoints.push(new XY(p.x * cos - p.y * sin, p.x * sin + p.y * cos));
        }
        return transPoints;
    }
    applyWorldMatrix(xyz, transform /* Transform */, transOW /* Matrix3 */, transWO /* Matrix3 */) {
        // TODO: Matrix3 operations not available
        return xyz;
    }
    getWorldMatrix(transform /* Transform */, normal, newNormal) {
        // TODO: Matrix3.ArbitraryAxis not available
        return { matrix: null, transOW: null, transWO: null };
    }
    _tableOnRemove(sender, e) {
        if (e.item === this.layer) {
            this.layer = this.document.layers.get(Layer.DefaultName);
        }
        if (e.item === this.lineType) {
            this.lineType = this.document.lineTypes.get(LineType.ByLayerName);
        }
    }
    transformNormal(transform /* Transform */, normal) {
        // TODO: transform.ApplyRotation and Normalize not available
        return normal;
    }
}
