// @node-projects/acad-ts/dist/Entities/Face3D.js
import { Entity } from './Entity.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { InvisibleEdgeFlags } from './InvisibleEdgeFlags.js';
import { XYZ } from '../Math/XYZ.js';
export class Face3D extends Entity {
    get objectType() {
        return ObjectType.FACE3D;
    }
    get objectName() {
        return DxfFileToken.Entity3DFace;
    }
    get subclassMarker() {
        return DxfSubclassMarker.Face3d;
    }
    firstCorner = new XYZ(0, 0, 0);
    secondCorner = new XYZ(0, 0, 0);
    thirdCorner = new XYZ(0, 0, 0);
    fourthCorner = new XYZ(0, 0, 0);
    flags = InvisibleEdgeFlags.None;
    constructor() {
        super();
    }
    getBoundingBox() {
        // TODO: BoundingBox.FromPoints not available
        return null;
    }
    applyTransform(transform) {
        // TODO: transform.ApplyTransform not available
    }
}
export { InvisibleEdgeFlags } from './InvisibleEdgeFlags.js';
