// @node-projects/acad-ts/dist/Entities/Hatch.js
import { Entity } from './Entity.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { HatchGradientPattern } from './HatchGradientPattern.js';
import { HatchPattern } from './HatchPattern.js';
import { HatchPatternType } from './HatchPatternType.js';
import { HatchStyleType } from './HatchStyleType.js';
import { BoundaryPathFlags } from './BoundaryPathFlags.js';
import { XYZ } from '../Math/XYZ.js';
import { XY } from '../Math/XY.js';
export var EdgeType;
(function (EdgeType) {
    EdgeType[EdgeType["Polyline"] = 0] = "Polyline";
    EdgeType[EdgeType["Line"] = 1] = "Line";
    EdgeType[EdgeType["CircularArc"] = 2] = "CircularArc";
    EdgeType[EdgeType["EllipticArc"] = 3] = "EllipticArc";
    EdgeType[EdgeType["Spline"] = 4] = "Spline";
})(EdgeType || (EdgeType = {}));
export class HatchBoundaryPathEdge {
    clone() {
        return Object.assign(Object.create(Object.getPrototypeOf(this)), this);
    }
}
export class HatchBoundaryPathArc extends HatchBoundaryPathEdge {
    center = new XY(0, 0);
    counterClockWise = false;
    endAngle = 0;
    radius = 0;
    startAngle = 0;
    get type() { return EdgeType.CircularArc; }
    applyTransform(transform) {
        // TODO: Transform
    }
    getBoundingBox() { return null; }
    polygonalVertexes(precision) {
        // TODO: Arc.PolygonalVertexes
        return [];
    }
    toEntity() {
        // TODO: Create Arc entity
        return null;
    }
}
export class HatchBoundaryPathEllipse extends HatchBoundaryPathEdge {
    center = new XY(0, 0);
    counterClockWise = false;
    endAngle = 0;
    majorAxisEndPoint = new XY(0, 0);
    minorToMajorRatio = 0;
    startAngle = 0;
    get type() { return EdgeType.EllipticArc; }
    applyTransform(transform) {
        // TODO: Transform
    }
    getBoundingBox() { return null; }
    polygonalVertexes(precision) {
        return [];
    }
    toEntity() {
        // TODO: Create Ellipse entity
        return null;
    }
}
export class HatchBoundaryPathLine extends HatchBoundaryPathEdge {
    end = new XY(0, 0);
    start = new XY(0, 0);
    get type() { return EdgeType.Line; }
    applyTransform(transform) {
        // TODO: Transform
    }
    getBoundingBox() { return null; }
    toEntity() {
        // TODO: Create Line entity
        return null;
    }
}
export class HatchBoundaryPathPolyline extends HatchBoundaryPathEdge {
    get bulges() { return this.vertices.map(v => v.z); }
    get hasBulge() { return this.bulges.some(b => b !== 0); }
    isClosed = false;
    get type() { return EdgeType.Polyline; }
    vertices = [];
    applyTransform(transform) {
        // TODO: Transform
    }
    clone() {
        const c = super.clone();
        c.vertices = [...this.vertices];
        return c;
    }
    getBoundingBox() { return null; }
    toEntity() {
        // TODO: Create Polyline2D entity
        return null;
    }
}
export class HatchBoundaryPathSpline extends HatchBoundaryPathEdge {
    controlPoints = [];
    degree = 0;
    endTangent = new XY(0, 0);
    fitPoints = [];
    knots = [];
    isPeriodic = false;
    isRational = false;
    startTangent = new XY(0, 0);
    get weights() { return this.controlPoints.map(c => c.z); }
    get type() { return EdgeType.Spline; }
    applyTransform(transform) {
        // TODO: Transform
    }
    clone() {
        const c = super.clone();
        c.controlPoints = [...this.controlPoints];
        c.fitPoints = [...this.fitPoints];
        c.knots = [...this.knots];
        return c;
    }
    getBoundingBox() { return null; }
    polygonalVertexes(precision) {
        return [];
    }
    toEntity() {
        // TODO: Create Spline entity
        return null;
    }
}
export class HatchBoundaryPath {
    edges = [];
    entities = [];
    get flags() {
        let f = this._flags;
        if (this.isPolyline) {
            f = f | BoundaryPathFlags.Polyline;
        }
        else {
            f = f & ~BoundaryPathFlags.Polyline;
        }
        return f;
    }
    set flags(value) {
        this._flags = value;
    }
    get isPolyline() {
        return this.edges.some(e => e instanceof HatchBoundaryPathPolyline);
    }
    _flags = 0;
    constructor(edges) {
        if (edges) {
            this.edges = [...edges];
        }
    }
    applyTransform(transform) {
        for (const e of this.edges) {
            e.applyTransform(transform);
        }
    }
    clone() {
        const path = new HatchBoundaryPath();
        path._flags = this._flags;
        path.entities = this.entities.map(e => e.clone());
        path.edges = this.edges.map(e => e.clone());
        return path;
    }
    getBoundingBox() {
        return null;
    }
    getPoints(precision = 256) {
        const pts = [];
        for (const edge of this.edges) {
            if (edge instanceof HatchBoundaryPathArc) {
                pts.push(...edge.polygonalVertexes(precision));
            }
            else if (edge instanceof HatchBoundaryPathEllipse) {
                pts.push(...edge.polygonalVertexes(precision));
            }
            else if (edge instanceof HatchBoundaryPathLine) {
                pts.push(new XYZ(edge.start.x, edge.start.y, 0));
                pts.push(new XYZ(edge.end.x, edge.end.y, 0));
            }
            else if (edge instanceof HatchBoundaryPathSpline) {
                pts.push(...edge.polygonalVertexes(precision));
            }
        }
        return pts;
    }
    updateEdges() {
        if (this.entities.length === 0)
            return;
        this.edges = [];
        // TODO: Convert entities to edges
    }
}
export class Hatch extends Entity {
    elevation = 0;
    gradientColor = new HatchGradientPattern();
    isAssociative = false;
    isDouble = false;
    isSolid = false;
    normal = new XYZ(0, 0, 1);
    get objectName() {
        return DxfFileToken.EntityHatch;
    }
    get objectType() {
        return ObjectType.HATCH;
    }
    paths = [];
    pattern = HatchPattern.Solid;
    get patternAngle() {
        return this._patternAngle;
    }
    set patternAngle(value) {
        this._patternAngle = value;
        this.pattern?.update(new XY(0, 0), this._patternAngle, 1);
    }
    get patternScale() {
        return this._patternScale;
    }
    set patternScale(value) {
        this._patternScale = value;
        this.pattern?.update(new XY(0, 0), 0, this._patternScale);
    }
    patternType = HatchPatternType.PatternFill;
    pixelSize = 0;
    seedPoints = [];
    style = HatchStyleType.Normal;
    get subclassMarker() {
        return DxfSubclassMarker.Hatch;
    }
    _patternAngle = 0;
    _patternScale = 0;
    applyTransform(transform) {
        for (const p of this.paths) {
            p.applyTransform(transform);
        }
        // TODO: Pattern angle/scale recalculation
    }
    clone() {
        const clone = super.clone();
        clone.gradientColor = this.gradientColor?.clone();
        clone.pattern = this.pattern?.clone();
        clone.paths = this.paths.map(p => p.clone());
        return clone;
    }
    *explode() {
        for (const b of this.paths) {
            for (const e of b.edges) {
                yield e.toEntity();
            }
        }
    }
    getBoundingBox() {
        return null;
    }
}
export { HatchPattern } from './HatchPattern.js';
export { HatchStyleType } from './HatchStyleType.js';
export { HatchPatternType } from './HatchPatternType.js';
export { BoundaryPathFlags } from './BoundaryPathFlags.js';
export { HatchGradientPattern } from './HatchGradientPattern.js';
