// @node-projects/acad-ts/dist/Entities/Insert.js
import { Entity } from './Entity.js';
import { SeqendCollection } from './SeqendCollection.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { AttributeEntity } from './AttributeEntity.js';
import { XYZ } from '../Math/XYZ.js';
export class Insert extends Entity {
    attributes = new SeqendCollection();
    block = null;
    columnCount = 1;
    columnSpacing = 0;
    get hasAttributes() {
        return this.attributes.length > 0;
    }
    get hasDynamicSubclass() {
        return true;
    }
    insertPoint = new XYZ(0, 0, 0);
    get isMultiple() {
        return this.rowCount > 1 || this.columnCount > 1;
    }
    normal = new XYZ(0, 0, 1);
    get objectName() {
        return DxfFileToken.EntityInsert;
    }
    get objectType() {
        if (this.rowCount > 1 || this.columnCount > 1) {
            return ObjectType.MINSERT;
        }
        return ObjectType.INSERT;
    }
    rotation = 0.0;
    rowCount = 1;
    rowSpacing = 0;
    get subclassMarker() {
        return this.isMultiple ? DxfSubclassMarker.MInsert : DxfSubclassMarker.Insert;
    }
    get xScale() {
        return this._xscale;
    }
    set xScale(value) {
        if (value === 0) {
            throw new Error('XScale value must be non-zero.');
        }
        this._xscale = value;
    }
    get yScale() {
        return this._yscale;
    }
    set yScale(value) {
        if (value === 0) {
            throw new Error('YScale value must be non-zero.');
        }
        this._yscale = value;
    }
    get zScale() {
        return this._zscale;
    }
    set zScale(value) {
        if (value === 0) {
            throw new Error('ZScale value must be non-zero.');
        }
        this._zscale = value;
    }
    _xscale = 1;
    _yscale = 1;
    _zscale = 1;
    constructor(block) {
        super();
        if (block) {
            if (block.document != null) {
                this.block = block.clone();
            }
            else {
                this.block = block;
            }
            const attDefs = block.attributeDefinitions;
            if (attDefs) {
                for (const item of attDefs) {
                    this.attributes.push(new AttributeEntity(item));
                }
            }
        }
    }
    applyTransform(transform) {
        // TODO: Complex transform with matrix operations
        for (const att of this.attributes) {
            att.applyTransform(transform);
        }
    }
    clone() {
        const clone = super.clone();
        clone.block = this.block?.clone() ?? null;
        clone.attributes = new SeqendCollection();
        for (const att of this.attributes) {
            clone.attributes.push(att.clone());
        }
        return clone;
    }
    *explode() {
        // TODO: Complex explode logic with transform
    }
    getBoundingBox() {
        // TODO: Block.GetBoundingBox with transform
        return null;
    }
    getTransform() {
        // TODO: Matrix4 operations not available
        return null;
    }
    updateAttributes() {
        if (!this.block)
            return;
        const attDefs = this.block.attributeDefinitions ?? [];
        const defTags = attDefs.map(d => d.tag);
        const attTags = this.attributes.map(a => a.tag);
        const filtered = new SeqendCollection(...this.attributes.filter(att => defTags.includes(att.tag)));
        filtered.Seqend = this.attributes.Seqend;
        this.attributes = filtered;
        for (const attdef of attDefs) {
            if (!attTags.includes(attdef.tag)) {
                this.attributes.push(new AttributeEntity(attdef));
            }
        }
    }
    /** @internal */
    assignDocument(doc) {
        super.assignDocument(doc);
        if (this.block == null)
            return;
        const existing = doc.blockRecords?.get(this.block.name);
        if (existing) {
            this.block = existing;
        }
        else {
            doc.blockRecords?.add(this.block);
        }
    }
    /** @internal */
    unassignDocument() {
        this.block = this.block?.clone() ?? null;
        super.unassignDocument();
    }
}
