// @node-projects/acad-ts/dist/Entities/Line.js
import { Entity } from './Entity.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { XYZ } from '../Math/XYZ.js';
export class Line extends Entity {
    endPoint = new XYZ(0, 0, 0);
    normal = new XYZ(0, 0, 1);
    get objectName() {
        return DxfFileToken.EntityLine;
    }
    get objectType() {
        return ObjectType.LINE;
    }
    startPoint = new XYZ(0, 0, 0);
    get subclassMarker() {
        return DxfSubclassMarker.Line;
    }
    thickness = 0.0;
    constructor(start, end) {
        super();
        if (start) {
            this.startPoint = start;
        }
        if (end) {
            this.endPoint = end;
        }
    }
    applyTransform(transform) {
        // TODO: transform.ApplyTransform not available
    }
    getBoundingBox() {
        const min = new XYZ(Math.min(this.startPoint.x, this.endPoint.x), Math.min(this.startPoint.y, this.endPoint.y), Math.min(this.startPoint.z, this.endPoint.z));
        const max = new XYZ(Math.max(this.startPoint.x, this.endPoint.x), Math.max(this.startPoint.y, this.endPoint.y), Math.max(this.startPoint.z, this.endPoint.z));
        return { min, max };
    }
}
