// @node-projects/acad-ts/dist/Entities/MLine.js
import { Entity } from './Entity.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { MLineFlags } from './MLineFlags.js';
import { MLineJustification } from './MLineJustification.js';
import { XYZ } from '../Math/XYZ.js';
export class MLineSegment {
    parameters = [];
    areaFillParameters = [];
    applyScale(scaleFactor) {
        this.parameters = this.parameters.map(p => p * scaleFactor);
        this.areaFillParameters = this.areaFillParameters.map(p => p * scaleFactor);
    }
    clone() {
        const c = new MLineSegment();
        c.parameters = [...this.parameters];
        c.areaFillParameters = [...this.areaFillParameters];
        return c;
    }
}
export class MLineVertex {
    position = new XYZ(0, 0, 0);
    direction = new XYZ(0, 0, 0);
    miter = new XYZ(0, 0, 0);
    segments = [];
    applyTransform(transform) {
        // TODO: Transform position, direction, miter
        // Apply miter length scale to segments
    }
    clone() {
        const c = new MLineVertex();
        c.position = new XYZ(this.position.x, this.position.y, this.position.z);
        c.direction = new XYZ(this.direction.x, this.direction.y, this.direction.z);
        c.miter = new XYZ(this.miter.x, this.miter.y, this.miter.z);
        c.segments = this.segments.map(s => s.clone());
        return c;
    }
}
export class MLine extends Entity {
    flags = MLineFlags.Has;
    justification = MLineJustification.Zero;
    normal = new XYZ(0, 0, 1);
    get objectName() {
        return DxfFileToken.EntityMLine;
    }
    get objectType() {
        return ObjectType.MLINE;
    }
    scaleFactor = 1;
    startPoint = new XYZ(0, 0, 0);
    get style() { return this._style; }
    set style(value) {
        if (!value)
            throw new Error('MLine style cannot be null');
        this._style = value;
    }
    get subclassMarker() {
        return DxfSubclassMarker.MLine;
    }
    vertices = [];
    _style = null;
    applyTransform(transform) {
        // TODO: Transform normal, startPoint, vertices
    }
    clone() {
        const clone = super.clone();
        // clone.style = this.style?.clone();
        clone.vertices = this.vertices.map(v => v.clone());
        return clone;
    }
    getBoundingBox() {
        return null;
    }
}
export { MLineFlags } from './MLineFlags.js';
export { MLineJustification } from './MLineJustification.js';
