// @node-projects/acad-ts/dist/Entities/ModelerGeometry.js
import { Entity } from './Entity.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { Color } from '../Color.js';
import { XYZ } from '../Math/XYZ.js';
export class ModelerGeometryWire {
    acisIndex = 0;
    applyTransformPresent = false;
    color = Color.ByLayer;
    hasRotation = false;
    hasShear = false;
    points = [];
    scale = 0;
    selectionMarker = 0;
    translation = new XYZ(0, 0, 0);
    type = 0;
    xAxis = new XYZ(0, 0, 0);
    yAxis = new XYZ(0, 0, 0);
    zAxis = new XYZ(0, 0, 0);
    /** @internal */
    hasReflection = false;
}
export class ModelerGeometrySilhouette {
    viewportDirectionFromTarget = new XYZ(0, 0, 0);
    viewportId = 0;
    viewportPerspective = false;
    viewportTarget = new XYZ(0, 0, 0);
    viewportUpDirection = new XYZ(0, 0, 0);
    wires = [];
}
export class ModelerGeometry extends Entity {
    point = new XYZ(0, 0, 0);
    silhouettes = [];
    get subclassMarker() {
        return DxfSubclassMarker.ModelerGeometry;
    }
    wires = [];
    /** @internal */
    guid = '';
    modelerFormatVersion = 0;
    proprietaryData = '';
    applyTransform(transform) {
        // No-op in C# source
    }
    getBoundingBox() {
        // BoundingBox.Null
        return null;
    }
}
