// @node-projects/acad-ts/dist/Entities/MultiLeader.js
import { Entity } from './Entity.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { Color } from '../Color.js';
import { LeaderContentType } from '../Objects/LeaderContentType.js';
import { MultiLeaderPathType } from '../MultiLeaderPathType.js';
import { TextAlignmentType } from '../TextAlignmentType.js';
import { TextAttachmentDirectionType } from '../TextAttachmentDirectionType.js';
import { TextAttachmentPointType } from '../TextAttachmentPoint.js';
import { TextAttachmentType } from '../TextAttachmentType.js';
import { XYZ } from '../Math/XYZ.js';
export class MultiLeaderBlockAttribute {
    attributeDefinition = null;
    index = 0;
    width = 0;
    text = '';
    clone() {
        const c = new MultiLeaderBlockAttribute();
        c.attributeDefinition = this.attributeDefinition;
        c.index = this.index;
        c.width = this.width;
        c.text = this.text;
        return c;
    }
}
export class MultiLeader extends Entity {
    arrowhead = null;
    arrowheadSize = 0;
    blockAttributes = [];
    blockContentColor = Color.ByBlock;
    blockContentConnection = 0;
    blockContentId = null;
    blockContentRotation = 0;
    blockContentScale = new XYZ(1, 1, 1);
    contentType = LeaderContentType.MText;
    contextData = null;
    enableAnnotationScale = false;
    enableDogleg = true;
    enableLanding = true;
    extendedToText = false;
    landingDistance = 0;
    leaderLineType = null;
    leaderLineWeight = 0;
    lineColor = Color.ByBlock;
    get objectName() {
        return DxfFileToken.EntityMultiLeader;
    }
    get objectType() {
        return ObjectType.UNLISTED;
    }
    pathType = MultiLeaderPathType.StraightLineSegments;
    propertyOverrideFlags = 0;
    scaleFactor = 1.0;
    style = null;
    get subclassMarker() {
        return DxfSubclassMarker.MultiLeader;
    }
    textAlignment = TextAlignmentType.Left;
    textAngle = 0;
    textAttachmentDirection = TextAttachmentDirectionType.Horizontal;
    textAttachmentPoint = TextAttachmentPointType.Center;
    textBottomAttachment = TextAttachmentType.CenterOfText;
    textColor = Color.ByBlock;
    textFrame = false;
    textLeftAttachment = TextAttachmentType.MiddleOfTopLine;
    textRightAttachment = TextAttachmentType.MiddleOfTopLine;
    textStyle = null;
    textTopAttachment = TextAttachmentType.CenterOfText;
    textDirectionNegative = false;
    textAligninIPE = false;
    applyTransform(transform) {
        // No-op
    }
    clone() {
        const clone = super.clone();
        clone.contextData = this.contextData; // TODO: deep clone
        clone.blockAttributes = this.blockAttributes.map(a => a.clone());
        return clone;
    }
    getBoundingBox() { return null; }
}
export { MultiLeaderPropertyOverrideFlags } from './MultiLeaderPropertyOverrideFlags.js';
export { MultiLeaderPathType } from '../MultiLeaderPathType.js';
export { LeaderContentType } from '../Objects/LeaderContentType.js';
export { TextAttachmentType } from '../TextAttachmentType.js';
export { TextAlignmentType } from '../TextAlignmentType.js';
export { TextAttachmentPointType } from '../TextAttachmentPoint.js';
export { TextAttachmentDirectionType } from '../TextAttachmentDirectionType.js';
