// @node-projects/acad-ts/dist/Entities/Shape.js
import { Entity } from './Entity.js';
import { CadObject } from '../CadObject.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { XYZ } from '../Math/XYZ.js';
export class Shape extends Entity {
    get objectType() {
        return ObjectType.SHAPE;
    }
    get objectName() {
        return DxfFileToken.EntityShape;
    }
    get subclassMarker() {
        return DxfSubclassMarker.Shape;
    }
    thickness = 0.0;
    insertionPoint = new XYZ(0, 0, 0);
    size = 1.0;
    get shapeStyle() {
        return this._style;
    }
    set shapeStyle(value) {
        if (value == null || !value.isShapeFile) {
            throw new Error('value cannot be null and must be a shape file');
        }
        if (this.document != null) {
            this._style = CadObject.updateCollection(value, this.document.textStyles);
        }
        else {
            this._style = value;
        }
    }
    rotation = 0;
    relativeXScale = 1;
    obliqueAngle = 0;
    normal = new XYZ(0, 0, 1);
    /** @internal */
    shapeIndex = 0;
    _style;
    /** @internal */
    constructor(textStyle) {
        super();
        if (textStyle) {
            this.shapeStyle = textStyle;
        }
    }
    clone() {
        const clone = super.clone();
        clone._style = this._style?.clone();
        return clone;
    }
    getBoundingBox() {
        return null;
    }
    applyTransform(transform) {
        // TODO: transform operations not available
    }
}
