// @node-projects/acad-ts/dist/Entities/Spline.js
import { Entity } from './Entity.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { SplineFlags } from './SplineFlags.js';
import { SplineFlags1 } from './SplineFlags1.js';
import { KnotParametrization } from './KnotParametrization.js';
import { XYZ } from '../Math/XYZ.js';
export class Spline extends Entity {
    controlPoints = [];
    controlPointTolerance = 0.0000001;
    degree = 3;
    endTangent = new XYZ(0, 0, 0);
    fitPoints = [];
    fitTolerance = 0.0000000001;
    get flags() {
        return this._flags;
    }
    set flags(value) {
        this._flags = value;
    }
    get flags1() {
        return this._flags1;
    }
    set flags1(value) {
        this._flags1 = value;
    }
    get hasValidKnotCount() {
        const expected = this.controlPoints.length + (this.isClosed ? 2 : 1) * this.degree + 1;
        return this.knots.length === expected;
    }
    get isClosed() {
        return (this._flags & SplineFlags.Closed) !== 0;
    }
    set isClosed(value) {
        if (value) {
            this._flags = this._flags | SplineFlags.Closed;
            this._flags1 = this._flags1 | SplineFlags1.Closed;
        }
        else {
            this._flags = this._flags & ~SplineFlags.Closed;
            this._flags1 = this._flags1 & ~SplineFlags1.Closed;
        }
    }
    get isPeriodic() {
        return (this._flags & SplineFlags.Periodic) !== 0;
    }
    set isPeriodic(value) {
        if (value) {
            this._flags = this._flags | SplineFlags.Periodic;
        }
        else {
            this._flags = this._flags & ~SplineFlags.Periodic;
        }
    }
    knotParametrization = KnotParametrization.Chord;
    knots = [];
    knotTolerance = 0.0000001;
    normal = new XYZ(0, 0, 1);
    get objectName() {
        return DxfFileToken.EntitySpline;
    }
    get objectType() {
        return ObjectType.SPLINE;
    }
    startTangent = new XYZ(0, 0, 0);
    get subclassMarker() {
        return DxfSubclassMarker.Spline;
    }
    weights = [];
    static MaxDegree = 10;
    _flags = SplineFlags.None;
    _flags1 = SplineFlags1.None;
    constructor() {
        super();
    }
    applyTransform(transform) {
        // TODO: transform operations not available
    }
    clone() {
        const clone = super.clone();
        clone.controlPoints = [...this.controlPoints];
        clone.fitPoints = [...this.fitPoints];
        clone.weights = [...this.weights];
        clone.knots = [...this.knots];
        return clone;
    }
    getBoundingBox() {
        // TODO: BoundingBox.FromPoints / PolygonalVertexes not available
        return null;
    }
    pointOnSpline(t) {
        if (t < 0 || t > 1) {
            throw new Error('The parametric position must be a value between 0 and 1.');
        }
        if (t === 1) {
            t -= Number.EPSILON;
        }
        const { controlPts, weights, knots } = this.prepare();
        const { uStart, uEnd } = this.getStartAndEndKnots(knots);
        const uDelta = (uEnd - uStart) * t;
        const u = uStart + uDelta;
        return Spline.c(controlPts, weights, knots, this.degree, u);
    }
    polygonalVertexes(precision) {
        if (precision < 2) {
            throw new Error('The precision must be equal or greater than two.');
        }
        const vertexes = [];
        const { controlPts, weights, knots } = this.prepare();
        const { uStart, uEnd } = this.getStartAndEndKnots(knots);
        if (!this.isClosed && !this.isPeriodic) {
            precision -= 1;
        }
        const uDelta = (uEnd - uStart) / precision;
        for (let i = 0; i < precision; i++) {
            const u = uStart + uDelta * i;
            vertexes.push(Spline.c(controlPts, weights, knots, this.degree, u));
        }
        if (!(this.isClosed || this.isPeriodic)) {
            vertexes.push(controlPts[controlPts.length - 1]);
        }
        return vertexes;
    }
    tryPointOnSpline(t) {
        try {
            const point = this.pointOnSpline(t);
            return { success: true, point };
        }
        catch {
            return { success: false, point: new XYZ(NaN, NaN, NaN) };
        }
    }
    tryPolygonalVertexes(precision) {
        try {
            const points = this.polygonalVertexes(precision);
            return { success: true, points };
        }
        catch {
            return { success: false, points: [] };
        }
    }
    updateFromFitPoints(iterationLimit = 255) {
        // TODO: Complex spline fitting algorithm - requires full XYZ vector math
        return false;
    }
    static c(ctrlPoints, weights, knots, degree, u) {
        let sumX = 0, sumY = 0, sumZ = 0;
        let denominatorSum = 0;
        for (let i = 0; i < ctrlPoints.length; i++) {
            const nurb = Spline.computeNurb(knots, i, degree, u);
            denominatorSum += nurb * weights[i];
            sumX += weights[i] * nurb * ctrlPoints[i].x;
            sumY += weights[i] * nurb * ctrlPoints[i].y;
            sumZ += weights[i] * nurb * ctrlPoints[i].z;
        }
        if (Math.abs(denominatorSum) < Number.EPSILON) {
            return new XYZ(0, 0, 0);
        }
        const inv = 1.0 / denominatorSum;
        return new XYZ(inv * sumX, inv * sumY, inv * sumZ);
    }
    static computeNurb(knots, i, p, u) {
        if (p <= 0) {
            if (knots[i] <= u && u < knots[i + 1]) {
                return 1;
            }
            return 0.0;
        }
        let leftCoefficient = 0.0;
        if (!(Math.abs(knots[i + p] - knots[i]) < Number.EPSILON)) {
            leftCoefficient = (u - knots[i]) / (knots[i + p] - knots[i]);
        }
        let rightCoefficient = 0.0;
        if (!(Math.abs(knots[i + p + 1] - knots[i + 1]) < Number.EPSILON)) {
            rightCoefficient = (knots[i + p + 1] - u) / (knots[i + p + 1] - knots[i + 1]);
        }
        return leftCoefficient * Spline.computeNurb(knots, i, p - 1, u) +
            rightCoefficient * Spline.computeNurb(knots, i + 1, p - 1, u);
    }
    static createKnotVector(numControlPoints, degree, isPeriodic) {
        let knots;
        if (!isPeriodic) {
            const numKnots = numControlPoints + degree + 1;
            knots = new Array(numKnots).fill(0);
            let i;
            for (i = 0; i <= degree; i++) {
                knots[i] = 0.0;
            }
            for (; i < numControlPoints; i++) {
                knots[i] = i - degree;
            }
            for (; i < numKnots; i++) {
                knots[i] = numControlPoints - degree;
            }
        }
        else {
            const numKnots = numControlPoints + 2 * degree + 1;
            knots = new Array(numKnots).fill(0);
            const factor = 1.0 / (numControlPoints - degree);
            for (let i = 0; i < numKnots; i++) {
                knots[i] = (i - degree) * factor;
            }
        }
        return knots;
    }
    getStartAndEndKnots(knots) {
        if (this.isClosed) {
            return { uStart: knots[0], uEnd: knots[knots.length - 1] };
        }
        else if (this.isPeriodic) {
            return { uStart: knots[this.degree], uEnd: knots[knots.length - this.degree - 1] };
        }
        else {
            return { uStart: knots[0], uEnd: knots[knots.length - 1] };
        }
    }
    prepare() {
        let c = [...this.controlPoints];
        let w = [...this.weights];
        let knots = [...this.knots];
        const numCtrlPoints = c.length;
        if (numCtrlPoints === 0) {
            throw new Error('A spline entity with control points is required.');
        }
        if (w.length === 0 || w.length !== numCtrlPoints) {
            w = new Array(numCtrlPoints).fill(1.0);
        }
        if (knots.length === 0 || !this.hasValidKnotCount) {
            knots = Spline.createKnotVector(numCtrlPoints, this.degree, this.isPeriodic);
        }
        let controlPts;
        let weights;
        if (this.isPeriodic) {
            controlPts = new Array(numCtrlPoints + this.degree);
            weights = new Array(numCtrlPoints + this.degree);
            for (let i = 0; i < this.degree; i++) {
                const index = numCtrlPoints - this.degree + i;
                controlPts[i] = c[index];
                weights[i] = w[index];
            }
            for (let i = 0; i < numCtrlPoints; i++) {
                controlPts[this.degree + i] = c[i];
                weights[this.degree + i] = w[i];
            }
        }
        else {
            controlPts = c;
            weights = w;
        }
        return { controlPts, weights, knots };
    }
}
export { SplineFlags } from './SplineFlags.js';
export { SplineFlags1 } from './SplineFlags1.js';
export { KnotParametrization } from './KnotParametrization.js';
