// @node-projects/acad-ts/dist/Entities/TableEntity.js
import { Insert } from './Insert.js';
import { Color } from '../Color.js';
import { CadValue } from '../CadValue.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { LineWeightType } from '../Types/LineWeightType.js';
import { XYZ } from '../Math/XYZ.js';
// === Enums ===
export var TableEntityBorderType;
(function (TableEntityBorderType) {
    TableEntityBorderType[TableEntityBorderType["Single"] = 1] = "Single";
    TableEntityBorderType[TableEntityBorderType["Double"] = 2] = "Double";
})(TableEntityBorderType || (TableEntityBorderType = {}));
export var BreakFlowDirection;
(function (BreakFlowDirection) {
    BreakFlowDirection[BreakFlowDirection["Right"] = 1] = "Right";
    BreakFlowDirection[BreakFlowDirection["Vertical"] = 2] = "Vertical";
    BreakFlowDirection[BreakFlowDirection["Left"] = 4] = "Left";
})(BreakFlowDirection || (BreakFlowDirection = {}));
export var BreakOptionFlags;
(function (BreakOptionFlags) {
    BreakOptionFlags[BreakOptionFlags["None"] = 0] = "None";
    BreakOptionFlags[BreakOptionFlags["EnableBreaks"] = 1] = "EnableBreaks";
    BreakOptionFlags[BreakOptionFlags["RepeatTopLabels"] = 2] = "RepeatTopLabels";
    BreakOptionFlags[BreakOptionFlags["RepeatBottomLabels"] = 4] = "RepeatBottomLabels";
    BreakOptionFlags[BreakOptionFlags["AllowManualPositions"] = 8] = "AllowManualPositions";
    BreakOptionFlags[BreakOptionFlags["AllowManualHeights"] = 16] = "AllowManualHeights";
})(BreakOptionFlags || (BreakOptionFlags = {}));
export var CellAlignmentType;
(function (CellAlignmentType) {
    CellAlignmentType[CellAlignmentType["None"] = 0] = "None";
    CellAlignmentType[CellAlignmentType["TopLeft"] = 1] = "TopLeft";
    CellAlignmentType[CellAlignmentType["TopCenter"] = 2] = "TopCenter";
    CellAlignmentType[CellAlignmentType["TopRight"] = 3] = "TopRight";
    CellAlignmentType[CellAlignmentType["MiddleLeft"] = 4] = "MiddleLeft";
    CellAlignmentType[CellAlignmentType["MiddleCenter"] = 5] = "MiddleCenter";
    CellAlignmentType[CellAlignmentType["MiddleRight"] = 6] = "MiddleRight";
    CellAlignmentType[CellAlignmentType["BottomLeft"] = 7] = "BottomLeft";
    CellAlignmentType[CellAlignmentType["BottomCenter"] = 8] = "BottomCenter";
    CellAlignmentType[CellAlignmentType["BottomRight"] = 9] = "BottomRight";
})(CellAlignmentType || (CellAlignmentType = {}));
export var CellEdgeFlags;
(function (CellEdgeFlags) {
    CellEdgeFlags[CellEdgeFlags["Unknown"] = 0] = "Unknown";
    CellEdgeFlags[CellEdgeFlags["Top"] = 1] = "Top";
    CellEdgeFlags[CellEdgeFlags["Right"] = 2] = "Right";
    CellEdgeFlags[CellEdgeFlags["Bottom"] = 4] = "Bottom";
    CellEdgeFlags[CellEdgeFlags["Left"] = 8] = "Left";
    CellEdgeFlags[CellEdgeFlags["InsideVertical"] = 16] = "InsideVertical";
    CellEdgeFlags[CellEdgeFlags["InsideHorizontal"] = 32] = "InsideHorizontal";
})(CellEdgeFlags || (CellEdgeFlags = {}));
export var CellStyleClass;
(function (CellStyleClass) {
    CellStyleClass[CellStyleClass["Data"] = 1] = "Data";
    CellStyleClass[CellStyleClass["Label"] = 2] = "Label";
})(CellStyleClass || (CellStyleClass = {}));
export var CellStyleType;
(function (CellStyleType) {
    CellStyleType[CellStyleType["Unknown"] = 0] = "Unknown";
    CellStyleType[CellStyleType["Cell"] = 1] = "Cell";
    CellStyleType[CellStyleType["Row"] = 2] = "Row";
    CellStyleType[CellStyleType["Column"] = 3] = "Column";
    CellStyleType[CellStyleType["FormattedTableData"] = 4] = "FormattedTableData";
    CellStyleType[CellStyleType["Table"] = 5] = "Table";
})(CellStyleType || (CellStyleType = {}));
export var CellType;
(function (CellType) {
    CellType[CellType["Text"] = 1] = "Text";
    CellType[CellType["Block"] = 2] = "Block";
})(CellType || (CellType = {}));
export var MarginFlags;
(function (MarginFlags) {
    MarginFlags[MarginFlags["None"] = 0] = "None";
    MarginFlags[MarginFlags["Override"] = 1] = "Override";
})(MarginFlags || (MarginFlags = {}));
export var TableBorderPropertyFlags;
(function (TableBorderPropertyFlags) {
    TableBorderPropertyFlags[TableBorderPropertyFlags["None"] = 0] = "None";
    TableBorderPropertyFlags[TableBorderPropertyFlags["BorderType"] = 1] = "BorderType";
    TableBorderPropertyFlags[TableBorderPropertyFlags["LineWeight"] = 2] = "LineWeight";
    TableBorderPropertyFlags[TableBorderPropertyFlags["LineType"] = 4] = "LineType";
    TableBorderPropertyFlags[TableBorderPropertyFlags["Color"] = 8] = "Color";
    TableBorderPropertyFlags[TableBorderPropertyFlags["Invisibility"] = 16] = "Invisibility";
    TableBorderPropertyFlags[TableBorderPropertyFlags["DoubleLineSpacing"] = 32] = "DoubleLineSpacing";
    TableBorderPropertyFlags[TableBorderPropertyFlags["All"] = 63] = "All";
})(TableBorderPropertyFlags || (TableBorderPropertyFlags = {}));
export var TableCellContentLayoutFlags;
(function (TableCellContentLayoutFlags) {
    TableCellContentLayoutFlags[TableCellContentLayoutFlags["None"] = 0] = "None";
    TableCellContentLayoutFlags[TableCellContentLayoutFlags["Flow"] = 1] = "Flow";
    TableCellContentLayoutFlags[TableCellContentLayoutFlags["StackedHorizontal"] = 2] = "StackedHorizontal";
    TableCellContentLayoutFlags[TableCellContentLayoutFlags["StackedVertical"] = 4] = "StackedVertical";
})(TableCellContentLayoutFlags || (TableCellContentLayoutFlags = {}));
export var TableCellContentType;
(function (TableCellContentType) {
    TableCellContentType[TableCellContentType["Unknown"] = 0] = "Unknown";
    TableCellContentType[TableCellContentType["Value"] = 1] = "Value";
    TableCellContentType[TableCellContentType["Field"] = 2] = "Field";
    TableCellContentType[TableCellContentType["Block"] = 4] = "Block";
})(TableCellContentType || (TableCellContentType = {}));
export var TableCellStateFlags;
(function (TableCellStateFlags) {
    TableCellStateFlags[TableCellStateFlags["None"] = 0] = "None";
    TableCellStateFlags[TableCellStateFlags["ContentLocked"] = 1] = "ContentLocked";
    TableCellStateFlags[TableCellStateFlags["ContentReadOnly"] = 2] = "ContentReadOnly";
    TableCellStateFlags[TableCellStateFlags["Linked"] = 4] = "Linked";
    TableCellStateFlags[TableCellStateFlags["ContentModifiedAfterUpdate"] = 8] = "ContentModifiedAfterUpdate";
    TableCellStateFlags[TableCellStateFlags["FormatLocked"] = 16] = "FormatLocked";
    TableCellStateFlags[TableCellStateFlags["FormatReadOnly"] = 32] = "FormatReadOnly";
    TableCellStateFlags[TableCellStateFlags["FormatModifiedAfterUpdate"] = 64] = "FormatModifiedAfterUpdate";
})(TableCellStateFlags || (TableCellStateFlags = {}));
export var TableCellStylePropertyFlags;
(function (TableCellStylePropertyFlags) {
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["None"] = 0] = "None";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["DataType"] = 1] = "DataType";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["DataFormat"] = 2] = "DataFormat";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["Rotation"] = 4] = "Rotation";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["BlockScale"] = 8] = "BlockScale";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["Alignment"] = 16] = "Alignment";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["ContentColor"] = 32] = "ContentColor";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["TextStyle"] = 64] = "TextStyle";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["TextHeight"] = 128] = "TextHeight";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["AutoScale"] = 256] = "AutoScale";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["BackgroundColor"] = 512] = "BackgroundColor";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["MarginLeft"] = 1024] = "MarginLeft";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["MarginTop"] = 2048] = "MarginTop";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["MarginRight"] = 4096] = "MarginRight";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["MarginBottom"] = 8192] = "MarginBottom";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["ContentLayout"] = 16384] = "ContentLayout";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["MarginHorizontalSpacing"] = 131072] = "MarginHorizontalSpacing";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["MarginVerticalSpacing"] = 262144] = "MarginVerticalSpacing";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["MergeAll"] = 32768] = "MergeAll";
    TableCellStylePropertyFlags[TableCellStylePropertyFlags["FlowDirectionBottomToTop"] = 65536] = "FlowDirectionBottomToTop";
})(TableCellStylePropertyFlags || (TableCellStylePropertyFlags = {}));
export var TableEntityValueUnitType;
(function (TableEntityValueUnitType) {
    TableEntityValueUnitType[TableEntityValueUnitType["NoUnits"] = 0] = "NoUnits";
    TableEntityValueUnitType[TableEntityValueUnitType["Distance"] = 1] = "Distance";
    TableEntityValueUnitType[TableEntityValueUnitType["Angle"] = 2] = "Angle";
    TableEntityValueUnitType[TableEntityValueUnitType["Area"] = 4] = "Area";
    TableEntityValueUnitType[TableEntityValueUnitType["Volume"] = 8] = "Volume";
    TableEntityValueUnitType[TableEntityValueUnitType["Currency"] = 16] = "Currency";
    TableEntityValueUnitType[TableEntityValueUnitType["Percentage"] = 32] = "Percentage";
})(TableEntityValueUnitType || (TableEntityValueUnitType = {}));
// Internal enums
export var BorderOverrideFlags;
(function (BorderOverrideFlags) {
    BorderOverrideFlags[BorderOverrideFlags["None"] = 0] = "None";
    BorderOverrideFlags[BorderOverrideFlags["TitleHorizontalTop"] = 1] = "TitleHorizontalTop";
    BorderOverrideFlags[BorderOverrideFlags["TitleHorizontalInsert"] = 2] = "TitleHorizontalInsert";
    BorderOverrideFlags[BorderOverrideFlags["TitleHorizontalBottom"] = 4] = "TitleHorizontalBottom";
    BorderOverrideFlags[BorderOverrideFlags["TitleVerticalLeft"] = 8] = "TitleVerticalLeft";
    BorderOverrideFlags[BorderOverrideFlags["TitleVerticalInsert"] = 16] = "TitleVerticalInsert";
    BorderOverrideFlags[BorderOverrideFlags["TitleVerticalRight"] = 32] = "TitleVerticalRight";
    BorderOverrideFlags[BorderOverrideFlags["HeaderHorizontalTop"] = 64] = "HeaderHorizontalTop";
    BorderOverrideFlags[BorderOverrideFlags["HeaderHorizontalInsert"] = 128] = "HeaderHorizontalInsert";
    BorderOverrideFlags[BorderOverrideFlags["HeaderHorizontalBottom"] = 256] = "HeaderHorizontalBottom";
    BorderOverrideFlags[BorderOverrideFlags["HeaderVerticalLeft"] = 512] = "HeaderVerticalLeft";
    BorderOverrideFlags[BorderOverrideFlags["HeaderVerticalInsert"] = 1024] = "HeaderVerticalInsert";
    BorderOverrideFlags[BorderOverrideFlags["HeaderVerticalRight"] = 2048] = "HeaderVerticalRight";
    BorderOverrideFlags[BorderOverrideFlags["DataHorizontalTop"] = 4096] = "DataHorizontalTop";
    BorderOverrideFlags[BorderOverrideFlags["DataHorizontalInsert"] = 8192] = "DataHorizontalInsert";
    BorderOverrideFlags[BorderOverrideFlags["DataHorizontalBottom"] = 16384] = "DataHorizontalBottom";
    BorderOverrideFlags[BorderOverrideFlags["DataVerticalLeft"] = 32768] = "DataVerticalLeft";
    BorderOverrideFlags[BorderOverrideFlags["DataVerticalInsert"] = 65536] = "DataVerticalInsert";
    BorderOverrideFlags[BorderOverrideFlags["DataVerticalRight"] = 131072] = "DataVerticalRight";
})(BorderOverrideFlags || (BorderOverrideFlags = {}));
export var TableOverrideFlags;
(function (TableOverrideFlags) {
    TableOverrideFlags[TableOverrideFlags["None"] = 0] = "None";
    TableOverrideFlags[TableOverrideFlags["TitleSuppressed"] = 1] = "TitleSuppressed";
    TableOverrideFlags[TableOverrideFlags["HeaderSuppressed"] = 2] = "HeaderSuppressed";
    TableOverrideFlags[TableOverrideFlags["FlowDirection"] = 4] = "FlowDirection";
    TableOverrideFlags[TableOverrideFlags["HorizontalCellMargin"] = 8] = "HorizontalCellMargin";
    TableOverrideFlags[TableOverrideFlags["VerticalCellMargin"] = 16] = "VerticalCellMargin";
    TableOverrideFlags[TableOverrideFlags["TitleRowColor"] = 32] = "TitleRowColor";
    TableOverrideFlags[TableOverrideFlags["HeaderRowColor"] = 64] = "HeaderRowColor";
    TableOverrideFlags[TableOverrideFlags["DataRowColor"] = 128] = "DataRowColor";
    TableOverrideFlags[TableOverrideFlags["TitleRowFillNone"] = 256] = "TitleRowFillNone";
    TableOverrideFlags[TableOverrideFlags["HeaderRowFillNone"] = 512] = "HeaderRowFillNone";
    TableOverrideFlags[TableOverrideFlags["DataRowFillNone"] = 1024] = "DataRowFillNone";
    TableOverrideFlags[TableOverrideFlags["TitleRowFillColor"] = 2048] = "TitleRowFillColor";
    TableOverrideFlags[TableOverrideFlags["HeaderRowFillColor"] = 4096] = "HeaderRowFillColor";
    TableOverrideFlags[TableOverrideFlags["DataRowFillColor"] = 8192] = "DataRowFillColor";
    TableOverrideFlags[TableOverrideFlags["TitleRowAlign"] = 16384] = "TitleRowAlign";
    TableOverrideFlags[TableOverrideFlags["HeaderRowAlign"] = 32768] = "HeaderRowAlign";
    TableOverrideFlags[TableOverrideFlags["DataRowAlign"] = 65536] = "DataRowAlign";
    TableOverrideFlags[TableOverrideFlags["TitleTextStyle"] = 131072] = "TitleTextStyle";
    TableOverrideFlags[TableOverrideFlags["HeaderTextStyle"] = 262144] = "HeaderTextStyle";
    TableOverrideFlags[TableOverrideFlags["DataTextStyle"] = 524288] = "DataTextStyle";
    TableOverrideFlags[TableOverrideFlags["TitleRowHeight"] = 1048576] = "TitleRowHeight";
    TableOverrideFlags[TableOverrideFlags["HeaderRowHeight"] = 2097152] = "HeaderRowHeight";
    TableOverrideFlags[TableOverrideFlags["DataRowHeight"] = 4194304] = "DataRowHeight";
})(TableOverrideFlags || (TableOverrideFlags = {}));
export var CellOverrideFlags;
(function (CellOverrideFlags) {
    CellOverrideFlags[CellOverrideFlags["None"] = 0] = "None";
    CellOverrideFlags[CellOverrideFlags["CellAlignment"] = 1] = "CellAlignment";
    CellOverrideFlags[CellOverrideFlags["BackgroundFillNone"] = 2] = "BackgroundFillNone";
    CellOverrideFlags[CellOverrideFlags["BackgroundColor"] = 4] = "BackgroundColor";
    CellOverrideFlags[CellOverrideFlags["ContentColor"] = 8] = "ContentColor";
    CellOverrideFlags[CellOverrideFlags["TextStyle"] = 16] = "TextStyle";
    CellOverrideFlags[CellOverrideFlags["TextHeight"] = 32] = "TextHeight";
    CellOverrideFlags[CellOverrideFlags["TopGridColor"] = 64] = "TopGridColor";
    CellOverrideFlags[CellOverrideFlags["TopGridLineWeight"] = 1024] = "TopGridLineWeight";
    CellOverrideFlags[CellOverrideFlags["TopVisibility"] = 16384] = "TopVisibility";
    CellOverrideFlags[CellOverrideFlags["RightGridColor"] = 128] = "RightGridColor";
    CellOverrideFlags[CellOverrideFlags["RightGridLineWeight"] = 2048] = "RightGridLineWeight";
    CellOverrideFlags[CellOverrideFlags["RightVisibility"] = 32768] = "RightVisibility";
    CellOverrideFlags[CellOverrideFlags["BottomGridColor"] = 256] = "BottomGridColor";
    CellOverrideFlags[CellOverrideFlags["BottomGridLineWeight"] = 4096] = "BottomGridLineWeight";
    CellOverrideFlags[CellOverrideFlags["BottomVisibility"] = 65536] = "BottomVisibility";
    CellOverrideFlags[CellOverrideFlags["LeftGridColor"] = 512] = "LeftGridColor";
    CellOverrideFlags[CellOverrideFlags["LeftGridLineWeight"] = 8192] = "LeftGridLineWeight";
    CellOverrideFlags[CellOverrideFlags["LeftVisibility"] = 131072] = "LeftVisibility";
})(CellOverrideFlags || (CellOverrideFlags = {}));
// === Data Classes ===
export class TableAttribute {
    value = '';
}
export class CustomDataEntry {
    name = '';
    value = new CadValue();
}
export class CellContentGeometry {
    distanceTopLeft = new XYZ(0, 0, 0);
    distanceCenter = new XYZ(0, 0, 0);
    contentWidth = 0;
    contentHeight = 0;
    width = 0;
    height = 0;
    flags = 0;
}
export class CellRange {
    bottomRowIndex = 0;
    leftColumnIndex = 0;
    rightColumnIndex = 0;
    topRowIndex = 0;
}
export class ContentFormat {
    alignment = 0;
    color = Color.ByBlock;
    hasData = false;
    propertyFlags = 0;
    propertyOverrideFlags = 0;
    rotation = 0;
    scale = 0;
    textHeight = 0;
    textStyle = null;
    valueDataType = 0;
    valueFormatString = '';
    valueUnitType = 0;
}
export class CellBorder {
    color = Color.ByBlock;
    doubleLineSpacing = 0;
    edgeFlags;
    isInvisible = false;
    lineWeight = LineWeightType.Default;
    propertyOverrideFlags = 0;
    type = TableEntityBorderType.Single;
    constructor(edgeFlags) {
        this.edgeFlags = edgeFlags;
    }
}
export class CellContent {
    contentType = TableCellContentType.Unknown;
    format = new ContentFormat();
    cadValue = new CadValue();
}
export class CellStyle extends ContentFormat {
    backgroundColor = Color.ByBlock;
    bottomBorder = new CellBorder(CellEdgeFlags.Bottom);
    bottomMargin = 0;
    cellAlignment = CellAlignmentType.None;
    contentColor = Color.ByBlock;
    contentLayoutFlags = 0;
    horizontalInsideBorder = new CellBorder(CellEdgeFlags.InsideHorizontal);
    horizontalMargin = 0.06;
    isFillColorOn = false;
    leftBorder = new CellBorder(CellEdgeFlags.Left);
    marginHorizontalSpacing = 0;
    marginOverrideFlags = 0;
    marginVerticalSpacing = 0;
    name = '';
    rightBorder = new CellBorder(CellEdgeFlags.Right);
    rightMargin = 0;
    styleClass = CellStyleClass.Data;
    tableCellStylePropertyFlags = 0;
    textColor = Color.ByBlock;
    topBorder = new CellBorder(CellEdgeFlags.Right);
    cellStyleType = CellStyleType.Unknown;
    verticalInsideBorder = new CellBorder(CellEdgeFlags.InsideVertical);
    verticalMargin = 0.06;
    id = 0;
}
export class TableEntityCell {
    autoFit = false;
    blockScale = 0;
    borderHeight = 0;
    borderWidth = 0;
    get content() {
        if (this.contents == null || this.hasMultipleContent) {
            return null;
        }
        return this.contents.length > 0 ? this.contents[0] : null;
    }
    contents = [];
    customData = 0;
    customDataCollection = [];
    edgeFlags = 0;
    geometry = null;
    hasLinkedData = false;
    get hasMultipleContent() {
        return this.contents != null && this.contents.length > 1;
    }
    mergedValue = 0;
    rotation = 0;
    stateFlags = 0;
    styleOverride = new CellStyle();
    toolTip = '';
    type = CellType.Text;
    virtualEdgeFlag = 0;
}
export class TableEntityColumn {
    name = '';
    width = 0;
    customData = 0;
    cellStyleOverride = new CellStyle();
    customDataCollection = [];
}
export class TableEntityRow {
    height = 0;
    customData = 0;
    cellStyleOverride = new CellStyle();
    cells = [];
    customDataCollection = [];
}
export class BreakHeight {
    position = new XYZ(0, 0, 0);
    height = 0;
}
export class BreakData {
    flags = BreakOptionFlags.None;
    flowDirection = BreakFlowDirection.Right;
    breakSpacing = 0;
    heights = [];
}
export class BreakRowRange {
    position = new XYZ(0, 0, 0);
    startRowIndex = 0;
    endRowIndex = 0;
}
// === Main Class ===
export class TableEntity extends Insert {
    get columns() {
        return this._content.columns;
    }
    horizontalDirection = new XYZ(0, 0, 0);
    get objectName() {
        return DxfFileToken.EntityTable;
    }
    get objectType() {
        return ObjectType.UNLISTED;
    }
    overrideBorderColor = false;
    overrideBorderLineWeight = false;
    overrideBorderVisibility = false;
    overrideFlag = false;
    get rows() {
        return this._content.rows;
    }
    get style() {
        return this._content.style;
    }
    set style(value) {
        this._content.style = value;
    }
    get subclassMarker() {
        return DxfSubclassMarker.TableEntity;
    }
    valueFlag = 0;
    version = 0;
    breakRowRanges = [];
    get content() { return this._content; }
    tableBreakData = new BreakData();
    _content = new TableContent();
    clone() {
        return super.clone();
    }
    getBoundingBox() {
        return null;
    }
    getCell(row, column) {
        return this.rows[row].cells[column];
    }
}
export class TableContent {
    columns = [];
    rows = [];
    style = null;
}
