// @node-projects/acad-ts/dist/Entities/TextEntity.js
import { Entity } from './Entity.js';
import { CadObject } from '../CadObject.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { TextStyle } from '../Tables/TextStyle.js';
import { TextHorizontalAlignment } from './TextHorizontalAlignment.js';
import { TextVerticalAlignmentType } from './TextVerticalAlignmentType.js';
import { TextMirrorFlag } from './TextMirrorFlag.js';
import { XYZ } from '../Math/XYZ.js';
export class TextEntity extends Entity {
    alignmentPoint = new XYZ(0, 0, 0);
    get height() {
        return this._height;
    }
    set height(value) {
        if (value <= 0) {
            throw new Error('The Text height must be greater than zero.');
        }
        this._height = value;
    }
    horizontalAlignment = TextHorizontalAlignment.Left;
    insertPoint = new XYZ(0, 0, 0);
    get mirror() {
        return this._mirror;
    }
    set mirror(value) {
        this._mirror = value;
    }
    normal = new XYZ(0, 0, 1);
    get objectName() {
        return DxfFileToken.EntityText;
    }
    get objectType() {
        return ObjectType.TEXT;
    }
    obliqueAngle = 0.0;
    rotation = 0;
    get style() {
        return this._style;
    }
    set style(value) {
        if (value == null) {
            throw new Error('value cannot be null');
        }
        if (this.document != null) {
            this._style = CadObject.updateCollection(value, this.document.textStyles);
        }
        else {
            this._style = value;
        }
    }
    get subclassMarker() {
        return DxfSubclassMarker.Text;
    }
    thickness = 0.0;
    get value() {
        return this._value;
    }
    set value(v) {
        if (v.length > 256) {
            throw new Error(`Text length cannot be superior to 256, current: ${v.length}`);
        }
        this._value = v;
    }
    verticalAlignment = TextVerticalAlignmentType.Baseline;
    widthFactor = 1.0;
    _height = 1.0;
    _mirror = TextMirrorFlag.None;
    _style = TextStyle.Default;
    _value = '';
    constructor() {
        super();
    }
    applyTransform(transform) {
        // TODO: Complex transform logic not available (Matrix3, XY operations)
    }
    clone() {
        const clone = super.clone();
        clone._style = this._style.clone();
        return clone;
    }
    getBoundingBox() {
        return null;
    }
    /** @internal */
    assignDocument(doc) {
        super.assignDocument(doc);
        this._style = CadObject.updateCollection(this._style, doc.textStyles);
    }
    /** @internal */
    unassignDocument() {
        super.unassignDocument();
        this._style = this._style.clone();
    }
    _tableOnRemove(sender, e) {
        super._tableOnRemove(sender, e);
        if (e.item === this._style) {
            this._style = this.document.textStyles.get(TextStyle.DefaultName);
        }
    }
}
export { TextHorizontalAlignment } from './TextHorizontalAlignment.js';
export { TextVerticalAlignmentType } from './TextVerticalAlignmentType.js';
export { TextMirrorFlag } from './TextMirrorFlag.js';
