// @node-projects/acad-ts/dist/Entities/Tolerance.js
import { Entity } from './Entity.js';
import { CadObject } from '../CadObject.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { DimensionStyle } from '../Tables/DimensionStyle.js';
import { XYZ } from '../Math/XYZ.js';
export class Tolerance extends Entity {
    direction = new XYZ(0, 0, 0);
    insertionPoint = new XYZ(0, 0, 0);
    normal = new XYZ(0, 0, 1);
    get objectName() {
        return DxfFileToken.EntityTolerance;
    }
    get objectType() {
        return ObjectType.TOLERANCE;
    }
    get style() {
        return this._style;
    }
    set style(value) {
        if (value == null) {
            throw new Error('value cannot be null');
        }
        if (this.document != null) {
            this._style = CadObject.updateCollection(value, this.document.dimensionStyles);
        }
        else {
            this._style = value;
        }
    }
    get subclassMarker() {
        return DxfSubclassMarker.Tolerance;
    }
    text = '';
    _style = DimensionStyle.Default;
    applyTransform(transform) {
        // TODO: transform operations not available
    }
    getBoundingBox() {
        return null;
    }
}
