// @node-projects/acad-ts/dist/Entities/UnderlayEntity.js
import { Entity } from './Entity.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { UnderlayDisplayFlags } from './UnderlayDisplayFlags.js';
import { XYZ } from '../Math/XYZ.js';
export class UnderlayEntity extends Entity {
    clipBoundaryVertices = [];
    get contrast() {
        return this._contrast;
    }
    set contrast(value) {
        if (value < 0 || value > 100) {
            throw new Error(`Invalid Contrast value: ${value}, must be in range 0-100`);
        }
        this._contrast = value;
    }
    definition = null;
    get fade() {
        return this._fade;
    }
    set fade(value) {
        if (value < 0 || value > 100) {
            throw new Error(`Invalid Fade value: ${value}, must be in range 0-100`);
        }
        this._fade = value;
    }
    flags = UnderlayDisplayFlags.Default;
    insertPoint = new XYZ(0, 0, 0);
    normal = new XYZ(0, 0, 1);
    rotation = 0.0;
    get subclassMarker() {
        return DxfSubclassMarker.Underlay;
    }
    get xScale() {
        return this._xscale;
    }
    set xScale(value) {
        if (value === 0) {
            throw new Error('XScale value must be non-zero.');
        }
        this._xscale = value;
    }
    get yScale() {
        return this._yscale;
    }
    set yScale(value) {
        if (value === 0) {
            throw new Error('YScale value must be non-zero.');
        }
        this._yscale = value;
    }
    get zScale() {
        return this._zscale;
    }
    set zScale(value) {
        if (value === 0) {
            throw new Error('ZScale value must be non-zero.');
        }
        this._zscale = value;
    }
    _contrast = 100;
    _fade = 0;
    _xscale = 1;
    _yscale = 1;
    _zscale = 1;
    constructor(definition) {
        super();
        if (definition) {
            this.definition = definition;
        }
    }
    applyTransform(transform) {
        // TODO: Complex transform with world matrix
    }
    clone() {
        const clone = super.clone();
        clone.definition = this.definition?.clone?.() ?? null;
        clone.clipBoundaryVertices = [...this.clipBoundaryVertices];
        return clone;
    }
    getBoundingBox() {
        return null;
    }
    /** @internal */
    assignDocument(doc) {
        super.assignDocument(doc);
    }
    /** @internal */
    unassignDocument() {
        super.unassignDocument();
        this.definition = this.definition?.clone?.() ?? null;
    }
}
