// @node-projects/acad-ts/dist/Entities/Viewport.js
import { Entity } from './Entity.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { LightingType } from './LightingType.js';
import { XYZ } from '../Math/XYZ.js';
import { XY } from '../Math/XY.js';
export class Viewport extends Entity {
    ambientLightColor = null;
    backClipPlane = 0;
    boundary = null;
    brightness = 0;
    center = new XYZ(0, 0, 0);
    circleZoomPercent = 0;
    contrast = 0;
    defaultLightingType = LightingType.OneDistantLight;
    displayUcsIcon = false;
    elevation = 0;
    frontClipPlane = 0;
    frozenLayers = [];
    gridSpacing = new XY(0, 0);
    height = 0;
    get id() {
        // TODO: Owner-based sequential ID calculation
        return 0;
    }
    lensLength = 0;
    majorGridLineFrequency = 0;
    get objectName() {
        return DxfFileToken.EntityViewport;
    }
    get objectType() {
        return ObjectType.VIEWPORT;
    }
    renderMode = 0;
    get representsPaper() {
        return this.id === Viewport.PaperViewId;
    }
    scale = null;
    get scaleFactor() {
        if (this.height === 0)
            return 1;
        return 1 / (this.viewHeight / this.height);
    }
    shadePlotMode = 0;
    snapAngle = 0;
    snapBase = new XY(0, 0);
    snapSpacing = new XY(0, 0);
    status = 0;
    styleSheetName = '';
    get subclassMarker() {
        return DxfSubclassMarker.Viewport;
    }
    twistAngle = 0;
    ucsOrigin = new XYZ(0, 0, 0);
    ucsOrthographicType = 0;
    ucsPerViewport = false;
    ucsXAxis = new XYZ(1, 0, 0);
    ucsYAxis = new XYZ(0, 1, 0);
    useDefaultLighting = false;
    viewCenter = new XY(0, 0);
    viewDirection = new XYZ(0, 0, 1);
    viewHeight = 0;
    viewTarget = new XYZ(0, 0, 0);
    get viewWidth() {
        if (this.height === 0)
            return 0;
        return this.viewHeight / this.height * this.width;
    }
    visualStyle = null;
    width = 0;
    static ASDK_XREC_ANNOTATION_SCALE_INFO = 'ASDK_XREC_ANNOTATION_SCALE_INFO';
    static PaperViewId = 1;
    applyTransform(transform) {
        if (this.boundary != null) {
            this.boundary.applyTransform(transform);
        }
        else {
            // TODO: Transform center and recalculate width/height
        }
    }
    clone() {
        const clone = super.clone();
        clone.boundary = this.boundary?.clone() ?? null;
        clone.visualStyle = this.visualStyle?.clone?.() ?? null;
        clone.scale = this.scale?.clone?.() ?? null;
        clone.frozenLayers = this.frozenLayers.map(l => l.clone());
        return clone;
    }
    getBoundingBox() {
        return {
            min: new XYZ(this.center.x - this.width / 2, this.center.y - this.height / 2, this.center.z),
            max: new XYZ(this.center.x + this.width / 2, this.center.y + this.height / 2, this.center.z),
        };
    }
    getModelBoundingBox() {
        return {
            min: new XYZ(this.viewCenter.x - this.viewWidth / 2, this.viewCenter.y - this.viewHeight / 2, 0),
            max: new XYZ(this.viewCenter.x + this.viewWidth / 2, this.viewCenter.y + this.viewHeight / 2, 0),
        };
    }
    selectEntities(includePartial = true) {
        if (this.document == null) {
            throw new Error('Viewport needs to be assigned to a document.');
        }
        // TODO: Entity selection by bounding box
        return [];
    }
    /** @internal */
    assignDocument(doc) {
        super.assignDocument(doc);
    }
    /** @internal */
    unassignDocument() {
        super.unassignDocument();
        this.scale = this.scale?.clone?.() ?? null;
    }
}
export { ViewportStatusFlags } from './ViewportStatusFlags.js';
