// @node-projects/acad-ts/dist/Blocks/Block.js
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { BlockTypeFlags } from './BlockTypeFlags.js';
import { Entity } from '../Entities/Entity.js';
import { XYZ } from '../Math/XYZ.js';
export const XYZ_Zero = new XYZ(0, 0, 0);
export class Block extends Entity {
    basePoint = XYZ_Zero;
    get blockOwner() {
        return this.owner;
    }
    comments = null;
    flags = BlockTypeFlags.None;
    isUnloaded = false;
    get name() {
        return this.blockOwner?.name ?? '';
    }
    set name(value) {
        if (this.blockOwner) {
            this.blockOwner.name = value;
        }
    }
    get objectName() {
        return DxfFileToken.Block;
    }
    get objectType() {
        return ObjectType.BLOCK;
    }
    get subclassMarker() {
        return DxfSubclassMarker.BlockBegin;
    }
    xRefPath = null;
    constructor(record /* BlockRecord */) {
        super();
        if (record) {
            this.owner = record;
        }
    }
    clone() {
        const clone = super.clone();
        // TODO: clone.owner = new BlockRecord(this.name);
        return clone;
    }
    applyTransform(transform) {
        // TODO: apply block transform when block geometry support is implemented
    }
    getBoundingBox() {
        return null;
    }
}
