// @spectrum-web-components/color-handle/src/ColorHandle.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, SpectrumElement, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/color-loupe/sp-color-loupe.js';
import styles from './color-handle.css.js';
export const extractHueAndSaturationRegExp = /^hs[v|l]a?\s?\((\d{1,3}\.?\d*?),?\s?(\d{1,3})/;
export const replaceHueAndSaturationRegExp = /(^hs[v|l]a?\s?\()\d{1,3}\.?\d*?(,?\s?)\d{1,3}/;
export const replaceHueRegExp = /(^hs[v|l]a?\()\d{1,3}/;
/**
 * @element sp-color-handle
 */
export class ColorHandle extends SpectrumElement {
    constructor() {
        super(...arguments);
        this.disabled = false;
        this.focused = false;
        this.open = false;
        this.color = 'rgba(255, 0, 0, 0.5)';
    }
    static get styles() {
        return [styles];
    }
    handlePointerdown(event) {
        if (event.pointerType === 'touch') {
            this.open = true;
        }
        this.setPointerCapture(event.pointerId);
    }
    handlePointerup(event) {
        this.open = false;
        this.releasePointerCapture(event.pointerId);
    }
    render() {
        return html `
            <div class="color" style="background-color: ${this.color}"></div>
            <sp-color-loupe
                color=${this.color}
                ?open=${this.open && !this.disabled}
            ></sp-color-loupe>
        `;
    }
    firstUpdated(changed) {
        super.firstUpdated(changed);
        this.addEventListener('pointerdown', this.handlePointerdown);
        this.addEventListener('pointerup', this.handlePointerup);
        this.addEventListener('pointercancel', this.handlePointerup);
    }
}
__decorate([
    property({ type: Boolean, reflect: true })
], ColorHandle.prototype, "disabled", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], ColorHandle.prototype, "focused", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], ColorHandle.prototype, "open", void 0);
__decorate([
    property({ type: String })
], ColorHandle.prototype, "color", void 0);
