// @papyrs/stylo/dist/collection/stores/config.store.js
import { createStore } from '@stencil/store';
import { code } from '../plugins/code.plugin.js';
import { h1 } from '../plugins/h1.plugin.js';
import { h2 } from '../plugins/h2.plugin.js';
import { h3 } from '../plugins/h3.plugin.js';
import { hr } from '../plugins/hr.plugin.js';
import { img } from '../plugins/img.plugin.js';
import { ul } from '../plugins/ul.plugin.js';
import { DEFAULT_PALETTE } from '../types/palette.js';
export const DEFAULT_PLUGINS = [h1, h2, h3, ul, img, code, hr];
export const DEFAULT_TOOLBAR = {
  palette: DEFAULT_PALETTE,
  command: 'native',
  style: {
    list: false,
    align: true,
    fontSize: true,
    backgroundColor: true
  }
};
export const DEFAULT_PLACEHOLDERS = ['div', 'p', 'span'];
export const DEFAULT_TEXT_PARAGRAPHS = ['h1', 'h2', 'h3', 'div', 'p'];
export const DEFAULT_EXCLUDE_ATTRIBUTES = [
  'placeholder',
  'data-gramm',
  'class',
  'spellcheck',
  'contenteditable'
];
export const DEFAULT_DONT_INJECT_HEAD_CSS = false;
const { state, onChange } = createStore({
  plugins: DEFAULT_PLUGINS,
  toolbar: DEFAULT_TOOLBAR,
  placeholders: DEFAULT_PLACEHOLDERS,
  textParagraphs: DEFAULT_TEXT_PARAGRAPHS,
  menus: undefined,
  excludeAttributes: DEFAULT_EXCLUDE_ATTRIBUTES
});
export default { state, onChange };
