// @spectrum-web-components/avatar/stories/avatar.stories.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { html } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import '../sp-avatar.js';
import { avatar } from './images.js';
export default {
    component: 'sp-avatar',
    title: 'Avatar',
    argTypes: {
        label: { control: 'text' },
        src: { control: 'text' },
    },
    args: {
        label: 'Place dog',
        src: avatar,
    },
};
const Template = ({ label = 'Place Dog', src = avatar, size = 100, } = {}) => {
    return html `
        <sp-avatar label=${label} src=${src} size=${size}></sp-avatar>
    `;
};
export const size50 = (args = {}) => Template(Object.assign(Object.assign({}, args), { size: 50 }));
export const size75 = (args = {}) => Template(Object.assign(Object.assign({}, args), { size: 75 }));
export const size100 = (args = {}) => Template(Object.assign(Object.assign({}, args), { size: 100 }));
export const size200 = (args = {}) => Template(Object.assign(Object.assign({}, args), { size: 200 }));
export const size300 = (args = {}) => Template(Object.assign(Object.assign({}, args), { size: 300 }));
export const size400 = (args = {}) => Template(Object.assign(Object.assign({}, args), { size: 400 }));
export const size500 = (args = {}) => Template(Object.assign(Object.assign({}, args), { size: 500 }));
export const size600 = (args = {}) => Template(Object.assign(Object.assign({}, args), { size: 600 }));
export const size700 = (args = {}) => Template(Object.assign(Object.assign({}, args), { size: 700 }));
