// @spectrum-web-components/menu/src/MenuItem.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property, query, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/icons-ui/icons/sp-icon-checkmark100.js';
import { LikeAnchor } from '/web-component-designer-demo/node_modules/@spectrum-web-components/shared/src/like-anchor.js';
import { Focusable } from '/web-component-designer-demo/node_modules/@spectrum-web-components/shared/src/focusable.js';
import menuItemStyles from './menu-item.css.js';
import checkmarkStyles from '/web-component-designer-demo/node_modules/@spectrum-web-components/icon/src/spectrum-icon-checkmark.css.js';
export class MenuItemRemovedEvent extends Event {
    constructor() {
        super('sp-menu-item-removed', {
            bubbles: true,
            composed: true,
        });
        this.focused = false;
    }
    get item() {
        return this._item;
    }
    reset(item) {
        this._item = item;
    }
}
export class MenuItemAddedOrUpdatedEvent extends Event {
    constructor() {
        super('sp-menu-item-added-or-updated', {
            bubbles: true,
            composed: true,
        });
    }
    set focusRoot(root) {
        this.item.menuData.focusRoot = this.item.menuData.focusRoot || root;
    }
    set selectionRoot(root) {
        this.item.menuData.selectionRoot =
            this.item.menuData.selectionRoot || root;
    }
    get item() {
        return this._item;
    }
    set currentAncestorWithSelects(ancestor) {
        this._currentAncestorWithSelects = ancestor;
    }
    get currentAncestorWithSelects() {
        return this._currentAncestorWithSelects;
    }
    reset(item) {
        this._item = item;
        this._currentAncestorWithSelects = undefined;
        item.menuData = {
            focusRoot: undefined,
            selectionRoot: undefined,
        };
    }
}
const addOrUpdateEvent = new MenuItemAddedOrUpdatedEvent();
const removeEvent = new MenuItemRemovedEvent();
/**
 * @element sp-menu-item
 *
 * @slot - text content to display within the Menu Item
 * @slot icon - icon element to be placed at the start of the Menu Item
 * @slot value - content placed at the end of the Menu Item like values, keyboard shortcuts, etc.
 * @fires sp-menu-item-added - announces the item has been added so a parent menu can take ownerships
 * @fires sp-menu-item-removed - announces when removed from the DOM so the parent menu can remove ownership and update selected state
 */
export class MenuItem extends LikeAnchor(Focusable) {
    constructor() {
        super();
        this.active = false;
        this.focused = false;
        this.selected = false;
        this._value = '';
        this.noWrap = false;
        this.menuData = {
            focusRoot: undefined,
            selectionRoot: undefined,
        };
        this.proxyFocus = this.proxyFocus.bind(this);
        this.addEventListener('click', this.handleClickCapture, {
            capture: true,
        });
    }
    static get styles() {
        return [menuItemStyles, checkmarkStyles];
    }
    get value() {
        return this._value || this.itemText;
    }
    set value(value) {
        if (value === this._value) {
            return;
        }
        this._value = value || '';
        if (this._value) {
            this.setAttribute('value', this._value);
        }
        else {
            this.removeAttribute('value');
        }
    }
    /**
     * @private
     */
    get itemText() {
        return (this.textContent || /* c8 ignore next */ '').trim();
    }
    get focusElement() {
        return this;
    }
    get itemChildren() {
        if (this._itemChildren) {
            return this._itemChildren;
        }
        const iconSlot = this.shadowRoot.querySelector('slot[name="icon"]');
        const icon = !iconSlot
            ? []
            : iconSlot.assignedElements().map((element) => {
                const newElement = element.cloneNode(true);
                newElement.removeAttribute('slot');
                newElement.classList.toggle('icon');
                return newElement;
            });
        const contentSlot = this.shadowRoot.querySelector('slot:not([name])');
        const content = !contentSlot
            ? []
            : contentSlot.assignedNodes().map((node) => node.cloneNode(true));
        this._itemChildren = { icon, content };
        return this._itemChildren;
    }
    click() {
        if (this.disabled) {
            return;
        }
        if (this.shouldProxyClick()) {
            return;
        }
        super.click();
    }
    handleClickCapture(event) {
        if (this.disabled) {
            event.preventDefault();
            event.stopImmediatePropagation();
            event.stopPropagation();
            return false;
        }
    }
    proxyFocus() {
        this.focus();
    }
    shouldProxyClick() {
        let handled = false;
        if (this.anchorElement) {
            this.anchorElement.click();
            handled = true;
        }
        return handled;
    }
    breakItemChildrenCache() {
        this._itemChildren = undefined;
        this.triggerUpdate();
    }
    render() {
        return html `
            <slot name="icon" @slotchange=${this.breakItemChildrenCache}></slot>
            <div id="label">
                <slot
                    id="slot"
                    @slotchange=${this.breakItemChildrenCache}
                ></slot>
            </div>
            <slot name="value"></slot>
            ${this.selected
            ? html `
                      <sp-icon-checkmark100
                          id="selected"
                          class="spectrum-UIIcon-Checkmark100 icon checkmark"
                      ></sp-icon-checkmark100>
                  `
            : html ``}
            ${this.href && this.href.length > 0
            ? super.renderAnchor({
                id: 'button',
                ariaHidden: true,
                className: 'button anchor hidden',
            })
            : html ``}
        `;
    }
    handleRemoveActive() {
        this.active = false;
    }
    handlePointerdown() {
        this.active = true;
    }
    firstUpdated(changes) {
        super.firstUpdated(changes);
        this.setAttribute('tabindex', '-1');
        this.addEventListener('pointerdown', this.handlePointerdown);
        if (!this.hasAttribute('id')) {
            this.id = `sp-menu-item-${MenuItem.instanceCount++}`;
        }
    }
    updateAriaSelected() {
        const role = this.getAttribute('role');
        if (role === 'option') {
            this.setAttribute('aria-selected', this.selected ? 'true' : 'false');
        }
        else if (role === 'menuitemcheckbox' || role === 'menuitemradio') {
            this.setAttribute('aria-checked', this.selected ? 'true' : 'false');
        }
    }
    setRole(role) {
        this.setAttribute('role', role);
        this.updateAriaSelected();
    }
    updated(changes) {
        super.updated(changes);
        if (changes.has('label')) {
            this.setAttribute('aria-label', this.label || '');
        }
        if (changes.has('active')) {
            if (this.active) {
                this.addEventListener('pointerup', this.handleRemoveActive);
                this.addEventListener('pointerleave', this.handleRemoveActive);
            }
            else {
                this.removeEventListener('pointerup', this.handleRemoveActive);
                this.removeEventListener('pointerleave', this.handleRemoveActive);
            }
        }
        if (this.anchorElement) {
            this.anchorElement.addEventListener('focus', this.proxyFocus);
            this.anchorElement.tabIndex = -1;
        }
        if (changes.has('selected')) {
            this.updateAriaSelected();
        }
    }
    connectedCallback() {
        super.connectedCallback();
        addOrUpdateEvent.reset(this);
        this.dispatchEvent(addOrUpdateEvent);
        this._parentElement = this.parentElement;
    }
    disconnectedCallback() {
        var _a;
        removeEvent.reset(this);
        (_a = this._parentElement) === null || _a === void 0 ? void 0 : _a.dispatchEvent(removeEvent);
        super.disconnectedCallback();
    }
    async triggerUpdate() {
        await new Promise((ready) => requestAnimationFrame(ready));
        addOrUpdateEvent.reset(this);
        this.dispatchEvent(addOrUpdateEvent);
    }
}
MenuItem.instanceCount = 0;
__decorate([
    property({ type: Boolean, reflect: true })
], MenuItem.prototype, "active", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MenuItem.prototype, "focused", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MenuItem.prototype, "selected", void 0);
__decorate([
    property({ type: String })
], MenuItem.prototype, "value", null);
__decorate([
    property({
        type: Boolean,
        reflect: true,
        attribute: 'no-wrap',
        hasChanged() {
            return false;
        },
    })
], MenuItem.prototype, "noWrap", void 0);
__decorate([
    query('.anchor')
], MenuItem.prototype, "anchorElement", void 0);
