// @spectrum-web-components/menu/src/MenuGroup.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { queryAssignedNodes } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import { Menu } from './Menu.js';
import '../sp-menu.js';
import menuGroupStyles from './menu-group.css.js';
/**
 * @element sp-menu-group
 *
 * @slot header - headline of the menu group
 * @slot - menu items to be listed in the group
 */
export class MenuGroup extends Menu {
    constructor() {
        super();
        MenuGroup.instances += 1;
        this.headerId = `sp-menu-group-label-${MenuGroup.instances}`;
    }
    static get styles() {
        return [...super.styles, menuGroupStyles];
    }
    get ownRole() {
        switch (this.selects) {
            case 'multiple':
            case 'single':
            case 'inherit':
                return 'group';
            default:
                return 'menu';
        }
    }
    updateLabel() {
        const headerElement = this.headerElements.length
            ? this.headerElements[0]
            : undefined;
        if (headerElement !== this.headerElement) {
            if (this.headerElement && this.headerElement.id === this.headerId) {
                this.headerElement.removeAttribute('id');
            }
            if (headerElement) {
                const headerId = headerElement.id || this.headerId;
                if (!headerElement.id) {
                    headerElement.id = headerId;
                }
                this.setAttribute('aria-labelledby', headerId);
            }
            else {
                this.removeAttribute('aria-labelledby');
            }
        }
        this.headerElement = headerElement;
    }
    render() {
        return html `
            <span class="header" aria-hidden="true">
                <slot name="header" @slotchange=${this.updateLabel}></slot>
            </span>
            <sp-menu role="none">
                <slot></slot>
            </sp-menu>
        `;
    }
}
MenuGroup.instances = 0;
__decorate([
    queryAssignedNodes('header', true)
], MenuGroup.prototype, "headerElements", void 0);
