// @node-projects/acad-ts/dist/IO/DXF/DxfDocumentBuilder.js
import { Entity } from '../../Entities/Entity.js';
import { CadDictionary } from '../../Objects/CadDictionary.js';
import { CadEntityTemplate } from '../Templates/CadEntityTemplate.js';
import { CadBlockRecordTemplate } from '../Templates/CadBlockRecordTemplate.js';
import { CadPolyLineTemplate } from '../Templates/CadPolyLineTemplate.js';
import { CadInsertTemplate } from '../Templates/CadInsertTemplate.js';
import { Vertex } from '../../Entities/Vertex.js';
import { Seqend } from '../../Entities/Seqend.js';
import { AttributeEntity } from '../../Entities/AttributeEntity.js';
import { BlockRecord } from '../../Tables/BlockRecord.js';
import { CadDocumentBuilder } from '../CadDocumentBuilder.js';
import { NotificationType } from '../NotificationEventHandler.js';
export class DxfDocumentBuilder extends CadDocumentBuilder {
    Configuration;
    get KeepUnknownEntities() {
        return this.Configuration.KeepUnknownEntities;
    }
    get KeepUnknownNonGraphicalObjects() {
        return this.Configuration.KeepUnknownNonGraphicalObjects;
    }
    ModelSpaceEntities = new Set();
    ModelSpaceTemplate = null;
    OrphanTemplates = [];
    constructor(version, document, configuration) {
        super(version, document);
        this.Configuration = configuration;
    }
    BuildDocument() {
        if (this.ModelSpaceTemplate === null) {
            const record = BlockRecord.ModelSpace;
            this.BlockRecords.add(record);
            this.ModelSpaceTemplate = new CadBlockRecordTemplate(record);
            this.AddTemplate(this.ModelSpaceTemplate);
        }
        this.createMissingHandles();
        for (const entity of this.ModelSpaceEntities) {
            this.ModelSpaceTemplate.OwnedObjectsHandlers.add(entity.handle);
        }
        this.RegisterTables();
        this.BuildTables();
        this.buildDictionaries();
        for (const template of this.OrphanTemplates) {
            this.assignOwner(template);
        }
        super.BuildDocument();
        if (this.Configuration.CreateDefaults) {
            this.DocumentToBuild.createDefaults();
        }
    }
    BuildEntities() {
        const entities = [];
        for (const item of this.cadObjectsTemplates.values()) {
            if (item instanceof CadEntityTemplate) {
                item.Build(this);
                item.SetUnlinkedReferences();
            }
        }
        for (const item of this.cadObjectsTemplates.values()) {
            if (item instanceof CadEntityTemplate && item.CadObject.owner === null) {
                item.CadObject.handle = 0;
                entities.push(item.CadObject);
            }
        }
        return entities;
    }
    assignOwner(template) {
        if (template.CadObject.owner !== null || template.CadObject instanceof CadDictionary || !template.OwnerHandle) {
            return;
        }
        const ownerResult = this.TryGetObjectTemplate(template.OwnerHandle);
        if (ownerResult) {
            const owner = ownerResult;
            if (owner instanceof CadBlockRecordTemplate && template.CadObject instanceof Entity) {
                // The entries should be assigned in the blocks or entities section
            }
            else if (owner instanceof CadPolyLineTemplate && template.CadObject instanceof Vertex) {
                owner.OwnedObjectsHandlers.add(template.CadObject.handle);
            }
            else if (owner instanceof CadPolyLineTemplate && template.CadObject instanceof Seqend) {
                owner.SeqendHandle = template.CadObject.handle;
            }
            else if (owner instanceof CadInsertTemplate && template.CadObject instanceof AttributeEntity) {
                owner.OwnedObjectsHandlers.add(template.CadObject.handle);
            }
            else if (owner instanceof CadInsertTemplate && template.CadObject instanceof Seqend) {
                owner.SeqendHandle = template.CadObject.handle;
            }
            else {
                this.Notify(`Owner ${owner.constructor.name} with handle ${template.OwnerHandle} assignation not implemented for ${template.CadObject.constructor.name} with handle ${template.CadObject.handle}`, NotificationType.Warning);
            }
        }
        else {
            this.Notify(`Owner ${template.OwnerHandle} not found for ${template.constructor.name} with handle ${template.CadObject.handle}`, NotificationType.Warning);
        }
    }
}
