// @node-projects/acad-ts/dist/IO/DXF/DxfWriter.js
import { CadWriterBase } from '../CadWriterBase.js';
import { DxfWriterConfiguration } from './DxfWriterConfiguration.js';
import { DxfAsciiWriter } from './DxfStreamWriter/DxfAsciiWriter.js';
import { DxfBinaryWriter } from './DxfStreamWriter/DxfBinaryWriter.js';
import { DxfHeaderSectionWriter } from './DxfStreamWriter/DxfHeaderSectionWriter.js';
import { DxfClassesSectionWriter } from './DxfStreamWriter/DxfClassesSectionWriter.js';
import { DxfTablesSectionWriter } from './DxfStreamWriter/DxfTablesSectionWriter.js';
import { DxfBlocksSectionWriter } from './DxfStreamWriter/DxfBlocksSectionWriter.js';
import { DxfEntitiesSectionWriter } from './DxfStreamWriter/DxfEntitiesSectionWriter.js';
import { DxfObjectsSectionWriter } from './DxfStreamWriter/DxfObjectsSectionWriter.js';
import { CadObjectHolder } from './CadObjectHolder.js';
import { DxfCode } from '../../DxfCode.js';
import { DxfFileToken } from '../../DxfFileToken.js';
class Uint8ArrayTextOutput {
    _stream;
    _encoder = new TextEncoder();
    _position = 0;
    constructor(stream) {
        this._stream = stream;
    }
    write(value) {
        const bytes = this._encoder.encode(value);
        this.ensureCapacity(bytes.length);
        this._stream.set(bytes, this._position);
        this._position += bytes.length;
    }
    flush() { }
    close() { }
    ensureCapacity(length) {
        if (this._position + length > this._stream.length) {
            throw new Error('DXF output buffer is too small.');
        }
    }
}
class Uint8ArrayBinaryOutput {
    _stream;
    _position = 0;
    constructor(stream) {
        this._stream = stream;
    }
    write(value) {
        this.ensureCapacity(value.length);
        this._stream.set(value, this._position);
        this._position += value.length;
    }
    flush() { }
    close() { }
    ensureCapacity(length) {
        if (this._position + length > this._stream.length) {
            throw new Error('DXF output buffer is too small.');
        }
    }
}
export class DxfWriter extends CadWriterBase {
    IsBinary;
    _writer;
    _objectHolder = new CadObjectHolder();
    constructor(stream, document, binary = false) {
        super(stream, document);
        this.IsBinary = binary;
        this.Configuration = new DxfWriterConfiguration();
    }
    Write() {
        super.Write();
        this.createStreamWriter();
        this._objectHolder.Objects.push(this._document.rootDictionary);
        this.writeHeader();
        this.writeDxfClasses();
        this.writeTables();
        this.writeBlocks();
        this.writeEntities();
        this.writeObjects();
        this.writeACDSData();
        this._writer.Write(DxfCode.Start, DxfFileToken.EndOfFile);
        this._writer.Flush();
        if (this.Configuration.CloseStream) {
            this._writer.Close();
        }
    }
    Dispose() {
        this._writer.Dispose();
    }
    static WriteToStream(stream, document, binary = false, configuration, notification) {
        const writer = new DxfWriter(stream, document, binary);
        if (configuration) {
            writer.Configuration = configuration;
        }
        writer.OnNotification = notification ?? null;
        writer.Write();
        writer.Dispose();
    }
    createStreamWriter() {
        if (this.IsBinary) {
            this._writer = new DxfBinaryWriter(this.createBinaryTarget());
        }
        else {
            this._writer = new DxfAsciiWriter(this.createTextTarget());
        }
        this._writer.WriteOptional = this.Configuration.WriteOptionalValues;
    }
    createTextTarget() {
        if (this._stream instanceof Uint8Array) {
            return new Uint8ArrayTextOutput(this._stream);
        }
        return this._stream;
    }
    createBinaryTarget() {
        if (this._stream instanceof Uint8Array) {
            return new Uint8ArrayBinaryOutput(this._stream);
        }
        return this._stream;
    }
    writeHeader() {
        const writer = new DxfHeaderSectionWriter(this._writer, this._document, this._objectHolder, this.Configuration);
        writer.OnNotification = this.triggerNotification.bind(this);
        writer.Write();
    }
    writeDxfClasses() {
        const writer = new DxfClassesSectionWriter(this._writer, this._document, this._objectHolder, this.Configuration);
        writer.OnNotification = this.triggerNotification.bind(this);
        writer.Write();
    }
    writeTables() {
        const writer = new DxfTablesSectionWriter(this._writer, this._document, this._objectHolder, this.Configuration);
        writer.OnNotification = this.triggerNotification.bind(this);
        writer.Write();
    }
    writeBlocks() {
        const writer = new DxfBlocksSectionWriter(this._writer, this._document, this._objectHolder, this.Configuration);
        writer.OnNotification = this.triggerNotification.bind(this);
        writer.Write();
    }
    writeEntities() {
        const writer = new DxfEntitiesSectionWriter(this._writer, this._document, this._objectHolder, this.Configuration);
        writer.OnNotification = this.triggerNotification.bind(this);
        writer.Write();
    }
    writeObjects() {
        const writer = new DxfObjectsSectionWriter(this._writer, this._document, this._objectHolder, this.Configuration);
        writer.OnNotification = this.triggerNotification.bind(this);
        writer.Write();
    }
    writeACDSData() {
        // not implemented
    }
    createDefaultConfiguration() {
        return new DxfWriterConfiguration();
    }
}
