// @node-projects/layout2vector/dist/index.js
/**
 * @node-projects/html-converter
 *
 * DOM → Geometry → DXF/PDF library.
 * Extracts rendered layout geometry from a live DOM and converts it to DXF or PDF.
 */
// Pipeline
export { extractIR, renderIR } from "./pipeline.js";
// Traversal (advanced usage)
export { traverseDOM, flattenStackingOrder, extractStyle, isVisible, createsStackingContext, isSVGElement, isSVGRoot, } from "./traversal.js";
// Geometry utilities
export { getElementQuad, getElementQuads, getElementOrigin, quadSize, getSvgScreenCtm, roundedQuadPath } from "./geometry.js";
// Extractors (advanced usage)
export { extractHTMLGeometry } from "./extractors/html-extractor.js";
export { extractSVGSubtree } from "./extractors/svg-extractor.js";
export { extractImageGeometry, isImageElement, hasBackgroundImage, extractBackgroundImage, preloadImages } from "./extractors/image-extractor.js";
export { isMathMLRoot, isMathMLElement, extractMathMLFeatures } from "./extractors/mathml-extractor.js";
// Writers
export { DXFWriter } from "./writers/dxf-writer.js";
export { EMFWriter } from "./writers/emf-writer.js";
export { PDFWriter } from "./writers/pdf-writer.js";
export { CanvasWriter } from "./writers/canvas-writer.js";
export { ImageWriter, ImageResult } from "./writers/image-writer.js";
/** @deprecated Use ImageWriter/ImageResult/ImageWriterOptions instead. */
export { PNGWriter, PNGResult } from "./writers/png-writer.js";
export { SVGWriter } from "./writers/svg-writer.js";
export { HTMLWriter } from "./writers/html-writer.js";
export { DWGWriter, AcadDXFWriter } from "./writers/acad-writer.js";
// Font utilities
export { parseTTF } from "./ttf-parser.js";
