// @node-projects/layout2vector/dist/index.js
/**
 * @node-projects/html-converter
 *
 * DOM → Geometry → DXF/PDF library.
 * Extracts rendered layout geometry from a live DOM and converts it to DXF or PDF.
 */
// Pipeline
export { extractIR, renderIR } from "./pipeline.js";
// Traversal (advanced usage)
export { traverseDOM, flattenStackingOrder, extractStyle, isVisible, createsStackingContext, isSVGElement, isSVGRoot, } from "./traversal.js";
// Geometry utilities
export { getElementQuad, getElementQuads, getElementOrigin, quadSize, getSvgScreenCtm, roundedQuadPath } from "./geometry.js";
// Extractors (advanced usage)
export { extractHTMLGeometry } from "./html-extractor.js";
export { extractSVGSubtree } from "./svg-extractor.js";
export { extractImageGeometry, isImageElement, hasBackgroundImage, extractBackgroundImage, preloadImages } from "./image-extractor.js";
export { isMathMLRoot, isMathMLElement, extractMathMLFeatures } from "./mathml-extractor.js";
// Writers
export { DXFWriter } from "./dxf-writer.js";
export { EMFWriter } from "./emf-writer.js";
export { PDFWriter } from "./pdf-writer.js";
export { ImageWriter, ImageResult } from "./image-writer.js";
/** @deprecated Use ImageWriter/ImageResult/ImageWriterOptions instead. */
export { PNGWriter, PNGResult } from "./png-writer.js";
export { SVGWriter } from "./svg-writer.js";
export { HTMLWriter } from "./html-writer.js";
// Font utilities
export { parseTTF } from "./ttf-parser.js";
