// @spectrum-web-components/tags/src/Tag.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, SizedMixin, SpectrumElement, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/button/sp-clear-button.js';
import styles from './tag.css.js';
/**
 * @element sp-tag
 *
 * @slot - text content for labeling the tag
 * @slot avatar - an avatar element to display within the Tag
 * @slot icon - an icon element to display within the Tag
 */
export class Tag extends SizedMixin(SpectrumElement, {
    validSizes: ['s', 'm', 'l'],
}) {
    constructor() {
        super();
        this.deletable = false;
        this.disabled = false;
        this.readonly = false;
        this.handleFocusin = () => {
            this.addEventListener('focusout', this.handleFocusout);
            this.addEventListener('keydown', this.handleKeydown);
        };
        this.handleFocusout = () => {
            this.removeEventListener('keydown', this.handleKeydown);
            this.removeEventListener('focusout', this.handleFocusout);
        };
        this.handleKeydown = (event) => {
            if (!this.deletable) {
                return;
            }
            const { code } = event;
            switch (code) {
                case 'Backspace':
                case 'Space':
                case 'Delete':
                    this.delete();
                    return;
                default:
                    return;
            }
        };
        this.addEventListener('focusin', this.handleFocusin);
    }
    static get styles() {
        return [styles];
    }
    get hasIcon() {
        return !!this.querySelector('[slot="icon"]');
    }
    get hasAvatar() {
        return !!this.querySelector('[slot="avatar"]');
    }
    delete() {
        if (this.readonly) {
            return;
        }
        this.dispatchEvent(new Event('delete', {
            bubbles: true,
        }));
    }
    render() {
        const slots = [];
        if (this.hasAvatar) {
            slots.push(html `
                    <slot name="avatar"></slot>
                `);
        }
        if (this.hasIcon) {
            slots.push(html `
                    <slot name="icon"></slot>
                `);
        }
        return html `
            ${slots}
            <span class="label"><slot></slot></span>
            ${this.deletable
            ? html `
                      <sp-clear-button
                          class="clear-button"
                          ?disabled=${this.disabled}
                          label="Remove"
                          size="s"
                          tabindex="-1"
                          @click=${this.delete}
                      ></sp-clear-button>
                  `
            : html ``}
        `;
    }
    firstUpdated(changes) {
        super.firstUpdated(changes);
        if (!this.hasAttribute('role')) {
            this.setAttribute('role', 'listitem');
        }
        if (this.deletable) {
            this.setAttribute('tabindex', !this.disabled && this.matches(':first-of-type:not([disabled])')
                ? '0'
                : '-1');
        }
    }
    updated(changes) {
        super.updated(changes);
        if (changes.has('disabled')) {
            if (this.disabled) {
                this.setAttribute('aria-disabled', 'true');
            }
            else {
                this.removeAttribute('aria-disabled');
            }
        }
    }
}
__decorate([
    property({ type: Boolean, reflect: true })
], Tag.prototype, "deletable", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], Tag.prototype, "disabled", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], Tag.prototype, "readonly", void 0);
