// @spectrum-web-components/tags/src/Tags.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, SpectrumElement, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { queryAssignedNodes } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import { FocusVisiblePolyfillMixin } from '/web-component-designer-demo/node_modules/@spectrum-web-components/shared/src/focus-visible.js';
import { RovingTabindexController } from '/web-component-designer-demo/node_modules/@spectrum-web-components/reactive-controllers/src/RovingTabindex.js';
import { Tag } from './Tag.js';
import styles from './tags.css.js';
/**
 * @element sp-tags
 *
 * @slot - Tag elements to manage as a group
 */
export class Tags extends FocusVisiblePolyfillMixin(SpectrumElement) {
    constructor() {
        super();
        this.rovingTabindexController = new RovingTabindexController(this, {
            focusInIndex: (elements) => {
                return elements.findIndex((el) => {
                    return !el.disabled && el.deletable;
                });
            },
            elements: () => this.tags,
            isFocusableElement: (el) => !el.disabled && el.deletable,
        });
        this.handleFocusin = () => {
            this.addEventListener('focusout', this.handleFocusout);
            this.addEventListener('keydown', this.handleKeydown);
        };
        this.handleKeydown = (event) => {
            const { code } = event;
            if (code !== 'PageUp' && code !== 'PageDown')
                return;
            const circularIndexedElement = (list, index) => list[(list.length + index) % list.length];
            const tagsSiblings = [
                ...this.getRootNode().querySelectorAll('sp-tags'),
            ];
            if (tagsSiblings.length < 2) {
                return;
            }
            event.preventDefault();
            const currentIndex = tagsSiblings.indexOf(this);
            const offset = code === 'PageUp' ? -1 : 1;
            let nextTagsIndex = currentIndex + offset;
            let nextTags = circularIndexedElement(tagsSiblings, nextTagsIndex);
            while (!nextTags.tags.length) {
                nextTagsIndex += offset;
                nextTags = circularIndexedElement(tagsSiblings, nextTagsIndex);
            }
            nextTags.focus();
        };
        this.handleFocusout = () => {
            this.removeEventListener('keydown', this.handleKeydown);
            this.removeEventListener('focusout', this.handleFocusout);
        };
        this.addEventListener('focusin', this.handleFocusin);
    }
    static get styles() {
        return [styles];
    }
    get tags() {
        return this.defaultNodes.filter((node) => node instanceof Tag);
    }
    focus() {
        this.rovingTabindexController.focus();
    }
    handleSlotchange() {
        this.rovingTabindexController.clearElementCache();
    }
    render() {
        return html `
            <slot @slotchange=${this.handleSlotchange}></slot>
        `;
    }
    firstUpdated() {
        if (!this.hasAttribute('role')) {
            this.setAttribute('role', 'list');
        }
        if (!this.hasAttribute('aria-label')) {
            this.setAttribute('aria-label', 'Tags');
        }
    }
}
__decorate([
    queryAssignedNodes('')
], Tags.prototype, "defaultNodes", void 0);
