// @node-projects/acad-ts/dist/IO/Templates/CadBlockVisibilityParameterTemplate.js
import { BlockVisibilityParameter } from '../../Objects/Evaluations/BlockVisibilityParameter.js';
import { BlockVisibilityState } from '../../Objects/Evaluations/BlockVisibilityParameter.js';
import { CadBlock1PtParameterTemplate } from './CadBlock1PtParameterTemplate.js';
export class CadBlockVisibilityParameterTemplate extends CadBlock1PtParameterTemplate {
    entityHandles = [];
    StateTemplates = [];
    constructor(cadObject) {
        super(cadObject ?? new BlockVisibilityParameter());
    }
    build(builder) {
        super.build(builder);
        const bvp = this.CadObject;
        for (const handle of this.entityHandles) {
            const entity = builder.TryGetCadObject(handle);
            if (entity) {
                bvp.entities.push(entity);
            }
            else {
                builder.Notify(`[${bvp.toString()}] entity with handle ${handle} not found.`);
            }
        }
        for (const item of this.StateTemplates) {
            item.Build(builder, this.entityHandles);
            bvp.addState(item.state);
        }
    }
}
(function (CadBlockVisibilityParameterTemplate) {
    class StateTemplate {
        state = new BlockVisibilityState();
        entityHandles = new Set();
        expressionHandles = new Set();
        constructor(state) {
            if (state) {
                this.state = state;
            }
        }
        Build(builder, parentEntityHandles) {
            this.setEntities(builder, this.state.entities, this.entityHandles, parentEntityHandles);
            this.setEntities(builder, this.state.expressions, this.expressionHandles, null);
        }
        setEntities(builder, subset, handles, entities) {
            for (const h of handles) {
                if (entities !== null && !entities.includes(h)) {
                    builder.Notify(`[${this.state.toString()}] parent does not contain handle ${h}.`);
                }
                const obj = builder.TryGetCadObject(h);
                if (obj) {
                    subset.push(obj);
                }
                else {
                    builder.Notify(`[${this.state.toString()}] object with handle ${h} not found.`);
                }
            }
        }
    }
    CadBlockVisibilityParameterTemplate.StateTemplate = StateTemplate;
})(CadBlockVisibilityParameterTemplate || (CadBlockVisibilityParameterTemplate = {}));
