// @node-projects/acad-ts/dist/IO/Templates/CadDimensionTemplate.js
import { Dimension, DimensionType } from '../../Entities/Dimension.js';
import { DimensionAligned } from '../../Entities/DimensionAligned.js';
import { DimensionLinear } from '../../Entities/DimensionLinear.js';
import { ObjectType } from '../../Types/ObjectType.js';
import { CadEntityTemplate } from './CadEntityTemplate.js';
export class CadDimensionTemplate extends CadEntityTemplate {
    StyleHandle = null;
    BlockHandle = null;
    BlockName = null;
    StyleName = null;
    constructor(dimension) {
        super(dimension ?? new DimensionPlaceholder());
    }
    build(builder) {
        super.build(builder);
        const dimension = this.CadObject;
        const style = this.getTableReference(builder, this.StyleHandle, this.StyleName ?? '');
        if (style) {
            dimension.style = style;
        }
        const block = this.getTableReference(builder, this.BlockHandle, this.BlockName ?? '');
        if (block) {
            dimension.block = block;
        }
    }
    SetDimensionFlags(flags) {
        const dimension = this.CadObject;
        dimension.flags = flags;
    }
    SetDimensionObject(dimension) {
        dimension.handle = this.CadObject.handle;
        dimension.owner = this.CadObject.owner;
        dimension.xDictionary = this.CadObject.xDictionary;
        dimension.color = this.CadObject.color;
        dimension.lineWeight = this.CadObject.lineWeight;
        dimension.lineTypeScale = this.CadObject.lineTypeScale;
        dimension.isInvisible = this.CadObject.isInvisible;
        dimension.transparency = this.CadObject.transparency;
        const source = this.CadObject;
        dimension.version = source.version;
        dimension.definitionPoint = source.definitionPoint;
        dimension.textMiddlePoint = source.textMiddlePoint;
        dimension.insertionPoint = source.insertionPoint;
        dimension.normal = source.normal;
        dimension.isTextUserDefinedLocation = source.isTextUserDefinedLocation;
        dimension.attachmentPoint = source.attachmentPoint;
        dimension.lineSpacingStyle = source.lineSpacingStyle;
        dimension.lineSpacingFactor = source.lineSpacingFactor;
        dimension.text = source.text;
        dimension.textRotation = source.textRotation;
        dimension.horizontalDirection = source.horizontalDirection;
        dimension.flags = source.flags;
        if (this.CadObject instanceof DimensionAligned &&
            dimension instanceof DimensionLinear) {
            const aligned = this.CadObject;
            const linear = dimension;
            linear.firstPoint = aligned.firstPoint;
            linear.secondPoint = aligned.secondPoint;
            linear.extLineRotation = aligned.extLineRotation;
        }
        this.CadObject = dimension;
    }
}
export class DimensionPlaceholder extends Dimension {
    get objectType() { return ObjectType.INVALID; }
    get measurement() { return 0; }
    getBoundingBox() { return null; }
    constructor() {
        super(DimensionType.Linear);
    }
}
