// @node-projects/acad-ts/dist/IO/Templates/CadPolyLineTemplate.js
import { Polyline2D } from '../../Entities/Polyline2D.js';
import { Polyline3D } from '../../Entities/Polyline3D.js';
import { PolyfaceMesh } from '../../Entities/PolyfaceMesh.js';
import { PolygonMesh } from '../../Entities/PolygonMesh.js';
import { Seqend } from '../../Entities/Seqend.js';
import { VertexFaceMesh } from '../../Entities/VertexFaceMesh.js';
import { VertexFaceRecord } from '../../Entities/VertexFaceRecord.js';
import { ObjectType } from '../../Types/ObjectType.js';
import { Polyline } from '../../Entities/Polyline.js';
import { NotificationType } from '../NotificationEventHandler.js';
import { CadEntityTemplate } from './CadEntityTemplate.js';
export class CadPolyLineTemplate extends CadEntityTemplate {
    FirstVertexHandle = null;
    LastVertexHandle = null;
    SeqendHandle = null;
    OwnedObjectsHandlers = new Set();
    constructor(entity) {
        super(entity ? entity : new PolyLinePlaceholder());
    }
    SetPolyLineObject(polyLine) {
        polyLine.handle = this.CadObject.handle;
        polyLine.color = this.CadObject.color;
        polyLine.lineWeight = this.CadObject.lineWeight;
        polyLine.lineTypeScale = this.CadObject.lineTypeScale;
        polyLine.isInvisible = this.CadObject.isInvisible;
        polyLine.transparency = this.CadObject.transparency;
        this.CadObject = polyLine;
    }
    addVertices(builder, ...vertices) {
        const obj = this.CadObject;
        if (obj instanceof Polyline2D) {
            for (const v of vertices) {
                obj.vertices.push(v);
                v.owner = obj;
            }
        }
        else if (obj instanceof Polyline3D) {
            for (const v of vertices) {
                obj.vertices.push(v);
                v.owner = obj;
            }
        }
        else if (obj instanceof PolyfaceMesh) {
            for (const item of vertices) {
                this.addPolyfaceMeshVertex(builder, obj, item);
            }
        }
        else if (obj instanceof PolygonMesh) {
            for (const v of vertices) {
                obj.vertices.push(v);
                v.owner = obj;
            }
        }
        else {
            builder.Notify(`Unknown polyline type ${this.CadObject.subclassMarker}`, NotificationType.Warning);
        }
    }
    build(builder) {
        super.build(builder);
        const seqend = builder.TryGetCadObject(this.SeqendHandle);
        if (seqend) {
            this.setSeqend(builder, seqend);
        }
        if (this.FirstVertexHandle != null) {
            const vertices = Array.from(this.getEntitiesCollection(builder, this.FirstVertexHandle, this.LastVertexHandle));
            this.addVertices(builder, ...vertices);
        }
        else {
            if (this.CadObject instanceof PolyfaceMesh) {
                this.buildPolyfaceMesh(this.CadObject, builder);
            }
            else {
                for (const handle of this.OwnedObjectsHandlers) {
                    const v = builder.TryGetCadObject(handle);
                    if (v) {
                        this.addVertices(builder, v);
                    }
                    else {
                        const s = builder.TryGetCadObject(handle);
                        if (s) {
                            this.setSeqend(builder, s);
                        }
                        else {
                            builder.Notify(`Vertex ${handle} not found for polyline ${this.CadObject.handle}`, NotificationType.Warning);
                        }
                    }
                }
            }
        }
    }
    setSeqend(builder, seqend) {
        const obj = this.CadObject;
        seqend.owner = obj;
        if (obj instanceof Polyline2D) {
            obj.vertices.Seqend = seqend;
        }
        else if (obj instanceof Polyline3D) {
            obj.vertices.Seqend = seqend;
        }
        else if (obj instanceof PolyfaceMesh) {
            obj.vertices.Seqend = seqend;
        }
        else if (obj instanceof PolygonMesh) {
            obj.vertices.Seqend = seqend;
        }
        else {
            builder.Notify(`Unknown polyline type ${this.CadObject.subclassMarker}`, NotificationType.Warning);
        }
    }
    buildPolyfaceMesh(polyfaceMesh, builder) {
        for (const handle of this.OwnedObjectsHandlers) {
            const e = builder.TryGetCadObject(handle);
            if (e) {
                this.addPolyfaceMeshVertex(builder, polyfaceMesh, e);
            }
        }
    }
    addPolyfaceMeshVertex(builder, polyfaceMesh, e) {
        if (e instanceof VertexFaceMesh) {
            polyfaceMesh.vertices.push(e);
            e.owner = polyfaceMesh;
        }
        else if (e instanceof VertexFaceRecord) {
            polyfaceMesh.faces.push(e);
            e.owner = polyfaceMesh;
        }
        else if (e instanceof Seqend) {
            polyfaceMesh.vertices.Seqend = e;
            e.owner = polyfaceMesh;
        }
        else {
            builder.Notify(`Unidentified type for PolyfaceMesh ${e.constructor.name}`);
        }
    }
}
class PolyLinePlaceholder extends Polyline {
    get objectType() { return ObjectType.INVALID; }
    get subclassMarker() { return 'PolyLinePlaceholder'; }
    getBoundingBox() { return null; }
}
