// @node-projects/acad-ts/dist/IO/Templates/CadTableEntityTemplate.js
import { TableEntity, TableEntityCell, CellStyle } from '../../Entities/TableEntity.js';
import { CadInsertTemplate } from './CadInsertTemplate.js';
export class CadTableEntityTemplate extends CadInsertTemplate {
    BlockOwnerHandle = null;
    cadTableCellTemplates = [];
    get CurrentCell() { return this.CurrentCellTemplate.Cell; }
    CurrentCellTemplate;
    HorizontalMargin = null;
    NullHandle = null;
    styleHandle = null;
    get TableEntity() { return this.CadObject; }
    _currCellIndex = 0;
    constructor(table) {
        super(table ?? new TableEntity());
    }
    CreateCell(type) {
        const rowIndex = Math.floor(this._currCellIndex / this.TableEntity.columns.length);
        const cell = new TableEntityCell();
        cell.type = type;
        this.TableEntity.rows[rowIndex].cells.push(cell);
        this.CurrentCellTemplate = new CadTableCellTemplate(cell);
        this.cadTableCellTemplates.push(this.CurrentCellTemplate);
        this._currCellIndex++;
    }
    build(builder) {
        super.build(builder);
        for (const cellTemplate of this.cadTableCellTemplates) {
            cellTemplate.Build(builder);
        }
    }
}
export class CadTableCellContentFormatTemplate {
    Format;
    TextStyleHandle = null;
    TextStyleName = null;
    constructor(format) {
        this.Format = format;
    }
    Build(builder) {
        throw new Error('Not implemented');
    }
}
export class CadCellStyleTemplate extends CadTableCellContentFormatTemplate {
    BorderLinetypePairs = [];
    get CellStyle() { return this.Format; }
    textStyleHandle = null;
    constructor(style) {
        super(style ?? new CellStyle());
    }
}
export class CadTableAttributeTemplate {
    AttDefHandle = null;
    _tableAtt;
    constructor(tableAtt) {
        this._tableAtt = tableAtt;
    }
    Build(builder) {
        throw new Error('Not implemented');
    }
}
export class CadTableCellContentTemplate {
    BlockRecordHandle = null;
    CadValueTemplate = null;
    Content;
    FieldHandle = null;
    constructor(content) {
        this.Content = content;
    }
    Build(builder) {
        throw new Error('Not implemented');
    }
}
export class CadTableCellTemplate {
    AttributeHandles = new Set();
    Cell;
    ContentTemplates = [];
    FormatTextHeight = null;
    StyleId = 0;
    TextStyleOverrideHandle = null;
    UnknownHandle = null;
    ValueHandle = null;
    constructor(cell) {
        this.Cell = cell;
    }
    Build(builder) {
        const cadObject = builder.TryGetCadObject(this.ValueHandle);
    }
}
