// @node-projects/acad-ts/dist/IO/Templates/CadTemplate.js
import { NotificationType } from '../NotificationEventHandler.js';
export class CadTemplate {
    CadObject;
    EDataTemplate = new Map();
    EDataTemplateByAppName = new Map();
    HasBeenBuilt = false;
    OwnerHandle = null;
    ReactorsHandles = new Set();
    XDictHandle = null;
    constructor(cadObject) {
        this.CadObject = cadObject;
    }
    Build(builder) {
        if (this.HasBeenBuilt) {
            return;
        }
        else {
            this.HasBeenBuilt = true;
        }
        this.build(builder);
    }
    toString() {
        return `${this.CadObject?.toString()}`;
    }
    build(builder) {
        const cadDictionary = builder.TryGetCadObject(this.XDictHandle);
        if (cadDictionary) {
            this.CadObject.xDictionary = cadDictionary;
        }
        for (const handle of this.ReactorsHandles) {
            const reactor = builder.TryGetCadObject(handle);
            if (reactor) {
                this.CadObject.addReactor(reactor);
            }
            else {
                builder.Notify(`Reactor with handle ${handle} not found`, NotificationType.Warning);
            }
        }
        for (const [key, value] of this.EDataTemplate) {
            const app = builder.TryGetCadObject(key);
            if (app) {
                this.CadObject.extendedData.set(app, value);
            }
            else {
                builder.Notify(`AppId in extended data with handle ${key} not found`, NotificationType.Warning);
            }
        }
        for (const [key, value] of this.EDataTemplateByAppName) {
            const app = builder.TryGetTableEntry(key);
            if (app) {
                this.CadObject.extendedData.set(app, value);
            }
            else {
                builder.Notify(`AppId in extended data with handle ${key} not found`, NotificationType.Warning);
            }
        }
    }
    *getEntitiesCollection(builder, firstHandle, endHandle) {
        let template = builder.GetObjectTemplate(firstHandle);
        if (!template) {
            builder.Notify(`Leading entity with handle ${firstHandle} not found.`, NotificationType.Warning);
            template = builder.GetObjectTemplate(endHandle);
        }
        while (template) {
            yield template.CadObject;
            if (template.CadObject.handle === endHandle) {
                break;
            }
            if (template.NextEntity != null) {
                template = builder.GetObjectTemplate(template.NextEntity);
            }
            else {
                template = builder.GetObjectTemplate(template.CadObject.handle + 1);
            }
        }
    }
    getTableReference(builder, handle, name) {
        const byHandle = builder.TryGetCadObject(handle);
        if (byHandle) {
            return byHandle;
        }
        const byName = builder.TryGetTableEntry(name);
        if (byName) {
            return byName;
        }
        if ((name && name.length > 0) || (handle != null && handle !== 0)) {
            builder.Notify(`Table reference with handle: ${handle} | name: ${name} not found for ${this.CadObject.constructor.name} with handle ${this.CadObject.handle}`, NotificationType.Warning);
        }
        return null;
    }
}
