// @node-projects/acad-ts/dist/Text/TextProcessor.js
export class TextProcessor {
    static parse(text) {
        const groups = [];
        if (!text) {
            return { result: text, groups };
        }
        let sb = '';
        let index = 0;
        let openGroup = false;
        while (index < text.length) {
            const prev = index > 0 ? text[index - 1] : undefined;
            const current = text[index];
            const next = index + 1 < text.length ? text[index + 1] : undefined;
            if (current === '\\' && next !== undefined) {
                switch (next) {
                    case '}':
                    case '{':
                    case '\\':
                        sb += next;
                        index += 2;
                        break;
                    case 'A':
                        index = TextProcessor.jump(text, index);
                        break;
                    case 'c':
                    case 'C':
                        index = TextProcessor.processColor(text, index);
                        break;
                    case 'f':
                    case 'F':
                        index = TextProcessor.processFont(text, index);
                        break;
                    case 'h':
                    case 'H':
                        index = TextProcessor.processHeight(text, index);
                        break;
                    case 'p':
                        index = TextProcessor.processJustification(text, index);
                        break;
                    case 'P':
                    case 'n':
                        sb += '\n';
                        index += 2;
                        break;
                    default:
                        index++;
                        break;
                }
            }
            else if (current === '{' && prev !== '\\') {
                openGroup = true;
                index++;
            }
            else if (current === '}' && prev !== '\\') {
                openGroup = false;
                index++;
            }
            else {
                sb += current;
                index++;
            }
        }
        return { result: sb, groups };
    }
    static unescape(text) {
        if (!text) {
            return text;
        }
        let sb = '';
        let index = 0;
        let openGroup = false;
        while (index < text.length) {
            let currIndex = text.indexOf('\\', index);
            if (currIndex <= 0) {
                let s = text.substring(index);
                if (openGroup && s.includes('}')) {
                    s = s.replace(/}/g, '');
                    openGroup = false;
                }
                sb += s;
                break;
            }
            const prev = text[currIndex - 1];
            const current = text[currIndex];
            const next = text[currIndex + 1];
            if (prev === '{') {
                currIndex--;
                openGroup = true;
            }
            if (currIndex > index) {
                let s = text.substring(index, currIndex);
                if (openGroup && s.includes('}')) {
                    s = s.replace(/}/g, '');
                    openGroup = false;
                }
                sb += s;
            }
            let f;
            switch (next) {
                case 'f':
                case 'F':
                    f = TextProcessor.processFont(text, currIndex);
                    currIndex = f;
                    break;
                case 'c':
                case 'C':
                    f = TextProcessor.processColor(text, currIndex);
                    currIndex = f;
                    break;
                case 'P':
                case 'n':
                    sb += '\n';
                    currIndex += 2;
                    break;
                case 'r':
                    break;
                case '}':
                case '{':
                case '\\':
                    sb += next;
                    break;
            }
            index = currIndex;
        }
        return sb;
    }
    static processFont(text, start) {
        let end = text.indexOf(';', start);
        end += 1;
        // const data = text.substring(start, end).split('|');
        // FontData would be: { name: data[0] }
        return end;
    }
    static jump(text, start) {
        let end = text.indexOf(';', start);
        end += 1;
        return end;
    }
    static processColor(text, start) {
        let end = text.indexOf(';', start);
        end += 1;
        return end;
    }
    static processHeight(text, start) {
        let end = text.indexOf(';', start);
        end += 1;
        return end;
    }
    static processJustification(text, start) {
        let end = text.indexOf(';', start);
        end += 1;
        return end;
    }
}
