// @spectrum-web-components/card/src/Card.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, SizedMixin, SpectrumElement, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { ifDefined } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/directives.js';
import { property, query, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import { FocusVisiblePolyfillMixin } from '/web-component-designer-demo/node_modules/@spectrum-web-components/shared/src/focus-visible.js';
import { ObserveSlotPresence } from '/web-component-designer-demo/node_modules/@spectrum-web-components/shared/src/observe-slot-presence.js';
import { LikeAnchor } from '/web-component-designer-demo/node_modules/@spectrum-web-components/shared/src/like-anchor.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/asset/sp-asset.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/checkbox/sp-checkbox.js';
import '/web-component-designer-demo/node_modules/@spectrum-web-components/quick-actions/sp-quick-actions.js';
import cardStyles from './card.css.js';
import headingStyles from '/web-component-designer-demo/node_modules/@spectrum-web-components/styles/heading.js';
import detailStyles from '/web-component-designer-demo/node_modules/@spectrum-web-components/styles/detail.js';
/**
 * @element sp-card
 *
 * @fires change - Announces a change in the `selected` property of a card
 * @slot preview - This is the preview image for Gallery Cards
 * @slot cover-photo - This is the cover photo for Default and Quiet Cards
 * @slot heading - HTML content to be listed as the heading
 * @slot subheading - HTML content to be listed as the subheading
 * @slot description - A description of the card
 * @slot actions - an `sp-action-menu` element outlining actions to take on the represened object
 * @slot footer - Footer text
 */
export class Card extends LikeAnchor(SizedMixin(ObserveSlotPresence(FocusVisiblePolyfillMixin(SpectrumElement), [
    '[slot="cover-photo"]',
    '[slot="preview"]',
]), {
    validSizes: ['s', 'm'],
})) {
    constructor() {
        super(...arguments);
        this.variant = 'standard';
        this.selected = false;
        this.heading = '';
        this.horizontal = false;
        this.focused = false;
        this.toggles = false;
        this.subheading = '';
        this.handleFocusin = (event) => {
            this.focused = true;
            const target = event.composedPath()[0];
            if (target !== this) {
                this.removeEventListener('keydown', this.handleKeydown);
                return;
            }
            this.addEventListener('keydown', this.handleKeydown);
        };
    }
    static get styles() {
        return [headingStyles, detailStyles, cardStyles];
    }
    get hasCoverPhoto() {
        return this.getSlotContentPresence('[slot="cover-photo"]');
    }
    get hasPreview() {
        return this.getSlotContentPresence('[slot="preview"]');
    }
    click() {
        var _a;
        (_a = this.likeAnchor) === null || _a === void 0 ? void 0 : _a.click();
    }
    handleFocusout(event) {
        this.focused = false;
        const target = event.composedPath()[0];
        if (target === this) {
            this.removeEventListener('keydown', this.handleKeydown);
        }
    }
    handleKeydown(event) {
        const { code } = event;
        switch (code) {
            case 'Space':
                this.toggleSelected();
                if (this.toggles) {
                    break;
                }
            case 'Enter':
            case 'NumpadEnter':
                this.click();
        }
    }
    handleSelectedChange({ target: { checked }, }) {
        this.selected = checked;
        this.announceChange();
    }
    toggleSelected() {
        if (!this.toggles) {
            this.dispatchEvent(new Event('click', {
                bubbles: true,
                composed: true,
            }));
            return;
        }
        this.selected = !this.selected;
        this.announceChange();
    }
    announceChange() {
        const applyDefault = this.dispatchEvent(new Event('change', {
            cancelable: true,
        }));
        if (!applyDefault) {
            this.selected = !this.selected;
        }
    }
    stopPropagationOnHref(event) {
        if (this.href) {
            event.stopPropagation();
        }
    }
    handlePointerdown(event) {
        const path = event.composedPath();
        const hasAnchor = path.some((el) => el.localName === 'a');
        if (hasAnchor)
            return;
        const start = +new Date();
        const handleEnd = () => {
            const end = +new Date();
            if (end - start < 200) {
                this.click();
            }
            this.removeEventListener('pointerup', handleEnd);
        };
        this.addEventListener('pointerup', handleEnd);
    }
    get renderHeading() {
        return html `
            <div
                class="title spectrum-Heading spectrum-Heading--sizeXS"
                id="heading"
            >
                <slot name="heading">${this.heading}</slot>
            </div>
        `;
    }
    get renderPreviewImage() {
        return html `
            <sp-asset id="preview" variant=${ifDefined(this.asset)}>
                <slot name="preview"></slot>
            </sp-asset>
        `;
    }
    get renderCoverImage() {
        return html `
            <sp-asset id="cover-photo" variant=${ifDefined(this.asset)}>
                <slot name="cover-photo"></slot>
            </sp-asset>
        `;
    }
    get images() {
        const images = [];
        if (this.hasPreview)
            images.push(this.renderPreviewImage);
        if (this.hasCoverPhoto)
            images.push(this.renderCoverImage);
        return images;
    }
    renderImage() {
        if (this.horizontal) {
            return this.images;
        }
        if (this.variant !== 'standard') {
            return [this.renderPreviewImage];
        }
        return this.images;
    }
    get renderSubtitleAndDescription() {
        return html `
            <div class="subtitle spectrum-Detail spectrum-Detail--sizeS">
                <slot name="subheading">${this.subheading}</slot>
            </div>
            <slot name="description"></slot>
        `;
    }
    render() {
        return html `
            <div class="body">
                <div class="header">
                    ${this.renderHeading}
                    ${this.variant === 'gallery'
            ? this.renderSubtitleAndDescription
            : html ``}
                    ${this.variant !== 'quiet' || this.size !== 's'
            ? html `
                              <div
                                  class="action-button"
                                  @pointerdown=${this.stopPropagationOnHref}
                              >
                                  <slot name="actions"></slot>
                              </div>
                          `
            : html ``}
                </div>
                ${this.variant !== 'gallery'
            ? html `
                          <div class="content">
                              ${this.renderSubtitleAndDescription}
                          </div>
                      `
            : html ``}
            </div>
            ${this.href
            ? this.renderAnchor({
                id: 'like-anchor',
                labelledby: 'heading',
            })
            : html ``}
            ${this.variant === 'standard'
            ? html `
                      <slot name="footer"></slot>
                  `
            : html ``}
            ${this.renderImage()}
            ${this.toggles
            ? html `
                      <sp-quick-actions
                          class="quick-actions"
                          @pointerdown=${this.stopPropagationOnHref}
                      >
                          <sp-checkbox
                              class="checkbox"
                              @change=${this.handleSelectedChange}
                              ?checked=${this.selected}
                          ></sp-checkbox>
                      </sp-quick-actions>
                  `
            : html ``}
            ${this.variant === 'quiet' && this.size === 's'
            ? html `
                      <sp-quick-actions
                          class="spectrum-QuickActions actions"
                          @pointerdown=${this.stopPropagationOnHref}
                      >
                          <slot name="actions"></slot>
                      </sp-quick-actions>
                  `
            : html ``}
        `;
    }
    firstUpdated(changes) {
        super.firstUpdated(changes);
        this.addEventListener('pointerdown', this.handlePointerdown);
        this.addEventListener('focusin', this.handleFocusin);
        this.shadowRoot.addEventListener('focusin', this.handleFocusin);
        this.addEventListener('focusout', this.handleFocusout);
    }
}
__decorate([
    property()
], Card.prototype, "asset", void 0);
__decorate([
    property({ reflect: true })
], Card.prototype, "variant", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], Card.prototype, "selected", void 0);
__decorate([
    property()
], Card.prototype, "heading", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], Card.prototype, "horizontal", void 0);
__decorate([
    query('#like-anchor')
], Card.prototype, "likeAnchor", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], Card.prototype, "focused", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], Card.prototype, "toggles", void 0);
__decorate([
    property()
], Card.prototype, "subheading", void 0);
