// @node-projects/acad-ts/dist/Objects/BookColor.js
import { NonGraphicalObject } from './NonGraphicalObject.js';
import { Color } from '../Color.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
export class BookColor extends NonGraphicalObject {
    get objectName() {
        return DxfFileToken.ObjectDBColor;
    }
    get subclassMarker() {
        return DxfSubclassMarker.DbColor;
    }
    get name() {
        if (!this.colorName) {
            return '';
        }
        return `${this.bookName}$${this.colorName}`;
    }
    set name(value) {
        if (value.includes('$')) {
            const parts = value.split('$');
            this.bookName = parts[0];
            this.colorName = parts[parts.length - 1];
        }
        else {
            this.colorName = value;
        }
    }
    colorName = '';
    bookName = '';
    color = Color.ByBlock;
    constructor(name, bookName) {
        super(name);
        if (bookName) {
            this.bookName = bookName;
        }
    }
}
