// @node-projects/acad-ts/dist/Objects/CadDictionary.js
import { NonGraphicalObject } from './NonGraphicalObject.js';
import { CollectionChangedEventArgs } from '../CollectionChangedEventArgs.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { DictionaryCloningFlags } from './DictionaryCloningFlags.js';
export class CadDictionary extends NonGraphicalObject {
    onAdd = null;
    onRemove = null;
    clonningFlags = DictionaryCloningFlags.NotApplicable;
    get entryHandles() { return [...this._entries.values()].map(c => c.handle); }
    get entryNames() { return [...this._entries.keys()]; }
    hardOwnerFlag = false;
    get objectName() { return DxfFileToken.ObjectDictionary; }
    get objectType() { return ObjectType.DICTIONARY; }
    get subclassMarker() { return DxfSubclassMarker.Dictionary; }
    static AcadColor = 'ACAD_COLOR';
    static AcadFieldList = 'ACAD_FIELDLIST';
    static AcadGroup = 'ACAD_GROUP';
    static AcadImageDict = 'ACAD_IMAGE_DICT';
    static AcadLayout = 'ACAD_LAYOUT';
    static AcadMaterial = 'ACAD_MATERIAL';
    static AcadMLeaderStyle = 'ACAD_MLEADERSTYLE';
    static AcadMLineStyle = 'ACAD_MLINESTYLE';
    static AcadPdfDefinitions = 'ACAD_PDFDEFINITIONS';
    static AcadPlotSettings = 'ACAD_PLOTSETTINGS';
    static AcadPlotStyleName = 'ACAD_PLOTSTYLENAME';
    static AcadScaleList = 'ACAD_SCALELIST';
    static AcadSortEnts = 'ACAD_SORTENTS';
    static AcadTableStyle = 'ACAD_TABLESTYLE';
    static AcadVisualStyle = 'ACAD_VISUALSTYLE';
    static GeographicData = 'ACAD_GEOGRAPHICDATA';
    static Root = 'ROOT';
    static VariableDictionary = 'AcDbVariableDictionary';
    _entries = new Map();
    constructor(name) {
        super(name);
    }
    static createDefaultEntries(root) {
        root.tryAdd(new CadDictionary(CadDictionary.AcadColor));
        root.tryAdd(new CadDictionary(CadDictionary.AcadGroup));
        root.tryAdd(new CadDictionary(CadDictionary.AcadLayout));
        root.tryAdd(new CadDictionary(CadDictionary.AcadMaterial));
        root.tryAdd(new CadDictionary(CadDictionary.AcadSortEnts));
        root.tryAdd(new CadDictionary(CadDictionary.AcadMLeaderStyle));
        root.tryAdd(new CadDictionary(CadDictionary.AcadMLineStyle));
        root.tryAdd(new CadDictionary(CadDictionary.AcadTableStyle));
        root.tryAdd(new CadDictionary(CadDictionary.AcadPlotSettings));
        root.tryAdd(new CadDictionary(CadDictionary.VariableDictionary));
        root.tryAdd(new CadDictionary(CadDictionary.AcadScaleList));
        root.tryAdd(new CadDictionary(CadDictionary.AcadVisualStyle));
        root.tryAdd(new CadDictionary(CadDictionary.AcadFieldList));
        root.tryAdd(new CadDictionary(CadDictionary.AcadImageDict));
    }
    static createRoot() {
        const root = new CadDictionary(CadDictionary.Root);
        CadDictionary.createDefaultEntries(root);
        return root;
    }
    addByKey(key, value) {
        if (!key) {
            throw new Error('NonGraphicalObject must have a name');
        }
        this._entries.set(key.toLowerCase(), value);
        value.owner = this;
        this.onAdd?.call(this, this, new CollectionChangedEventArgs(value));
    }
    add(value) {
        this.addByKey(value.name, value);
    }
    clear() {
        for (const [key] of this._entries) {
            this.remove(key);
        }
    }
    clone() {
        const clone = super.clone();
        clone.onAdd = null;
        clone.onRemove = null;
        clone._entries = new Map();
        for (const item of this._entries.values()) {
            clone.add(item.clone());
        }
        return clone;
    }
    containsKey(key) {
        return this._entries.has(key.toLowerCase());
    }
    getEntry(name) {
        const entry = this._entries.get(name.toLowerCase());
        return entry ?? null;
    }
    remove(key) {
        const item = this._entries.get(key.toLowerCase());
        if (item) {
            this._entries.delete(key.toLowerCase());
            item.owner = null;
            this.onRemove?.call(this, this, new CollectionChangedEventArgs(item));
            return true;
        }
        return false;
    }
    tryAdd(value) {
        if (!this._entries.has(value.name.toLowerCase())) {
            this.add(value);
            return true;
        }
        return false;
    }
    get(key) {
        return this._entries.get(key.toLowerCase());
    }
    [Symbol.iterator]() {
        return this._entries.values();
    }
}
export { DictionaryCloningFlags } from './DictionaryCloningFlags.js';
