// @node-projects/acad-ts/dist/Objects/DimensionAssociation.js
import { NonGraphicalObject } from './NonGraphicalObject.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { AssociativityFlags } from './AssociativityFlags.js';
import { ObjectOsnapType } from './ObjectOsnapType.js';
import { RotatedDimensionType } from './RotatedDimensionType.js';
import { SubentType } from './SubentType.js';
import { XYZ } from '../Math/XYZ.js';
export class OsnapPointRef {
    geometryParameter = 0;
    objectOsnapType = ObjectOsnapType.None;
    osnapPoint = new XYZ(0, 0, 0);
    subentType = SubentType.Unknown;
    gsMarker = 0;
    intersectionSubType = SubentType.Unknown;
    intersectionGsMarker = 0;
    hasLastPointRef = false;
    geometry = null;
}
export class DimensionAssociation extends NonGraphicalObject {
    associativityFlags = AssociativityFlags.None;
    dimension = null;
    firstPointRef = null;
    fourthPointRef = null;
    isTransSpace = false;
    get objectName() {
        return DxfFileToken.ObjectDimensionAssociation;
    }
    get objectType() {
        return ObjectType.UNLISTED;
    }
    rotatedDimensionType = RotatedDimensionType.Unknown;
    secondPointRef = null;
    get subclassMarker() {
        return DxfSubclassMarker.DimensionAssociation;
    }
    thirdPointRef = null;
    static OsnapPointRefClassName = 'AcDbOsnapPointRef';
}
export { AssociativityFlags } from './AssociativityFlags.js';
export { RotatedDimensionType } from './RotatedDimensionType.js';
export { ObjectOsnapType } from './ObjectOsnapType.js';
