// @node-projects/acad-ts/dist/Objects/Field.js
import { NonGraphicalObject } from './NonGraphicalObject.js';
import { CadValue } from '../CadValue.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { EvaluationOptionFlags } from './EvaluationOptionFlags.js';
import { EvaluationStatusFlags } from './EvaluationStatusFlags.js';
import { FieldStateFlags } from './FieldStateFlags.js';
import { FilingOptionFlags } from './FilingOptionFlags.js';
export class Field extends NonGraphicalObject {
    cadObjects = [];
    children = [];
    evaluationErrorCode = 0;
    evaluationErrorMessage = '';
    evaluationOptionFlags = EvaluationOptionFlags.Never;
    evaluationStatusFlags = EvaluationStatusFlags.NotEvaluated;
    evaluatorId = '';
    fieldCode = '';
    fieldStateFlags = FieldStateFlags.Unknown;
    filingOptionFlags = FilingOptionFlags.None;
    formatString = '';
    get objectName() {
        return DxfFileToken.ObjectField;
    }
    get subclassMarker() {
        return DxfSubclassMarker.Field;
    }
    value = new CadValue();
    values = new Map();
}
export { EvaluationOptionFlags } from './EvaluationOptionFlags.js';
export { FilingOptionFlags } from './FilingOptionFlags.js';
export { FieldStateFlags } from './FieldStateFlags.js';
export { EvaluationStatusFlags } from './EvaluationStatusFlags.js';
