// @node-projects/acad-ts/dist/Objects/GeoData.js
import { NonGraphicalObject } from './NonGraphicalObject.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { DesignCoordinatesType } from './DesignCoordinatesType.js';
import { GeoDataVersion } from './GeoDataVersion.js';
import { ScaleEstimationType } from './ScaleEstimationType.js';
import { XYZ } from '../Math/XYZ.js';
import { XY } from '../Math/XY.js';
export class GeoMeshFace {
    index1 = 0;
    index2 = 0;
    index3 = 0;
}
export class GeoMeshPoint {
    source = new XY(0, 0);
    destination = new XY(0, 0);
    toString() {
        return `src:${this.source.x},${this.source.y} dest:${this.destination.x},${this.destination.y}`;
    }
}
export class GeoData extends NonGraphicalObject {
    get objectType() { return ObjectType.UNLISTED; }
    get objectName() { return DxfFileToken.ObjectGeoData; }
    get subclassMarker() { return DxfSubclassMarker.GeoData; }
    version = GeoDataVersion.R2013;
    coordinatesType = DesignCoordinatesType.LocalGrid;
    hostBlock = null;
    designPoint = new XYZ(0, 0, 0);
    referencePoint = new XYZ(0, 0, 0);
    northDirection = new XY(0, 1);
    horizontalUnitScale = 1;
    verticalUnitScale = 1;
    horizontalUnits = 0;
    verticalUnits = 0;
    upDirection = new XYZ(0, 0, 1);
    scaleEstimationMethod = ScaleEstimationType.None;
    enableSeaLevelCorrection = false;
    userSpecifiedScaleFactor = 0;
    seaLevelElevation = 0;
    coordinateProjectionRadius = 0;
    coordinateSystemDefinition = '';
    geoRssTag = '';
    observationFromTag = '';
    observationToTag = '';
    observationCoverageTag = '';
    points = [];
    faces = [];
}
export { GeoDataVersion } from './GeoDataVersion.js';
export { DesignCoordinatesType } from './DesignCoordinatesType.js';
export { ScaleEstimationType } from './ScaleEstimationType.js';
