// @node-projects/acad-ts/dist/Objects/Group.js
import { NonGraphicalObject } from './NonGraphicalObject.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
export class Group extends NonGraphicalObject {
    description = '';
    get entities() { return this._entities; }
    get isUnnamed() {
        return !this.name || this.name.startsWith('*');
    }
    get objectName() {
        return DxfFileToken.TableGroup;
    }
    get objectType() {
        return ObjectType.GROUP;
    }
    selectable = true;
    get subclassMarker() {
        return DxfSubclassMarker.Group;
    }
    _entities = [];
    constructor(name) {
        super(name);
    }
    add(entity) {
        if (this.document !== entity.document) {
            throw new Error('The Group and the entity must belong to the same document.');
        }
        this._entities.push(entity);
        entity.addReactor(this);
    }
    addRange(entities) {
        for (const e of entities) {
            this.add(e);
        }
    }
    clear() {
        for (const e of this._entities) {
            e.removeReactor(this);
        }
        this._entities = [];
    }
    remove(entity) {
        entity.removeReactor(this);
        const idx = this._entities.indexOf(entity);
        if (idx >= 0) {
            this._entities.splice(idx, 1);
            return true;
        }
        return false;
    }
    clone() {
        const clone = super.clone();
        clone._entities = [];
        return clone;
    }
}
