// @node-projects/acad-ts/dist/Objects/Layout.js
import { PlotSettings } from './PlotSettings.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { LayoutFlags } from './LayoutFlags.js';
import { XYZ } from '../Math/XYZ.js';
import { XY } from '../Math/XY.js';
export class Layout extends PlotSettings {
    _blockRecord = null;
    _lastViewport = null;
    get associatedBlock() {
        return this._blockRecord;
    }
    set associatedBlock(value) {
        if (value == null) {
            throw new Error('value cannot be null');
        }
        this._blockRecord = value;
        this._blockRecord.layout = this;
    }
    baseUCS = null;
    elevation = 0.0;
    insertionBasePoint = new XYZ(0, 0, 0);
    get isPaperSpace() {
        return this.name?.toLowerCase() !== Layout.ModelLayoutName.toLowerCase();
    }
    get lastActiveViewport() {
        return this._lastViewport;
    }
    set lastActiveViewport(value) {
        this._lastViewport = value;
    }
    layoutFlags = LayoutFlags.PaperSpaceLinetypeScaling;
    maxExtents = new XYZ(231.3, 175.5, 0.0);
    maxLimits = new XY(277.0, 202.5);
    minExtents = new XYZ(25.7, 19.5, 0.0);
    minLimits = new XY(-20.0, -7.5);
    get objectName() { return DxfFileToken.ObjectLayout; }
    get objectType() { return ObjectType.LAYOUT; }
    origin = new XYZ(0, 0, 0);
    get subclassMarker() { return DxfSubclassMarker.Layout; }
    tabOrder = 0;
    ucs = null;
    ucsOrthographicType = 0;
    get viewports() {
        return this.associatedBlock?.viewports ?? null;
    }
    xAxis = new XYZ(1, 0, 0);
    yAxis = new XYZ(0, 1, 0);
    static ModelLayoutName = 'Model';
    static PaperLayoutName = 'Layout1';
    constructor(name, blockName) {
        super();
        if (name != null) {
            this.name = name;
            // TODO: Create BlockRecord with blockName or name
        }
    }
    addViewport(viewport) {
        this.associatedBlock?.entities?.push(viewport);
    }
    clone() {
        const clone = super.clone();
        // TODO: clone._blockRecord = this._blockRecord?.clone();
        return clone;
    }
    toString() {
        return `${this.objectName}:${this.name}`;
    }
    updatePaperViewport() {
        if (!this.isPaperSpace || this.associatedBlock == null) {
            return;
        }
        // TODO: find or create viewport in associated block
    }
}
export { LayoutFlags } from './LayoutFlags.js';
