// @node-projects/acad-ts/dist/Objects/Material.js
import { NonGraphicalObject } from './NonGraphicalObject.js';
import { Color } from '../Color.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { AutoTransformMethodFlags } from './AutoTransformMethodFlags.js';
import { ColorMethod } from './ColorMethod.js';
import { MapSource } from './MapSource.js';
import { ProjectionMethod } from './ProjectionMethod.js';
import { TilingMethod } from './TilingMethod.js';
export class Material extends NonGraphicalObject {
    ambientColor = Color.ByLayer;
    _ambientColorFactor = 1.0;
    get ambientColorFactor() { return this._ambientColorFactor; }
    set ambientColorFactor(value) {
        if (value < 0 || value > 1)
            throw new Error('Value must be in range 0 to 1');
        this._ambientColorFactor = value;
    }
    ambientColorMethod = ColorMethod.Current;
    bumpAutoTransform = AutoTransformMethodFlags.NoAutoTransform;
    bumpMapBlendFactor = 1.0;
    bumpMapFileName = '';
    bumpMapSource = MapSource.UseImageFile;
    bumpMatrix = []; // Matrix4
    bumpProjectionMethod = ProjectionMethod.Planar;
    bumpTilingMethod = TilingMethod.Tile;
    channelFlags = 0;
    description = '';
    diffuseAutoTransform = AutoTransformMethodFlags.NoAutoTransform;
    diffuseColor = Color.ByLayer;
    _diffuseColorFactor = 1.0;
    get diffuseColorFactor() { return this._diffuseColorFactor; }
    set diffuseColorFactor(value) {
        if (value < 0 || value > 1)
            throw new Error('Value must be in range 0 to 1');
        this._diffuseColorFactor = value;
    }
    diffuseColorMethod = ColorMethod.Current;
    diffuseMapBlendFactor = 1.0;
    diffuseMapFileName = '';
    diffuseMapSource = MapSource.UseImageFile;
    diffuseMatrix = []; // Matrix4
    diffuseProjectionMethod = ProjectionMethod.Planar;
    diffuseTilingMethod = TilingMethod.Tile;
    illuminationModel = 0;
    get objectName() { return DxfFileToken.ObjectMaterial; }
    get objectType() { return ObjectType.UNLISTED; }
    opacity = 1.0;
    opacityAutoTransform = AutoTransformMethodFlags.NoAutoTransform;
    opacityMapBlendFactor = 1.0;
    opacityMapFileName = '';
    opacityMapSource = MapSource.UseImageFile;
    opacityMatrix = []; // Matrix4
    opacityProjectionMethod = ProjectionMethod.Planar;
    opacityTilingMethod = TilingMethod.Tile;
    reflectionAutoTransform = AutoTransformMethodFlags.NoAutoTransform;
    reflectionMapBlendFactor = 1.0;
    reflectionMapFileName = '';
    reflectionMapSource = MapSource.UseImageFile;
    reflectionMatrix = []; // Matrix4
    reflectionProjectionMethod = ProjectionMethod.Planar;
    reflectionTilingMethod = TilingMethod.Tile;
    reflectivity = 0.0;
    refractionAutoTransform = AutoTransformMethodFlags.NoAutoTransform;
    refractionIndex = 1.0;
    refractionMapBlendFactor = 1.0;
    refractionMapFileName = '';
    refractionMapSource = MapSource.UseImageFile;
    refractionMatrix = []; // Matrix4
    refractionProjectionMethod = ProjectionMethod.Planar;
    refractionTilingMethod = TilingMethod.Tile;
    specularAutoTransform = AutoTransformMethodFlags.NoAutoTransform;
    specularColor = Color.ByLayer;
    _specularColorFactor = 1.0;
    get specularColorFactor() { return this._specularColorFactor; }
    set specularColorFactor(value) {
        if (value < 0 || value > 1)
            throw new Error('Value must be in range 0 to 1');
        this._specularColorFactor = value;
    }
    specularColorMethod = ColorMethod.Current;
    specularGlossFactor = 0.5;
    specularMapBlendFactor = 1.0;
    specularMapFileName = '';
    specularMapSource = MapSource.UseImageFile;
    specularMatrix = []; // Matrix4
    specularProjectionMethod = ProjectionMethod.Planar;
    specularTilingMethod = TilingMethod.Tile;
    get subclassMarker() { return DxfSubclassMarker.Material; }
    translucence = 0.0;
    static ByBlockName = 'ByBlock';
    static ByLayerName = 'ByLayer';
    static GlobalName = 'Global';
    constructor(name) {
        super(name);
    }
}
export { ColorMethod } from './ColorMethod.js';
export { ProjectionMethod } from './ProjectionMethod.js';
export { TilingMethod } from './TilingMethod.js';
export { AutoTransformMethodFlags } from './AutoTransformMethodFlags.js';
export { MapSource } from './MapSource.js';
