// @node-projects/acad-ts/dist/Objects/MLineStyle.js
import { NonGraphicalObject } from './NonGraphicalObject.js';
import { CadObject } from '../CadObject.js';
import { Color } from '../Color.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { MLineStyleFlags } from './MLineStyleFlags.js';
export class MLineStyleElement {
    color = Color.ByLayer;
    _lineType = null; // LineType
    get lineType() { return this._lineType; }
    set lineType(value) {
        this._lineType = CadObject.updateCollectionStatic(value, this.owner?.document?.lineTypes);
    }
    offset = 0;
    owner = null;
    clone() {
        const clone = new MLineStyleElement();
        clone.color = this.color;
        clone.offset = this.offset;
        clone.owner = null;
        clone._lineType = this._lineType; // TODO: clone lineType
        return clone;
    }
    assignDocument(doc) {
        this._lineType = CadObject.updateCollectionStatic(this._lineType, doc.lineTypes);
    }
    unassignDocument() {
        // TODO: clone lineType on unassign
    }
}
export class MLineStyle extends NonGraphicalObject {
    static get default_() {
        const def = new MLineStyle(MLineStyle.DefaultName);
        def.startAngle = Math.PI / 2;
        def.endAngle = Math.PI / 2;
        const e1 = new MLineStyleElement();
        e1.offset = 0.5;
        def.addElement(e1);
        const e2 = new MLineStyleElement();
        e2.offset = -0.5;
        def.addElement(e2);
        return def;
    }
    description = '';
    get elements() {
        return this._elements;
    }
    endAngle = Math.PI / 2;
    fillColor = Color.ByLayer;
    flags = MLineStyleFlags.None;
    get objectName() { return DxfFileToken.ObjectMLineStyle; }
    get objectType() { return ObjectType.MLINESTYLE; }
    startAngle = Math.PI / 2;
    get subclassMarker() { return DxfSubclassMarker.MLineStyle; }
    static DefaultName = 'Standard';
    _elements = [];
    constructor(name) {
        super(name);
    }
    addElement(element) {
        if (element.owner != null) {
            throw new Error(`Element already assigned to a MLineStyle: ${element.owner.name}`);
        }
        element.lineType = CadObject.updateCollection(element.lineType, this.document?.lineTypes);
        element.owner = this;
        this._elements.push(element);
    }
    clone() {
        const clone = super.clone();
        clone._elements = [];
        for (const element of this._elements) {
            clone.addElement(element.clone());
        }
        return clone;
    }
}
export { MLineStyleFlags } from './MLineStyleFlags.js';
