// @node-projects/acad-ts/dist/Objects/MultiLeaderObjectContextData.js
import { AnnotScaleObjectContextData } from './AnnotScaleObjectContextData.js';
import { Color } from '../Color.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { XYZ } from '../Math/XYZ.js';
import { Matrix4 } from '../Math/Matrix4.js';
export class StartEndPointPair {
    startPoint;
    endPoint;
    constructor(startPoint = new XYZ(0, 0, 0), endPoint = new XYZ(0, 0, 0)) {
        this.startPoint = startPoint;
        this.endPoint = endPoint;
    }
    clone() {
        return new StartEndPointPair(new XYZ(this.startPoint.x, this.startPoint.y, this.startPoint.z), new XYZ(this.endPoint.x, this.endPoint.y, this.endPoint.z));
    }
}
export class LeaderLine {
    arrowhead = null;
    arrowheadSize = 0;
    breakInfoCount = 0;
    index = 0;
    lineColor = Color.ByLayer;
    _lineType = null;
    get lineType() { return this._lineType; }
    set lineType(value) { this._lineType = value; }
    lineWeight = 0;
    overrideFlags = 0;
    pathType = 0;
    points = [];
    segmentIndex = 0;
    startEndPoints = [];
    document = null;
    assignDocument(doc) {
        this.document = doc;
        // TODO: update line type from doc.lineTypes
    }
    clone() {
        const clone = new LeaderLine();
        clone.arrowheadSize = this.arrowheadSize;
        clone.breakInfoCount = this.breakInfoCount;
        clone.index = this.index;
        clone.lineColor = this.lineColor;
        clone.lineWeight = this.lineWeight;
        clone.overrideFlags = this.overrideFlags;
        clone.pathType = this.pathType;
        clone.segmentIndex = this.segmentIndex;
        clone.points = this.points.map(p => new XYZ(p.x, p.y, p.z));
        clone.startEndPoints = this.startEndPoints.map(s => s.clone());
        return clone;
    }
    unassignDocument() {
        this.document = null;
    }
}
export class LeaderRoot {
    breakStartEndPointsPairs = [];
    connectionPoint = new XYZ(0, 0, 0);
    contentValid = false;
    direction = new XYZ(0, 0, 0);
    landingDistance = 0;
    leaderIndex = 0;
    lines = [];
    textAttachmentDirection = 0;
    unknown = false;
    clone() {
        const clone = new LeaderRoot();
        clone.connectionPoint = new XYZ(this.connectionPoint.x, this.connectionPoint.y, this.connectionPoint.z);
        clone.contentValid = this.contentValid;
        clone.direction = new XYZ(this.direction.x, this.direction.y, this.direction.z);
        clone.landingDistance = this.landingDistance;
        clone.leaderIndex = this.leaderIndex;
        clone.textAttachmentDirection = this.textAttachmentDirection;
        clone.unknown = this.unknown;
        clone.breakStartEndPointsPairs = this.breakStartEndPointsPairs.map(s => s.clone());
        clone.lines = this.lines.map(l => l.clone());
        return clone;
    }
}
export class MultiLeaderObjectContextData extends AnnotScaleObjectContextData {
    arrowheadSize = 0;
    backgroundFillColor = Color.ByLayer;
    backgroundFillEnabled = false;
    backgroundMaskFillOn = false;
    backgroundScaleFactor = 0;
    backgroundTransparency = 0;
    baseDirection = new XYZ(0, 0, 0);
    basePoint = new XYZ(0, 0, 0);
    baseVertical = new XYZ(0, 0, 0);
    _blockContent = null;
    get blockContent() { return this._blockContent; }
    set blockContent(value) {
        this._blockContent = value;
    }
    blockContentColor = Color.ByLayer;
    blockContentConnection = 0;
    blockContentLocation = new XYZ(0, 0, 0);
    blockContentNormal = new XYZ(0, 0, 0);
    blockContentRotation = 0;
    blockContentScale = new XYZ(1, 1, 1);
    boundaryHeight = 0;
    boundaryWidth = 0;
    columnFlowReversed = false;
    columnGutter = 0;
    columnSizes = [];
    columnType = 0;
    columnWidth = 0;
    contentBasePoint = new XYZ(0, 0, 0);
    direction = new XYZ(0, 0, 0);
    flowDirection = 0;
    hasContentsBlock = false;
    hasTextContents = false;
    landingGap = 0;
    leaderRoots = [];
    lineSpacing = 0;
    lineSpacingFactor = 0;
    normalReversed = false;
    get objectName() { return DxfFileToken.ObjectMLeaderContextData; }
    get objectType() { return ObjectType.UNLISTED; }
    scaleFactor = 0;
    get subclassMarker() { return DxfSubclassMarker.MultiLeaderObjectContextData; }
    textAlignment = 0;
    textAttachmentPoint = 0;
    textBottomAttachment = 0;
    textColor = Color.ByLayer;
    textHeight = 0;
    textHeightAutomatic = false;
    textLabel = '';
    textLeftAttachment = 0;
    textLocation = new XYZ(0, 0, 0);
    textNormal = new XYZ(0, 0, 0);
    textRightAttachment = 0;
    textRotation = 0;
    _textStyle = null;
    get textStyle() { return this._textStyle; }
    set textStyle(value) {
        if (value == null)
            throw new Error('value cannot be null');
        this._textStyle = value;
    }
    textTopAttachment = 0;
    transformationMatrix = Matrix4.identity();
    wordBreak = false;
    clone() {
        const clone = super.clone();
        clone.leaderRoots = this.leaderRoots.map(r => r.clone());
        // TODO: clone textStyle, blockContent
        return clone;
    }
}
