// @node-projects/acad-ts/dist/Objects/ObjectOsnapType.js
export var ObjectOsnapType;
(function (ObjectOsnapType) {
    ObjectOsnapType[ObjectOsnapType["None"] = 0] = "None";
    ObjectOsnapType[ObjectOsnapType["Endpoint"] = 1] = "Endpoint";
    ObjectOsnapType[ObjectOsnapType["Midpoint"] = 2] = "Midpoint";
    ObjectOsnapType[ObjectOsnapType["Center"] = 3] = "Center";
    ObjectOsnapType[ObjectOsnapType["Node"] = 4] = "Node";
    ObjectOsnapType[ObjectOsnapType["Quadrant"] = 5] = "Quadrant";
    ObjectOsnapType[ObjectOsnapType["Intersection"] = 6] = "Intersection";
    ObjectOsnapType[ObjectOsnapType["Insertion"] = 7] = "Insertion";
    ObjectOsnapType[ObjectOsnapType["Perpendicular"] = 8] = "Perpendicular";
    ObjectOsnapType[ObjectOsnapType["Tangent"] = 9] = "Tangent";
    ObjectOsnapType[ObjectOsnapType["Nearest"] = 10] = "Nearest";
    ObjectOsnapType[ObjectOsnapType["ApparentIntersection"] = 11] = "ApparentIntersection";
    ObjectOsnapType[ObjectOsnapType["Parallel"] = 12] = "Parallel";
    ObjectOsnapType[ObjectOsnapType["StartPoint"] = 13] = "StartPoint";
})(ObjectOsnapType || (ObjectOsnapType = {}));
