// @node-projects/acad-ts/dist/Objects/PlotFlags.js
export var PlotFlags;
(function (PlotFlags) {
    PlotFlags[PlotFlags["None"] = 0] = "None";
    PlotFlags[PlotFlags["PlotViewportBorders"] = 1] = "PlotViewportBorders";
    PlotFlags[PlotFlags["ShowPlotStyles"] = 2] = "ShowPlotStyles";
    PlotFlags[PlotFlags["PlotCentered"] = 4] = "PlotCentered";
    PlotFlags[PlotFlags["PlotHidden"] = 8] = "PlotHidden";
    PlotFlags[PlotFlags["UseStandardScale"] = 16] = "UseStandardScale";
    PlotFlags[PlotFlags["PlotPlotStyles"] = 32] = "PlotPlotStyles";
    PlotFlags[PlotFlags["ScaleLineweights"] = 64] = "ScaleLineweights";
    PlotFlags[PlotFlags["PrintLineweights"] = 128] = "PrintLineweights";
    PlotFlags[PlotFlags["DrawViewportsFirst"] = 512] = "DrawViewportsFirst";
    PlotFlags[PlotFlags["ModelType"] = 1024] = "ModelType";
    PlotFlags[PlotFlags["UpdatePaper"] = 2048] = "UpdatePaper";
    PlotFlags[PlotFlags["ZoomToPaperOnUpdate"] = 4096] = "ZoomToPaperOnUpdate";
    PlotFlags[PlotFlags["Initializing"] = 8192] = "Initializing";
    PlotFlags[PlotFlags["PrevPlotInit"] = 16384] = "PrevPlotInit";
})(PlotFlags || (PlotFlags = {}));
