// @node-projects/acad-ts/dist/Objects/PlotSettings.js
import { NonGraphicalObject } from './NonGraphicalObject.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { PlotFlags } from './PlotFlags.js';
import { PlotPaperUnits } from './PlotPaperUnits.js';
import { PlotRotation } from './PlotRotation.js';
import { ShadePlotMode } from './ShadePlotMode.js';
import { ShadePlotResolutionMode } from './ShadePlotResolutionMode.js';
import { PaperMargin } from './PaperMargin.js';
import { XY } from '../Math/XY.js';
export var PlotType;
(function (PlotType) {
    PlotType[PlotType["DrawingExtents"] = 0] = "DrawingExtents";
    PlotType[PlotType["Limits"] = 1] = "Limits";
    PlotType[PlotType["View"] = 2] = "View";
    PlotType[PlotType["Window"] = 3] = "Window";
    PlotType[PlotType["Layout"] = 4] = "Layout";
    PlotType[PlotType["Display"] = 5] = "Display";
})(PlotType || (PlotType = {}));
export var ScaledType;
(function (ScaledType) {
    ScaledType[ScaledType["ScaleToFit"] = 0] = "ScaleToFit";
    ScaledType[ScaledType["UserDefined"] = 1] = "UserDefined";
})(ScaledType || (ScaledType = {}));
export class PlotSettings extends NonGraphicalObject {
    _denominatorScale = 1.0;
    _numeratorScale = 1.0;
    _shadePlotDPI = 300;
    get denominatorScale() {
        return this._denominatorScale;
    }
    set denominatorScale(value) {
        if (value <= 0.0) {
            throw new Error('Value must be greater than zero');
        }
        this._denominatorScale = value;
    }
    flags = PlotFlags.DrawViewportsFirst | PlotFlags.PrintLineweights | PlotFlags.PlotPlotStyles | PlotFlags.UseStandardScale;
    get numeratorScale() {
        return this._numeratorScale;
    }
    set numeratorScale(value) {
        if (value <= 0.0) {
            throw new Error('Value must be greater than zero');
        }
        this._numeratorScale = value;
    }
    get objectName() { return DxfFileToken.ObjectPlotSettings; }
    get objectType() { return ObjectType.UNLISTED; }
    pageName = 'none_device';
    paperHeight = 0;
    paperImageOrigin = new XY(0, 0);
    paperImageOriginX = 0;
    paperImageOriginY = 0;
    paperRotation = PlotRotation.NoRotation;
    paperSize = 'ISO_A4_(210.00_x_297.00_MM)';
    paperUnits = PlotPaperUnits.Millimeters;
    paperWidth = 0;
    plotOriginX = 0;
    plotOriginY = 0;
    plotType = PlotType.DrawingExtents;
    plotViewName = '';
    get printScale() {
        return this.numeratorScale / this.denominatorScale;
    }
    scaledFit = ScaledType.ScaleToFit;
    get shadePlotDPI() {
        return this._shadePlotDPI;
    }
    set shadePlotDPI(value) {
        if (value < 100 || value > 32767) {
            throw new Error('The valid shade plot DPI values range from 100 to 32767.');
        }
        this._shadePlotDPI = value;
    }
    shadePlotIDHandle = 0;
    shadePlotMode = ShadePlotMode.AsDisplayed;
    shadePlotResolutionMode = ShadePlotResolutionMode.Draft;
    standardScale = 1.0;
    styleSheet = '';
    get subclassMarker() { return DxfSubclassMarker.PlotSettings; }
    systemPrinterName = '';
    unprintableMargin = new PaperMargin();
    windowLowerLeftX = 0;
    windowLowerLeftY = 0;
    windowUpperLeftX = 0;
    windowUpperLeftY = 0;
    constructor(name) {
        super(name);
    }
}
export { PlotFlags } from './PlotFlags.js';
export { PlotPaperUnits } from './PlotPaperUnits.js';
export { PlotRotation } from './PlotRotation.js';
export { ShadePlotMode } from './ShadePlotMode.js';
export { ShadePlotResolutionMode } from './ShadePlotResolutionMode.js';
export { PaperMargin } from './PaperMargin.js';
