// @node-projects/acad-ts/dist/Objects/SortEntitiesTable.js
import { NonGraphicalObject } from './NonGraphicalObject.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
export class Sorter {
    sortHandle = 0;
    entity = null;
    constructor(entity, handle) {
        this.entity = entity;
        this.sortHandle = handle;
    }
    toString() {
        return `${this.sortHandle} | ${this.entity?.toString()}`;
    }
    compareTo(other) {
        if (this.sortHandle < other.sortHandle)
            return -1;
        if (this.sortHandle > other.sortHandle)
            return 1;
        return 0;
    }
}
export class SortEntitiesTable extends NonGraphicalObject {
    blockOwner = null;
    get objectName() { return DxfFileToken.ObjectSortEntsTable; }
    get objectType() { return ObjectType.UNLISTED; }
    get subclassMarker() { return DxfSubclassMarker.SortentsTable; }
    static DictionaryEntryName = 'ACAD_SORTENTS';
    _sorters = [];
    constructor(owner) {
        super(SortEntitiesTable.DictionaryEntryName);
        if (owner) {
            this.blockOwner = owner;
        }
    }
    add(entity, sorterHandle) {
        this._sorters.push(new Sorter(entity, sorterHandle));
    }
    clear() {
        this._sorters.length = 0;
    }
    clone() {
        const clone = super.clone();
        clone._sorters = [];
        return clone;
    }
    getSorterHandle(entity) {
        const sorter = this._sorters.find(s => s.entity === entity);
        if (sorter) {
            return sorter.sortHandle;
        }
        return entity.handle;
    }
    remove(entity) {
        const idx = this._sorters.findIndex(s => s.entity === entity);
        if (idx < 0)
            return false;
        this._sorters.splice(idx, 1);
        return true;
    }
    [Symbol.iterator]() {
        const sorted = [...this._sorters].sort((a, b) => a.compareTo(b));
        return sorted[Symbol.iterator]();
    }
}
