// @node-projects/acad-ts/dist/Objects/XRecrod.js
import { NonGraphicalObject } from './NonGraphicalObject.js';
import { CadObject } from '../CadObject.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { DictionaryCloningFlags } from './DictionaryCloningFlags.js';
import { GroupCodeValue } from '../GroupCodeValue.js';
export class XRecordEntry {
    code;
    value;
    owner;
    get groupCode() {
        return GroupCodeValue.transformValue(this.code);
    }
    get hasLinkedObject() {
        const gc = this.groupCode;
        // Handle, ObjectId, ExtendedDataHandle
        return gc === 5 || gc === 105;
    }
    constructor(code, value, owner) {
        this.code = code;
        this.value = value;
        this.owner = owner;
    }
    getReference() {
        if (!this.hasLinkedObject)
            return null;
        if (this.value instanceof CadObject) {
            if (this.value.document !== this.owner.document) {
                return null;
            }
            return this.value;
        }
        return null;
    }
    toString() {
        return `${this.code}:${this.value}`;
    }
}
export class XRecord extends NonGraphicalObject {
    cloningFlags = DictionaryCloningFlags.NotApplicable;
    get entries() { return this._entries; }
    get objectName() {
        return DxfFileToken.ObjectXRecord;
    }
    get objectType() {
        return ObjectType.XRECORD;
    }
    get subclassMarker() {
        return DxfSubclassMarker.XRecord;
    }
    _entries = [];
    constructor(name) {
        super(name);
    }
    createEntry(code, value) {
        this._entries.push(new XRecordEntry(code, value, this));
    }
    CreateEntry(code, value) {
        this.createEntry(code, value);
    }
}
