// @spectrum-web-components/number-field/test/helpers.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { html } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { elementUpdated, fixture, nextFrame } from '@open-wc/testing';
import { sendMouse } from '../../../test/plugins/browser.js';
export async function getElFrom(test) {
    const wrapped = await fixture(html `
        <div style="--spectrum-alias-ui-icon-chevron-size-75: 20px;">
            ${test}
        </div>
    `);
    const el = wrapped.querySelector('sp-number-field');
    await elementUpdated(el);
    return el;
}
export async function clickBySelector(el, selector, options = {}) {
    const target = el.shadowRoot.querySelector(selector);
    const targetRect = target.getBoundingClientRect();
    await sendMouse({
        steps: [
            {
                type: 'move',
                position: [
                    targetRect.x + targetRect.width / 2,
                    targetRect.y + targetRect.height / 2,
                ],
                options,
            },
            {
                type: 'down',
                options,
            },
        ],
    });
    await nextFrame();
    await sendMouse({
        steps: [
            {
                type: 'up',
                options,
            },
        ],
    });
    await elementUpdated(el);
}
export function createLanguageContext(lang) {
    const langResolvers = [];
    const createLangResolver = (event) => {
        langResolvers.push(event.detail.callback);
        resolveLanguage();
    };
    const resolveLanguage = () => {
        langResolvers.forEach((resolver) => resolver(lang));
    };
    return createLangResolver;
}
