// @node-projects/acad-ts/dist/Tables/BlockRecord.js
import { Block } from '../Blocks/Block.js';
import { BlockEnd } from '../Blocks/BlockEnd.js';
import { BlockTypeFlags } from '../Blocks/BlockTypeFlags.js';
import { CadObjectCollection } from '../CadObjectCollection.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { UnitsType } from '../Types/Units/UnitsType.js';
import { TableEntry } from './TableEntry.js';
// TODO: Entity, Layout, EvaluationGraph, SortEntitiesTable, Viewport, AttributeDefinition
// These types are not yet converted; using 'any' placeholders.
export class BlockRecord extends TableEntry {
    static get ModelSpace() {
        const record = new BlockRecord(BlockRecord.ModelSpaceName);
        // TODO: Create Layout and assign when Layout type is available
        return record;
    }
    static get PaperSpace() {
        const record = new BlockRecord(BlockRecord.PaperSpaceName);
        // TODO: Create Layout and assign when Layout type is available
        return record;
    }
    get attributeDefinitions() {
        return Array.from(this.entities).filter((e) => e.constructor?.name === 'AttributeDefinition');
    }
    get blockEnd() {
        return this._blockEnd;
    }
    set blockEnd(value) {
        this._blockEnd = value;
        this._blockEnd.owner = this;
    }
    get blockEntity() {
        return this._blockEntity;
    }
    set blockEntity(value) {
        this._blockEntity = value;
        this._blockEntity.owner = this;
    }
    canScale = true;
    entities;
    get evaluationGraph() {
        // TODO: Implement when XDictionary/EvaluationGraph types are available
        return null;
    }
    get blockFlags() {
        return this.blockEntity.flags;
    }
    set blockFlags(value) {
        this.blockEntity.flags = value;
    }
    get hasAttributes() {
        return this.attributeDefinitions.length > 0;
    }
    get isAnonymous() {
        return (this.blockFlags & BlockTypeFlags.Anonymous) !== 0;
    }
    set isAnonymous(value) {
        if (value) {
            this.blockFlags |= BlockTypeFlags.Anonymous;
        }
        else {
            this.blockFlags &= ~BlockTypeFlags.Anonymous;
        }
    }
    get isDynamic() {
        return this.evaluationGraph != null;
    }
    isExplodable = false;
    get isUnloaded() {
        return this.blockEntity.isUnloaded;
    }
    set isUnloaded(value) {
        this.blockEntity.isUnloaded = value;
    }
    get layout() {
        return this._layout;
    }
    set layout(value /* Layout */) {
        this._layout = value;
    }
    get objectName() {
        return DxfFileToken.TableBlockRecord;
    }
    get objectType() {
        return ObjectType.BLOCK_HEADER;
    }
    preview = null;
    get sortEntitiesTable() {
        // TODO: Implement when XDictionary/SortEntitiesTable types are available
        return null;
    }
    get source() {
        // TODO: Implement when ExtendedData resolution is available
        return null;
    }
    get subclassMarker() {
        return DxfSubclassMarker.BlockRecord;
    }
    units = UnitsType.Unitless;
    get viewports() {
        return Array.from(this.entities).filter((e) => e.constructor?.name === 'Viewport');
    }
    static AnonymousPrefix = '*A';
    static ModelSpaceName = '*Model_Space';
    static PaperSpaceName = '*Paper_Space';
    _blockEnd;
    _blockEntity;
    _layout = null;
    constructor(name) {
        super(name);
        this._blockEntity = new Block(this);
        this._blockEnd = new BlockEnd(this);
        this.entities = new CadObjectCollection(this);
    }
    // TODO: xref constructor (name, xrefFile, isOverlay)
    clone() {
        const clone = super.clone();
        clone._layout = null;
        // TODO: SortEntitiesTable cloning when available
        clone.entities = new CadObjectCollection(clone);
        for (const item of this.entities) {
            const e = item.clone();
            clone.entities.add(e);
        }
        clone.blockEntity = this.blockEntity.clone();
        clone.blockEntity.owner = clone;
        clone.blockEnd = this.blockEnd.clone();
        clone.blockEnd.owner = clone;
        return clone;
    }
}
export { BlockTypeFlags } from '../Blocks/BlockTypeFlags.js';
