// @node-projects/acad-ts/dist/Tables/DimensionStyle.js
import { Color } from '../Color.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { LineWeightType } from '../Types/LineWeightType.js';
import { AngularUnitFormat } from '../Types/Units/AngularUnitFormat.js';
import { LinearUnitFormat } from '../Types/Units/LinearUnitFormat.js';
import { UnitStyleFormat } from '../Types/Units/UnitStyleFormat.js';
import { ArcLengthSymbolPosition } from './ArcLengthSymbolPosition.js';
import { DimensionTextBackgroundFillMode } from './DimensionTextBackgroundFillMode.js';
import { DimensionTextHorizontalAlignment } from './DimensionTextHorizontalAlignment.js';
import { DimensionTextVerticalAlignment } from './DimensionTextVerticalAlignment.js';
import { FractionFormat } from './FractionFormat.js';
import { TableEntry } from './TableEntry.js';
import { TextArrowFitType } from './TextArrowFitType.js';
import { TextDirection } from './TextDirection.js';
import { TextMovement } from './TextMovement.js';
import { TextStyle } from './TextStyle.js';
import { ToleranceAlignment } from './ToleranceAlignment.js';
import { ZeroHandling, AngularZeroHandling } from './ZeroHandling.js';
export class DimensionStyle extends TableEntry {
    static get Default() {
        return new DimensionStyle(DimensionStyle.DefaultName);
    }
    alternateDimensioningSuffix = '[]';
    alternateUnitDecimalPlaces = 3;
    alternateUnitDimensioning = false;
    alternateUnitFormat = LinearUnitFormat.Decimal;
    alternateUnitRounding = 0.0;
    alternateUnitScaleFactor = 25.4;
    alternateUnitToleranceDecimalPlaces = 3;
    alternateUnitToleranceZeroHandling = ZeroHandling.SuppressZeroFeetAndInches;
    alternateUnitZeroHandling = ZeroHandling.SuppressZeroFeetAndInches;
    angularDecimalPlaces = 0;
    angularUnit = AngularUnitFormat.DecimalDegrees;
    angularZeroHandling = AngularZeroHandling.DisplayAll;
    arcLengthSymbolPosition = ArcLengthSymbolPosition.BeforeDimensionText;
    get arrowBlock() {
        return this._dimArrowBlock;
    }
    set arrowBlock(value) {
        this._dimArrowBlock = value;
    }
    get arrowSize() {
        return this._arrowSize;
    }
    set arrowSize(value) {
        if (value < 0) {
            throw new Error(`The arrowSize must be equals or greater than zero.`);
        }
        this._arrowSize = value;
    }
    centerMarkSize = 0.0900;
    cursorUpdate = false;
    decimalPlaces = 2;
    decimalSeparator = '.';
    get dimArrow1() {
        return this._dimArrow1;
    }
    set dimArrow1(value) {
        this._dimArrow1 = value;
    }
    get dimArrow2() {
        return this._dimArrow2;
    }
    set dimArrow2(value) {
        this._dimArrow2 = value;
    }
    dimensionFit = 0;
    dimensionLineColor = Color.ByBlock;
    dimensionLineExtension = 0.0;
    dimensionLineGap = 0.6250;
    dimensionLineIncrement = 3.75;
    dimensionLineWeight = LineWeightType.ByBlock;
    dimensionTextArrowFit = TextArrowFitType.BestFit;
    dimensionUnit = 2;
    extensionLineColor = Color.ByBlock;
    extensionLineExtension = 1.2500;
    extensionLineOffset = 0.6250;
    extensionLineWeight = LineWeightType.ByBlock;
    fixedExtensionLineLength = 1.0;
    fractionFormat = FractionFormat.Horizontal;
    generateTolerances = false;
    isExtensionLineLengthFixed = false;
    get joggedRadiusDimensionTransverseSegmentAngle() {
        return this._joggedRadiusDimensionTransverseSegmentAngle;
    }
    set joggedRadiusDimensionTransverseSegmentAngle(value) {
        const rounded = Math.round(value * 1e6) / 1e6;
        const deg5Rad = (5 * Math.PI) / 180;
        const halfPI = Math.PI / 2;
        if (rounded <= deg5Rad || rounded >= halfPI) {
            throw new Error(`The joggedRadiusDimensionTransverseSegmentAngle must be in range of 5 to 90 degrees.`);
        }
        this._joggedRadiusDimensionTransverseSegmentAngle = value;
    }
    get leaderArrow() {
        return this._leaderArrow;
    }
    set leaderArrow(value) {
        this._leaderArrow = value;
    }
    limitsGeneration = false;
    linearScaleFactor = 1.0;
    linearUnitFormat = LinearUnitFormat.Decimal;
    get lineType() {
        return this._lineType;
    }
    set lineType(value) {
        this._lineType = value;
    }
    get lineTypeExt1() {
        return this._lineTypeExt1;
    }
    set lineTypeExt1(value) {
        this._lineTypeExt1 = value;
    }
    get lineTypeExt2() {
        return this._lineTypeExt2;
    }
    set lineTypeExt2(value) {
        this._lineTypeExt2 = value;
    }
    minusTolerance = 0.0;
    get objectName() {
        return DxfFileToken.TableDimstyle;
    }
    get objectType() {
        return ObjectType.DIMSTYLE;
    }
    plusTolerance = 0.0;
    postFix = '<>';
    get prefix() {
        const { prefix } = this.getDimStylePrefixAndSuffix(this.postFix, '<', '>');
        return prefix;
    }
    set prefix(value) {
        const { suffix } = this.getDimStylePrefixAndSuffix(this.postFix, '<', '>');
        this.postFix = `${value}${this.postFix}${suffix}`;
    }
    rounding = 0.0;
    get scaleFactor() {
        return this._scaleFactor;
    }
    set scaleFactor(value) {
        if (value < 0) {
            throw new Error(`The scaleFactor must be equals or greater than zero.`);
        }
        this._scaleFactor = value;
    }
    separateArrowBlocks = true;
    get style() {
        return this._style;
    }
    set style(value) {
        if (value == null) {
            throw new Error('Style cannot be null.');
        }
        this._style = value;
    }
    get subclassMarker() {
        return DxfSubclassMarker.DimensionStyle;
    }
    get suffix() {
        const { suffix } = this.getDimStylePrefixAndSuffix(this.postFix, '<', '>');
        return suffix;
    }
    set suffix(value) {
        const { prefix } = this.getDimStylePrefixAndSuffix(this.postFix, '<', '>');
        this.postFix = `${prefix}${this.postFix}${value}`;
    }
    suppressFirstDimensionLine = false;
    suppressFirstExtensionLine = false;
    suppressOutsideExtensions = false;
    suppressSecondDimensionLine = false;
    suppressSecondExtensionLine = false;
    textBackgroundColor = Color.ByBlock;
    textBackgroundFillMode = DimensionTextBackgroundFillMode.NoBackground;
    textColor = Color.ByBlock;
    textDirection = TextDirection.LeftToRight;
    get textHeight() {
        return this._textHeight;
    }
    set textHeight(value) {
        if (value <= 0) {
            throw new Error(`The textHeight must be greater than zero.`);
        }
        this._textHeight = value;
    }
    textHorizontalAlignment = DimensionTextHorizontalAlignment.Centered;
    textInsideExtensions = false;
    textInsideHorizontal = false;
    textMovement = TextMovement.MoveLineWithText;
    textOutsideExtensions = false;
    textOutsideHorizontal = false;
    textVerticalAlignment = DimensionTextVerticalAlignment.Above;
    textVerticalPosition = 0.0;
    tickSize = 0.0;
    toleranceAlignment = ToleranceAlignment.Bottom;
    toleranceDecimalPlaces = 2;
    toleranceScaleFactor = 1.0;
    toleranceZeroHandling = ZeroHandling.SuppressDecimalTrailingZeroes;
    zeroHandling = ZeroHandling.SuppressDecimalTrailingZeroes;
    static DefaultName = 'Standard';
    static StyleOverrideEntryName = 'DSTYLE';
    _arrowSize = 0.18;
    _dimArrow1 = null;
    _dimArrow2 = null;
    _dimArrowBlock = null;
    _joggedRadiusDimensionTransverseSegmentAngle = Math.PI / 4.0;
    _leaderArrow = null;
    _lineType = null;
    _lineTypeExt1 = null;
    _lineTypeExt2 = null;
    _scaleFactor = 1.0;
    _style = TextStyle.Default;
    _textHeight = 0.18;
    constructor(name) {
        super(name);
    }
    applyRounding(value, isAlternate = false) {
        const rounding = isAlternate ? this.alternateUnitRounding : this.rounding;
        if (rounding !== 0.0) {
            value = rounding * Math.round(value / rounding);
        }
        return value;
    }
    clone() {
        const clone = super.clone();
        clone.style = this.style?.clone();
        clone.leaderArrow = this.leaderArrow?.clone() ?? null;
        clone.arrowBlock = this.arrowBlock?.clone() ?? null;
        clone.dimArrow1 = this.dimArrow1?.clone() ?? null;
        clone.dimArrow2 = this.dimArrow2?.clone() ?? null;
        clone.lineType = this.lineType?.clone() ?? null;
        clone.lineTypeExt1 = this.lineTypeExt1?.clone() ?? null;
        clone.lineTypeExt2 = this.lineTypeExt2?.clone() ?? null;
        return clone;
    }
    getAlternateUnitStyleFormat() {
        const format = new UnitStyleFormat();
        format.linearDecimalPlaces = this.alternateUnitDecimalPlaces;
        format.angularDecimalPlaces = this.alternateUnitDecimalPlaces;
        format.decimalSeparator = this.decimalSeparator;
        format.fractionHeightScale = this.toleranceScaleFactor;
        format.fractionType = this.fractionFormat;
        format.linearZeroHandling = this.alternateUnitZeroHandling;
        format.angularZeroHandling = this.angularZeroHandling;
        return format;
    }
    getUnitStyleFormat() {
        const format = new UnitStyleFormat();
        format.linearDecimalPlaces = this.decimalPlaces;
        format.angularDecimalPlaces = this.angularDecimalPlaces === -1 ? this.decimalPlaces : this.angularDecimalPlaces;
        format.decimalSeparator = this.decimalSeparator;
        format.fractionHeightScale = this.toleranceScaleFactor;
        format.fractionType = this.fractionFormat;
        format.linearZeroHandling = this.zeroHandling;
        format.angularZeroHandling = this.angularZeroHandling;
        return format;
    }
    getDimStylePrefixAndSuffix(text, start, end) {
        let index = -1;
        for (let i = 0; i < text.length; i++) {
            if (text[i] === start) {
                if (i + 1 < text.length && text[i + 1] === end) {
                    index = i;
                    break;
                }
            }
        }
        if (index < 0) {
            return { prefix: '', suffix: text };
        }
        else {
            return {
                prefix: text.substring(0, index),
                suffix: text.substring(index + 2),
            };
        }
    }
    altMzf = 0;
    altMzs = "";
    mzf = 0;
    mzs = "";
}
export { ZeroHandling } from './ZeroHandling.js';
export { ToleranceAlignment } from './ToleranceAlignment.js';
export { DimensionTextHorizontalAlignment } from './DimensionTextHorizontalAlignment.js';
export { DimensionTextVerticalAlignment } from './DimensionTextVerticalAlignment.js';
export { AngularZeroHandling } from './ZeroHandling.js';
export { ArcLengthSymbolPosition } from './ArcLengthSymbolPosition.js';
export { DimensionTextBackgroundFillMode } from './DimensionTextBackgroundFillMode.js';
export { FractionFormat } from './FractionFormat.js';
export { TextMovement } from './TextMovement.js';
export { TextDirection } from './TextDirection.js';
