// @node-projects/acad-ts/dist/Tables/Layer.js
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { Color } from '../Color.js';
import { LineWeightType } from '../Types/LineWeightType.js';
import { CadObject } from '../CadObject.js';
import { TableEntry } from './TableEntry.js';
import { LineType } from './LineType.js';
export class Layer extends TableEntry {
    static get Default() {
        return new Layer(Layer.DefaultName);
    }
    static get Defpoints() {
        const l = new Layer(Layer.DefpointsName);
        l.plotFlag = false;
        return l;
    }
    get color() {
        return this._color;
    }
    set color(value) {
        if (value.isByLayer || value.isByBlock) {
            throw new Error('The layer color cannot be ByLayer or ByBlock');
        }
        this._color = value;
    }
    get layerFlags() {
        return this.flags;
    }
    set layerFlags(value) {
        this.flags = value;
    }
    isOn = true;
    lineType = null;
    lineWeight = LineWeightType.Default;
    material = null;
    get objectName() {
        return DxfFileToken.TableLayer;
    }
    get objectType() {
        return ObjectType.LAYER;
    }
    get plotFlag() {
        if (this._name.toLowerCase() === Layer.DefpointsName.toLowerCase()) {
            return false;
        }
        return this._plotFlag;
    }
    set plotFlag(value) {
        this._plotFlag = value;
    }
    plotStyleName = 0;
    get subclassMarker() {
        return DxfSubclassMarker.Layer;
    }
    static DefaultName = '0';
    static DefpointsName = 'defpoints';
    _color = new Color(7);
    _plotFlag = true;
    constructor(name) {
        super(name);
    }
    clone() {
        const clone = super.clone();
        // TODO: clone.lineType = this.lineType?.clone();
        // TODO: clone.material = this.material?.clone();
        return clone;
    }
    /** @internal */
    assignDocument(doc) {
        super.assignDocument(doc);
        this.lineType = CadObject.updateCollection(this.lineType ?? LineType.Continuous, doc.lineTypes);
    }
    /** @internal */
    unassignDocument() {
        super.unassignDocument();
        this.lineType = this.lineType?.clone?.() ?? LineType.Continuous;
    }
}
export { LayerFlags } from './LayerFlags.js';
