// @node-projects/acad-ts/dist/Tables/LineType.js
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { TableEntry } from './TableEntry.js';
import { LineTypeShapeFlags } from './LinetypeShapeFlags.js';
export class LineTypeSegment {
    length = 0;
    shapeFlags = LineTypeShapeFlags.None;
    shapeNumber = 0;
    offset = { x: 0, y: 0 };
    rotation = 0;
    scale = 0;
    text = '';
    style = null;
    owner = null;
    get isShape() {
        return (this.shapeFlags & LineTypeShapeFlags.Shape) !== 0;
    }
    clone() {
        const clone = new LineTypeSegment();
        clone.length = this.length;
        clone.shapeFlags = this.shapeFlags;
        clone.shapeNumber = this.shapeNumber;
        clone.offset = { ...this.offset };
        clone.rotation = this.rotation;
        clone.scale = this.scale;
        clone.text = this.text;
        clone.style = this.style;
        clone.owner = null;
        return clone;
    }
}
export class LineType extends TableEntry {
    static get ByBlock() {
        return new LineType('ByBlock');
    }
    static get ByLayer() {
        return new LineType('ByLayer');
    }
    static get Continuous() {
        return new LineType('Continuous');
    }
    alignment = 'A';
    description = null;
    get hasShapes() {
        return this._segments.some(s => s.isShape);
    }
    get isComplex() {
        return this._segments.length > 0;
    }
    get objectName() {
        return DxfFileToken.TableLinetype;
    }
    get objectType() {
        return ObjectType.LTYPE;
    }
    get patternLength() {
        return this._segments.reduce((sum, s) => sum + Math.abs(s.length), 0);
    }
    get segments() {
        return this._segments;
    }
    get subclassMarker() {
        return DxfSubclassMarker.Linetype;
    }
    static ByBlockName = 'ByBlock';
    static ByLayerName = 'ByLayer';
    static ContinuousName = 'Continuous';
    _segments = [];
    constructor(name) {
        super(name);
    }
    addSegment(segment) {
        if (segment.owner != null) {
            throw new Error(`Segment already assigned to a LineType: ${segment.owner.name}`);
        }
        segment.owner = this;
        this._segments.push(segment);
    }
    clone() {
        const clone = super.clone();
        clone._segments = [];
        for (const segment of this._segments) {
            clone.addSegment(segment.clone());
        }
        return clone;
    }
}
export { LineTypeShapeFlags } from './LinetypeShapeFlags.js';
