// @node-projects/acad-ts/dist/Tables/TableEntry.js
import { CadObject } from '../CadObject.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { StandardFlags } from './StandardFlags.js';
import { OnNameChangedArgs } from '../OnNameChangedArgs.js';
export class TableEntry extends CadObject {
    onNameChanged = null;
    flags = StandardFlags.None;
    get name() {
        return this._name;
    }
    set name(value) {
        if (!value || !value.trim()) {
            throw new Error(`Table entry [${this.constructor.name}] must have a name`);
        }
        if (this.onNameChanged) {
            this.onNameChanged(this, new OnNameChangedArgs(this._name, value));
        }
        this._name = value;
    }
    get subclassMarker() {
        return DxfSubclassMarker.TableRecord;
    }
    _name = '';
    constructor(name) {
        super();
        if (name !== undefined) {
            if (!name) {
                throw new Error(`${this.constructor.name} must have a name.`);
            }
            this.name = name;
        }
    }
    clone() {
        const clone = super.clone();
        clone.onNameChanged = null;
        return clone;
    }
    toString() {
        return `${this.objectName}:${this.name}`;
    }
}
export { StandardFlags } from './StandardFlags.js';
