// @node-projects/acad-ts/dist/Tables/TextStyle.js
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { TableEntry } from './TableEntry.js';
import { StyleFlags } from './StyleFlags.js';
import { FontFlags } from './FontFlags.js';
// TODO: TextMirrorFlag from Entities not yet converted
export var TextMirrorFlag;
(function (TextMirrorFlag) {
    TextMirrorFlag[TextMirrorFlag["None"] = 0] = "None";
    TextMirrorFlag[TextMirrorFlag["FlipX"] = 2] = "FlipX";
    TextMirrorFlag[TextMirrorFlag["FlipY"] = 4] = "FlipY";
})(TextMirrorFlag || (TextMirrorFlag = {}));
export class TextStyle extends TableEntry {
    static DefaultName = 'Standard';
    get objectType() {
        return ObjectType.STYLE;
    }
    get objectName() {
        return DxfFileToken.TableStyle;
    }
    get subclassMarker() {
        return DxfSubclassMarker.TextStyle;
    }
    static get Default() {
        return new TextStyle(TextStyle.DefaultName);
    }
    get styleFlags() {
        return this.flags;
    }
    set styleFlags(value) {
        this.flags = value;
    }
    filename = '';
    bigFontFilename = null;
    height = 0;
    width = 1.0;
    lastHeight = 0;
    obliqueAngle = 0.0;
    mirrorFlag = TextMirrorFlag.None;
    trueType = FontFlags.Regular;
    get isShapeFile() {
        return (this.styleFlags & StyleFlags.IsShape) !== 0;
    }
    constructor(name) {
        super(name);
    }
}
export { StyleFlags } from './StyleFlags.js';
