// @node-projects/acad-ts/dist/Tables/View.js
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { OrthographicType } from '../Types/OrthographicType.js';
import { RenderMode } from '../Types/RenderMode.js';
import { TableEntry } from './TableEntry.js';
import { ViewModeType } from './ViewModeType.js';
import { XYZ } from '../Math/XYZ.js';
import { XY } from '../Math/XY.js';
export class View extends TableEntry {
    get objectType() {
        return ObjectType.VIEW;
    }
    get objectName() {
        return DxfFileToken.TableView;
    }
    get subclassMarker() {
        return DxfSubclassMarker.View;
    }
    height = 0;
    width = 0;
    lensLength = 0;
    frontClipping = 0;
    backClipping = 0;
    angle = 0;
    viewMode = ViewModeType.Off;
    isUcsAssociated = false;
    isPlottable = false;
    renderMode = RenderMode.Optimized2D;
    center = new XY(0, 0);
    direction = new XYZ(0, 0, 0);
    target = new XYZ(0, 0, 0);
    visualStyle = null;
    ucsOrigin = new XYZ(0, 0, 0);
    ucsXAxis = new XYZ(0, 0, 0);
    ucsYAxis = new XYZ(0, 0, 0);
    ucsElevation = 0;
    ucsOrthographicType = OrthographicType.None;
    constructor(name) {
        super(name);
    }
    clone() {
        const clone = super.clone();
        clone.visualStyle = this.visualStyle?.clone?.() ?? null;
        return clone;
    }
}
export { ViewModeType } from './ViewModeType.js';
export { RenderMode } from '../Types/RenderMode.js';
