// @node-projects/acad-ts/dist/Tables/VPort.js
import { Color } from '../Color.js';
import { DxfFileToken } from '../DxfFileToken.js';
import { DxfSubclassMarker } from '../DxfSubclassMarker.js';
import { ObjectType } from '../Types/ObjectType.js';
import { OrthographicType } from '../Types/OrthographicType.js';
import { RenderMode } from '../Types/RenderMode.js';
import { DefaultLightingType } from './DefaultLightingType.js';
import { GridFlags } from './GridFlags.js';
import { TableEntry } from './TableEntry.js';
import { UscIconType } from './UscIconType.js';
import { ViewModeType } from './ViewModeType.js';
import { XYZ } from '../Math/XYZ.js';
import { XY } from '../Math/XY.js';
function normalizeXYZ(v) {
    const len = Math.sqrt(v.x * v.x + v.y * v.y + v.z * v.z);
    if (len === 0)
        return new XYZ(0, 0, 0);
    return new XYZ(v.x / len, v.y / len, v.z / len);
}
export class VPort extends TableEntry {
    static DefaultName = '*Active';
    get objectType() {
        return ObjectType.VPORT;
    }
    get objectName() {
        return DxfFileToken.TableVport;
    }
    get subclassMarker() {
        return DxfSubclassMarker.VPort;
    }
    static get Default() {
        return new VPort(VPort.DefaultName);
    }
    bottomLeft = new XY(0, 0);
    topRight = new XY(1, 1);
    center = new XY(0, 0);
    snapBasePoint = new XY(0, 0);
    snapSpacing = new XY(0.5, 0.5);
    gridSpacing = new XY(10, 10);
    get direction() {
        return this._direction;
    }
    set direction(value) {
        this._direction = normalizeXYZ(value);
    }
    target = new XYZ(0, 0, 0);
    viewHeight = 10;
    aspectRatio = 1.0;
    lensLength = 50.0;
    frontClippingPlane = 0.0;
    backClippingPlane = 0;
    snapRotation = 0;
    twistAngle = 0;
    circleZoomPercent = 1000;
    renderMode = RenderMode.Optimized2D;
    viewMode = ViewModeType.FrontClippingZ;
    ucsIconDisplay = UscIconType.OnOrigin;
    snapOn = false;
    showGrid = true;
    isometricSnap = false;
    snapIsoPair = 0;
    origin = new XYZ(0, 0, 0);
    xAxis = new XYZ(1, 0, 0);
    yAxis = new XYZ(0, 1, 0);
    namedUcs = null;
    baseUcs = null;
    orthographicType = OrthographicType.None;
    elevation = 0;
    gridFlags = GridFlags._1 | GridFlags._2;
    minorGridLinesPerMajorGridLine = 5;
    visualStyle = null;
    useDefaultLighting = true;
    defaultLighting = DefaultLightingType.TwoDistantLights;
    brightness = 0;
    contrast = 0;
    ambientColor = new Color(0);
    _direction = new XYZ(0, 0, 1);
    constructor(name) {
        super(name);
    }
    clone() {
        const clone = super.clone();
        clone.baseUcs = this.baseUcs ? this.baseUcs.clone() : null;
        clone.namedUcs = this.namedUcs ? this.namedUcs.clone() : null;
        return clone;
    }
}
export { DefaultLightingType } from './DefaultLightingType.js';
export { GridFlags } from './GridFlags.js';
export { UscIconType } from './UscIconType.js';
