// @node-projects/acad-ts/dist/Tables/ZeroHandling.js
export var ZeroHandling;
(function (ZeroHandling) {
    ZeroHandling[ZeroHandling["SuppressZeroFeetAndInches"] = 0] = "SuppressZeroFeetAndInches";
    ZeroHandling[ZeroHandling["ShowZeroFeetAndInches"] = 1] = "ShowZeroFeetAndInches";
    ZeroHandling[ZeroHandling["ShowZeroFeetSuppressZeroInches"] = 2] = "ShowZeroFeetSuppressZeroInches";
    ZeroHandling[ZeroHandling["SuppressZeroFeetShowZeroInches"] = 3] = "SuppressZeroFeetShowZeroInches";
    ZeroHandling[ZeroHandling["SuppressDecimalLeadingZeroes"] = 4] = "SuppressDecimalLeadingZeroes";
    ZeroHandling[ZeroHandling["SuppressDecimalTrailingZeroes"] = 8] = "SuppressDecimalTrailingZeroes";
    ZeroHandling[ZeroHandling["SuppressDecimalLeadingAndTrailingZeroes"] = 12] = "SuppressDecimalLeadingAndTrailingZeroes";
})(ZeroHandling || (ZeroHandling = {}));
export var AngularZeroHandling;
(function (AngularZeroHandling) {
    AngularZeroHandling[AngularZeroHandling["DisplayAll"] = 0] = "DisplayAll";
    AngularZeroHandling[AngularZeroHandling["SuppressLeadingZeroes"] = 1] = "SuppressLeadingZeroes";
    AngularZeroHandling[AngularZeroHandling["SupressTrailingZeroes"] = 2] = "SupressTrailingZeroes";
    AngularZeroHandling[AngularZeroHandling["SupressAll"] = 3] = "SupressAll";
})(AngularZeroHandling || (AngularZeroHandling = {}));
