// @node-projects/layout2vector/dist/writers/acad-writer.js
/**
 * Shared AutoCAD document builder for DWG and DXF writers using @node-projects/acad-ts.
 * Builds a CadDocument from IR nodes with LwPolyline, Hatch, TextEntity entities.
 */
import { roundedQuadPath } from "../geometry.js";
import { normalizeWhitespaceAwareText } from "../shared/text-whitespace.js";
import { parseCssColor } from "./shared/css-color.js";
import { getVisibleStroke, isAxisAlignedRect, parseAverageBorderRadius as parseBorderRadius } from "./shared/writer-utils.js";
// Lazy-loaded dependencies
let CadDocument;
let ACadVersion;
let LwPolyline;
let LwPolylineVertex;
let Hatch;
let HatchBoundaryPath;
let HatchBoundaryPathPolyline;
let HatchPattern;
let HatchPatternTypeEnum;
let TextEntity;
let Color;
let XY;
let XYZ;
let DwgWriter;
let AcadDxfWriter;
async function ensureAcadLoaded() {
    if (!CadDocument) {
        const acad = await import("@node-projects/acad-ts");
        CadDocument = acad.CadDocument;
        ACadVersion = acad.ACadVersion;
        LwPolyline = acad.LwPolyline;
        LwPolylineVertex = acad.LwPolylineVertex;
        Hatch = acad.Hatch;
        HatchBoundaryPath = acad.HatchBoundaryPath;
        HatchBoundaryPathPolyline = acad.HatchBoundaryPathPolyline;
        HatchPattern = acad.HatchPattern;
        HatchPatternTypeEnum = acad.HatchPatternType;
        TextEntity = acad.TextEntity;
        Color = acad.Color;
        XY = acad.XY;
        XYZ = acad.XYZ;
        DwgWriter = acad.DwgWriter;
        AcadDxfWriter = acad.DxfWriter;
    }
}
/** Convert CSS color to acad-ts Color (true color). Returns undefined for invisible colors. */
function cssToAcadColor(color) {
    const parsed = parseCssColor(color);
    if (!parsed || parsed.a <= 0)
        return undefined;
    return new Color(parsed.r, parsed.g, parsed.b);
}
/** Generate arc points for a rounded corner. */
function arcPoints(cx, cy, r, startAngle, endAngle, segments) {
    const pts = [];
    for (let i = 0; i <= segments; i++) {
        const a = startAngle + (endAngle - startAngle) * (i / segments);
        pts.push({ x: cx + r * Math.cos(a), y: cy + r * Math.sin(a) });
    }
    return pts;
}
/**
 * Internal shared writer that builds a CadDocument from IR.
 * Used by both AcadDXFWriter and DWGWriter.
 */
class AcadDocumentBuilder {
    doc;
    maxY;
    acadVersion;
    constructor(options) {
        const z = options?.zoom ?? 1;
        this.maxY = (options?.maxY ?? 1000) * z;
        this.acadVersion = options?.acadVersion ?? ACadVersion?.AC1018;
    }
    async begin() {
        await ensureAcadLoaded();
        this.doc = new CadDocument(this.acadVersion);
        this.doc.createDefaults();
    }
    /** Flip Y coordinate for DXF/DWG (Y-up) coordinate system. */
    flipY(y) {
        return this.maxY - y;
    }
    /** Create a LwPolyline from vertices. */
    createPolyline(verts, closed, color) {
        const xyVerts = verts.map(v => new XY(v.x, this.flipY(v.y)));
        const poly = new LwPolyline(xyVerts);
        poly.isClosed = closed;
        if (color)
            poly.color = color;
        this.doc.modelSpace.entities.add(poly);
    }
    /** Add a solid-filled HATCH for the given vertices. */
    addSolidHatch(verts, fillColor) {
        const hatch = new Hatch();
        hatch.isSolid = true;
        hatch.patternType = HatchPatternTypeEnum.SolidFill;
        hatch.pattern = HatchPattern.Solid;
        hatch.color = fillColor;
        const boundaryEdge = new HatchBoundaryPathPolyline();
        boundaryEdge.isClosed = true;
        boundaryEdge.vertices = verts.map(v => new XYZ(v.x, this.flipY(v.y), 0));
        const path = new HatchBoundaryPath([boundaryEdge]);
        hatch.paths.push(path);
        this.doc.modelSpace.entities.add(hatch);
    }
    async drawPolygon(points, style) {
        const fillColor = cssToAcadColor(style.fill);
        const stroke = getVisibleStroke(style, (c) => cssToAcadColor(c));
        const strokeColor = stroke?.color;
        const trueColor = strokeColor ?? fillColor;
        const fillVisible = fillColor !== undefined;
        if (!fillVisible && !stroke)
            return;
        const dxfElW = Math.abs(points[1].x - points[0].x);
        const dxfElH = Math.abs(points[3].y - points[0].y);
        const radius = parseBorderRadius(style.borderRadius, dxfElW, dxfElH);
        if (radius > 0 && isAxisAlignedRect(points)) {
            const x = Math.min(points[0].x, points[1].x, points[2].x, points[3].x);
            const y = Math.min(points[0].y, points[1].y, points[2].y, points[3].y);
            const w = Math.abs(points[1].x - points[0].x);
            const h = Math.abs(points[3].y - points[0].y);
            const r = Math.min(radius, w / 2, h / 2);
            const ARC_SEGS = 8;
            const verts = [];
            verts.push({ x: x + r, y: y });
            verts.push({ x: x + w - r, y: y });
            verts.push(...arcPoints(x + w - r, y + r, r, -Math.PI / 2, 0, ARC_SEGS));
            verts.push({ x: x + w, y: y + h - r });
            verts.push(...arcPoints(x + w - r, y + h - r, r, 0, Math.PI / 2, ARC_SEGS));
            verts.push({ x: x + r, y: y + h });
            verts.push(...arcPoints(x + r, y + h - r, r, Math.PI / 2, Math.PI, ARC_SEGS));
            verts.push({ x: x, y: y + r });
            verts.push(...arcPoints(x + r, y + r, r, Math.PI, Math.PI * 1.5, ARC_SEGS));
            if (fillVisible && fillColor)
                this.addSolidHatch(verts, fillColor);
            this.createPolyline(verts, true, trueColor);
            return;
        }
        if (radius > 0 && !isAxisAlignedRect(points)) {
            const segs = roundedQuadPath(points, radius);
            const verts = [];
            for (const s of segs) {
                if (s.type === "M" || s.type === "L") {
                    verts.push({ x: s.x, y: s.y });
                }
                else if (s.type === "Q") {
                    const prev = verts[verts.length - 1];
                    if (prev) {
                        for (let t = 0.25; t <= 1; t += 0.25) {
                            const u = 1 - t;
                            verts.push({
                                x: u * u * prev.x + 2 * u * t * s.cx + t * t * s.x,
                                y: u * u * prev.y + 2 * u * t * s.cy + t * t * s.y,
                            });
                        }
                    }
                }
            }
            if (fillVisible && fillColor)
                this.addSolidHatch(verts, fillColor);
            this.createPolyline(verts, true, trueColor);
            return;
        }
        const verts = points.map(p => ({ x: p.x, y: p.y }));
        if (fillVisible && fillColor)
            this.addSolidHatch(verts, fillColor);
        this.createPolyline(verts, true, trueColor);
    }
    async drawPolyline(points, closed, style) {
        const fillColor = cssToAcadColor(style.fill);
        const stroke = getVisibleStroke(style, (c) => cssToAcadColor(c));
        const strokeColor = stroke?.color;
        const fillVisible = fillColor !== undefined;
        if (!fillVisible && !stroke)
            return;
        if (closed && fillVisible && fillColor && points.length >= 3) {
            this.addSolidHatch(points.map(p => ({ x: p.x, y: p.y })), fillColor);
        }
        const trueColor = strokeColor ?? fillColor;
        this.createPolyline(points.map(p => ({ x: p.x, y: p.y })), closed, trueColor);
    }
    async drawText(quad, text, style) {
        const sanitized = normalizeWhitespaceAwareText(text, style);
        if (sanitized.length === 0)
            return;
        const dx = quad[1].x - quad[0].x;
        const dy = quad[1].y - quad[0].y;
        const angleRad = Math.atan2(dy, dx);
        const angleDeg = -angleRad * (180 / Math.PI);
        const ldx = quad[3].x - quad[0].x;
        const ldy = quad[3].y - quad[0].y;
        const quadHeight = Math.sqrt(ldx * ldx + ldy * ldy) || 12;
        const styleFontSize = style.fontSize ? parseFloat(style.fontSize) : 12;
        const height = quadHeight > 0 ? Math.min(styleFontSize, quadHeight) : styleFontSize;
        const halfLeading = Math.max(0, (quadHeight - height) / 2);
        const ascentRatio = 0.75;
        const baselineT = quadHeight > 0 ? (halfLeading + ascentRatio * height) / quadHeight : 1;
        const bottomLeft = {
            x: quad[0].x + (quad[3].x - quad[0].x) * baselineT,
            y: quad[0].y + (quad[3].y - quad[0].y) * baselineT,
        };
        const textColor = cssToAcadColor(style.color) ?? cssToAcadColor(style.fill);
        const entity = new TextEntity();
        entity.insertPoint = new XYZ(bottomLeft.x, this.flipY(bottomLeft.y), 0);
        entity.height = height;
        entity.value = sanitized;
        if (Math.abs(angleDeg) > 0.1)
            entity.rotation = angleDeg;
        if (textColor)
            entity.color = textColor;
        this.doc.modelSpace.entities.add(entity);
    }
    async drawImage(_quad, _dataUrl, _width, _height, _style) {
        // Image embedding in acad-ts CadDocument is not straightforward — skip for now.
        // The polyline/hatch/text entities cover the main vector use case.
    }
    async end() {
        return this.doc;
    }
}
/**
 * DWG Writer using @node-projects/acad-ts.
 * Maps IR nodes to a CadDocument and serializes to DWG binary format.
 */
export class DWGWriter {
    builder;
    constructor(options) {
        this.builder = new AcadDocumentBuilder(options);
    }
    async begin() {
        await this.builder.begin();
    }
    async drawPolygon(points, style) {
        await this.builder.drawPolygon(points, style);
    }
    async drawPolyline(points, closed, style) {
        await this.builder.drawPolyline(points, closed, style);
    }
    async drawText(quad, text, style) {
        await this.builder.drawText(quad, text, style);
    }
    async drawImage(quad, dataUrl, width, height, style) {
        await this.builder.drawImage(quad, dataUrl, width, height, style);
    }
    async end() {
        await ensureAcadLoaded();
        const doc = await this.builder.end();
        // Estimate a generous buffer size for the DWG output
        const buffer = new ArrayBuffer(10 * 1024 * 1024);
        const writer = new DwgWriter(buffer, doc);
        writer.Write();
        const bytesWritten = writer.bytesWritten;
        writer.Dispose();
        return new Uint8Array(buffer, 0, bytesWritten);
    }
}
/**
 * DXF Writer using @node-projects/acad-ts.
 * Maps IR nodes to a CadDocument and serializes to DXF ASCII format.
 * This is an alternative to the existing DXFWriter (which uses @tarikjabiri/dxf).
 */
export class AcadDXFWriter {
    builder;
    constructor(options) {
        this.builder = new AcadDocumentBuilder(options);
    }
    async begin() {
        await this.builder.begin();
    }
    async drawPolygon(points, style) {
        await this.builder.drawPolygon(points, style);
    }
    async drawPolyline(points, closed, style) {
        await this.builder.drawPolyline(points, closed, style);
    }
    async drawText(quad, text, style) {
        await this.builder.drawText(quad, text, style);
    }
    async drawImage(quad, dataUrl, width, height, style) {
        await this.builder.drawImage(quad, dataUrl, width, height, style);
    }
    async end() {
        await ensureAcadLoaded();
        const doc = await this.builder.end();
        // Collect DXF output as text
        let result = "";
        const target = {
            write(value) { result += value; },
        };
        const writer = new AcadDxfWriter(target, doc);
        writer.Write();
        writer.Dispose();
        return result;
    }
}
