// @node-projects/layout2vector/dist/writers/dxf-writer.js
/**
 * DXF Writer using @tarikjabiri/dxf.
 * Maps IR nodes to DXF entities.
 */
// Lazy-loaded DXF dependencies (typed, possibly undefined)
let DxfWriter;
let point3d;
let point2d;
let HatchBoundaryPaths;
let HatchPolylineBoundary;
let HatchPredefinedPatterns;
let pattern;
async function ensureDxfLoaded() {
    if (!DxfWriter) {
        const dxf = await import("@tarikjabiri/dxf");
        DxfWriter = dxf.DxfWriter;
        point3d = dxf.point3d;
        point2d = dxf.point2d;
        HatchBoundaryPaths = dxf.HatchBoundaryPaths;
        HatchPolylineBoundary = dxf.HatchPolylineBoundary;
        HatchPredefinedPatterns = dxf.HatchPredefinedPatterns;
        pattern = dxf.pattern;
    }
}
import { roundedQuadPath } from "../geometry.js";
import { normalizeWhitespaceAwareText } from "../shared/text-whitespace.js";
import { cssColorToTrueColor } from "./shared/css-color.js";
import { getVisibleStroke, isAxisAlignedRect, parseAverageBorderRadius as parseBorderRadius } from "./shared/writer-utils.js";
/** Determine file extension from a data URL MIME type. */
function dataUrlToExtension(dataUrl) {
    const match = dataUrl.match(/^data:image\/([^;,]+)/);
    if (match) {
        const mime = match[1].toLowerCase();
        if (mime === "jpeg" || mime === "jpg")
            return "jpg";
        if (mime === "png")
            return "png";
        if (mime === "gif")
            return "gif";
        if (mime === "webp")
            return "webp";
        if (mime === "bmp")
            return "bmp";
        if (mime === "tiff")
            return "tiff";
    }
    return "jpg";
}
/** Generate arc points for a rounded corner. */
function arcPoints(cx, cy, r, startAngle, endAngle, segments) {
    const pts = [];
    for (let i = 0; i <= segments; i++) {
        const a = startAngle + (endAngle - startAngle) * (i / segments);
        pts.push({ x: cx + r * Math.cos(a), y: cy + r * Math.sin(a) });
    }
    return pts;
}
export class DXFWriter {
    dxf;
    maxY;
    embedImages;
    imageCounter = 0;
    fontStyles = new Map();
    /**
     * Image files referenced by the DXF output.
     * Maps relative file paths (as used in the DXF IMAGE entities) to data URL strings.
     * After calling `end()`, save these files alongside the DXF to display raster images.
     * Empty when `embedImages` is true.
     */
    imageFiles = new Map();
    /**
     * @param optionsOrMaxY Options object, or the maximum Y coordinate for Y-axis flipping (positional form).
     * @param zoom Scale factor applied to the maxY coordinate (positional form).
     */
    constructor(optionsOrMaxY, zoom) {
        if (typeof optionsOrMaxY === "object") {
            const z = optionsOrMaxY.zoom ?? 1;
            this.maxY = (optionsOrMaxY.maxY ?? 1000) * z;
            this.embedImages = optionsOrMaxY.embedImages ?? false;
        }
        else {
            const z = zoom ?? 1;
            this.maxY = (optionsOrMaxY ?? 1000) * z;
            this.embedImages = false;
        }
    }
    async begin() {
        await ensureDxfLoaded();
        this.dxf = new DxfWriter();
        this.imageCounter = 0;
        this.imageFiles.clear();
        this.fontStyles.clear();
    }
    /** Get or create a DXF text style for a given font family. Returns the style name. */
    getTextStyle(fontFamily) {
        // DXF dependencies are loaded in begin()
        if (!fontFamily)
            return undefined;
        // Extract the first font name from the CSS font-family list
        const name = fontFamily.split(",")[0]?.trim().replace(/['"]/g, "");
        if (!name || name === "serif" || name === "sans-serif" || name === "monospace")
            return undefined;
        if (this.fontStyles.has(name))
            return this.fontStyles.get(name);
        // Create a DXF text style referencing this TrueType font
        const style = this.dxf.tables.addStyle(name);
        style.fontFileName = name + ".ttf";
        this.fontStyles.set(name, name);
        return name;
    }
    /** Add a HATCH entity with SOLID fill for the given vertices. */
    addSolidHatch(verts, fillColor) {
        const BoundaryPaths = HatchBoundaryPaths;
        const PolylineBoundary = HatchPolylineBoundary;
        const PredefinedPatterns = HatchPredefinedPatterns;
        const Pattern = pattern;
        const boundary = new BoundaryPaths();
        const polyBoundary = new PolylineBoundary();
        for (const v of verts) {
            polyBoundary.add({ x: v.x, y: this.flipY(v.y) });
        }
        boundary.addPolylineBoundary(polyBoundary);
        this.dxf.addHatch(boundary, Pattern({ name: PredefinedPatterns.SOLID }), { trueColor: String(fillColor) });
    }
    async drawPolygon(points, style) {
        const fillColor = cssColorToTrueColor(style.fill);
        const stroke = getVisibleStroke(style, cssColorToTrueColor);
        const trueColor = stroke?.color ?? fillColor;
        const fillVisible = fillColor !== undefined;
        // Skip fully transparent elements
        if (!fillVisible && !stroke)
            return;
        const opts = trueColor !== undefined ? { trueColor: String(trueColor) } : undefined;
        // Check for rounded rect
        const dxfElW = Math.abs(points[1].x - points[0].x);
        const dxfElH = Math.abs(points[3].y - points[0].y);
        const radius = parseBorderRadius(style.borderRadius, dxfElW, dxfElH);
        if (radius > 0 && isAxisAlignedRect(points)) {
            const x = Math.min(points[0].x, points[1].x, points[2].x, points[3].x);
            const y = Math.min(points[0].y, points[1].y, points[2].y, points[3].y);
            const w = Math.abs(points[1].x - points[0].x);
            const h = Math.abs(points[3].y - points[0].y);
            const r = Math.min(radius, w / 2, h / 2);
            const ARC_SEGS = 8;
            // Build rounded rectangle polyline
            const verts = [];
            // Top edge (left to right)
            verts.push({ x: x + r, y: y });
            verts.push({ x: x + w - r, y: y });
            // Top-right corner
            verts.push(...arcPoints(x + w - r, y + r, r, -Math.PI / 2, 0, ARC_SEGS));
            // Right edge
            verts.push({ x: x + w, y: y + h - r });
            // Bottom-right corner
            verts.push(...arcPoints(x + w - r, y + h - r, r, 0, Math.PI / 2, ARC_SEGS));
            // Bottom edge
            verts.push({ x: x + r, y: y + h });
            // Bottom-left corner
            verts.push(...arcPoints(x + r, y + h - r, r, Math.PI / 2, Math.PI, ARC_SEGS));
            // Left edge
            verts.push({ x: x, y: y + r });
            // Top-left corner
            verts.push(...arcPoints(x + r, y + r, r, Math.PI, Math.PI * 1.5, ARC_SEGS));
            const vertices = verts.map((p) => ({
                point: point2d(p.x, this.flipY(p.y)),
            }));
            // Close
            vertices.push({ point: point2d(verts[0].x, this.flipY(verts[0].y)) });
            this.dxf.addLWPolyline(vertices, opts);
            return;
        }
        // Non-axis-aligned quad with border-radius: use rounded quad path as polyline approximation
        if (radius > 0 && !isAxisAlignedRect(points)) {
            const segs = roundedQuadPath(points, radius);
            const verts = [];
            for (const s of segs) {
                if (s.type === "M" || s.type === "L") {
                    verts.push({ x: s.x, y: s.y });
                }
                else if (s.type === "Q") {
                    // Approximate quadratic bezier with line segments
                    const prev = verts[verts.length - 1];
                    if (prev) {
                        for (let t = 0.25; t <= 1; t += 0.25) {
                            const u = 1 - t;
                            verts.push({
                                x: u * u * prev.x + 2 * u * t * s.cx + t * t * s.x,
                                y: u * u * prev.y + 2 * u * t * s.cy + t * t * s.y,
                            });
                        }
                    }
                }
            }
            if (fillVisible) {
                const fillColor2 = cssColorToTrueColor(style.fill);
                if (fillColor2 !== undefined)
                    this.addSolidHatch(verts, fillColor2);
            }
            const dxfVerts = verts.map(p => ({ point: point2d(p.x, this.flipY(p.y)) }));
            dxfVerts.push({ point: point2d(verts[0].x, this.flipY(verts[0].y)) });
            this.dxf.addLWPolyline(dxfVerts, opts);
            return;
        }
        const vertices = points.map((p) => ({
            point: point2d(p.x, this.flipY(p.y)),
        }));
        // Close the polygon
        vertices.push({
            point: point2d(points[0].x, this.flipY(points[0].y)),
        });
        this.dxf.addLWPolyline(vertices, opts);
    }
    async drawPolyline(points, closed, style) {
        const fillColor = cssColorToTrueColor(style.fill);
        const stroke = getVisibleStroke(style, cssColorToTrueColor);
        const strokeColor = stroke?.color;
        const fillVisible = fillColor !== undefined;
        // Skip fully transparent elements
        if (!fillVisible && !stroke)
            return;
        // Solid fill via HATCH for closed shapes with a visible fill
        if (closed && fillVisible && fillColor !== undefined && points.length >= 3) {
            this.addSolidHatch(points.map(p => ({ x: p.x, y: p.y })), fillColor);
        }
        // Always draw the polyline outline — use stroke color if available, otherwise fill color
        const trueColor = strokeColor ?? fillColor;
        const opts = trueColor !== undefined ? { trueColor: String(trueColor) } : undefined;
        const vertices = points.map((p) => ({
            point: point2d(p.x, this.flipY(p.y)),
        }));
        if (closed && points.length > 0) {
            vertices.push({
                point: point2d(points[0].x, this.flipY(points[0].y)),
            });
        }
        this.dxf.addLWPolyline(vertices, opts);
    }
    async drawText(quad, text, style) {
        const sanitized = normalizeWhitespaceAwareText(text, style);
        if (sanitized.length === 0)
            return;
        // Compute rotation angle from quad top edge (topLeft → topRight)
        const dx = quad[1].x - quad[0].x;
        const dy = quad[1].y - quad[0].y;
        const angleRad = Math.atan2(dy, dx);
        // DXF Y-up vs browser Y-down — negate the angle
        const angleDeg = -angleRad * (180 / Math.PI);
        // Compute text height from left edge of quad
        const ldx = quad[3].x - quad[0].x;
        const ldy = quad[3].y - quad[0].y;
        const quadHeight = Math.sqrt(ldx * ldx + ldy * ldy) || 12;
        // Use actual font size (em square) as DXF text height, not line-height
        const styleFontSize = style.fontSize ? parseFloat(style.fontSize) : 12;
        const height = quadHeight > 0 ? Math.min(styleFontSize, quadHeight) : styleFontSize;
        // Position at baseline: offset from quad[0] by halfLeading + ascent
        const halfLeading = Math.max(0, (quadHeight - height) / 2);
        const ascentRatio = 0.75; // approximate for most Latin fonts
        const baselineT = quadHeight > 0 ? (halfLeading + ascentRatio * height) / quadHeight : 1;
        const bottomLeft = {
            x: quad[0].x + (quad[3].x - quad[0].x) * baselineT,
            y: quad[0].y + (quad[3].y - quad[0].y) * baselineT,
        };
        // Text color: prefer style.color (CSS color), then style.fill
        const trueColor = cssColorToTrueColor(style.color) ?? cssColorToTrueColor(style.fill);
        const opts = {};
        if (trueColor !== undefined)
            opts.trueColor = String(trueColor);
        if (Math.abs(angleDeg) > 0.1)
            opts.rotation = angleDeg;
        const Point3d = point3d;
        const textEntity = this.dxf.addText(Point3d(bottomLeft.x, this.flipY(bottomLeft.y)), height, sanitized, Object.keys(opts).length > 0 ? opts : undefined);
        // Assign font-specific text style if needed
        const textStyleName = this.getTextStyle(style.fontFamily);
        if (textStyleName) {
            textEntity.textStyle = textStyleName;
        }
    }
    async drawImage(quad, dataUrl, width, height, _style) {
        const idx = ++this.imageCounter;
        let imagePath;
        if (this.embedImages) {
            // Embed the image data directly in the DXF as a data URL
            imagePath = dataUrl;
        }
        else {
            // Reference as external file
            const ext = dataUrlToExtension(dataUrl);
            const fileName = `images/image${idx}.${ext}`;
            this.imageFiles.set(fileName, dataUrl);
            imagePath = fileName;
        }
        // Compute rotation angle from the top edge of the quad (topLeft → topRight)
        const dx = quad[1].x - quad[0].x;
        const dy = quad[1].y - quad[0].y;
        const angleRad = Math.atan2(dy, dx);
        // Browser Y is down, DXF Y is up — negate the angle
        const angleDeg = -angleRad * (180 / Math.PI);
        // Compute display size from edge lengths (works for rotated quads)
        const displayWidth = Math.sqrt(dx * dx + dy * dy);
        // The scale parameter is the display size in DXF units (not a multiplier)
        const scale = displayWidth || 1;
        // Insertion point is bottom-left in DXF coords.
        // For DXF IMAGE, the insertion point is the top-left corner before rotation,
        // but in DXF Y-up coords that's quad[0] with flipped Y.
        const x = quad[0].x;
        const y = this.flipY(quad[0].y);
        const Point3d = point3d;
        this.dxf.addImage(imagePath, `image${idx}`, Point3d(x, y, 0), width, height, scale, angleDeg);
    }
    async end() {
        return this.dxf.stringify();
    }
    /** Flip Y coordinate for DXF (Y-up) coordinate system. */
    flipY(y) {
        return this.maxY - y;
    }
}
