// @node-projects/acad-ts/dist/XData/ExtendedDataDictionary.js
import { ExtendedData } from './ExtendedData.js';
export class ExtendedDataDictionary {
    get document() {
        return this.owner?.document ?? null;
    }
    owner;
    _data = new Map();
    constructor(owner) {
        this.owner = owner;
    }
    addByAppId(app, extendedData) {
        if (extendedData === undefined) {
            extendedData = new ExtendedData();
        }
        const doc = this.document;
        if (doc != null) {
            // TODO: Integrate with document's AppIds table once Tables are converted
            this._data.set(app, extendedData);
        }
        else {
            this._data.set(app, extendedData);
        }
    }
    addByName(appName, extendedData) {
        this.addByAppId({ name: appName }, extendedData);
    }
    addWithRecords(app, records) {
        this._data.set(app, new ExtendedData(records));
    }
    clear() {
        this._data.clear();
    }
    containsKey(app) {
        return this._data.has(app);
    }
    containsKeyName(name) {
        for (const key of this._data.keys()) {
            if (key.name === name)
                return true;
        }
        return false;
    }
    getByName(name) {
        const byName = this.getExtendedDataByName();
        const result = byName.get(name.toUpperCase());
        if (!result) {
            throw new Error(`AppId '${name}' not found`);
        }
        return result;
    }
    get(app) {
        const result = this._data.get(app);
        if (!result) {
            throw new Error(`AppId not found`);
        }
        return result;
    }
    get size() { return this._data.size; }
    entries() { return this._data.entries(); }
    add(key, value) { this._data.set(key, value); }
    set(key, value) { this._data.set(key, value); }
    [Symbol.iterator]() {
        return this._data.entries();
    }
    getExtendedDataByName() {
        const result = new Map();
        for (const [key, value] of this._data) {
            result.set(key.name.toUpperCase(), value);
        }
        return result;
    }
    tryAdd(appName, extendedData) {
        const existing = this.tryGetByName(appName);
        if (existing.found) {
            return existing.value;
        }
        else {
            this.addByName(appName, extendedData);
            return extendedData;
        }
    }
    tryGet(app) {
        const result = this._data.get(app);
        if (result) {
            return { value: result, found: true };
        }
        return { value: null, found: false };
    }
    tryGetByName(name) {
        const byName = this.getExtendedDataByName();
        const result = byName.get(name.toUpperCase());
        if (result) {
            return { value: result, found: true };
        }
        return { value: null, found: false };
    }
}
