// @node-projects/acad-ts/dist/Types/ObjectType.js
export var ObjectType;
(function (ObjectType) {
    ObjectType[ObjectType["UNLISTED"] = -999] = "UNLISTED";
    ObjectType[ObjectType["INVALID"] = -1] = "INVALID";
    ObjectType[ObjectType["UNDEFINED"] = 0] = "UNDEFINED";
    ObjectType[ObjectType["TEXT"] = 1] = "TEXT";
    ObjectType[ObjectType["ATTRIB"] = 2] = "ATTRIB";
    ObjectType[ObjectType["ATTDEF"] = 3] = "ATTDEF";
    ObjectType[ObjectType["BLOCK"] = 4] = "BLOCK";
    ObjectType[ObjectType["ENDBLK"] = 5] = "ENDBLK";
    ObjectType[ObjectType["SEQEND"] = 6] = "SEQEND";
    ObjectType[ObjectType["INSERT"] = 7] = "INSERT";
    ObjectType[ObjectType["MINSERT"] = 8] = "MINSERT";
    ObjectType[ObjectType["UNKNOW_9"] = 9] = "UNKNOW_9";
    ObjectType[ObjectType["VERTEX_2D"] = 10] = "VERTEX_2D";
    ObjectType[ObjectType["VERTEX_3D"] = 11] = "VERTEX_3D";
    ObjectType[ObjectType["VERTEX_MESH"] = 12] = "VERTEX_MESH";
    ObjectType[ObjectType["VERTEX_PFACE"] = 13] = "VERTEX_PFACE";
    ObjectType[ObjectType["VERTEX_PFACE_FACE"] = 14] = "VERTEX_PFACE_FACE";
    ObjectType[ObjectType["POLYLINE_2D"] = 15] = "POLYLINE_2D";
    ObjectType[ObjectType["POLYLINE_3D"] = 16] = "POLYLINE_3D";
    ObjectType[ObjectType["ARC"] = 17] = "ARC";
    ObjectType[ObjectType["CIRCLE"] = 18] = "CIRCLE";
    ObjectType[ObjectType["LINE"] = 19] = "LINE";
    ObjectType[ObjectType["DIMENSION_ORDINATE"] = 20] = "DIMENSION_ORDINATE";
    ObjectType[ObjectType["DIMENSION_LINEAR"] = 21] = "DIMENSION_LINEAR";
    ObjectType[ObjectType["DIMENSION_ALIGNED"] = 22] = "DIMENSION_ALIGNED";
    ObjectType[ObjectType["DIMENSION_ANG_3_Pt"] = 23] = "DIMENSION_ANG_3_Pt";
    ObjectType[ObjectType["DIMENSION_ANG_2_Ln"] = 24] = "DIMENSION_ANG_2_Ln";
    ObjectType[ObjectType["DIMENSION_RADIUS"] = 25] = "DIMENSION_RADIUS";
    ObjectType[ObjectType["DIMENSION_DIAMETER"] = 26] = "DIMENSION_DIAMETER";
    ObjectType[ObjectType["POINT"] = 27] = "POINT";
    ObjectType[ObjectType["FACE3D"] = 28] = "FACE3D";
    ObjectType[ObjectType["POLYLINE_PFACE"] = 29] = "POLYLINE_PFACE";
    ObjectType[ObjectType["POLYLINE_MESH"] = 30] = "POLYLINE_MESH";
    ObjectType[ObjectType["SOLID"] = 31] = "SOLID";
    ObjectType[ObjectType["TRACE"] = 32] = "TRACE";
    ObjectType[ObjectType["SHAPE"] = 33] = "SHAPE";
    ObjectType[ObjectType["VIEWPORT"] = 34] = "VIEWPORT";
    ObjectType[ObjectType["ELLIPSE"] = 35] = "ELLIPSE";
    ObjectType[ObjectType["SPLINE"] = 36] = "SPLINE";
    ObjectType[ObjectType["REGION"] = 37] = "REGION";
    ObjectType[ObjectType["SOLID3D"] = 38] = "SOLID3D";
    ObjectType[ObjectType["BODY"] = 39] = "BODY";
    ObjectType[ObjectType["RAY"] = 40] = "RAY";
    ObjectType[ObjectType["XLINE"] = 41] = "XLINE";
    ObjectType[ObjectType["DICTIONARY"] = 42] = "DICTIONARY";
    ObjectType[ObjectType["OLEFRAME"] = 43] = "OLEFRAME";
    ObjectType[ObjectType["MTEXT"] = 44] = "MTEXT";
    ObjectType[ObjectType["LEADER"] = 45] = "LEADER";
    ObjectType[ObjectType["TOLERANCE"] = 46] = "TOLERANCE";
    ObjectType[ObjectType["MLINE"] = 47] = "MLINE";
    ObjectType[ObjectType["BLOCK_CONTROL_OBJ"] = 48] = "BLOCK_CONTROL_OBJ";
    ObjectType[ObjectType["BLOCK_HEADER"] = 49] = "BLOCK_HEADER";
    ObjectType[ObjectType["LAYER_CONTROL_OBJ"] = 50] = "LAYER_CONTROL_OBJ";
    ObjectType[ObjectType["LAYER"] = 51] = "LAYER";
    ObjectType[ObjectType["STYLE_CONTROL_OBJ"] = 52] = "STYLE_CONTROL_OBJ";
    ObjectType[ObjectType["STYLE"] = 53] = "STYLE";
    ObjectType[ObjectType["UNKNOW_36"] = 54] = "UNKNOW_36";
    ObjectType[ObjectType["UNKNOW_37"] = 55] = "UNKNOW_37";
    ObjectType[ObjectType["LTYPE_CONTROL_OBJ"] = 56] = "LTYPE_CONTROL_OBJ";
    ObjectType[ObjectType["LTYPE"] = 57] = "LTYPE";
    ObjectType[ObjectType["UNKNOW_3A"] = 58] = "UNKNOW_3A";
    ObjectType[ObjectType["UNKNOW_3B"] = 59] = "UNKNOW_3B";
    ObjectType[ObjectType["VIEW_CONTROL_OBJ"] = 60] = "VIEW_CONTROL_OBJ";
    ObjectType[ObjectType["VIEW"] = 61] = "VIEW";
    ObjectType[ObjectType["UCS_CONTROL_OBJ"] = 62] = "UCS_CONTROL_OBJ";
    ObjectType[ObjectType["UCS"] = 63] = "UCS";
    ObjectType[ObjectType["VPORT_CONTROL_OBJ"] = 64] = "VPORT_CONTROL_OBJ";
    ObjectType[ObjectType["VPORT"] = 65] = "VPORT";
    ObjectType[ObjectType["APPID_CONTROL_OBJ"] = 66] = "APPID_CONTROL_OBJ";
    ObjectType[ObjectType["APPID"] = 67] = "APPID";
    ObjectType[ObjectType["DIMSTYLE_CONTROL_OBJ"] = 68] = "DIMSTYLE_CONTROL_OBJ";
    ObjectType[ObjectType["DIMSTYLE"] = 69] = "DIMSTYLE";
    ObjectType[ObjectType["VP_ENT_HDR_CTRL_OBJ"] = 70] = "VP_ENT_HDR_CTRL_OBJ";
    ObjectType[ObjectType["VP_ENT_HDR"] = 71] = "VP_ENT_HDR";
    ObjectType[ObjectType["GROUP"] = 72] = "GROUP";
    ObjectType[ObjectType["MLINESTYLE"] = 73] = "MLINESTYLE";
    ObjectType[ObjectType["OLE2FRAME"] = 74] = "OLE2FRAME";
    ObjectType[ObjectType["DUMMY"] = 75] = "DUMMY";
    ObjectType[ObjectType["LONG_TRANSACTION"] = 76] = "LONG_TRANSACTION";
    ObjectType[ObjectType["LWPOLYLINE"] = 77] = "LWPOLYLINE";
    ObjectType[ObjectType["HATCH"] = 78] = "HATCH";
    ObjectType[ObjectType["XRECORD"] = 79] = "XRECORD";
    ObjectType[ObjectType["ACDBPLACEHOLDER"] = 80] = "ACDBPLACEHOLDER";
    ObjectType[ObjectType["VBA_PROJECT"] = 81] = "VBA_PROJECT";
    ObjectType[ObjectType["LAYOUT"] = 82] = "LAYOUT";
    ObjectType[ObjectType["ACAD_PROXY_ENTITY"] = 498] = "ACAD_PROXY_ENTITY";
    ObjectType[ObjectType["ACAD_PROXY_OBJECT"] = 499] = "ACAD_PROXY_OBJECT";
})(ObjectType || (ObjectType = {}));
