// @node-projects/acad-ts/dist/Math/Matrix4.js
export class Matrix4 {
    m00 = 1;
    m01 = 0;
    m02 = 0;
    m03 = 0;
    m10 = 0;
    m11 = 1;
    m12 = 0;
    m13 = 0;
    m20 = 0;
    m21 = 0;
    m22 = 1;
    m23 = 0;
    m30 = 0;
    m31 = 0;
    m32 = 0;
    m33 = 1;
    constructor(values) {
        if (!values || values.length === 0) {
            return;
        }
        if (values.length !== 16) {
            throw new Error('Matrix4 requires 16 values.');
        }
        [
            this.m00, this.m01, this.m02, this.m03,
            this.m10, this.m11, this.m12, this.m13,
            this.m20, this.m21, this.m22, this.m23,
            this.m30, this.m31, this.m32, this.m33,
        ] = values;
    }
    // PascalCase aliases
    get M00() { return this.m00; }
    set M00(v) { this.m00 = v; }
    get M01() { return this.m01; }
    set M01(v) { this.m01 = v; }
    get M02() { return this.m02; }
    set M02(v) { this.m02 = v; }
    get M03() { return this.m03; }
    set M03(v) { this.m03 = v; }
    get M10() { return this.m10; }
    set M10(v) { this.m10 = v; }
    get M11() { return this.m11; }
    set M11(v) { this.m11 = v; }
    get M12() { return this.m12; }
    set M12(v) { this.m12 = v; }
    get M13() { return this.m13; }
    set M13(v) { this.m13 = v; }
    get M20() { return this.m20; }
    set M20(v) { this.m20 = v; }
    get M21() { return this.m21; }
    set M21(v) { this.m21 = v; }
    get M22() { return this.m22; }
    set M22(v) { this.m22 = v; }
    get M23() { return this.m23; }
    set M23(v) { this.m23 = v; }
    get M30() { return this.m30; }
    set M30(v) { this.m30 = v; }
    get M31() { return this.m31; }
    set M31(v) { this.m31 = v; }
    get M32() { return this.m32; }
    set M32(v) { this.m32 = v; }
    get M33() { return this.m33; }
    set M33(v) { this.m33 = v; }
    static identity() {
        return new Matrix4();
    }
    get(row, col) {
        const key = `m${row}${col}`;
        return this[key];
    }
    setElement(row, col, value) {
        const key = `m${row}${col}`;
        this[key] = value;
    }
}
