// @node-projects/acad-ts/dist/Math/XY.js
export class XY {
    x;
    y;
    get dimension() { return 2; }
    get [0]() { return this.x; }
    set [0](v) { this.x = v; }
    get [1]() { return this.y; }
    set [1](v) { this.y = v; }
    // PascalCase aliases for C# compat
    get X() { return this.x; }
    set X(v) { this.x = v; }
    get Y() { return this.y; }
    set Y(v) { this.y = v; }
    constructor(x = 0, y = 0) {
        this.x = x;
        this.y = y;
    }
    static Zero = new XY(0, 0);
    static AxisX = new XY(1, 0);
    static AxisY = new XY(0, 1);
    getLength() {
        return Math.sqrt(this.x * this.x + this.y * this.y);
    }
    normalize() {
        const len = this.getLength();
        if (len === 0)
            return new XY();
        return new XY(this.x / len, this.y / len);
    }
    getAngle() {
        return Math.atan2(this.y, this.x);
    }
    dot(other) {
        return this.x * other.x + this.y * other.y;
    }
    cross(other) {
        return this.x * other.y - this.y * other.x;
    }
    static Cross(a, b) {
        return a.x * b.y - a.y * b.x;
    }
    static Polar(point, angle, distance) {
        return new XY(point.x + distance * Math.cos(angle), point.y + distance * Math.sin(angle));
    }
    static Rotate(point, angle) {
        const cos = Math.cos(angle);
        const sin = Math.sin(angle);
        return new XY(point.x * cos - point.y * sin, point.x * sin + point.y * cos);
    }
    equals(other) {
        return this.x === other.x && this.y === other.y;
    }
}
