// @node-projects/acad-ts/dist/Math/XYZ.js
export class XYZ {
    x;
    y;
    z;
    get dimension() { return 3; }
    get [0]() { return this.x; }
    set [0](v) { this.x = v; }
    get [1]() { return this.y; }
    set [1](v) { this.y = v; }
    get [2]() { return this.z; }
    set [2](v) { this.z = v; }
    // PascalCase aliases for C# compat
    get X() { return this.x; }
    set X(v) { this.x = v; }
    get Y() { return this.y; }
    set Y(v) { this.y = v; }
    get Z() { return this.z; }
    set Z(v) { this.z = v; }
    constructor(x = 0, y = 0, z = 0) {
        this.x = x;
        this.y = y;
        this.z = z;
    }
    static Zero = new XYZ(0, 0, 0);
    static AxisX = new XYZ(1, 0, 0);
    static AxisY = new XYZ(0, 1, 0);
    static AxisZ = new XYZ(0, 0, 1);
    static NaN = new XYZ(Number.NaN, Number.NaN, Number.NaN);
    getLength() {
        return Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
    }
    normalize() {
        const len = this.getLength();
        if (len === 0)
            return new XYZ();
        return new XYZ(this.x / len, this.y / len, this.z / len);
    }
    getAngle() {
        return Math.atan2(this.y, this.x);
    }
    dot(other) {
        return this.x * other.x + this.y * other.y + this.z * other.z;
    }
    cross(other) {
        return new XYZ(this.y * other.z - this.z * other.y, this.z * other.x - this.x * other.z, this.x * other.y - this.y * other.x);
    }
    static Cross(a, b) {
        return new XYZ(a.y * b.z - a.z * b.y, a.z * b.x - a.x * b.z, a.x * b.y - a.y * b.x);
    }
    static equals(a, b) {
        return a.x === b.x && a.y === b.y && a.z === b.z;
    }
    equals(other) {
        return this.x === other.x && this.y === other.y && this.z === other.z;
    }
}
