// @spectrum-web-components/radio/src/Radio.js
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
import { __decorate } from '/web-component-designer-demo/node_modules/tslib/tslib.es6.js';
import { html, SpectrumElement, } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/./src/index.js';
import { property } from '/web-component-designer-demo/node_modules/@spectrum-web-components/base/src/decorators.js';
import { FocusVisiblePolyfillMixin } from '/web-component-designer-demo/node_modules/@spectrum-web-components/shared/src/focus-visible.js';
import radioStyles from './radio.css.js';
/**
 * @element sp-radio
 *
 * @slot - text label of the Radio button
 * @attr label-below - Moves the label below the radio button
 * @attr invalid - Uses the invalid style
 * @attr disabled - Uses the disabled style
 * @attr checked - Represents when the input is checked
 * @attr value - Identifies this radio button within its radio group
 *
 * @fires change - When the input is interacted with and its state is changed
 */
export class Radio extends FocusVisiblePolyfillMixin(SpectrumElement) {
    constructor() {
        super(...arguments);
        /**
         * When this control is rendered, focus it automatically
         * @private
         */
        this.autofocus = false;
        this.value = '';
        this.checked = false;
        this.disabled = false;
        this.emphasized = false;
        this.invalid = false;
        this.readonly = false;
    }
    static get styles() {
        return [radioStyles];
    }
    click() {
        if (this.disabled) {
            return;
        }
        this.activate();
    }
    manageAutoFocus() {
        if (this.autofocus) {
            /**
             * Trick :focus-visible polyfill into thinking keyboard based focus
             *
             * @private
             **/
            this.dispatchEvent(new KeyboardEvent('keydown', {
                code: 'Tab',
            }));
            this.focus();
        }
    }
    activate() {
        this.checked = true;
        this.dispatchEvent(new Event('change', {
            bubbles: true,
            composed: true,
        }));
    }
    handleKeyup(event) {
        if (event.code === 'Space') {
            this.activate();
        }
    }
    render() {
        return html `
            <div id="input"></div>
            <span id="button"></span>
            <span id="label" role="presentation"><slot></slot></span>
        `;
    }
    firstUpdated(changes) {
        super.firstUpdated(changes);
        this.setAttribute('role', 'radio');
        if (!this.hasAttribute('tabindex')) {
            this.tabIndex = 0;
        }
        this.manageAutoFocus();
        this.addEventListener('click', this.activate);
        this.addEventListener('keyup', this.handleKeyup);
    }
    updated(changes) {
        super.updated(changes);
        if (changes.has('invalid')) {
            if (this.invalid) {
                this.setAttribute('aria-invalid', 'true');
            }
            else {
                this.removeAttribute('aria-invalid');
            }
        }
        if (changes.has('checked')) {
            if (this.checked) {
                this.setAttribute('aria-checked', 'true');
            }
            else {
                this.setAttribute('aria-checked', 'false');
            }
        }
        if (changes.has('disabled')) {
            if (this.disabled) {
                this.setAttribute('aria-disabled', 'true');
            }
            else {
                this.removeAttribute('aria-disabeld');
            }
        }
    }
}
__decorate([
    property({ type: Boolean })
], Radio.prototype, "autofocus", void 0);
__decorate([
    property({ type: String, reflect: true })
], Radio.prototype, "value", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], Radio.prototype, "checked", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], Radio.prototype, "disabled", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], Radio.prototype, "emphasized", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], Radio.prototype, "invalid", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], Radio.prototype, "readonly", void 0);
